<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="UM Bioinformatics Core Workshop Team" />


<title>Cell Ranger in Action</title>

<script src="site_libs/header-attrs-2.25/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/paper.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.13.2/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/default.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<!--
Favicon dervied from
https://twemoji.twitter.com/
https://favicon.io/emoji-favicons/dna/
-->
<link rel="apple-touch-icon" sizes="180x180" href="/apple-touch-icon.png">
<link rel="icon" type="image/png" sizes="32x32" href="/favicon-32x32.png">
<link rel="icon" type="image/png" sizes="16x16" href="/favicon-16x16.png">
<link rel="shortcut icon" href="favicon-16x16.png" />
<link rel="manifest" href="/site.webmanifest">
<link rel="mask-icon" href="/safari-pinned-tab.svg" color="#5bbad5">
<meta name="msapplication-TileColor" content="#da532c">
<meta name="theme-color" content="#ffffff">

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>









<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-inverse  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">Intro to scRNA-Seq</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="workshop_intro.html">Intro</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Day 1
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="00A-OrientingOnScRNASeq.html">Orienting on scRNA-Seq</a>
    </li>
    <li>
      <a href="01-GettingStarted.html">Getting Started with Seurat</a>
    </li>
    <li>
      <a href="00B-CellRangerInAction.html">Cell Ranger in Action</a>
    </li>
    <li>
      <a href="02-QCandFiltering.html">Secondary QC &amp; Filtering</a>
    </li>
    <li>
      <a href="03-Normalization.html">Normalization</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Day 2
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="04-PCAandIntegration.html">PCA &amp; Integration</a>
    </li>
    <li>
      <a href="05-ProjectionAndClustering.html">Projection &amp; Clustering</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Day 3
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="06-MarkerVisualization.html">Marker identification and visualization</a>
    </li>
    <li>
      <a href="07-CellTypeAnnos.html">Cell type annotation</a>
    </li>
    <li>
      <a href="08-DifferentialExpression.html">Differential expression analysis</a>
    </li>
  </ul>
</li>
<li>
  <a href="workshop_wrap_up.html">Wrap up</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Cell Ranger in Action</h1>
<h4 class="author">UM Bioinformatics Core Workshop Team</h4>

</div>


<style type="text/css">

body {
   font-size: 18px;
}
code.r{
  font-size: 12px;
}
pre {
  font-size: 12px
}

table.fig, th.fig, td.fig {
  border: 1px solid black;
  border-collapse: collapse;
  padding: 15px;
}
</style>
<div id="workflow-overview" class="section level1 unlisted unnumbered">
<h1 class="unlisted unnumbered">Workflow Overview</h1>
<p><br/> <img src="images/wayfinder/wayfinder.png" alt="wayfinder" style="height: 400px;"/> <br/> <br/></p>
<div id="objectives" class="section level2">
<h2>Objectives</h2>
<ul>
<li>Understand what Cell Ranger does and also how to run <code>cellranger count</code></li>
<li>Describe the key inputs to Cell Ranger.</li>
<li>Describe the purpose and overall structure of key Cell Ranger outputs.</li>
<li>Interpret a <code>cellranger count</code> web_summary.html report to determine sample quality and inform decisions about additional sequencing.</li>
<li>Understand when you would need to run Cell Ranger and the relevant system requirements. <br/> <br/></li>
</ul>
</div>
<div id="cell-ranger-overview" class="section level2">
<h2>Cell Ranger Overview</h2>
<p>The initial data processing steps in scRNA-Seq transform sample FASTQ files to gene expression counts. Some of these steps are similar to bulk RNA-Seq and some are distinct. 10x Genomics provides a pipeline tool, Cell Ranger, to expedite these initial transformations <a href="#references">[1]</a>.</p>
<table class="fig">
<tr>
<th class="fig">
Cell Ranger inputs and outputs
</th>
</tr>
<tr>
<td class="fig">
<img src="images/cellranger/cellranger-30k.png" />
</td>
</tr>
<tr>
<td class="fig">
Cell Ranger accepts a set of sample FASTQ files and a set of reference files. It produces a QC web summary (detailed below), a set of matrix files (also detailed below) along with a large collection of supporting files. The matrix files can be used for downstream analysis in Seurat and other tools.
</td>
</tr>
</table>
<br/><br/>
<table class="fig">
<tr>
<th class="fig">
A high level view of Cell Ranger steps
</th>
</tr>
<tr>
<td class="fig">
<img src="images/cellranger/cellranger-5k.png" />
</td>
</tr>
<tr>
<td class="fig">
A. For each sequence, the barcode and UMI (on read 1) is used to label and bin the sequences by their barcodes.<br/> B. The transcript sequences are aligned to a genome reference using a modified version of the STAR aligner.<br/> C. For each transcript that falls in a gene, count the pile-up of sequences aligned to each gene; deduplicate the reads based on UMI. Use these deduplicated counts to build a feature barcode matrix.<br/> D. Apply a cell-calling algorithm to distinguish putative cells from background <a href="#references">[3,4]</a><br/> E. Subset the raw matrix to only putative cells to create the filtered matrix.<br/> F. Use information from all previous steps to build a web summary QC report as well as basic gene expression clustering visualizations and other supporting files.<br/>
</td>
</tr>
</table>
<p><br/> <br/></p>
</div>
<div id="key-inputs" class="section level2">
<h2>Key inputs</h2>
<p>The basic data package from the UM Advanced Genomics Core includes:</p>
<ul>
<li>an <strong>*.md5</strong> file to validate your data transfer.</li>
<li>a <strong>DemuxStats_*.csv</strong> that has some basic metrics about how your samples performed on the sequencer.</li>
<li>the <strong>fastq_*</strong> folder containing the fastq.gz files</li>
<li>a <strong>README.txt</strong> including details about how your samples were processed.</li>
</ul>
<p>It looks like the tree below.</p>
<pre><code>0000-SR
├── 0000-SR.md5
├── DemuxStats_0000-SR.csv
├── fastqs_0000-SR
│   ├── 0000-SR-1-GEX_S25_R1_001.fastq.gz
│   ├── 0000-SR-1-GEX_S25_R2_001.fastq.gz
│   ├── 0000-SR-2-GEX_S26_R1_001.fastq.gz
│   ├── 0000-SR-2-GEX_S26_R2_001.fastq.gz
│   ├── 0000-SR-3-GEX_S27_R1_001.fastq.gz
│   ├── 0000-SR-3-GEX_S27_R2_001.fastq.gz
│   ├── 0000-SR-4-GEX_S28_R1_001.fastq.gz
│   └── 0000-SR-4-GEX_S28_R2_001.fastq.gz
└── README.txt</code></pre>
<p>All you need to run 10x cellranger count are the above 10x sample FASTQ files and the correctly formatted <a href="https://www.10xgenomics.com/support/software/cell-ranger/downloads#reference-downloads" target="_blank">reference genome files</a>.</p>
<table class="fig">
<tr>
<td class="fig">
<strong>As a researcher, what parts of these outputs should I download and save?</strong>
</td>
</tr>
<tr>
<td class="fig">
<ol style="list-style-type: decimal">
<li>If we sent it to you, it is probably important for either downstream analysis or publication of your sequencing data. So we recomend you download <strong>all of it</strong>.</li>
<li>You can verify all of your files are intact using the *.md5 file.</li>
<li>Keep these data safe as you are responsible for the storage of your data for publication and archival purposes. (Typically sequencing providers will regularly delete project data following delivery.)</li>
</ol>
</td>
</tr>
</table>
<p><br/> <br/></p>
</div>
<div id="how-to-run-cell-ranger" class="section level2">
<h2>How to run Cell Ranger</h2>
<p>Cell Ranger is run from a Linux/Unix command shell <a href="#references">[2]</a>. You can print the usage statement to see what is needed to build the command: <code>cellranger count --help</code></p>
<pre><code>cellranger-count
Count gene expression (targeted or whole-transcriptome) and/or feature barcode reads from a single sample and GEM well
 
USAGE:
    cellranger count [FLAGS] [OPTIONS] --id &amp;lt;ID&amp;gt; --transcriptome &amp;lt;PATH&amp;gt;
     
FLAGS:
          --no-bam                  Do not generate a bam file
          --nosecondary             Disable secondary analysis, e.g. clustering. Optional
          --include-introns         Include intronic reads in count
          --no-libraries            Proceed with processing using a --feature-ref but no Feature Barcode libraries
                                    specified with the &amp;#x27;libraries&amp;#x27; flag
          --no-target-umi-filter    Turn off the target UMI filtering subpipeline. Only applies when --target-panel is
                            used
          --dry                     Do not execute the pipeline. Generate a pipeline invocation (.mro) file and stop
          --disable-ui              Do not serve the web UI
          --noexit                  Keep web UI running after pipestance completes or fails
          --nopreflight             Skip preflight checks
      -h, --help                    Prints help information
  ...</code></pre>
<p>To run <code>cellranger count</code>, you need to specify:</p>
<ul>
<li><strong><code>--id</code></strong>. This can be any string less than 64 characters. Cell Ranger creates an output directory named for the id string.</li>
<li><strong><code>--fastqs</code></strong> path to the directory containing the FASTQ files.</li>
<li><strong><code>--sample</code></strong> argument to specify which subset of FASTQ files to use (specified sample string should match the beginning of the FASTQ file name).</li>
<li><strong><code>--transcriptome</code></strong> a path to the reference transcriptome. Typically you would download a specific transcript for your organism from <a href="https://www.10xgenomics.com/support/software/cell-ranger/downloads#reference-downloads" target="_blank">10x Genomics</a>.</li>
</ul>
<p>Recall the directory of outputs from the sequencer (from above):</p>
<pre><code>0000-SR
├── 0000-SR.md5
├── DemuxStats_0000-SR.csv
├── fastqs_0000-SR
│   ├── 0000-SR-1-GEX_S25_R1_001.fastq.gz
│   ├── 0000-SR-1-GEX_S25_R2_001.fastq.gz
│   ├── 0000-SR-2-GEX_S26_R1_001.fastq.gz
│   ├── 0000-SR-2-GEX_S26_R2_001.fastq.gz
│   ├── 0000-SR-3-GEX_S27_R1_001.fastq.gz
│   ├── 0000-SR-3-GEX_S27_R2_001.fastq.gz
│   ├── 0000-SR-4-GEX_S28_R1_001.fastq.gz
│   └── 0000-SR-4-GEX_S28_R2_001.fastq.gz
└── README.txt</code></pre>
<p>We can create the <code>cellranger count</code> command:</p>
<pre><code>cellranger count --id=Sample_0000-SR-1 \
   --fastqs=/nfs/turbo/path/to/0000-SR/0000-SR_fastqs \
   --sample=0000-SR-1\
   --transcriptome=path/to/refereces/cellranger_count/refdata-gex-GRCh38-2020-A`</code></pre>
<p>Data from a typical sample can take several hours to complete, so we’ve completed this step for you.</p>
<!-- This content will not appear in the rendered Markdown 
 but we're using a smaller data set that should process quickly.

The `cellranger count` screen output is similar to the following:

```
/mnt/yard/user.name/yard/apps/cellranger-7.2.0/bin
cellranger count (7.2.0)

cellranger count (7.2.0)
Copyright (c) 2021 10x Genomics, Inc.  All rights reserved.
-------------------------------------------------------------------------------

Martian Runtime - v4.0.6
...
2021-10-15 17:12:42 [perform] Serializing pipestance performance data.
Waiting 6 seconds for UI to do final refresh.
Pipestance completed successfully!
```

When the output of the `cellranger count` command says, “Pipestance completed successfully!”, this means the job is done.-->
<p><br/> <br/></p>
</div>
<div id="key-outputs" class="section level2">
<h2>Key outputs</h2>
<p>The <code>cellranger count</code> pipeline outputs are in the Sample directory in the outs folder (aka <code>~/ISC_Shell/cellranger_outputs/count_run_HODay0replicate1</code>).</p>
<p>The output is similar to the following:</p>
<pre><code>count_run_HODay0replicate1
├──...
└──outs
    ├── analysis 
    ├── cloupe.cloupe
    ├── filtered_feature_bc_matrix
    ├── filtered_feature_bc_matrix.h5
    ├── metrics_summary.csv
    ├── molecule_info.h5
    ├── possorted_genome_bam.bam
    ├── possorted_genome_bam.bam.bai
    ├── raw_feature_bc_matrix
    ├── raw_feature_bc_matrix.h5
    └── web_summary.html</code></pre>
<p>10x Genomics provides excellent documentation on all the Cell Ranger <a href="https://www.10xgenomics.com/support/software/cell-ranger/latest/analysis/outputs/cr-outputs-gex-overview" target="_blank">outputs</a>. Some of the most important outputs include:</p>
<ul>
<li><strong>web_summary.html</strong> : A comprehensive report that includes quality control information and basic analysis.</li>
<li><strong>metrics_summary.csv</strong> : Same data as the web_summary.html but in csv format.</li>
<li><strong>raw_feature_bc_matrix</strong>: A set of three files that represent the feature x barcode count table; includes all detected barcodes. (Format details elaborated below.)</li>
<li><strong>filtered_feature_bc_matrix</strong> : As above, but filtered to include only barcodes that passed Cell Ranger’s cell-calling algorithm <a href="#references">[3]</a>.</li>
<li><strong>molecule.h5</strong> : An HDF5 (binary) file all containing information for all molecules assigned to a feature/barcode. See references <a href="#references">[6]</a> for more info on this format.</li>
<li><strong>cloupe file</strong> : Used with the 10x Loupe Browser visualization software.</li>
</ul>
<div id="web-summary-html-report" class="section level3">
<h3>Web Summary html report</h3>
<p>It is useful to look at the web_summary.html files to get an idea of how each of your samples performed in the 10x pipeline, epecially if there were errors/wanrings and/or if your research could benefit from additional sequencing. <a href="https://cdn.10xgenomics.com/image/upload/v1660261286/support-documents/CG000329_TechnicalNote_InterpretingCellRangerWebSummaryFiles_RevA.pdf" target="_blank">10x provides a guide to interpreting this report</a>.</p>
<p>Let’s traverse to the outs directory for sample HODay0replicate1 and open the <code>web_summary.html</code>. <strong><a href="images/cellranger/HODay0replicate1_web_summary.html">Example count_run_HODay0replicate1 web_summary.html</a></strong></p>
<p>The <code>cellranger count</code> web summary has errors/warnings at the top of the report (if any) with additional metrics in the <code>Summary</code> and <code>Gene Expression</code> tabs.</p>
<p>The important parts of the web_summary.html we look at to determine the quality of the 10x analaysis include:</p>
<ul>
<li><strong>Warnings/Errors</strong></li>
<li><strong>Barcode Rank Plot</strong></li>
<li><strong>Sequencing Saturation Plot</strong></li>
<li>Cell count</li>
<li>reads/cell</li>
<li>UMI count</li>
<li>gene count</li>
<li>mapping to reference</li>
</ul>
<div id="warningserrors-remediations." class="section level4">
<h4>Warnings/Errors &amp; remediations.</h4>
<a href="https://www.10xgenomics.com/support/software/cell-ranger/latest/resources/cr-troubleshooting#alerts" target="_blank">Alerts</a> are generally the result of factors inherent in library preparation and sequencing or sample quality issues. Alerts do not affect the operation of the pipeline, but highlight potential causes for abnormal or missing data:
</p>
<ul>
<li><strong><code>WARNING</code></strong> alerts indicate that some parameter is suboptimal.</li>
<li><strong><code>ERROR</code></strong> alerts indicate a major issue.</li>
</ul>
<p><img src="images/cellranger/HODay0_alerts.png" /></p>
<p>Intron mode was used. <a href="https://www.10xgenomics.com/support/software/cell-ranger/latest/miscellaneous/cr-intron-mode-rec" target="_blank">This helps with cell calling and is on by default.</a></p>
</div>
<div id="other-alert-examples" class="section level4">
<h4>Other alert examples</h4>
<div class="figure">
<img src="images/cellranger/cellranger-warn-alert-3.1.png" alt="" />
<p class="caption">Report with alerts</p>
</div>
<p><img src="images/cellranger/cellranger-error-alert-3.1.png" /></p>
</div>
<div id="summary-tab" class="section level4">
<h4>Summary tab</h4>
<p>The run summary from <code>cellranger count</code> can be viewed by clicking <code>Summary</code> in the top left tab of the HTML file. The summary metrics describe sequencing quality and various characteristics of the detected cells.</p>
<p><img src="images/cellranger/HODay0_summary.png" /></p>
</div>
<div id="more-on-the-barcode-rank-plot" class="section level4">
<h4>More on the Barcode Rank Plot</h4>
<div class="figure">
<img src="images/cellranger/BRP-1_with_legend.png" alt="" />
<p class="caption">The y-axis is the number of UMI counts mapped to each barcode and the x-axis is the number of barcodes below that value.</p>
</div>
<p>The <a href="https://www.10xgenomics.com/support/software/cell-ranger/latest/advanced/cr-barcode-rank-plot" target="_blank">GEX Barcode Rank Plot</a> under the Cells dashboard is an interactive plot that shows all barcodes detected in an experiment, ranked from highest to lowest UMI count and which barcodes were inferred to be associated with cells. It is useful for understanding Cell Ranger’s cell calling algorithm and its performance on your data, and for providing insights into your sample quality.</p>
<p>Basically, all 10x Barcodes detected during sequencing (~100k) are plotted in decreasing order of the number of UMIs associated with that particular barcode. The number of UMIs detected in each GEM is then used by Cell Ranger to identify barcodes/GEMs that are likely to contain an intact cell based on the expected cell number and UMI counts. Then distinguish low RNA content cells from empty droplets, since GEMs containing cells are expected to have a greater number of transcripts (and thus UMIs) associated with them than non-cell containing GEMs.</p>
<p>Since barcodes can be associated with cells based on their UMI count or by their RNA profiles, some regions of the graph can contain both cell-associated and background-associated barcodes. The color of the graph represents the local density of barcodes that are cell-associated.</p>
<p>A steep drop-off is indicative of good separation between the cell-associated barcodes and the barcodes associated with empty partitions.</p>
</div>
<div id="gene-expression-tab" class="section level4">
<h4>Gene Expression tab</h4>
The automated secondary analysis results can be viewed by clicking the <code>Gene Expression</code> tab in the top left corner. Click the <code>?</code> icons next to each section title to display information about the secondary analyses shown in the dashboard.
</p>
The t-SNE Projection section shows the data reduced to two dimensions, colored by UMI count (left) or clustering (right). It is a good starting point to explore structure in the data. The projection colored by UMI counts is indicative of the RNA content of the cells and often correlates with cell size - redder points are cells with more RNA in them. For the projection colored by clustering results, select the type of clustering analysis to display from the drop-down button on the upper right (Graph-based by default) - change the category to vary the type of clustering and/or number of clusters (K=2-10) that are assigned to the data.
</p>
The Top Features By Cluster table shows which genes are differentially expressed in each cluster relative to all other clusters (Graph-based by default). To find the genes associated with a particular cluster, click the L2FC or p-value column headers associated with a given cluster number to sort the table by a specific cluster
</p>
<p><img src="images/cellranger/HODay0_tsne.png" /></p>
<div id="the-sequencing-saturation-plot" class="section level5">
<h5>The Sequencing Saturation plot</h5>
The <a href="https://kb.10xgenomics.com/hc/en-us/articles/115005062366-What-is-sequencing-saturation" target="_blank">Sequecing Saturation</a> plot shows the effect of decreased sequencing depth on sequencing saturation, which is a measure of the fraction of library complexity that was observed. The right-most point on the line is the full sequencing depth obtained in this run.
</p>
Similarly, the Median Genes per Cell plot shows the effect of decreased sequencing depth on median genes per cell, which is a way of measuring data yield as a function of depth. The right-most point on the line is the full sequencing depth obtained in this run.
</p>
<p><img src="images/cellranger/HODay0_seqsat.png" /></p>
<p><strong>The recommended sequencing saturation is &gt;60%</strong>. If you are below this threshold, you would probably benefit from additional sequencing for increased depth.</p>
<p>To think of this another way, if you hit 60% saturation, for every additional 100 reads/cell you could expect to see 40 more genes. At 40% saturation you could expect to see 60 additional genes for every additional 100 reads/cell. At 20% saturation you could expect to see 80 additional genes for every additional 100 reads/cell and so on. Therefore if your sequencing saturation is high enough, you are unlikely to gain additional biologically relevant data with additional sequencing. If your sequecing saturation is low, additional sequencing will improve your downstream analysis and reveal more about the underlying biology.</p>
<p>Assuming the Advanced Genomics Core has your samples it is relatively easy to request additional sequencing through the core.</p>
<p><br/> <br/></p>
</div>
</div>
</div>
<div id="feature-barcode-matrices" class="section level3">
<h3>Feature Barcode Matrices</h3>
<pre><code>cd ~/ISC_Shell/cellranger_outputs/count_run_HODay0replicate1/outs
tree raw_feature_bc_matrix
raw_feature_bc_matrix
    ├── barcodes.tsv.gz
    ├── features.tsv.gz
    └── matrix.mtx.gz
0 directories, 3 files</code></pre>
<table class="fig">
<tr>
<td class="fig">
<ol style="list-style-type: upper-alpha">
<li><br/> <img src="images/cellranger/feature_bc_matrix.png" /></li>
<li><br/> <img src="images/cellranger/matricesBRP.png" /></li>
</ol>
</td>
</tr>
<tr>
<td class="fig">
<ol style="list-style-type: upper-alpha">
<li>Each element of the feature-barcode matrix is the count of UMIs associated with a feature (row) and a barcode (column). Because the matrix is very sparse, storing the matrix as three individual, compressed files saves siginificant space. <a href="#references">[6]</a><br/></li>
<li>In relation to the Barcode Rank Plot, raw matrices include cells and background and the filtered matrix files only contain data for cells. Downstream analysis typically uses the <strong>filtered</strong> matrix files.</li>
</ol>
</td>
</tr>
</table>
<!--
<details>
<summary> Click to see more details about these files.</summary>

##### features.tsv.gz

Features correspond to row indices. For each feature, the feature ID and name are stored in the first and second column of the (unzipped) `features.tsv.gz` file, respectively. The third column identifies the type of feature, which will be one of `Gene Expression`, `Antibody Capture`, `CRISPR Guide Capture`, `Multiplexing Capture`, or `CUSTOM`, depending on the feature type. Below is a minimal example `features.tsv.gz` file showing data collected for three genes.</p>


```
ENSG00000141510       TP53         Gene Expression
ENSG00000012048       BRCA1        Gene Expression
ENSG00000139687       RB1          Gene Expression
```

Gene Expression</code> data, the ID corresponds to <code class="css-q0dnij">gene\_id</code> in the annotation field of the reference GTF. Similarly, the name corresponds to <code class="css-q0dnij">gene\_name</code> in the annotation field of the reference GTF. If no <code class="css-q0dnij">gene_name</code> field is present in the reference GTF, gene name is equivalent to gene ID. 

##### barcodes.tsv.gz

Barcode sequences correspond to column indices:</p>

```
head filtered_feature_bc_matrices/barcodes.tsv.gz

AAACCCAAGGAGAGTA-1      
AAACGCTTCAdvanced Genomics CoreCCAG-1
AAAGAACAGACGACTG-1
AAAGAACCAATGGCAG-1    
AAAGAACGTCTGCAAT-1    
AAAGGATAGTAGACAT-1    
AAAGGATCACCGGCTA-1    
AAAGGATTCAdvanced Genomics CoreTTGA-1    
AAAGGATTCCGTTTCG-1    
AAAGGGCTCATGCCCT-1
```

Each barcode sequence includes a suffix with a dash separator followed by a number:</p>

The matrix file (UMI counts) uses the [Market Exchange Format (MEX)](https://math.nist.gov/MatrixMarket/formats.html){target="_blank"}. It contains gzipped TSV files with feature and barcode sequences corresponding to row and column indices respectively.

<p class="css-d17snu">More details on the barcode sequence format are available in the <a href="/support/software/cell-ranger/latest/analysis/outputs/cr-outputs-bam" class="css-k0f5v edy1dh10" target="_blank" >barcoded BAM section</a>.</p>
</details>
-->
<p><br/> <br/></p>
</div>
</div>
<div id="re-running-cell-ranger" class="section level2">
<h2>Re-running Cell Ranger</h2>
<p>It’s most common for the sequence provider to run Cell Ranger on your behalf. Cell Ranger’s preliminary QC outputs give the provider key information about whether the prep and sequencing run worked as expected. With that in mind it’s entirely possible you may never have to run Cell Ranger yourself.</p>
<p>There are a few scenarios when you might consider re-running Cell Ranger:</p>
<ol style="list-style-type: decimal">
<li>You need to combine new samples with previously processed samples and you want all samples to be run through the same version of cell ranger</li>
<li>You need to adjust the run parameters (e.g. you want to exclude/include introns in the alignment).</li>
<li>You need to align against a custom reference ( e.g. one that includes a transgene)</li>
</ol>
<p>Also, keep in mind that if you or your collaborators have to re-run Cell Ranger for any reason you will need the original FASTQ files. (<strong>Again, you are repsonsible for your data.</strong> Your sequencing provider may not retain a copy and even if they do there may be charge and/or delay for reprocessing).</p>
<div id="system-requirements-for-running-cell-ranger" class="section level3">
<h3>System Requirements for running Cell Ranger</h3>
<p><img src="images/cellranger/onedoesntlaptop.jpg" /></p>
<p>Running Cell Ranger can be challenging because the <a href="https://www.10xgenomics.com/support/software/cell-ranger/downloads/cr-system-requirements" target="_blank">system requirements</a> exceed the specifications for a typical personal computer/laptop. In particiular, Cell Ranger pipelines run on Linux systems that meet these minimum requirements:</p>
<ul>
<li>64-bit CentOS/RedHat 7.0 or Ubuntu 14.04</li>
<li>8-core Intel or AMD processor (16 cores recommended).</li>
<li>64GB RAM (128GB recommended).<br />
</li>
<li>1TB free disk space.</li>
</ul>
<p>Note that UM Advanced Genomics Core and Bioinformatics Core use the <a href="https://arc.umich.edu/greatlakes/" target="_blank">Great Lakes Computing cluster</a> to run Cell Ranger. If your lab has signed up for the <a href="https://arc.umich.edu/umrcp/" target="_blank">University of Michigan Research Computing Package (UMRCP)</a>, you can also get access to the cluster and the available software.</p>
<p><br/> <br/></p>
<p><img src="images/cellranger/iceberg.jpg" /></p>
<p><br/> <br/></p>
</div>
</div>
<div id="conclusions" class="section level2">
<h2>Conclusions</h2>
<ul>
<li>To run <strong><code>cellranger count</code></strong> you need a computing cluster, the <strong><code>--id</code></strong> and <strong><code>--fastqs</code></strong> for each <strong><code>--sample</code></strong>, and the path to the cellranger formated <strong><code>--transcriptome</code></strong> reference file.</li>
<li>The Barcode Rank Plot shows how the software determined what were cells compared to background. The filtered matrix files only contain data for cells on the barcode rank plot, wheras the raw matrices include cells and background.</li>
<li>Additional sequencing is suggested if your sequencing saturation is low.</li>
</ul>
<p><br/> <br/></p>
</div>
<div id="references" class="section level2">
<h2>References</h2>
<ol style="list-style-type: decimal">
<li>Sections of this lesson were adapted from 10x training materials. See the <a href="https://www.10xgenomics.com/support/software/cell-ranger/latest" target="_blank">10x Genomics support website</a> for additional information.</li>
<li><a href="https://www.10xgenomics.com/support/software/cell-ranger/latest/tutorials/cr-tutorial-ct">10x Genomics Cell Ranger count tutorial</a></li>
<li><a href="https://www.10xgenomics.com/support/software/cell-ranger/latest/algorithms-overview/cr-gex-algorithm#cell_calling" target="_blank">10x Genomics cell-calling algorithm</a></li>
<li><a href="https://genomebiology.biomedcentral.com/articles/10.1186/s13059-019-1662-y" target="_blank">EmptyDrops method</a></li>
<li><a href="https://portal.hdfgroup.org/documentation/index.html" target="_blank">Intro to HDF5 format</a></li>
<li><a href="https://www.10xgenomics.com/support/software/cell-ranger/latest/analysis/outputs/cr-outputs-molecule-info" target="_blank">Detailed explanation of 10x molecule HDF5 format</a></li>
<li><a href="https://math.nist.gov/MatrixMarket/formats.html" target="_blank">Market Exchange Format (MEX)</a>.</li>
</ol>
<br/> <br/>
<hr/>
<table>
<thead>
<tr class="header">
<th align="left"><a href="01-GettingStarted.html">Previous lesson</a></th>
<th align="center"><a href="#top">Top of this lesson</a></th>
<th align="right"><a href="02-QCandFiltering.html">Next lesson</a></th>
</tr>
</thead>
<tbody>
</tbody>
</table>
</div>
</div>



</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3,h4",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
