<!DOCTYPE html>

<html lang="en" xml:lang="en">

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="UM Bioinformatics Core Workshop Team" />

<meta name="date" content="2025-04-11" />

<title>Getting Started with Seurat</title>

<script src="site_libs/header-attrs-2.28/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/paper.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.13.2/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/default.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<script src="site_libs/clipboard-1.7.1/clipboard.min.js"></script>
<link href="site_libs/primer-tooltips-1.4.0/build.css" rel="stylesheet" />
<link href="site_libs/klippy-0.0.0.9500/css/klippy.min.css" rel="stylesheet" />
<script src="site_libs/klippy-0.0.0.9500/js/klippy.min.js"></script>
<!--
Favicon dervied from
https://twemoji.twitter.com/
https://favicon.io/emoji-favicons/dna/
-->
<link rel="apple-touch-icon" sizes="180x180" href="apple-touch-icon.png">
<link rel="icon" type="image/png" sizes="32x32" href="favicon-32x32.png">
<link rel="icon" type="image/png" sizes="16x16" href="favicon-16x16.png">
<link rel="shortcut icon" href="favicon-16x16.png" />
<link rel="manifest" href="site.webmanifest">
<link rel="mask-icon" href="safari-pinned-tab.svg" color="#5bbad5">
<meta name="msapplication-TileColor" content="#da532c">
<meta name="theme-color" content="#ffffff">

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>









<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-inverse  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">Intro to scRNA-Seq</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="workshop_intro.html">Intro</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Day 1
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="00A-OrientingOnScRNASeq.html">Orienting on scRNA-Seq</a>
    </li>
    <li>
      <a href="01-GettingStarted.html">Getting Started with Seurat</a>
    </li>
    <li>
      <a href="00B-CellRangerInAction.html">Cell Ranger in Action</a>
    </li>
    <li>
      <a href="02-QCandFiltering.html">Secondary QC &amp; Filtering</a>
    </li>
    <li>
      <a href="03-Normalization.html">Normalization</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Day 2
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="04-PCAandIntegration.html">PCA &amp; Integration</a>
    </li>
    <li>
      <a href="05-ProjectionAndClustering.html">Projection &amp; Clustering</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Day 3
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="06-MarkerVisualization.html">Marker identification and visualization</a>
    </li>
    <li>
      <a href="07-CellTypeAnnos.html">Cell type annotation</a>
    </li>
    <li>
      <a href="08-DifferentialExpression.html">Differential expression analysis</a>
    </li>
  </ul>
</li>
<li>
  <a href="workshop_wrap_up.html">Wrap up</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Getting Started with Seurat</h1>
<h4 class="author">UM Bioinformatics Core Workshop Team</h4>
<h4 class="date">2025-04-11</h4>

</div>


<style type="text/css">
body, td {
   font-size: 18px;
}
code.r{
  font-size: 12px;
}
pre {
  font-size: 12px
}

table{
   width:100%;
}

table.fig, th.fig, td.fig {
  border: 1px solid lightgray;
  border-collapse: collapse;
  padding: 12px;
}

</style>
<script>
  addClassKlippyTo("pre.r, pre.markdown, pre.bash");
  addKlippy('right', 'top', 'auto', '1', 'Copy code', 'Copied!');
</script>
<style type="text/css">
.doNotRun {
background-color: CornSilk;
}
</style>
<p><br/> <img src="images/wayfinder/01-GettingStart-Wayfinder.png" />
<br/> <br/></p>
<div id="introduction" class="section level1">
<h1>Introduction</h1>
<table class="fig">
<tr>
<td class="fig">
<img src="images/graphical_abstracts/01-GettingStarted-Abstract.png" />
</td>
</tr>
<tr>
<td class="fig">
Cell Ranger outputs can be read into R via the triple of files directly,
or via a memory saving route with the BPCells package. We provide code
for both routes in this section.
</td>
</tr>
</table>
<p><br/></p>
<p>One of the goals of this workshop is to work through an example from
start to end, and to highlight some of the decision points and
rationales behind the decisions. The first step of any single-cell
project will be to get the data into R for analysis. <br/></p>
<div id="objectives" class="section level2">
<h2>Objectives</h2>
<ul>
<li>Orient on RStudio.</li>
<li>Create an RStudio project for analysis.</li>
<li>Create directory structure for analysis.</li>
<li>Learn how to read Cell Ranger data into Seurat.</li>
<li>Introduce the Seurat object, and how to access parts of it.</li>
</ul>
<hr />
</div>
</div>
<div id="orienting-on-rstudio" class="section level1">
<h1>Orienting on RStudio</h1>
<p>RStudio is an integrated development environment where you can write,
execute, and see the results of your code. The interface is arranged in
different panes:</p>
<ul>
<li>The Console pane along the left where you can enter commands and
execute them.</li>
<li>The Environment pane in the upper right shows any variables you have
created, along with their values.</li>
<li>The pane in the lower right has a few functions:
<ul>
<li>The Files tab let’s you navigate the file system.</li>
<li>The Plots tab displays any plots from code run in the Console.</li>
<li>The Help tab displays the documentation of functions.</li>
</ul></li>
</ul>
<div id="commands-in-the-console" class="section level2">
<h2>Commands in the Console</h2>
<p>We can input commands directly into the Console and press
<KBD>Enter</KBD> to execute them. Try it with:</p>
<pre><code>&gt; 2+2
[1] 4</code></pre>
<p>If we enter commands directly into the Console and execute them,
<strong>there is no record</strong> that persists after we close
RStudio.</p>
<p><img src="images/03_rstudio_console_execution.png" alt="rstudio console execution" style="width:1000px;"/></p>
<p><strong>Checkpoint</strong></p>
</div>
<div id="commands-in-the-script" class="section level2">
<h2>Commands in the Script</h2>
<p>To keep that record, we will enter our commands in a script. When
first opening RStudio, there is no pane to write code to save as a file.
But by clicking on the icon in the upper-right-most of the interface (a
blank piece of paper with a + sign), and selecting <KBD>R Script</KBD> a
new pane opens. This last pane is the Source (or Script) pane, and it is
here that we’ll keep a running record of the commands we learn in this
workshop.</p>
<p><img src="images/04_rstudio_create_script.png" alt="rstudio create script" style="width:1000px;"/></p>
<p>Which will swing open the Script pane:</p>
<p><img src="images/05_rstudio_script_pane.png" alt="rstudio script pane" style="width:1000px;"/></p>
<p>In the Script pane, enter:</p>
<pre><code>3+2</code></pre>
<p>Notice that if we press <KBD>Enter</KBD> in the Source pane, we get a
new line. <strong>The code does note execute.</strong> In order to
execute the code we press <KBD>Ctrl + Enter</KBD> <strong>on the line of
code we want to execute</strong>. We then see that code executed, along
with its result in in the Console pane.</p>
<p><img src="images/06_rstudio_script_execution.png" alt="rstudio script execution" style="width:1000px;"/></p>
<blockquote>
<h2 id="note-script-pane-and-console-pane"
class="unlisted unnumbered">Note: Script Pane and Console Pane</h2>
<p>The code written in the Script Pane is just text until it is executed
with <KBD>Control + Enter</KBD>. The Console Pane is the record of the
code that has been executed. <br></p>
</blockquote>
</div>
<div id="configuring-rstudio" class="section level2">
<h2>Configuring RStudio</h2>
<p>All of the panes in RStudio have configuration options. For example,
you can minimize/maximize a pane or resize panes by dragging the
borders. The most important customization options for pane layout are in
the <KBD>View</KBD> menu. Other options such as font sizes,
colors/themes, and more are in the <KBD>Tools</KBD> menu under
<KBD>Global Options</KBD>.</p>
<p>We can enable soft-wrapping of code by selecting <KBD>Code</KBD> and
then <KBD>Soft Wrap Long Lines</KBD>.</p>
<p><img src="images/07_rstudio_soft_wrap.png" alt="rstudio code execution" style="width:1000px;"/></p>
</div>
</div>
<div id="workshop-flow" class="section level1">
<h1>Workshop Flow</h1>
<p>The code we execute together is available in three forms:</p>
<ol style="list-style-type: decimal">
<li>By viewing the instructor’s Zoom screen share,</li>
<li>By code pasted into Slack, and</li>
<li>By code blocks on these web pages.</li>
</ol>
<p>As a learner you are welcome to get the same code into your Script
Pane by:</p>
<ol style="list-style-type: decimal">
<li>Typing the code from the Zoom screen share into your Script
Pane,</li>
<li>Copying the code from Slack and pasting into your Script Pane,
or</li>
<li>Copying the code from the code blocks on these web pages and pasting
into your Script Pane.</li>
</ol>
<p><strong>Questions?</strong></p>
<div id="creating-a-project" class="section level2">
<h2>Creating a project</h2>
<p>We will create an RStudio Project to keep our files and code
organized. See the <a
href="https://r4ds.hadley.nz/workflow-scripts.html#projects">Projects</a>
section of <a href="https://r4ds.hadley.nz/">R for Data Science</a> for
a more in-depth description of what a project is and how it’s
helpful.</p>
<p>To create a Project for this workshop, click <kbd>File</kbd> then
<kbd>New Project…</kbd>. In the New Project Wizard window that opens,
select <kbd>Existing Directory</kbd>, then <kbd>Browse…</kbd>. In the
Choose Directory window, select the <code>ISC_R</code> folder by
clicking it once, and then click the <kbd>Choose</kbd> button. Finally,
click <kbd>Create Project</kbd>.</p>
<p>Once we do this, RStudio will restart and the Files pane (lower
right) should put us in the <code>~/ISC_R</code> folder where there is
an <code>inputs/</code> folder and an <code>ISC_R.Rproj</code> file.</p>
</div>
<div id="directory-structure" class="section level2">
<h2>Directory structure</h2>
<p>We have included the data to be used in the workshop in the
<code>inputs/</code> folder. However, the project will need to include
folders for our analysis and our analysis scripts. Let’s create that
directory structure with the <code>dir.create()</code> function. We will
start by putting the following commands into the Console pane
directly.</p>
<pre class="r"><code># Create project directories
dir.create(&#39;scripts&#39;, showWarnings = FALSE, recursive = TRUE)
dir.create(&#39;results/figures&#39;, showWarnings = FALSE, recursive = TRUE)
dir.create(&#39;results/tables&#39;, showWarnings = FALSE, recursive = TRUE)
dir.create(&#39;results/rdata&#39;, showWarnings = FALSE, recursive = TRUE)</code></pre>
<p>In the Files pane we should see the new <code>results/</code> and
<code>scripts/</code> folders.</p>
</div>
<div id="analysis-script" class="section level2">
<h2>Analysis script</h2>
<p>The two most important artifacts of our analysis are the data from
Cell Ranger, and the script to analyze the data. There will be outputs
in <code>results/</code>, and these will be important, but if the
contents of <code>results/</code> are ever lost, the script will be able
to re-generate them if we’ve captured all our steps as code, which we
aim to do.</p>
<p>To create the analysis script, click <kbd>File</kbd>, hover over
<kbd>New File</kbd>, and click on <kbd>R Script</kbd>. A new pane in the
upper left slides into view and is the Untitled script file. Save this
file, and name it, by clicking <kbd>File</kbd> then <kbd>Save</kbd>.
Double click the <code>scripts/</code> folder, and in the File name:
text box type “analysis.R”. Then click <kbd>Save</kbd>.</p>
<p>As we proceed through the workshop, we should save this file (by
clicking the Floppy disk, clicking <kbd>File</kbd> then <kbd>Save</kbd>,
or by typing <kbd>Control + S</kbd>).</p>
<blockquote>
<p><strong>Good scripting practices</strong></p>
<p>In any analysis script, we recommend using comments (lines preceded
by a “#”) to provide additional information about code that may not be
self-evident. This is to the benefit of others that may look at the
code, but also to your future-self. <br></p>
</blockquote>
<p>For completeness, insert the <code>dir.create</code> commands at the
top of our new script.</p>
<pre class="r"><code># =========================================================================
# Getting Started with Seurat
# =========================================================================

# -------------------------------------------------------------------------
# Create project directories
dir.create(&#39;scripts&#39;, showWarnings = FALSE, recursive = TRUE)
dir.create(&#39;results/figures&#39;, showWarnings = FALSE, recursive = TRUE)
dir.create(&#39;results/tables&#39;, showWarnings = FALSE, recursive = TRUE)
dir.create(&#39;results/rdata&#39;, showWarnings = FALSE, recursive = TRUE)</code></pre>
</div>
</div>
<div id="loading-libraries" class="section level1">
<h1>Loading libraries</h1>
<p>We begin our analysis script by loading the libraries we expect to
use. It’s generally good practice to include all <code>library()</code>
calls at the top of a script for visibility.</p>
<pre class="r"><code># -------------------------------------------------------------------------
# Load libraries
library(Seurat)
library(BPCells)
library(tidyverse)

options(future.globals.maxSize = 1e9)</code></pre>
<p>The libraries that we are loading are:</p>
<ul>
<li>The <code>Seurat</code> library, developed by the <a
href="https://satijalab.org/">Satija lab</a>, which will provide the
essential functions used in our single-cell analysis. The <a
href="https://satijalab.org/seurat/">Seurat documentation</a> is
extensive and indispensible.</li>
<li>The <code>BPCells</code> library, developed by <a
href="https://bnprks.github.io/BPCells/authors.html#citation">Benjamin
Parks</a>, is a recent package with the primary goal of efficiently
storing single-cell data to reduce its memory footprint. The <a
href="https://bnprks.github.io/BPCells/index.html">BPCells
documentation</a> includes many useful tutorials.</li>
<li>The <code>tidyverse</code> library, developed by Posit, is an
essential package for data manipulation and plotting. The <a
href="https://www.tidyverse.org/">tidyverse documentation</a> is
essential for getting a handle on the array of functions in the many
packages contained therein.</li>
</ul>
</div>
<div id="read-in-data" class="section level1">
<h1>Read in data</h1>
<p>The <code>inputs/10x_cellranger_filtered_triples/</code> folder is
closer to what AGC would generate with Cell Ranger, where each sample
has a folder, and within that folder there are three files:</p>
<ul>
<li><code>barcodes.tsv.gz</code></li>
<li><code>features.tsv.gz</code></li>
<li><code>matrix.mtx.gz</code></li>
</ul>
<p>Note, these files are filtered matrices, AGC will touch on filtered
and unfiltered Cell Ranger outputs.</p>
<div id="read10x" class="section level2">
<h2>Read10X</h2>
<p>The <code>Read10X()</code> function will read in the triples
organized within sample folders:</p>
<pre class="r"><code># -------------------------------------------------------------------------
# To load data from 10X Cell Ranger

# Collect the input directories
#   Each sample dir contains barcodes.tsv.gz, features.tsv.gz, matrix.mtx.gz.
#   Naming the sample_dirs vector makes Seurat name the
#     samples in the corresponding manner, which is nice for us.
sample_dirs = list(
  HODay0replicate1  = &quot;inputs/10x_cellranger_filtered_triples/count_run_HODay0replicate1&quot;, 
  HODay0replicate2  = &quot;inputs/10x_cellranger_filtered_triples/count_run_HODay0replicate2&quot;,
  HODay0replicate3  = &quot;inputs/10x_cellranger_filtered_triples/count_run_HODay0replicate3&quot;,
  HODay0replicate4  = &quot;inputs/10x_cellranger_filtered_triples/count_run_HODay0replicate4&quot;,
  HODay7replicate1  = &quot;inputs/10x_cellranger_filtered_triples/count_run_HODay7replicate1&quot;,
  HODay7replicate2  = &quot;inputs/10x_cellranger_filtered_triples/count_run_HODay7replicate2&quot;,
  HODay7replicate3  = &quot;inputs/10x_cellranger_filtered_triples/count_run_HODay7replicate3&quot;,
  HODay7replicate4  = &quot;inputs/10x_cellranger_filtered_triples/count_run_HODay7replicate4&quot;,
  HODay21replicate1 = &quot;inputs/10x_cellranger_filtered_triples/count_run_HODay21replicate1&quot;,
  HODay21replicate2 = &quot;inputs/10x_cellranger_filtered_triples/count_run_HODay21replicate2&quot;,
  HODay21replicate3 = &quot;inputs/10x_cellranger_filtered_triples/count_run_HODay21replicate3&quot;,
  HODay21replicate4 = &quot;inputs/10x_cellranger_filtered_triples/count_run_HODay21replicate4&quot;)

# Create the expression matrix from sample dirs
#   Read10X needs a *vector* instead of a *list*, so we use *unlist* to convert
geo_mat = Read10X(data.dir = unlist(sample_dirs))</code></pre>
</div>
<div id="write-bpcells" class="section level2">
<h2>Write BPCells</h2>
<p>We could create a Seurat object with
<code>CreateSeuratObject(counts = geo_mat)</code>, but we will first
output <code>geo_mat</code> using the BPCells package to save
memory:</p>
<pre class="r"><code># -------------------------------------------------------------------------
# Build BPCells input dir
# To use BPCells (to save some memory), you can transform 
# the expression matrix data structure above into BPCells files.
# BPCells uses these files for processing, but you typically never look at 
# their contents directly
write_matrix_dir(mat = geo_mat, dir = &#39;~/ISC_R/bpcells&#39;, overwrite = TRUE)</code></pre>
<pre><code>32285 x 35269 IterableMatrix object with class MatrixDir

Row names: Xkr4, Gm1992 ... AC149090.1
Col names: HODay0replicate1_AAACCTGAGAGAACAG-1, HODay0replicate1_AAACCTGGTCATGCAT-1 ... HODay21replicate4_TTTGTCACAGGGAGAG-1

Data type: double
Storage order: column major

Queued Operations:
1. Load compressed matrix from directory /home/workshop/cgates/ISC_R/bpcells</code></pre>
<pre class="r"><code># -------------------------------------------------------------------------
# Cleanup
# Since we&#39;ll now be reading in from BPCells files, we will remove geo_mat
# from the environment and then prompt RStudio to run a &quot;garbage collection&quot;
# to free up unused memory
rm(geo_mat)
gc()</code></pre>
<pre><code>           used  (Mb) gc trigger   (Mb)   max used   (Mb)
Ncells  9159071 489.2   17466748  932.9   11848562  632.8
Vcells 41077899 313.4  686073055 5234.4 1071451558 8174.6</code></pre>
</div>
<div id="read-bpcells" class="section level2">
<h2>Read BPCells</h2>
<p>Let’s read the BPCells files in, and note that the
<code>geo_mat</code> object takes up less memory than when we created it
with <code>Read10X()</code>:</p>
<pre class="r"><code># -------------------------------------------------------------------------
# Create expression matrix and Seurat object from BPCells files
geo_mat = open_matrix_dir(dir = &#39;~/ISC_R/bpcells&#39;)</code></pre>
</div>
<div id="create-a-seurat-object" class="section level2">
<h2>Create a Seurat object</h2>
<p>The expression matrix is the precursor to creating the
<code>Seurat</code> object upon which all our analysis will be done. To
create the <code>Seurat</code> object:</p>
<pre class="r"><code># -------------------------------------------------------------------------
# Create seurat object
geo_so = CreateSeuratObject(counts = geo_mat, min.cells = 1, min.features = 50)
geo_so</code></pre>
<pre><code>An object of class Seurat 
26489 features across 35216 samples within 1 assay 
Active assay: RNA (26489 features, 0 variable features)
 1 layer present: counts</code></pre>
<p>We have specified some parameters to remove genes and cells which do
not contain very much information. Specifically a gene is removed if it
is expressed in 1 or fewer cells, and a cell is removed if it contains
reads for 50 or fewer genes. In the context of this workshop, this helps
us minimize memory usage.</p>
</div>
<div id="molecule_info.h5-as-an-alternative-input-to-seurat"
class="section level2">
<h2>molecule_info.h5 as an alternative input to Seurat</h2>
<details>
<summary>
<em>H5 (aka HDF5) is an alternative file format</em>
</summary>
<p>Some programs (like Seurat) need only the <strong>barcodes, features,
and matrix (counts)</strong> and those data are conveniently provided as
three separate files for each sample. Other programs (like Cell Ranger)
need to store more information (e.g.  probe info about certain library
preps or the chip or channel that cell ran on). Instead of adding more
files to complement the three base files above, 10x provides a single
binary file, <strong>molecule_info.h5</strong>, which contains <em>all
information</em> for all molecules. (Strictly only the molecules with
valid barcode/UMI assigned to a gene.)</p>
<ul>
<li>molecule_info.h5 is a single file, but with really acts like basket
of related structures (think tables) for a single sample.</li>
<li>The “.h5” extension stands for <a
href="https://www.hdfgroup.org/solutions/hdf5/"
target="_blank">HDF5</a>, a broadly adopted, platform agnostic,
scalable, high-performance file format for storing complex data.</li>
<li>You can view the data within the .h5 file with special tools like <a
href="https://www.hdfgroup.org/downloads/hdfview"
target="_blank">HDF5View</a> or <a
href="https://support.hdfgroup.org/documentation/hdf5/latest/_h5_t_o_o_l__d_p__u_g.html"
target="_blank">h5dump</a>. Here is an excerpt of the structures inside
that file:</li>
</ul>
<pre><code>molecule_info.h5
    ├─ barcodes [HDF5 group]
    ├─ count
    ├─ features [HDF5 group]
    ├─ gem_group
    ├─ library_info
    ├─ metrics_json
    ├─ probes [HDF5 group]
    ├─ ...
    └─ umi</code></pre>
<p>You can see the first structures above are barcodes, features, and
counts. So <em>if you have the <a
href="https://github.com/hhoeflin/hdf5r" target="_blank">hdf5 libraries
installed</a></em>, you can use the .h5 as input to Seurat like so:</p>
<pre class="default doNotRun"><code># ==========================================================================
### DO NOT RUN ###
# To load data from raw .h5 file
### DO NOT RUN ###

HODay0replicate1 = &quot;~/ISC_Shell/cellranger_outputs/count_run_HODay0replicate1/outs/molecule_info.h5&quot;
get_mat = Read10X_h5(filename = HODay0replicate1)
geo_so = CreateSeuratObject(counts = geo_mat, min.cells = 1, min.features = 50)
</code></pre>
<p>If you want to load multiple samples from .h5 files, see the function
<a
href="https://samuel-marsh.github.io/scCustomize/reference/Read10X_h5_Multi_Directory.html"
target="_blank">Read10X_h5_Multi_Directory</a> in the <a
href="https://samuel-marsh.github.io/scCustomize/index.html"
target="_blank">scCustomize</a> R library.</p>
<p>FYI</p>
<ul>
<li>Be aware, using HDF5 files from R support requires the hdf5 library
which entails a few more steps to install than a typical library. Brew a
nice cup of (decaf) tea and review how to <a
href="https://github.com/hhoeflin/hdf5r" target="_blank">install
hdf5</a>.</li>
<li>See the Seurat function <a
href="https://satijalab.org/seurat/reference/read10x"
target="_blank">Read10X</a>.</li>
<li>See <a
href="https://satijalab.org/seurat/articles/seurat5_bpcells_interaction_vignette#load-data-from-one-h5-file"
target="_blank">BPCells</a> docs for details on loading from h5 to a
BPCells directory.</li>
<li><a
href="https://www.10xgenomics.com/support/software/cell-ranger/latest/analysis/outputs/cr-outputs-molecule-info"
target="_blank">10x Genomics</a> details the full contents of
molecule_info.h5.</li>
<li>Checkout the <a
href="https://support.hdfgroup.org/documentation/hdf5/latest/_getting_started.html"
target="_blank">HDF5 reference docs</a> for more information about the
HDF5 format.</li>
</ul>
<hr/>
</details>
<p><br/></p>
</div>
</div>
<div id="structure-of-a-seurat-object" class="section level1">
<h1>Structure of a Seurat object</h1>
<p>The <code>Seurat</code> object is a complex data type, so let’s get a
birds eye view with an image from <a
href="https://swbioinf.github.io/scRNAseqInR_Doco/seuratobject.html#discussion-the-seurat-object-in-r">this
tutorial</a> on single-cell analysis.</p>
<div class="float">
<img src="images/seurat_schematic/Slide1.png"
alt="Image: Schematic of Seurat object." />
<div class="figcaption">Image: Schematic of <code>Seurat</code>
object.</div>
</div>
<p>The three main “slots” in the object are:</p>
<ol style="list-style-type: decimal">
<li>The <code>assays</code> slot stores the expression data as
<code>Assay</code> objects.</li>
<li>The <code>meta.data</code> slot which stores cell-level information,
including technical and phenotypic data.</li>
<li>The <code>reductions</code> slot stores the results of dimension
reduction applied to the <code>assays</code>.</li>
</ol>
<p>There are other slots which store information that becomes relevant
as we progress through the analysis. We will highlight the other slots
as they come up.</p>
<p>After reading the data in and creating the Seurat object above, we
can imagine the following schematic representing our object:</p>
<div class="float">
<img src="images/seurat_schematic/Slide2.png"
alt="Image: Schematic after creating the Seurat object." />
<div class="figcaption">Image: Schematic after creating the Seurat
object.</div>
</div>
<p>Note the RNA assay contains a <code>count</code> layer consisting of
a raw count matrix where the rows are genes (features, more
generically), and the columns are all cells across all samples. Note
also the presence of a <code>meta.data</code> table giving information
about each cell. We’ll pay close attention to this as we proceed. The
other slots include information about the <code>active.assay</code> and
<code>active.ident</code> which tell Seurat which expression data to use
and how the cells are to be identified.</p>
</div>
<div id="accessing-parts-of-the-object" class="section level1">
<h1>Accessing parts of the object</h1>
<p>The only slot of the <code>Seurat</code> object that we’ll typically
access or modify by hand–that is, without a function from the
<code>Seurat</code> package–is the <code>meta.data</code> object. In R,
slots are accessed with the <code>@</code> symbol, as in:</p>
<pre class="r"><code># -------------------------------------------------------------------------
# Examine Seurat object
head(geo_so@meta.data)</code></pre>
<pre><code>                                          orig.ident nCount_RNA nFeature_RNA
HODay0replicate1_AAACCTGAGAGAACAG-1 HODay0replicate1      10234         3226
HODay0replicate1_AAACCTGGTCATGCAT-1 HODay0replicate1       3158         1499
HODay0replicate1_AAACCTGTCAGAGCTT-1 HODay0replicate1      13464         4102
HODay0replicate1_AAACGGGAGAGACTTA-1 HODay0replicate1        577          346
HODay0replicate1_AAACGGGAGGCCCGTT-1 HODay0replicate1       1189          629
HODay0replicate1_AAACGGGCAACTGGCC-1 HODay0replicate1       7726         2602</code></pre>
<p>Here, each row is a cell, and each column is information about that
cell. The rows of the table are named according to the uniquely
identifiable name for the cell. In this case, the day and replicate, as
well as the barcode for that cell. As we continue the workshop, we will
check in on the <code>meta.data</code> slot and observe changes we want
to make, and that other functions will make. We’ll also observe the
other assays and layers and note their changes.</p>
</div>
<div id="save-our-progress" class="section level1">
<h1>Save our progress</h1>
<p>Let’s Save our progress as an RDS file with <code>saveRDS()</code>;
this allows us to have a copy of the object that we can read back into
our session with the <code>readRDS()</code> commmand. Periodically we
will be saving our Seurat object so that we can have a version of it at
different steps of the analysis. These will also help us get untangled
if we get into an odd state.</p>
<pre class="r"><code># -------------------------------------------------------------------------
# Save the Seurat object
saveRDS(geo_so, file = &#39;results/rdata/geo_so_unfiltered.rds&#39;)</code></pre>
<p>Now that the state of our analysis is saved on disk, we’ll
demonstrate how to power down the RStudio session, restart the session,
and re-load the data we just saved.</p>
<ol style="list-style-type: decimal">
<li>To power down the session, click the orange power button in the
upper right corner of the RStudio window. Note: There is no confirmation
dialogue, the session will simply end.</li>
<li>Click the “Start New Session” button to restart the RStudio
session.</li>
<li>Re-load the <code>geo_so</code> object with:</li>
</ol>
<pre class="r"><code># -------------------------------------------------------------------------
# Load the Seurat object
geo_so = readRDS(&#39;results/rdata/geo_so_unfiltered.rds&#39;)</code></pre>
<p><br/> <br/></p>
</div>
<div id="summary" class="section level1">
<h1>Summary</h1>
<table class="fig">
<tr>
<td class="fig">
<img src="images/graphical_abstracts/01-GettingStarted-Abstract.png" />
</td>
</tr>
<tr>
<td class="fig">
Cell Ranger outputs can be read into R via the triple of files directly,
or via a memory saving route with the BPCells package. We provide code
for both routes in this section.
</td>
</tr>
</table>
<p><br/></p>
<p>In this section we:</p>
<ul>
<li>Created an RStudio project for analysis.</li>
<li>Created the directory structure for analysis.</li>
<li>Learned how to read 10X data into Seurat.</li>
<li>Introduced the Seurat object, and how to access parts of it.</li>
</ul>
<hr />
<p>These materials have been adapted and extended from materials listed
above. These are open access materials distributed under the terms of
the <a href="http://creativecommons.org/licenses/by/4.0/">Creative
Commons Attribution license (CC BY 4.0)</a>, which permits unrestricted
use, distribution, and reproduction in any medium, provided the original
author and source are credited.</p>
<p><br/> <br/></p>
<hr />
<table style="width:100%;">
<colgroup>
<col width="28%" />
<col width="42%" />
<col width="28%" />
</colgroup>
<thead>
<tr class="header">
<th align="left"><a href="00A-OrientingOnScRNASeq.html">Previous
lesson</a></th>
<th align="center"><a href="#top">Top of this lesson</a></th>
<th align="right"><a href="00B-CellRangerInAction.html">Next
lesson</a></th>
</tr>
</thead>
<tbody>
</tbody>
</table>
</div>



</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3,h4",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
