<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="UM Bioinformatics Core" />

<meta name="date" content="2024-02-22" />

<title>Normalization</title>

<script src="site_libs/header-attrs-2.21/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/paper.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<script src="site_libs/navigation-1.1/codefolding.js"></script>
<script src="site_libs/navigation-1.1/sourceembed.js"></script>
<link href="site_libs/highlightjs-9.12.0/default.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<!--
Favicon dervied from
https://twemoji.twitter.com/
https://favicon.io/emoji-favicons/dna/
-->
<link rel="apple-touch-icon" sizes="180x180" href="/apple-touch-icon.png">
<link rel="icon" type="image/png" sizes="32x32" href="/favicon-32x32.png">
<link rel="icon" type="image/png" sizes="16x16" href="/favicon-16x16.png">
<link rel="shortcut icon" href="favicon-16x16.png" />
<link rel="manifest" href="/site.webmanifest">
<link rel="mask-icon" href="/safari-pinned-tab.svg" color="#5bbad5">
<meta name="msapplication-TileColor" content="#da532c">
<meta name="theme-color" content="#ffffff">

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>




<style type="text/css">
#rmd-source-code {
  display: none;
}
</style>





<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->
<style type="text/css">
.code-folding-btn { margin-bottom: 4px; }
</style>



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-inverse  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">Intro to scRNA-Seq</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="workshop_intro.html">Intro</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Day 1
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="00A-OrientingOnScRNASeq.html">Orienting on scRNA-Seq</a>
    </li>
    <li>
      <a href="00B-CellRangerInAction.html">Cell Ranger in Action</a>
    </li>
    <li>
      <a href="01-GettingStarted.html">Seurat</a>
    </li>
    <li>
      <a href="02-QCandFiltering.html">Secondary QC &amp; Filtering</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Day 2
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="03-Integration.html">Normalization</a>
    </li>
    <li>
      <a href="04-PCAandDimReduction.html">PCA &amp; Integration</a>
    </li>
    <li>
      <a href="05-ProjectionAndClustering.html">Projection &amp; Clustering</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Day 3
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="06-MarkerVisualization.html">Marker identification and visualization</a>
    </li>
    <li>
      <a href="07-CellTypeAnnos.html">Cell type annotation</a>
    </li>
    <li>
      <a href="08-DifferentialExpression.html">Differential expression analysis</a>
    </li>
  </ul>
</li>
<li>
  <a href="workshop_wrap_up.html">Wrap up</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">

<div class="btn-group pull-right float-right">
<button type="button" class="btn btn-default btn-xs btn-secondary btn-sm dropdown-toggle" data-toggle="dropdown" data-bs-toggle="dropdown" aria-haspopup="true" aria-expanded="false"><span>Code</span> <span class="caret"></span></button>
<ul class="dropdown-menu dropdown-menu-right" style="min-width: 50px;">
<li><a id="rmd-download-source" href="#">Download Rmd</a></li>
</ul>
</div>



<h1 class="title toc-ignore">Normalization</h1>
<h4 class="author">UM Bioinformatics Core</h4>
<h4 class="date">2024-02-22</h4>

</div>


<style type="text/css">
body, td {
   font-size: 18px;
}
code.r{
  font-size: 12px;
}
pre {
  font-size: 12px
}
</style>
<div id="workflow-overview" class="section level1 unlisted unnumbered">
<h1 class="unlisted unnumbered">Workflow Overview</h1>
<p><br/>
<img src="images/wayfinder/wayfinder.png" alt="wayfinder" style="height: 400px;"/>
<br/> <br/></p>
</div>
<div id="introduction" class="section level1">
<h1>Introduction</h1>
<p>For larger scRNA-seq projects, data may unavoidably be generated
across multiple batches. As with bulk RNA-seq, some steps during library
preparation can be taken to avoid confounding batch effects with
biological variables of interest. But it will remain necessary to
correct for technical variation in the sample integration step.
<!-- More here? --></p>
<p>In this section, we will demonstrate the steps to evaluate whether
integration is needed (it usually is), and will walk through integrating
the data and evaluating the effectiveness thereof.</p>
<p>Like in the previous section, one integration method might be
reported in a research paper but several options might have been tested
before determining which approach sufficiently corrected batch effects
without removing expected biological variability.</p>
<div id="objectives" class="section level2">
<h2>Objectives</h2>
<ul>
<li>Understand why normalization is needed.</li>
<li>Understand some options for normalization.</li>
<li>Normalize the counts with <code>SCTransform()</code>.</li>
</ul>
</div>
</div>
<div id="normalization" class="section level1">
<h1>Normalization</h1>
<p>Bulk RNA-seq suffers from two issues that must be addressed in a
differential expression analysis:</p>
<ol style="list-style-type: decimal">
<li>Differences in sequencing depth contributes to technical
variation.</li>
<li>There is a confounding relationship between gene abundance and gene
variation.</li>
</ol>
<p>These also affect single-cell RNA-seq! Moreover, the sparsity of
scRNA-seq data poses additional issues that require new methods for
normalization.</p>
<p>We will use the <code>SCTransform()</code> method that is part of the
Seurat package. <a
href="https://genomebiology-biomedcentral-com/articles/10.1186/s13059-021-02584-9">This
manuscript</a> describes the initial approach, and <a
href="https://link.springer.com/article/10.1186/s13059-021-02584-9">this
updated manuscript</a> describes v2 of the algorithm.</p>
<p>The authors go to great lengths to demonstrate that there is a
“strong linear relationship between unnormalized expression (gene UMI
count) and cellular sequencing depth.” The relationship holds across a
range of abundance levels, and suggests that a single scaling factor to
account for cellular sequencing depth will not be sufficient, and that a
solution will depend on the abundance level.</p>
<blockquote>
<p><strong>Other normalizations</strong></p>
<p>From the Seurat documentation, the <code>SCTransform()</code>
function “replaces <code>NormalizeData()</code> (this is a
log-normalization), <code>ScaleData()</code>, and
<code>FindVariableFeatures()</code>”. You may see these three commands
in other vignettes, indeed even in other Seurat vignettes (<a
href="https://satijalab.org/seurat/articles/sctransform_vignette">source</a>).</p>
</blockquote>
<p>As we discussed when describing the structure of a
<code>Seurat</code> object, the assay data is stored in layers. To
conform to this convention, we will run the following command, which
splits the “RNA” assay into a list of samples, each with its own
layers.</p>
<pre class="r"><code>geo_so[[&#39;RNA&#39;]] = split(geo_so[[&#39;RNA&#39;]], f = geo_so$orig.ident)
geo_so</code></pre>
<pre><code>An object of class Seurat 
47037 features across 29615 samples within 1 assay
Active assay: RNA (26489 features, 0 variable features)
 12 layers present: counts.HO.Day0.replicate1, counts.HO.Day0.replicate2, counts.HO.Day0.replicate3, counts.HO.Day0.replicate4, counts.HO.Day21.replicate1, counts.HO.Day21.replicate2, counts.HO.Day21.replicate3, counts.HO.Day21.replicate4, counts.HO.Day7.replicate1, counts.HO.Day7.replicate2, counts.HO.Day7.replicate3, counts.HO.Day7.replicate4</code></pre>
<p>We will now run <code>SCTransform()</code>, which will result in a
new assay, the <code>SCT</code> assay.</p>
<pre class="r"><code>geo_so = SCTransform(geo_so)</code></pre>
<p>And we can see the additional assay by viewing the
<code>geo_so</code> object:</p>
<pre class="r"><code>geo_so</code></pre>
<pre><code>An object of class Seurat 
38503 features across 22320 samples within 2 assays 
Active assay: SCT (17129 features, 3000 variable features)
 3 layers present: counts, data, scale.data
 1 other assay present: RNA</code></pre>
<blockquote>
<p><strong>Regressing out sources of variation</strong></p>
<p>In the <code>SCTransform()</code> function there is a parameter,
<code>vars.to.regress</code>, which allows us to regress out sources of
variation that we don’t want to play a role in downstream analyses. A
common variable to regress out is <code>percent.mt</code>.</p>
</blockquote>
<p>We observe at this point that <code>SCTransform()</code> performed
its normalization and variable feature identification for each batch
independently, and a consensus set of variable features is automatically
identified for later use (<a
href="https://satijalab.org/seurat/articles/seurat5_integration">source</a>).</p>
<p>We also note that within the “SCT” assay, the <code>counts</code>
layer contains the corrected counts, the <code>data</code> layer
contains the log(1 + counts), and the <code>scale.data</code> layer
contains the pearson residuals from the SCT models.</p>
</div>
<div id="summary" class="section level1">
<h1>Summary</h1>
<p>Next steps: Projection and clustering</p>
<hr />
<p>These materials have been adapted and extended from materials listed
above. These are open access materials distributed under the terms of
the <a href="http://creativecommons.org/licenses/by/4.0/">Creative
Commons Attribution license (CC BY 4.0)</a>, which permits unrestricted
use, distribution, and reproduction in any medium, provided the original
author and source are credited.</p>
<br/> <br/>
<hr/>
<table style="width:100%;">
<colgroup>
<col width="28%" />
<col width="42%" />
<col width="28%" />
</colgroup>
<thead>
<tr class="header">
<th align="left"><a href="02-QCandFiltering.html">Previous
lesson</a></th>
<th align="center"><a href="#top">Top of this lesson</a></th>
<th align="right"><a href="PCAandDimReduction.html">Next lesson</a></th>
</tr>
</thead>
<tbody>
</tbody>
</table>
</div>

<div id="rmd-source-code">LS0tCnRpdGxlOiAiTm9ybWFsaXphdGlvbiIKYXV0aG9yOiAiVU0gQmlvaW5mb3JtYXRpY3MgQ29yZSIKZGF0ZTogImByIFN5cy5EYXRlKClgIgpvdXRwdXQ6CiAgICAgICAgaHRtbF9kb2N1bWVudDoKICAgICAgICAgICAgaW5jbHVkZXM6CiAgICAgICAgICAgICAgICBpbl9oZWFkZXI6IGhlYWRlci5odG1sCiAgICAgICAgICAgIHRoZW1lOiBwYXBlcgogICAgICAgICAgICB0b2M6IHRydWUKICAgICAgICAgICAgdG9jX2RlcHRoOiA0CiAgICAgICAgICAgIHRvY19mbG9hdDogdHJ1ZQogICAgICAgICAgICBudW1iZXJfc2VjdGlvbnM6IGZhbHNlCiAgICAgICAgICAgIGZpZ19jYXB0aW9uOiB0cnVlCiAgICAgICAgICAgIG1hcmtkb3duOiBHRk0KICAgICAgICAgICAgY29kZV9kb3dubG9hZDogdHJ1ZQotLS0KCjxzdHlsZSB0eXBlPSJ0ZXh0L2NzcyI+CmJvZHksIHRkIHsKICAgZm9udC1zaXplOiAxOHB4Owp9CmNvZGUucnsKICBmb250LXNpemU6IDEycHg7Cn0KcHJlIHsKICBmb250LXNpemU6IDEycHgKfQo8L3N0eWxlPgoKYGBge3IsIGluY2x1ZGUgPSBGQUxTRX0Kc291cmNlKCIuLi9iaW4vY2h1bmstb3B0aW9ucy5SIikKa25pdHJfZmlnX3BhdGgoIlhYIikKYGBgCgojIFdvcmtmbG93IE92ZXJ2aWV3IHsudW5saXN0ZWQgLnVubnVtYmVyZWR9Cgo8YnIvPgo8aW1nIHNyYz0iaW1hZ2VzL3dheWZpbmRlci93YXlmaW5kZXIucG5nIiBhbHQ9IndheWZpbmRlciIgc3R5bGU9ImhlaWdodDogNDAwcHg7Ii8+Cjxici8+Cjxici8+IAoKIyBJbnRyb2R1Y3Rpb24gCgpGb3IgbGFyZ2VyIHNjUk5BLXNlcSBwcm9qZWN0cywgZGF0YSBtYXkgdW5hdm9pZGFibHkgYmUgZ2VuZXJhdGVkIGFjcm9zcyBtdWx0aXBsZSBiYXRjaGVzLiBBcyB3aXRoIGJ1bGsgUk5BLXNlcSwgc29tZSBzdGVwcyBkdXJpbmcgbGlicmFyeSBwcmVwYXJhdGlvbiBjYW4gYmUgdGFrZW4gdG8gYXZvaWQgY29uZm91bmRpbmcgYmF0Y2ggZWZmZWN0cyB3aXRoIGJpb2xvZ2ljYWwgdmFyaWFibGVzIG9mIGludGVyZXN0LiBCdXQgaXQgd2lsbCByZW1haW4gbmVjZXNzYXJ5IHRvIGNvcnJlY3QgZm9yIHRlY2huaWNhbCB2YXJpYXRpb24gaW4gdGhlIHNhbXBsZSBpbnRlZ3JhdGlvbiBzdGVwLiA8IS0tIE1vcmUgaGVyZT8gLS0+IAoKSW4gdGhpcyBzZWN0aW9uLCB3ZSB3aWxsIGRlbW9uc3RyYXRlIHRoZSBzdGVwcyB0byBldmFsdWF0ZSB3aGV0aGVyIGludGVncmF0aW9uIGlzIG5lZWRlZCAoaXQgdXN1YWxseSBpcyksIGFuZCB3aWxsIHdhbGsgdGhyb3VnaCBpbnRlZ3JhdGluZyB0aGUgZGF0YSBhbmQgZXZhbHVhdGluZyB0aGUgZWZmZWN0aXZlbmVzcyB0aGVyZW9mLiAKCkxpa2UgaW4gdGhlIHByZXZpb3VzIHNlY3Rpb24sIG9uZSBpbnRlZ3JhdGlvbiBtZXRob2QgbWlnaHQgYmUgcmVwb3J0ZWQgaW4gYSByZXNlYXJjaCBwYXBlciBidXQgc2V2ZXJhbCBvcHRpb25zIG1pZ2h0IGhhdmUgYmVlbiB0ZXN0ZWQgYmVmb3JlIGRldGVybWluaW5nIHdoaWNoIGFwcHJvYWNoIHN1ZmZpY2llbnRseSBjb3JyZWN0ZWQgYmF0Y2ggZWZmZWN0cyB3aXRob3V0IHJlbW92aW5nIGV4cGVjdGVkIGJpb2xvZ2ljYWwgdmFyaWFiaWxpdHkuCgojIyBPYmplY3RpdmVzCgotIFVuZGVyc3RhbmQgd2h5IG5vcm1hbGl6YXRpb24gaXMgbmVlZGVkLgotIFVuZGVyc3RhbmQgc29tZSBvcHRpb25zIGZvciBub3JtYWxpemF0aW9uLgotIE5vcm1hbGl6ZSB0aGUgY291bnRzIHdpdGggYFNDVHJhbnNmb3JtKClgLgoKIyBOb3JtYWxpemF0aW9uCgpCdWxrIFJOQS1zZXEgc3VmZmVycyBmcm9tIHR3byBpc3N1ZXMgdGhhdCBtdXN0IGJlIGFkZHJlc3NlZCBpbiBhIGRpZmZlcmVudGlhbCBleHByZXNzaW9uIGFuYWx5c2lzOgoKMS4gRGlmZmVyZW5jZXMgaW4gc2VxdWVuY2luZyBkZXB0aCBjb250cmlidXRlcyB0byB0ZWNobmljYWwgdmFyaWF0aW9uLgoyLiBUaGVyZSBpcyBhIGNvbmZvdW5kaW5nIHJlbGF0aW9uc2hpcCBiZXR3ZWVuIGdlbmUgYWJ1bmRhbmNlIGFuZCBnZW5lIHZhcmlhdGlvbi4KClRoZXNlIGFsc28gYWZmZWN0IHNpbmdsZS1jZWxsIFJOQS1zZXEhIE1vcmVvdmVyLCB0aGUgc3BhcnNpdHkgb2Ygc2NSTkEtc2VxIGRhdGEgcG9zZXMgYWRkaXRpb25hbCBpc3N1ZXMgdGhhdCByZXF1aXJlIG5ldyBtZXRob2RzIGZvciBub3JtYWxpemF0aW9uLgoKV2Ugd2lsbCB1c2UgdGhlIGBTQ1RyYW5zZm9ybSgpYCBtZXRob2QgdGhhdCBpcyBwYXJ0IG9mIHRoZSBTZXVyYXQgcGFja2FnZS4gW1RoaXMgbWFudXNjcmlwdF0oaHR0cHM6Ly9nZW5vbWViaW9sb2d5LWJpb21lZGNlbnRyYWwtY29tL2FydGljbGVzLzEwLjExODYvczEzMDU5LTAyMS0wMjU4NC05KSBkZXNjcmliZXMgdGhlIGluaXRpYWwgYXBwcm9hY2gsIGFuZCBbdGhpcyB1cGRhdGVkIG1hbnVzY3JpcHRdKGh0dHBzOi8vbGluay5zcHJpbmdlci5jb20vYXJ0aWNsZS8xMC4xMTg2L3MxMzA1OS0wMjEtMDI1ODQtOSkgZGVzY3JpYmVzIHYyIG9mIHRoZSBhbGdvcml0aG0uCgpUaGUgYXV0aG9ycyBnbyB0byBncmVhdCBsZW5ndGhzIHRvIGRlbW9uc3RyYXRlIHRoYXQgdGhlcmUgaXMgYSAic3Ryb25nIGxpbmVhciByZWxhdGlvbnNoaXAgYmV0d2VlbiB1bm5vcm1hbGl6ZWQgZXhwcmVzc2lvbiAoZ2VuZSBVTUkgY291bnQpIGFuZCBjZWxsdWxhciBzZXF1ZW5jaW5nIGRlcHRoLiIgVGhlIHJlbGF0aW9uc2hpcCBob2xkcyBhY3Jvc3MgYSByYW5nZSBvZiBhYnVuZGFuY2UgbGV2ZWxzLCBhbmQgc3VnZ2VzdHMgdGhhdCBhIHNpbmdsZSBzY2FsaW5nIGZhY3RvciB0byBhY2NvdW50IGZvciBjZWxsdWxhciBzZXF1ZW5jaW5nIGRlcHRoIHdpbGwgbm90IGJlIHN1ZmZpY2llbnQsIGFuZCB0aGF0IGEgc29sdXRpb24gd2lsbCBkZXBlbmQgb24gdGhlIGFidW5kYW5jZSBsZXZlbC4KCj4gKipPdGhlciBub3JtYWxpemF0aW9ucyoqCj4gCj4gRnJvbSB0aGUgU2V1cmF0IGRvY3VtZW50YXRpb24sIHRoZSBgU0NUcmFuc2Zvcm0oKWAgZnVuY3Rpb24gInJlcGxhY2VzIGBOb3JtYWxpemVEYXRhKClgICh0aGlzIGlzIGEgbG9nLW5vcm1hbGl6YXRpb24pLCBgU2NhbGVEYXRhKClgLCBhbmQgYEZpbmRWYXJpYWJsZUZlYXR1cmVzKClgIi4gWW91IG1heSBzZWUgdGhlc2UgdGhyZWUgY29tbWFuZHMgaW4gb3RoZXIgdmlnbmV0dGVzLCBpbmRlZWQgZXZlbiBpbiBvdGhlciBTZXVyYXQgdmlnbmV0dGVzIChbc291cmNlXShodHRwczovL3NhdGlqYWxhYi5vcmcvc2V1cmF0L2FydGljbGVzL3NjdHJhbnNmb3JtX3ZpZ25ldHRlKSkuCgpBcyB3ZSBkaXNjdXNzZWQgd2hlbiBkZXNjcmliaW5nIHRoZSBzdHJ1Y3R1cmUgb2YgYSBgU2V1cmF0YCBvYmplY3QsIHRoZSBhc3NheSBkYXRhIGlzIHN0b3JlZCBpbiBsYXllcnMuIFRvIGNvbmZvcm0gdG8gdGhpcyBjb252ZW50aW9uLCB3ZSB3aWxsIHJ1biB0aGUgZm9sbG93aW5nIGNvbW1hbmQsIHdoaWNoIHNwbGl0cyB0aGUgIlJOQSIgYXNzYXkgaW50byBhIGxpc3Qgb2Ygc2FtcGxlcywgZWFjaCB3aXRoIGl0cyBvd24gbGF5ZXJzLgoKYGBge3IsIGV2YWwgPSBGQUxTRX0KZ2VvX3NvW1snUk5BJ11dID0gc3BsaXQoZ2VvX3NvW1snUk5BJ11dLCBmID0gZ2VvX3NvJG9yaWcuaWRlbnQpCmdlb19zbwpgYGAKfn5+CkFuIG9iamVjdCBvZiBjbGFzcyBTZXVyYXQgCjQ3MDM3IGZlYXR1cmVzIGFjcm9zcyAyOTYxNSBzYW1wbGVzIHdpdGhpbiAxIGFzc2F5CkFjdGl2ZSBhc3NheTogUk5BICgyNjQ4OSBmZWF0dXJlcywgMCB2YXJpYWJsZSBmZWF0dXJlcykKIDEyIGxheWVycyBwcmVzZW50OiBjb3VudHMuSE8uRGF5MC5yZXBsaWNhdGUxLCBjb3VudHMuSE8uRGF5MC5yZXBsaWNhdGUyLCBjb3VudHMuSE8uRGF5MC5yZXBsaWNhdGUzLCBjb3VudHMuSE8uRGF5MC5yZXBsaWNhdGU0LCBjb3VudHMuSE8uRGF5MjEucmVwbGljYXRlMSwgY291bnRzLkhPLkRheTIxLnJlcGxpY2F0ZTIsIGNvdW50cy5ITy5EYXkyMS5yZXBsaWNhdGUzLCBjb3VudHMuSE8uRGF5MjEucmVwbGljYXRlNCwgY291bnRzLkhPLkRheTcucmVwbGljYXRlMSwgY291bnRzLkhPLkRheTcucmVwbGljYXRlMiwgY291bnRzLkhPLkRheTcucmVwbGljYXRlMywgY291bnRzLkhPLkRheTcucmVwbGljYXRlNAp+fn4KCldlIHdpbGwgbm93IHJ1biBgU0NUcmFuc2Zvcm0oKWAsIHdoaWNoIHdpbGwgcmVzdWx0IGluIGEgbmV3IGFzc2F5LCB0aGUgYFNDVGAgYXNzYXkuCgpgYGB7ciwgZXZhbCA9IEZBTFNFfQpnZW9fc28gPSBTQ1RyYW5zZm9ybShnZW9fc28pCmBgYAoKQW5kIHdlIGNhbiBzZWUgdGhlIGFkZGl0aW9uYWwgYXNzYXkgYnkgdmlld2luZyB0aGUgYGdlb19zb2Agb2JqZWN0OgoKYGBge3IsIGV2YWwgPSBGQUxTRX0KZ2VvX3NvCmBgYAp+fn4KQW4gb2JqZWN0IG9mIGNsYXNzIFNldXJhdCAKMzg1MDMgZmVhdHVyZXMgYWNyb3NzIDIyMzIwIHNhbXBsZXMgd2l0aGluIDIgYXNzYXlzIApBY3RpdmUgYXNzYXk6IFNDVCAoMTcxMjkgZmVhdHVyZXMsIDMwMDAgdmFyaWFibGUgZmVhdHVyZXMpCiAzIGxheWVycyBwcmVzZW50OiBjb3VudHMsIGRhdGEsIHNjYWxlLmRhdGEKIDEgb3RoZXIgYXNzYXkgcHJlc2VudDogUk5BCn5+fgoKPiAqKlJlZ3Jlc3Npbmcgb3V0IHNvdXJjZXMgb2YgdmFyaWF0aW9uKioKPiAKPiBJbiB0aGUgYFNDVHJhbnNmb3JtKClgIGZ1bmN0aW9uIHRoZXJlIGlzIGEgcGFyYW1ldGVyLCBgdmFycy50by5yZWdyZXNzYCwgd2hpY2ggYWxsb3dzIHVzIHRvIHJlZ3Jlc3Mgb3V0IHNvdXJjZXMgb2YgdmFyaWF0aW9uIHRoYXQgd2UgZG9uJ3Qgd2FudCB0byBwbGF5IGEgcm9sZSBpbiBkb3duc3RyZWFtIGFuYWx5c2VzLiBBIGNvbW1vbiB2YXJpYWJsZSB0byByZWdyZXNzIG91dCBpcyBgcGVyY2VudC5tdGAuCgpXZSBvYnNlcnZlIGF0IHRoaXMgcG9pbnQgdGhhdCBgU0NUcmFuc2Zvcm0oKWAgcGVyZm9ybWVkIGl0cyBub3JtYWxpemF0aW9uIGFuZCB2YXJpYWJsZSBmZWF0dXJlIGlkZW50aWZpY2F0aW9uIGZvciBlYWNoIGJhdGNoIGluZGVwZW5kZW50bHksIGFuZCBhIGNvbnNlbnN1cyBzZXQgb2YgdmFyaWFibGUgZmVhdHVyZXMgaXMgYXV0b21hdGljYWxseSBpZGVudGlmaWVkIGZvciBsYXRlciB1c2UgKFtzb3VyY2VdKGh0dHBzOi8vc2F0aWphbGFiLm9yZy9zZXVyYXQvYXJ0aWNsZXMvc2V1cmF0NV9pbnRlZ3JhdGlvbikpLgoKV2UgYWxzbyBub3RlIHRoYXQgd2l0aGluIHRoZSAiU0NUIiBhc3NheSwgdGhlIGBjb3VudHNgIGxheWVyIGNvbnRhaW5zIHRoZSBjb3JyZWN0ZWQgY291bnRzLCB0aGUgYGRhdGFgIGxheWVyIGNvbnRhaW5zIHRoZSBsb2coMSArIGNvdW50cyksIGFuZCB0aGUgYHNjYWxlLmRhdGFgIGxheWVyIGNvbnRhaW5zIHRoZSBwZWFyc29uIHJlc2lkdWFscyBmcm9tIHRoZSBTQ1QgbW9kZWxzLgoKIyBTdW1tYXJ5CgpOZXh0IHN0ZXBzOiBQcm9qZWN0aW9uIGFuZCBjbHVzdGVyaW5nCgotLS0tCgpUaGVzZSBtYXRlcmlhbHMgaGF2ZSBiZWVuIGFkYXB0ZWQgYW5kIGV4dGVuZGVkIGZyb20gbWF0ZXJpYWxzIGxpc3RlZCBhYm92ZS4gVGhlc2UgYXJlIG9wZW4gYWNjZXNzIG1hdGVyaWFscyBkaXN0cmlidXRlZCB1bmRlciB0aGUgdGVybXMgb2YgdGhlIFtDcmVhdGl2ZSBDb21tb25zIEF0dHJpYnV0aW9uIGxpY2Vuc2UgKENDIEJZIDQuMCldKGh0dHA6Ly9jcmVhdGl2ZWNvbW1vbnMub3JnL2xpY2Vuc2VzL2J5LzQuMC8pLCB3aGljaCBwZXJtaXRzIHVucmVzdHJpY3RlZCB1c2UsIGRpc3RyaWJ1dGlvbiwgYW5kIHJlcHJvZHVjdGlvbiBpbiBhbnkgbWVkaXVtLCBwcm92aWRlZCB0aGUgb3JpZ2luYWwgYXV0aG9yIGFuZCBzb3VyY2UgYXJlIGNyZWRpdGVkLgoKPGJyLz4KPGJyLz4KPGhyLz4KfCBbUHJldmlvdXMgbGVzc29uXSgwMi1RQ2FuZEZpbHRlcmluZy5odG1sKSB8IFtUb3Agb2YgdGhpcyBsZXNzb25dKCN0b3ApIHwgW05leHQgbGVzc29uXShQQ0FhbmREaW1SZWR1Y3Rpb24uaHRtbCkgfAp8IDotLS0gfCA6LS0tLTogfCAtLS06IHwKCgoKCg==</div>


</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->
<script>
$(document).ready(function () {
  window.initializeSourceEmbed("03-Integration.Rmd");
});
</script>

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3,h4",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
