<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="UM Bioinformatics Core" />

<meta name="date" content="2024-02-19" />

<title>Integration and batch correction</title>

<script src="site_libs/header-attrs-2.21/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/paper.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<script src="site_libs/navigation-1.1/codefolding.js"></script>
<script src="site_libs/navigation-1.1/sourceembed.js"></script>
<link href="site_libs/highlightjs-9.12.0/default.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<!--
Favicon dervied from
https://twemoji.twitter.com/
https://favicon.io/emoji-favicons/dna/
-->
<link rel="apple-touch-icon" sizes="180x180" href="/apple-touch-icon.png">
<link rel="icon" type="image/png" sizes="32x32" href="/favicon-32x32.png">
<link rel="icon" type="image/png" sizes="16x16" href="/favicon-16x16.png">
<link rel="shortcut icon" href="favicon-16x16.png" />
<link rel="manifest" href="/site.webmanifest">
<link rel="mask-icon" href="/safari-pinned-tab.svg" color="#5bbad5">
<meta name="msapplication-TileColor" content="#da532c">
<meta name="theme-color" content="#ffffff">

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>




<style type="text/css">
#rmd-source-code {
  display: none;
}
</style>





<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->
<style type="text/css">
.code-folding-btn { margin-bottom: 4px; }
</style>



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-inverse  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">Intro to scRNA-Seq</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="workshop_intro.html">Intro</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Day 1
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="00A-OrientingOnScRNASeq.html">Orienting on scRNA-Seq</a>
    </li>
    <li>
      <a href="00B-CellRangerInAction.html">Cell Ranger in Action</a>
    </li>
    <li>
      <a href="01-GettingStarted.html">Seurat</a>
    </li>
    <li>
      <a href="02-QCandFiltering.html">Secondary QC &amp; Filtering</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Day 2
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="03-Integration.html">Normalization</a>
    </li>
    <li>
      <a href="04-PCAandDimReduction.html">PCA &amp; Integration</a>
    </li>
    <li>
      <a href="05-ProjectionAndClustering.html">Projection &amp; Clustering</a>
    </li>
    <li>
      <a href="06-MarkerVisualization.html">Marker identification and visualization</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Day 3
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="07-CellTypeAnnos.html">Cell type annotation</a>
    </li>
    <li>
      <a href="08-DifferentialExpression.html">Differential expression analysis</a>
    </li>
    <li>
      <a href="09-IndependentExercise.html">Independent exercise</a>
    </li>
  </ul>
</li>
<li>
  <a href="workshop_wrap_up.html">Wrap up</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">

<div class="btn-group pull-right float-right">
<button type="button" class="btn btn-default btn-xs btn-secondary btn-sm dropdown-toggle" data-toggle="dropdown" data-bs-toggle="dropdown" aria-haspopup="true" aria-expanded="false"><span>Code</span> <span class="caret"></span></button>
<ul class="dropdown-menu dropdown-menu-right" style="min-width: 50px;">
<li><a id="rmd-download-source" href="#">Download Rmd</a></li>
</ul>
</div>



<h1 class="title toc-ignore">Integration and batch correction</h1>
<h4 class="author">UM Bioinformatics Core</h4>
<h4 class="date">2024-02-19</h4>

</div>


<style type="text/css">
body, td {
   font-size: 18px;
}
code.r{
  font-size: 12px;
}
pre {
  font-size: 12px
}
</style>
<div id="workflow-overview" class="section level1 unlisted unnumbered">
<h1 class="unlisted unnumbered">Workflow Overview</h1>
<p><br/>
<img src="images/wayfinder/wayfinder.png" alt="wayfinder" style="height: 400px;"/>
<br/> <br/></p>
</div>
<div id="objectives" class="section level1">
<h1>Objectives</h1>
<ul>
<li>Normalize the counts.</li>
<li>Perform dimension reduction and determine number of PCs to
keep.</li>
<li>Understand when integration of samples might be needed.</li>
<li>Execute integration of samples, and evaluate if it was
“successful”.</li>
<li>Discuss different integration techniques offered in Seurat v5.</li>
</ul>
</div>
<div id="introduction" class="section level1">
<h1>Introduction</h1>
<p>For larger scRNA-seq projects, data may unavoidably by generated
across multiple batches. As with bulk RNA-seq, some steps during library
preparation can be taken to avoid confounding batch effects with
biological variables of interest. But it will remain necessary to
correct for technical variation in the sample integration step.
<!-- More here? --></p>
<p>In this section, we will demonstrate the steps to evaluate whether
integration is needed (it usually is), and will walk through integrating
the data and evaluating the effectiveness thereof.</p>
<p>Like in the previous section, one integration method might be
reported in a research paper but several options might have been tested
before determining which approach sufficiently corrected batch effects
without removing expected biological variability.</p>
</div>
<div id="normalization" class="section level1">
<h1>Normalization</h1>
<p>As with bulk RNA-seq, differences in sequencing depth is a
contributing factor to technical variation in single-cell RNA-seq data.
Moreover, variance stabilization in gene counts can greatly improve
downstream analysis steps. Regardless of whether or not scRNA-seq data
is to be integrated, it is important to normalize the count data.</p>
<p>Here we will use the <code>SCTransform()</code> method that is part
of the Seurat package. <a
href="https://genomebiology-biomedcentral-com/articles/10.1186/s13059-021-02584-9">This
manuscript</a> describes the initial approach, and <a
href="https://link.springer.com/article/10.1186/s13059-021-02584-9">this
updated manuscript</a> describes the v2 version of
<code>SCTransform()</code>.</p>
<blockquote>
<h4 id="other-normalizations" class="unlisted unnumbered">Other
normalizations</h4>
<p>From the Seurat documentation, the <code>SCTransform()</code>
function “replaces <code>NormalizeData()</code>,
<code>ScaleData()</code>, and <code>FindVariableFeatures()</code>”. You
may see these three commands in other vignettes, indeed even in other
Seurat vignettes (<a
href="https://satijalab.org/seurat/articles/sctransform_vignette">source</a>).</p>
</blockquote>
<p>As we discussed when describing the structure of a
<code>Seurat</code> object, the assay data is stored in layers. To
conform to this convention, we will run the following command, which
splits the “RNA” assay into a list of samples, each with its own
layers.</p>
<pre class="r"><code>geo_so[[&#39;RNA&#39;]] = split(geo_so[[&#39;RNA&#39;]], f = geo_so$orig.ident)</code></pre>
<p>We will now run <code>SCTransform()</code>, which will result in a
new assay, the <code>SCT</code> assay.</p>
<pre class="r"><code>geo_so = SCTransform(geo_so)</code></pre>
<p>And we can see the additional assay by viewing the
<code>geo_so</code> object:</p>
<pre class="r"><code>geo_so</code></pre>
<pre><code>An object of class Seurat 
38503 features across 22320 samples within 2 assays 
Active assay: SCT (17129 features, 3000 variable features)
 3 layers present: counts, data, scale.data
 1 other assay present: RNA</code></pre>
<blockquote>
<h4 id="regressing-out-sources-of-variation"
class="unlisted unnumbered">Regressing out sources of variation</h4>
<p>In the <code>SCTransform()</code> function there is a parameter,
<code>vars.to.regress</code>, which allows us to regress out sources of
variation that we don’t want to play a role in downstream analyses. A
common variable to regress out is <code>percent.mt</code>.</p>
</blockquote>
<p>We observe at this point that <code>SCTransform()</code> performed
its normalization and variable feature identification for each batch
independently, and a consensus set of variable features is automatically
identified for later use (<a
href="https://satijalab.org/seurat/articles/seurat5_integration">source</a>).</p>
<p>We also note that within the “SCT” assay, the <code>counts</code>
layer contains the corrected counts, the <code>data</code> layer
contains the log(1 + counts), and the <code>scale.data</code> layer
contains the pearson residuals from the SCT models.</p>
</div>
<div id="dimension-reduction" class="section level1">
<h1>Dimension reduction</h1>
<p>In our data, we have seen our 12 samples contain 1K - 3K cells each,
with measurements across 21K genes. One of our ultimate goals is to
cluster the cells according to their expression profiles, and determine
what cell types correspond to these clusters for the purposes of
differential expression between clusters and/or between conditions
within clusters.</p>
<p>If we only had two genes, we would make a two-dimensional plot where
each axis is the expression of a gene, and each point represents a cell.
We could use the Euclidean distances to determine cells that were more
similar to each other than others. But as we’ve observed, our data
contains 21K genes, which doesn’t immediately lend itself to a
two-dimensional plot.</p>
<p>Dimension reduction helps us reduce the number of separate dimensions
by grouping multiple features into a single dimension. This makes
downstream analyses like nearest neighbor finding and clustering more
computationally feasible.</p>
<div id="principal-component-analysis" class="section level2">
<h2>Principal component analysis</h2>
<p>Principal component analysis (PCA) determines axes in
high-dimensional space that capture the most variation. By definition,
the PCs capture the greatest factors of heterogeneity in the data set,
and we can use the top PCs to approximate the full data set in
downstream analysis.</p>
<p>Here we’re assuming that biological processes affect multiple genes
in a coordinated way. Therefore the top PCs likely represent biological
structure rather than random tehcnical or biological noise, which
affects each gene independently [(source)[<a
href="https://bioconductor.org/books/3.12/OSCA/dimensionality-reduction.html#principal-components-analysis"
class="uri">https://bioconductor.org/books/3.12/OSCA/dimensionality-reduction.html#principal-components-analysis</a>]].</p>
<!-- An image of how PCA works? -->
<p>Let’s run PCA on the SCT normalized assay. We will name the reduction
in an informative way to keep them clear for us in the future. In this
case, our data has not yet been integrated, but it has been SCT
normalized. So we will name the reduction
<code>unintegrated.sct.pca</code>. Note that <code>SCTTransform()</code>
returned a set of highly variable genes, and the <code>RunPCA()</code>
function will use this subset to determine the PCs and genes associated
therewith.</p>
<pre class="r"><code>geo_so = RunPCA(geo_so, reduction.name = &#39;unintegrated.sct.pca&#39;)</code></pre>
<p>In the console, the first 5 PCs are listed along with the associated
genes. If we look at the <code>geo_so</code> object, we’ll see our
dimension reduction:</p>
<pre class="r"><code>geo_so</code></pre>
<pre><code>An object of class Seurat 
38503 features across 22320 samples within 2 assays 
Active assay: SCT (17129 features, 3000 variable features)
 3 layers present: counts, data, scale.data
 1 other assay present: RNA
 1 dimensional reductions calculated: unintegrated.sct.pca</code></pre>
<div id="choosing-the-number-of-pcs" class="section level3">
<h3>Choosing the number of PCs</h3>
<p>For downstream analysis, we need to make the decision of how many PCs
to include. The trade-off is that including more PCs includes more of
the biological signal but at the cost of including more technical
variation. One way to determine how many PCs to include is by looking at
an elbow plot, which shows the percent variance explained by successive
PCs.</p>
<pre class="r"><code>ElbowPlot(geo_so, ndims = 50, reduction = &#39;unintegrated.sct.pca&#39;)
ggsave(filename = &#39;ISC_R/results/figures/qc_sct_elbow_plot.png&#39;)</code></pre>
<p><img src="./images/curriculum/qc_sct_elbow_plot.png" /></p>
<p>In this plot, we could arbitrarily choose a number along the x-axis
that looks like a sharp change in the variance from one PC to the next,
that is, an elbow. Of course, the choice is not always obvious, and this
plot is no different. We could also quantify our choice a bit with a
helpful function:</p>
<pre class="r"><code># Function to determine optimal PCs after RunPCA(). Thanks Dana!
optimal_pcs = function(so, reduction) {
  # quantitative check for number of PCs to include
  pct = so@reductions[[reduction]]@stdev / sum(so@reductions[[reduction]]@stdev) * 100
  cum = cumsum(pct)
  co1 = which(cum &gt; 90 &amp; pct &lt; 5)[1]
  co2 = sort(which((pct[1:length(pct)-1] - pct[2:length(pct)]) &gt; .1), decreasing = T)[1] + 1
  pcs = min(co1, co2) 
  
  return(pcs)
}</code></pre>
<p>This function takes in a <code>Seurat</code> object and the name of a
reduction. It then determines two quantities:</p>
<ol style="list-style-type: decimal">
<li>At which PC is 90% of the variation cumulatively explained, while
also explaining &lt;5% itself?</li>
<li>Where is the last difference in variation explained from one PC to
the next greater than 0.1?</li>
</ol>
<p>Of these two, the minimum number can be selected as the “elbow” in
the plot. In our case, that number is 23, which we can round up to 25
for good measure.</p>
<pre class="r"><code>optimal_pcs(geo_so, &#39;unintegrated.sct.pca&#39;)</code></pre>
<pre><code>23</code></pre>
</div>
</div>
</div>
<div id="integrate-layers" class="section level1">
<h1>Integrate Layers</h1>
<p>Having normalized the data with <code>SCTransform()</code> and
performed the dimension reduction with <code>RunPCA()</code>, we are now
able to run <code>IntegrateLayers()</code>:</p>
<pre class="r"><code>geo_so = IntegrateLayers(
    object = geo_so, 
    method = CCAIntegration, 
    orig.reduction = &#39;unintegrated.sct.pca&#39;,
    normalization.method = &#39;SCT&#39;,
    new.reduction = &#39;integrated.sct.cca&#39;)</code></pre>
<p>Note we have specified the unintegrated reduction
<code>unintegrated.sct.pca</code>, which is what
<code>IntegrateLayers()</code> operates on, along with the
<code>SCT</code> assay. Let’s take a look to see what’s different about
the <code>Seurat</code> object:</p>
<pre class="r"><code>geo_so</code></pre>
<pre><code>An object of class Seurat 
38503 features across 22320 samples within 2 assays 
Active assay: SCT (17129 features, 3000 variable features)
 3 layers present: counts, data, scale.data
 1 other assay present: RNA
 2 dimensional reductions calculated: unintegrated.sct.pca, integrated.sct.cca</code></pre>
<p>Observe that we now have a new reduction,
<code>integrated.sct.cca</code>, which we will use downstream.</p>
<div id="alternate-integration-methods" class="section level2">
<h2>Alternate integration methods</h2>
<p>New to Seurat v5 is the improved <code>IntegrateLayers()</code>
function which makes selecting different integration methods much
easier. The results of each alternative integration method are stored
within the same <code>Seurat</code> object to make comparing downstream
effects much easier. <a
href="https://satijalab.org/seurat/articles/seurat5_integration#perform-streamlined-one-line-integrative-analysis">This</a>
Seurat vignette on integrative analysis in Seurat v5 provides examples
of each integration method. Note, that the code in this vignette assumes
the
<code>NormalizeData() &gt; ScaleData() &gt; FindVariableFeatures()</code>
pipeline was used. So, for example, if we wanted to run the slightly
faster <code>RPCAIntegration</code> method, we would run:</p>
<pre class="r"><code>geo_so = IntegrateLayers(
  object = geo_so, 
  method = RPCAIntegration, 
  orig.reduction = &#39;unintegrated.sct.pca&#39;,
  normalization.method = &#39;SCT&#39;,
  new.reduction = &#39;integrated.sct.rpca&#39;)</code></pre>
<p>Note the sepcification of <code>normalization.method = 'SCT'</code>
is different than the vignette.</p>
</div>
</div>
<div id="move-to-later-sections" class="section level1">
<h1>Move to later sections?</h1>
<!-- This stuff belongs in Dana's section on projection and clustering... But we'll keep it here for now... -->
<p>We now run
<code>FindNeighbors() &gt; FindClusters() &gt; RunUMAP()</code> for the
new reduction <code>integrated.sct.cca</code>:</p>
<pre class="r"><code>geo_so = FindNeighbors(geo_so, dims = 1:25, reduction = &#39;integrated.sct.cca&#39;)
geo_so = FindClusters(geo_so, resolution = 1, cluster.name = &#39;integrated.sct.cca.clusters&#39;)
geo_so = RunUMAP(geo_so, dims = 1:25, reduction = &#39;integrated.sct.cca&#39;, reduction.name = &#39;umap.sct.integrated.cca&#39;)</code></pre>
<p>Which sets us up to see the UMAP and evaluate whether we’ve reduced
the condition effect on the cluster membership.</p>
<pre class="r"><code>post_integration_umap_plot_orig.ident = DimPlot(geo_so, group.by = &#39;orig.ident&#39;, label = FALSE, reduction = &#39;umap.sct.integrated.cca&#39;) + NoLegend()
ggsave(filename = &#39;ISC_R/results/figures/umap_integrated_sct_orig.ident.png&#39;, plot = post_integration_umap_plot_orig.ident)</code></pre>
<p><img
src="images/curriculum/umap_integrated_sct_orig.ident.png" /></p>
<pre class="r"><code>post_integration_umap_plot_day = DimPlot(geo_so, group.by = &#39;day&#39;, label = FALSE, reduction = &#39;umap.sct.integrated.cca&#39;) + NoLegend()
ggsave(filename = &#39;ISC_R/results/figures/umap_integrated_sct_day.png&#39;, plot = post_integration_umap_plot_day)</code></pre>
<p><img src="images/curriculum/umap_integrated_sct_day.png" /></p>
<p>Indeed, it looks like we’ve succeeded in reducing the effect of
condition on cluster membership by integrating the data.</p>
</div>
<div id="rewind-pre-integration-evaluation" class="section level1">
<h1>Rewind: Pre-integration evaluation</h1>
<p>Prior to integration, we would like to know if there are any batch
effects that carry through to the cell clusters. Most importantly, we
would like to know if clusters tend to be determined by sample or
condition (in this case, the day). In order to figure this out we have
already run the following three commands, which we will describe in more
detail later.</p>
<pre class="r"><code># DefaultAssay(geo_so) = &#39;RNA&#39;

geo_so = RunPCA(geo_so, reduction.name = &#39;unintegrated.sct.pca&#39;)
geo_so = FindNeighbors(geo_so, dims = 1:25, reduction = &#39;unintegrated.sct.pca&#39;)
geo_so = FindClusters(geo_so, resolution = 0.5, cluster.name = &#39;unintegrated.sct.clusters&#39;)
geo_so = RunUMAP(geo_so, dims = 1:25, reduction = &#39;unintegrated.sct.pca&#39;, reduction.name = &#39;umap.unintegrated.sct.pca&#39;)</code></pre>
<p>After running these commands, we can visualize the cells in a UMAP
plot, which projects the high-dimensional gene expression into two
dimensions.</p>
<pre class="r"><code>pre_integration_umap_plot = DimPlot(geo_so, group.by = &#39;orig.ident&#39;, label = FALSE, reduction = &#39;umap.unintegrated.sct.pca&#39;) + NoLegend()
ggsave(filename = &#39;ISC_R/results/figures/umap_unintegrated_sct_orig.ident.png&#39;, plot = pre_integration_umap_plot)</code></pre>
<p><img
src="images/curriculum/umap_unintegrated_sct_orig.ident.png" /></p>
<p>From this image, we can qualitatively see that some clusters seem to
be driven by the original identity. We can see this even more clearly if
we use <code>group.by = 'day'</code>:</p>
<pre class="r"><code>pre_integration_umap_plot = DimPlot(geo_so, group.by = &#39;day&#39;, label = FALSE, reduction = &#39;umap.unintegrated.sct.pca&#39;) + NoLegend()
ggsave(filename = &#39;ISC_R/results/figures/umap_unintegrated_sct_day.png&#39;, plot = pre_integration_umap_plot)</code></pre>
<p><img src="images/curriculum/umap_unintegrated_sct_day.png" /></p>
<p>We can also evaluate cluster membership by <code>orig.ident</code> or
by <code>day</code> using the <code>table()</code> function:</p>
<pre class="r"><code>table(geo_so@meta.data$day, geo_so@meta.data$unintegrated.sct.clusters)</code></pre>
<pre><code>           0    1    2    3    4    5    6    7    8    9   10   11   12   13   14   15   16   17   18
  Day0     0   17    4 1147   50   85   46 1022  311   36   26  270  145  106   37   66   34  159    0
  Day21  108   46  195 1346   50 1239  685   15  604  384   60   95  195  176   40   41   59    5    2
  Day7  3190 2894 2428  132 1827   51  613    2   35  470  684  166   87   78  274  223  117    9  134</code></pre>
<p>We can see that many clusters appear to be wholly made up of cells
from one or two days. Integrating the data may solve this issue.</p>
<pre class="r"><code>DefaultAssay(geo_so) = &#39;SCT&#39;</code></pre>
</div>
<div id="summary" class="section level1">
<h1>Summary</h1>
<p>Next steps: Projection and clustering</p>
<hr />
<p>These materials have been adapted and extended from materials listed
above. These are open access materials distributed under the terms of
the <a href="http://creativecommons.org/licenses/by/4.0/">Creative
Commons Attribution license (CC BY 4.0)</a>, which permits unrestricted
use, distribution, and reproduction in any medium, provided the original
author and source are credited.</p>
<br/> <br/>
<hr/>
<table style="width:100%;">
<colgroup>
<col width="28%" />
<col width="42%" />
<col width="28%" />
</colgroup>
<thead>
<tr class="header">
<th align="left"><a href="02-QCandFiltering.html">Previous
lesson</a></th>
<th align="center"><a href="#top">Top of this lesson</a></th>
<th align="right"><a href="PCAandDimReduction.html">Next lesson</a></th>
</tr>
</thead>
<tbody>
</tbody>
</table>
</div>

<div id="rmd-source-code">---
title: "Integration and batch correction"
author: "UM Bioinformatics Core"
date: "`r Sys.Date()`"
output:
        html_document:
            includes:
                in_header: header.html
            theme: paper
            toc: true
            toc_depth: 4
            toc_float: true
            number_sections: false
            fig_caption: true
            markdown: GFM
            code_download: true
---

<style type="text/css">
body, td {
   font-size: 18px;
}
code.r{
  font-size: 12px;
}
pre {
  font-size: 12px
}
</style>

```{r, include = FALSE}
source("../bin/chunk-options.R")
knitr_fig_path("XX")
```

# Workflow Overview {.unlisted .unnumbered}

<br/>
<img src="images/wayfinder/wayfinder.png" alt="wayfinder" style="height: 400px;"/>
<br/>
<br/> 

# Objectives

- Normalize the counts.
- Perform dimension reduction and determine number of PCs to keep.
- Understand when integration of samples might be needed.
- Execute integration of samples, and evaluate if it was "successful".
- Discuss different integration techniques offered in Seurat v5.

# Introduction 

For larger scRNA-seq projects, data may unavoidably by generated across multiple batches. As with bulk RNA-seq, some steps during library preparation can be taken to avoid confounding batch effects with biological variables of interest. But it will remain necessary to correct for technical variation in the sample integration step. <!-- More here? --> 

In this section, we will demonstrate the steps to evaluate whether integration is needed (it usually is), and will walk through integrating the data and evaluating the effectiveness thereof. 

Like in the previous section, one integration method might be reported in a research paper but several options might have been tested before determining which approach sufficiently corrected batch effects without removing expected biological variability.

# Normalization

As with bulk RNA-seq, differences in sequencing depth is a contributing factor to technical variation in single-cell RNA-seq data. Moreover, variance stabilization in gene counts can greatly improve downstream analysis steps. Regardless of whether or not scRNA-seq data is to be integrated, it is important to normalize the count data.

Here we will use the `SCTransform()` method that is part of the Seurat package. [This manuscript](https://genomebiology-biomedcentral-com/articles/10.1186/s13059-021-02584-9) describes the initial approach, and [this updated manuscript](https://link.springer.com/article/10.1186/s13059-021-02584-9) describes the v2 version of `SCTransform()`.

> #### Other normalizations {.unlisted .unnumbered}
> 
> From the Seurat documentation, the `SCTransform()` function "replaces `NormalizeData()`, `ScaleData()`, and `FindVariableFeatures()`". You may see these three commands in other vignettes, indeed even in other Seurat vignettes ([source](https://satijalab.org/seurat/articles/sctransform_vignette)).

As we discussed when describing the structure of a `Seurat` object, the assay data is stored in layers. To conform to this convention, we will run the following command, which splits the "RNA" assay into a list of samples, each with its own layers.

```{r, eval = FALSE}
geo_so[['RNA']] = split(geo_so[['RNA']], f = geo_so$orig.ident)
```

We will now run `SCTransform()`, which will result in a new assay, the `SCT` assay.

```{r, eval = FALSE}
geo_so = SCTransform(geo_so)
```

And we can see the additional assay by viewing the `geo_so` object:

```{r, eval = FALSE}
geo_so
```
~~~
An object of class Seurat 
38503 features across 22320 samples within 2 assays 
Active assay: SCT (17129 features, 3000 variable features)
 3 layers present: counts, data, scale.data
 1 other assay present: RNA
~~~

> #### Regressing out sources of variation {.unlisted .unnumbered}
> 
> In the `SCTransform()` function there is a parameter, `vars.to.regress`, which allows us to regress out sources of variation that we don't want to play a role in downstream analyses. A common variable to regress out is `percent.mt`.

We observe at this point that `SCTransform()` performed its normalization and variable feature identification for each batch independently, and a consensus set of variable features is automatically identified for later use ([source](https://satijalab.org/seurat/articles/seurat5_integration)).

We also note that within the "SCT" assay, the `counts` layer contains the corrected counts, the `data` layer contains the log(1 + counts), and the `scale.data` layer contains the pearson residuals from the SCT models.

# Dimension reduction

In our data, we have seen our 12 samples contain 1K - 3K cells each, with measurements across 21K genes. One of our ultimate goals is to cluster the cells according to their expression profiles, and determine what cell types correspond to these clusters for the purposes of differential expression between clusters and/or between conditions within clusters.

If we only had two genes, we would make a two-dimensional plot where each axis is the expression of a gene, and each point represents a cell. We could use the Euclidean distances to determine cells that were more similar to each other than others. But as we've observed, our data contains 21K genes, which doesn't immediately lend itself to a two-dimensional plot.

Dimension reduction helps us reduce the number of separate dimensions by grouping multiple features into a single dimension. This makes downstream analyses like nearest neighbor finding and clustering more computationally feasible.

## Principal component analysis

Principal component analysis (PCA) determines axes in high-dimensional space that capture the most variation. By definition, the PCs capture the greatest factors of heterogeneity in the data set, and we can use the top PCs to approximate the full data set in downstream analysis.

Here we're assuming that biological processes affect multiple genes in a coordinated way. Therefore the top PCs likely represent biological structure rather than random tehcnical or biological noise, which affects each gene independently [(source)[https://bioconductor.org/books/3.12/OSCA/dimensionality-reduction.html#principal-components-analysis]].

<!-- An image of how PCA works? -->

Let's run PCA on the SCT normalized assay. We will name the reduction in an informative way to keep them clear for us in the future. In this case, our data has not yet been integrated, but it has been SCT normalized. So we will name the reduction `unintegrated.sct.pca`. Note that `SCTTransform()` returned a set of highly variable genes, and the `RunPCA()` function will use this subset to determine the PCs and genes associated therewith.

```{r, eval = FALSE}
geo_so = RunPCA(geo_so, reduction.name = 'unintegrated.sct.pca')
```

In the console, the first 5 PCs are listed along with the associated genes. If we look at the `geo_so` object, we'll see our dimension reduction:

```{r, eval = FALSE}
geo_so
```
~~~
An object of class Seurat 
38503 features across 22320 samples within 2 assays 
Active assay: SCT (17129 features, 3000 variable features)
 3 layers present: counts, data, scale.data
 1 other assay present: RNA
 1 dimensional reductions calculated: unintegrated.sct.pca
~~~

### Choosing the number of PCs

For downstream analysis, we need to make the decision of how many PCs to include. The trade-off is that including more PCs includes more of the biological signal but at the cost of including more technical variation. One way to determine how many PCs to include is by looking at an elbow plot, which shows the percent variance explained by successive PCs.

```{r, eval = FALSE}
ElbowPlot(geo_so, ndims = 50, reduction = 'unintegrated.sct.pca')
ggsave(filename = 'ISC_R/results/figures/qc_sct_elbow_plot.png')
```

![](./images/curriculum/qc_sct_elbow_plot.png)

In this plot, we could arbitrarily choose a number along the x-axis that looks like a sharp change in the variance from one PC to the next, that is, an elbow. Of course, the choice is not always obvious, and this plot is no different. We could also quantify our choice a bit with a helpful function:


```{r, eval = FALSE}
# Function to determine optimal PCs after RunPCA(). Thanks Dana!
optimal_pcs = function(so, reduction) {
  # quantitative check for number of PCs to include
  pct = so@reductions[[reduction]]@stdev / sum(so@reductions[[reduction]]@stdev) * 100
  cum = cumsum(pct)
  co1 = which(cum > 90 & pct < 5)[1]
  co2 = sort(which((pct[1:length(pct)-1] - pct[2:length(pct)]) > .1), decreasing = T)[1] + 1
  pcs = min(co1, co2) 
  
  return(pcs)
}
```

This function takes in a `Seurat` object and the name of a reduction. It then determines two quantities:

1. At which PC is 90% of the variation cumulatively explained, while also explaining <5% itself?
2. Where is the last difference in variation explained from one PC to the next greater than 0.1?

Of these two, the minimum number can be selected as the "elbow" in the plot. In our case, that number is 23, which we can round up to 25 for good measure.

```{r, eval = FALSE}
optimal_pcs(geo_so, 'unintegrated.sct.pca')
```
~~~
23
~~~

# Integrate Layers

Having normalized the data with `SCTransform()` and performed the dimension reduction with `RunPCA()`, we are now able to run `IntegrateLayers()`:

```{r, eval = FALSE}
geo_so = IntegrateLayers(
    object = geo_so, 
    method = CCAIntegration, 
    orig.reduction = 'unintegrated.sct.pca',
    normalization.method = 'SCT',
    new.reduction = 'integrated.sct.cca')
```

Note we have specified the unintegrated reduction `unintegrated.sct.pca`, which is what `IntegrateLayers()` operates on, along with the `SCT` assay. Let's take a look to see what's different about the `Seurat` object:

```{r, eval = FALSE}
geo_so
```
~~~
An object of class Seurat 
38503 features across 22320 samples within 2 assays 
Active assay: SCT (17129 features, 3000 variable features)
 3 layers present: counts, data, scale.data
 1 other assay present: RNA
 2 dimensional reductions calculated: unintegrated.sct.pca, integrated.sct.cca
~~~

Observe that we now have a new reduction, `integrated.sct.cca`, which we will use downstream.

## Alternate integration methods

New to Seurat v5 is the improved `IntegrateLayers()` function which makes selecting different integration methods much easier. The results of each alternative integration method are stored within the same `Seurat` object to make comparing downstream effects much easier. [This](https://satijalab.org/seurat/articles/seurat5_integration#perform-streamlined-one-line-integrative-analysis) Seurat vignette on integrative analysis in Seurat v5 provides examples of each integration method. Note, that the code in this vignette assumes the `NormalizeData() > ScaleData() > FindVariableFeatures()` pipeline was used. So, for example, if we wanted to run the slightly faster `RPCAIntegration` method, we would run:

```{r, eval = FALSE}
geo_so = IntegrateLayers(
  object = geo_so, 
  method = RPCAIntegration, 
  orig.reduction = 'unintegrated.sct.pca',
  normalization.method = 'SCT',
  new.reduction = 'integrated.sct.rpca')
```

Note the sepcification of `normalization.method = 'SCT'` is different than the vignette.

# Move to later sections?

<!-- This stuff belongs in Dana's section on projection and clustering... But we'll keep it here for now... -->

We now run `FindNeighbors() > FindClusters() > RunUMAP()` for the new reduction `integrated.sct.cca`:

```{r, eval = FALSE}
geo_so = FindNeighbors(geo_so, dims = 1:25, reduction = 'integrated.sct.cca')
geo_so = FindClusters(geo_so, resolution = 1, cluster.name = 'integrated.sct.cca.clusters')
geo_so = RunUMAP(geo_so, dims = 1:25, reduction = 'integrated.sct.cca', reduction.name = 'umap.sct.integrated.cca')
```

Which sets us up to see the UMAP and evaluate whether we've reduced the condition effect on the cluster membership.

```{r, eval = FALSE}
post_integration_umap_plot_orig.ident = DimPlot(geo_so, group.by = 'orig.ident', label = FALSE, reduction = 'umap.sct.integrated.cca') + NoLegend()
ggsave(filename = 'ISC_R/results/figures/umap_integrated_sct_orig.ident.png', plot = post_integration_umap_plot_orig.ident)
```

![](./images/curriculum/umap_integrated_sct_orig.ident.png)

```{r, eval = FALSE}
post_integration_umap_plot_day = DimPlot(geo_so, group.by = 'day', label = FALSE, reduction = 'umap.sct.integrated.cca') + NoLegend()
ggsave(filename = 'ISC_R/results/figures/umap_integrated_sct_day.png', plot = post_integration_umap_plot_day)
```

![](./images/curriculum/umap_integrated_sct_day.png)

Indeed, it looks like we've succeeded in reducing the effect of condition on cluster membership by integrating the data.

# Rewind: Pre-integration evaluation

Prior to integration, we would like to know if there are any batch effects that carry through to the cell clusters. Most importantly, we would like to know if clusters tend to be determined by sample or condition (in this case, the day). In order to figure this out we have already run the following three commands, which we will describe in more detail later.

```{r, eval = FALSE}
# DefaultAssay(geo_so) = 'RNA'

geo_so = RunPCA(geo_so, reduction.name = 'unintegrated.sct.pca')
geo_so = FindNeighbors(geo_so, dims = 1:25, reduction = 'unintegrated.sct.pca')
geo_so = FindClusters(geo_so, resolution = 0.5, cluster.name = 'unintegrated.sct.clusters')
geo_so = RunUMAP(geo_so, dims = 1:25, reduction = 'unintegrated.sct.pca', reduction.name = 'umap.unintegrated.sct.pca')
```

After running these commands, we can visualize the cells in a UMAP plot, which projects the high-dimensional gene expression into two dimensions.

```{r, eval = FALSE}
pre_integration_umap_plot = DimPlot(geo_so, group.by = 'orig.ident', label = FALSE, reduction = 'umap.unintegrated.sct.pca') + NoLegend()
ggsave(filename = 'ISC_R/results/figures/umap_unintegrated_sct_orig.ident.png', plot = pre_integration_umap_plot)
```

![](./images/curriculum/umap_unintegrated_sct_orig.ident.png)

From this image, we can qualitatively see that some clusters seem to be driven by the original identity. We can see this even more clearly if we use `group.by = 'day'`:

```{r, eval = FALSE}
pre_integration_umap_plot = DimPlot(geo_so, group.by = 'day', label = FALSE, reduction = 'umap.unintegrated.sct.pca') + NoLegend()
ggsave(filename = 'ISC_R/results/figures/umap_unintegrated_sct_day.png', plot = pre_integration_umap_plot)
```

![](./images/curriculum/umap_unintegrated_sct_day.png)

We can also evaluate cluster membership by `orig.ident` or by `day` using the `table()` function:

```{r, eval = FALSE}
table(geo_so@meta.data$day, geo_so@meta.data$unintegrated.sct.clusters)
```

~~~
           0    1    2    3    4    5    6    7    8    9   10   11   12   13   14   15   16   17   18
  Day0     0   17    4 1147   50   85   46 1022  311   36   26  270  145  106   37   66   34  159    0
  Day21  108   46  195 1346   50 1239  685   15  604  384   60   95  195  176   40   41   59    5    2
  Day7  3190 2894 2428  132 1827   51  613    2   35  470  684  166   87   78  274  223  117    9  134
~~~

We can see that many clusters appear to be wholly made up of cells from one or two days. Integrating the data may solve this issue.

```{r, eval = FALSE}
DefaultAssay(geo_so) = 'SCT'
```

# Summary

Next steps: Projection and clustering

----

These materials have been adapted and extended from materials listed above. These are open access materials distributed under the terms of the [Creative Commons Attribution license (CC BY 4.0)](http://creativecommons.org/licenses/by/4.0/), which permits unrestricted use, distribution, and reproduction in any medium, provided the original author and source are credited.

<br/>
<br/>
<hr/>
| [Previous lesson](02-QCandFiltering.html) | [Top of this lesson](#top) | [Next lesson](PCAandDimReduction.html) |
| :--- | :----: | ---: |




</div>


</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->
<script>
$(document).ready(function () {
  window.initializeSourceEmbed("03-Integration.Rmd");
});
</script>

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3,h4",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
