<!DOCTYPE html>

<html lang="en" xml:lang="en">

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="UM Bioinformatics Core Workshop Team" />

<meta name="date" content="2024-12-17" />

<title>Normalization</title>

<script src="site_libs/header-attrs-2.28/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/paper.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.13.2/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/default.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<script src="site_libs/clipboard-1.7.1/clipboard.min.js"></script>
<link href="site_libs/primer-tooltips-1.4.0/build.css" rel="stylesheet" />
<link href="site_libs/klippy-0.0.0.9500/css/klippy.min.css" rel="stylesheet" />
<script src="site_libs/klippy-0.0.0.9500/js/klippy.min.js"></script>
<!--
Favicon dervied from
https://twemoji.twitter.com/
https://favicon.io/emoji-favicons/dna/
-->
<link rel="apple-touch-icon" sizes="180x180" href="apple-touch-icon.png">
<link rel="icon" type="image/png" sizes="32x32" href="favicon-32x32.png">
<link rel="icon" type="image/png" sizes="16x16" href="favicon-16x16.png">
<link rel="shortcut icon" href="favicon-16x16.png" />
<link rel="manifest" href="site.webmanifest">
<link rel="mask-icon" href="safari-pinned-tab.svg" color="#5bbad5">
<meta name="msapplication-TileColor" content="#da532c">
<meta name="theme-color" content="#ffffff">

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>









<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-inverse  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">Intro to scRNA-Seq</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="workshop_intro.html">Intro</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Day 1
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="00A-OrientingOnScRNASeq.html">Orienting on scRNA-Seq</a>
    </li>
    <li>
      <a href="01-GettingStarted.html">Getting Started with Seurat</a>
    </li>
    <li>
      <a href="00B-CellRangerInAction.html">Cell Ranger in Action</a>
    </li>
    <li>
      <a href="02-QCandFiltering.html">Secondary QC &amp; Filtering</a>
    </li>
    <li>
      <a href="03-Normalization.html">Normalization</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Day 2
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="04-PCAandIntegration.html">PCA &amp; Integration</a>
    </li>
    <li>
      <a href="05-ProjectionAndClustering.html">Projection &amp; Clustering</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Day 3
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="06-MarkerVisualization.html">Marker identification and visualization</a>
    </li>
    <li>
      <a href="07-CellTypeAnnos.html">Cell type annotation</a>
    </li>
    <li>
      <a href="08-DifferentialExpression.html">Differential expression analysis</a>
    </li>
  </ul>
</li>
<li>
  <a href="workshop_wrap_up.html">Wrap up</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Normalization</h1>
<h4 class="author">UM Bioinformatics Core Workshop Team</h4>
<h4 class="date">2024-12-17</h4>

</div>


<style type="text/css">
body, td {
   font-size: 18px;
}
code.r{
  font-size: 12px;
}
pre {
  font-size: 12px
}

table.fig, th.fig, td.fig {
  border: 1px solid black;
  border-collapse: collapse;
  padding: 15px;
}
</style>
<script>
  addClassKlippyTo("pre.r, pre.markdown, pre.bash");
  addKlippy('right', 'top', 'auto', '1', 'Copy code', 'Copied!');
</script>
<div id="workflow-overview" class="section level1 unlisted unnumbered">
<h1 class="unlisted unnumbered">Workflow Overview</h1>
<p><br/>
<img src="images/wayfinder/wayfinder.png" alt="wayfinder" style="height: 400px;"/>
<br/> <br/></p>
</div>
<div id="introduction" class="section level1">
<h1>Introduction</h1>
<p>After removing low-quality cells from the data, the next task is the
normalization and variance stabilization of the counts for downstream
analysis.</p>
<p>Variation in scRNA-seq data comes from biological sources:</p>
<ul>
<li>Differences in cell type or state</li>
<li>Differences in response to treatment</li>
</ul>
<p>And from technical sources:</p>
<ul>
<li>Fluctuations in cellular RNA content</li>
<li>Efficiency in lysis and reverse transcription</li>
<li>Stochastic sampling during sequencing</li>
</ul>
<br/>
<table class="fig">
<tr class="fig">
<td class="fig">
<img
src="images/graphical_abstracts/graphical_abstract_normalization.png" />
</td>
</tr>
<tr class="fig">
<td class="fig">
A. The counts in the feature barcode matrix are a blend of technical and
biological effects. <br/> B. Technical effects can distort or mask the
biological effects of interest; this can confound downstream analyses.
<br/> C. Seurat can model the technical effects based on overall
patterns in expression across all cells. <br/> D. Once these technical
effects are minimized, the remaining signal is primarily biological
variance.
</td>
</tr>
</table>
<p><br/></p>
<p>It is important to control for the technical sources of variation
without removing the biological sources of variation. A key driver of
technical variation is cellular sequencing depth (that is, the number of
UMIs sequenced per cell). In the figure below, Sample A (left, red
reads) is more deeply sequenced than Sample B (right, green reads). In a
test for differential expression, we want to account for the difference
in sequencing depth to avoid erroneously calling a gene differentially
expressed.</p>
<div class="float">
<img src="images/curriculum/03-Normalization/normalization_depth.png"
alt="Image: Different sequencing depths can erroneously imply differential expression. Source: HBC training materials." />
<div class="figcaption">Image: Different sequencing depths can
erroneously imply differential expression. Source: <a
href="https://hbctraining.github.io/DGE_workshop/lessons/02_DGE_count_normalization.html">HBC
training materials</a>.</div>
</div>
<div id="objectives" class="section level2">
<h2>Objectives</h2>
<ul>
<li>Understand why normalization is needed.</li>
<li>Describe the essence of the <code>SCTransform()</code> method.</li>
<li>Understand normalization options.</li>
<li>Normalize the counts with <code>SCTransform()</code>.</li>
</ul>
<hr />
</div>
</div>
<div id="normalization" class="section level1">
<h1>Normalization</h1>
<p>We will use the <code>SCTransform()</code> function, which uses a
generalized linear model (GLM) framework to account for cell-level
sequencing depth while also stabilizing the variance of the counts.
Let’s get the normalization started, as it takes a little time, and then
we can explain what it’s doing and why it is an improvement on alternate
methods.</p>
<div id="layers-of-a-seurat-object" class="section level2">
<h2>Layers of a Seurat object</h2>
<p>The <code>assays</code> in a Seurat v5 object store data in “layers”.
When we first read the data in, we saw there was only one layer in the
RNA assay. In order to run <code>SCTransform()</code> in Seurat v5, we
have to separate the sample-wise data into layers with the following
command:</p>
<pre class="r"><code>##### Day 1 - Normalization 

# Separate sample data into layers  ---------------------------------------
geo_so[[&#39;RNA&#39;]] = split(geo_so[[&#39;RNA&#39;]], f = geo_so$orig.ident)
geo_so</code></pre>
<pre><code>An object of class Seurat 
26489 features across 35216 samples within 1 assay 
Active assay: RNA (26489 features, 0 variable features)
 12 layers present: counts.HODay0replicate1, counts.HODay0replicate2, counts.HODay0replicate3, counts.HODay0replicate4, counts.HODay21replicate1, counts.HODay21replicate2, counts.HODay21replicate3, counts.HODay21replicate4, counts.HODay7replicate1, counts.HODay7replicate2, counts.HODay7replicate3, counts.HODay7replicate4</code></pre>
<p>We see the 12 layers containing the count data for each of the
samples. Our running schematic has now changed:</p>
<div class="float">
<img src="images/seurat_schematic/Slide4.png"
alt="Image: Schematic splitting the counts layer of the RNA assay by sample." />
<div class="figcaption">Image: Schematic splitting the counts layer of
the RNA assay by sample.</div>
</div>
<p>Next, run <code>SCTransform()</code>:</p>
<pre class="r"><code># Normalize the data with SCTransform  ------------------------------------
geo_so = SCTransform(geo_so)</code></pre>
</div>
<div id="normalization-explained" class="section level2">
<h2>Normalization explained</h2>
<p>To get the idea of what <code>SCTransform()</code> is doing, we’ll
consider a simplified example.</p>
<p><img src="images/curriculum/03-Normalization/Slide1.png" /></p>
<p>Consider two cells that are identical in terms of their type,
expression, etc. Imagine that we put them through the microfluidic
device and then sequenced the RNA content.</p>
<p><img src="images/curriculum/03-Normalization/Slide2.png" /></p>
<p>If we were to plot the total cell UMI count against a particular gene
UMI count, in an ideal world, the points should be directly on top of
each other because the cells had identical expression and everything
done to measure that expression went perfectly.</p>
<p><img src="images/curriculum/03-Normalization/Slide3.png" /></p>
<p>However, we don’t live in a perfect world. We will likely observe the
cells have different total cell UMI counts as well as different gene UMI
counts. This difference, given that the cells were identical, can be
attributed to technical factors. For example, efficiency in lysis or in
reverse transcription, or the stoachastic sampling that occurs during
sequencing.</p>
<p><strong>It is these technical factors that normalization seeks to
correct for, getting us back to the “true” expression
state.</strong></p>
<p><img src="images/curriculum/03-Normalization/Slide4.png" /></p>
<p>Imagine doing this for thousands of cells. We would get a point cloud
like the above. Importantly, that point cloud <strong>has
structure</strong>. There is a relationship between the total cell UMI
count and the gene UMI count for each gene.</p>
<p><img src="images/curriculum/03-Normalization/Slide5.png" /></p>
<p>We could fit a line through the point cloud, where we estimate the
intercept, the slope, and the error.</p>
<p><img src="images/curriculum/03-Normalization/Slide6.png" /></p>
<p>The residuals, or the distance from the point to the line, represents
the expression of the gene less the total cell UMI count influence.</p>
<p><img src="images/curriculum/03-Normalization/Slide7.png" /></p>
<p>In other words, the residuals represent the biological variance, and
the regression removes the technical variance. Note now that the
residuals are the about the same for the two cells.</p>
</div>
<div id="additional-details" class="section level2">
<h2>Additional details</h2>
<p>The full description and justification of the
<code>SCTransform()</code> function are provided in two excellent
papers:</p>
<ul>
<li><p>Hafemeister &amp; Satija, <em>Normalization and variance
stabilization of single-cell RNA-seq data using regularized negative
binomial regression</em>, 2019, Genome Biology (<a
href="https://genomebiology.biomedcentral.com/articles/10.1186/s13059-019-1874-1#availability-of-data-and-materials">link</a>)</p></li>
<li><p>Choudhary &amp; Satija, <em>Comparison and evaluation of
statistical error models for scRNA-seq</em>, 2022, Genome Biology (<a
href="https://genomebiology.biomedcentral.com/articles/10.1186/s13059-021-02584-9">link</a>)</p></li>
</ul>
<p>A benefit of the this framework for normalization is that the model
can include other terms which can account for unwanted technical
variation. One example of this is the percent mitochondrial reads
(<code>percent.mt</code>). The <code>vars.to.regress</code> parameter of
<code>SCTransform()</code> is used for this purpose. See the <a
href="https://satijalab.org/seurat/reference/sctransform">documentation</a>
for details.</p>
</div>
<div id="other-normalizations" class="section level2">
<h2>Other normalizations</h2>
<p>From the Seurat documentation, the <code>SCTransform()</code>
function “replaces <code>NormalizeData()</code>,
<code>ScaleData()</code>, and <code>FindVariableFeatures()</code>”. This
chain of functions is referred to as the “log-normalization procedure”.
You may see these three commands in other vignettes, and even in other
Seurat vignettes (<a
href="https://satijalab.org/seurat/articles/sctransform_vignette">source</a>).
In the two papers referenced above, the authors show how the
log-normalization procedure does not always fully account for cell
sequencing depth and overdispersion. Therefore, we urge you to use this
alternative pipeline with caution.</p>
</div>
</div>
<div id="normalization-continued" class="section level1">
<h1>Normalization, continued</h1>
<p>By now, <code>SCTransform()</code> should have finished running, so
let’s take a look at the result:</p>
<pre class="r"><code># Examine Seurat Object  --------------------------------------------------
geo_so</code></pre>
<pre><code>An object of class Seurat 
46955 features across 31560 samples within 2 assays 
Active assay: SCT (20466 features, 3000 variable features)
 3 layers present: counts, data, scale.data
 1 other assay present: RNA</code></pre>
<p>We now observe changes to our running schematic:</p>
<div class="float">
<img src="images/seurat_schematic/Slide5.png"
alt="Image: Schematic after SCTransform()." />
<div class="figcaption">Image: Schematic after SCTransform().</div>
</div>
<p>Note that we have a new assay, <code>SCT</code> which has three
layers: <code>counts</code>, <code>data</code>, and
<code>scale.data</code>. Also note additional columns in
<code>meta.data</code> and the change of the <code>active.assay</code>
to <code>SCT</code>. <code>SCTransform()</code> has also determined the
common variable features across the cells to be used in our downstream
analysis. Viewed in our running schematic:</p>
</div>
<div id="save-our-progress" class="section level1">
<h1>Save our progress</h1>
<p>Let’s save this normalized form of our Seurat object.</p>
<pre class="r"><code># Save Seurat object  -----------------------------------------------------
saveRDS(geo_so, file = &#39;results/rdata/geo_so_sct_normalized.rds&#39;)</code></pre>
</div>
<div id="summary" class="section level1">
<h1>Summary</h1>
<p>In this section we have run the <code>SCTransform()</code> function
to account for variation in cell sequencing depth and to stabilize the
variance of the counts.</p>
<p>Next steps: PCA and integration</p>
<hr />
<p>These materials have been adapted and extended from materials listed
above. These are open access materials distributed under the terms of
the <a href="http://creativecommons.org/licenses/by/4.0/">Creative
Commons Attribution license (CC BY 4.0)</a>, which permits unrestricted
use, distribution, and reproduction in any medium, provided the original
author and source are credited.</p>
<br/> <br/>
<hr/>
<table style="width:100%;">
<colgroup>
<col width="28%" />
<col width="42%" />
<col width="28%" />
</colgroup>
<thead>
<tr class="header">
<th align="left"><a href="02-QCandFiltering.html">Previous
lesson</a></th>
<th align="center"><a href="#top">Top of this lesson</a></th>
<th align="right"><a href="04-PCAandIntegration.html">Next
lesson</a></th>
</tr>
</thead>
<tbody>
</tbody>
</table>
</div>



</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3,h4",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
