<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="UM Bioinformatics Core" />

<meta name="date" content="2024-02-20" />

<title>PCA and dimensionality reduction</title>

<script src="site_libs/header-attrs-2.25/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/paper.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.13.2/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<script src="site_libs/navigation-1.1/codefolding.js"></script>
<script src="site_libs/navigation-1.1/sourceembed.js"></script>
<link href="site_libs/highlightjs-9.12.0/default.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<!--
Favicon dervied from
https://twemoji.twitter.com/
https://favicon.io/emoji-favicons/dna/
-->
<link rel="apple-touch-icon" sizes="180x180" href="/apple-touch-icon.png">
<link rel="icon" type="image/png" sizes="32x32" href="/favicon-32x32.png">
<link rel="icon" type="image/png" sizes="16x16" href="/favicon-16x16.png">
<link rel="shortcut icon" href="favicon-16x16.png" />
<link rel="manifest" href="/site.webmanifest">
<link rel="mask-icon" href="/safari-pinned-tab.svg" color="#5bbad5">
<meta name="msapplication-TileColor" content="#da532c">
<meta name="theme-color" content="#ffffff">

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>




<style type="text/css">
#rmd-source-code {
  display: none;
}
</style>





<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->
<style type="text/css">
.code-folding-btn { margin-bottom: 4px; }
</style>



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-inverse  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">Intro to scRNA-Seq</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="workshop_intro.html">Intro</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Day 1
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="00A-OrientingOnScRNASeq.html">Orienting on scRNA-Seq</a>
    </li>
    <li>
      <a href="00B-CellRangerInAction.html">Cell Ranger in Action</a>
    </li>
    <li>
      <a href="01-GettingStarted.html">Seurat</a>
    </li>
    <li>
      <a href="02-QCandFiltering.html">Secondary QC &amp; Filtering</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Day 2
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="03-Integration.html">Normalization</a>
    </li>
    <li>
      <a href="04-PCAandDimReduction.html">PCA &amp; Integration</a>
    </li>
    <li>
      <a href="05-ProjectionAndClustering.html">Projection &amp; Clustering</a>
    </li>
    <li>
      <a href="06-MarkerVisualization.html">Marker identification and visualization</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Day 3
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="07-CellTypeAnnos.html">Cell type annotation</a>
    </li>
    <li>
      <a href="08-DifferentialExpression.html">Differential expression analysis</a>
    </li>
    <li>
      <a href="09-IndependentExercise.html">Independent exercise</a>
    </li>
  </ul>
</li>
<li>
  <a href="workshop_wrap_up.html">Wrap up</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">

<div class="btn-group pull-right float-right">
<button type="button" class="btn btn-default btn-xs btn-secondary btn-sm dropdown-toggle" data-toggle="dropdown" data-bs-toggle="dropdown" aria-haspopup="true" aria-expanded="false"><span>Code</span> <span class="caret"></span></button>
<ul class="dropdown-menu dropdown-menu-right" style="min-width: 50px;">
<li><a id="rmd-download-source" href="#">Download Rmd</a></li>
</ul>
</div>



<h1 class="title toc-ignore">PCA and dimensionality reduction</h1>
<h4 class="author">UM Bioinformatics Core</h4>
<h4 class="date">2024-02-20</h4>

</div>


<style type="text/css">
body, td {
   font-size: 18px;
}
code.r{
  font-size: 12px;
}
pre {
  font-size: 12px
}
</style>
<div id="workflow-overview" class="section level1 unlisted unnumbered">
<h1 class="unlisted unnumbered">Workflow Overview</h1>
<p><br/>
<img src="images/wayfinder/wayfinder.png" alt="wayfinder" style="height: 400px;"/>
<br/> <br/></p>
</div>
<div id="introduction" class="section level1">
<h1>Introduction</h1>
<p>One of our ultimate goals is to cluster the cells according to their
expression profiles, and determine what cell types correspond to these
clusters for the purposes of differential expression between clusters
and/or between conditions within clusters. However, for our data we saw
that our 12 samples contain 1K - 3K cells each with measurements across
21K genes, and single cell experiments can generate up to 10K -12K cells
per sample.</p>
<p>This means that even after filtering the data, we are working very
“high dimensional” data. High dimensional data presents some major
challenges and requires something called dimensionality reduction.
Dimension reduction helps us reduce the number of separate dimensions by
grouping correlated features into a single dimension. This makes
downstream analysis steps more computationally feasible.</p>
<p>Similar to the previous sections, the process of selecting
informative dimensions in a dataset is often iterative only a single
value is likely to be reported, even when multiple parameter choices may
have lead to similar results/conclusions.</p>
<p>In this section, we will demonstrate the steps to perform
dimensionality reduction on our data using principal component analysis
(PCA).</p>
<div id="objectives" class="section level2">
<h2>Objectives</h2>
<ul>
<li>Understand why we use PCA for dimensionality reduction</li>
<li>Choose an appropriate number of principal components to cluster our
data</li>
</ul>
<hr />
</div>
</div>
<div id="dimensionality-reduction" class="section level1">
<h1>Dimensionality reduction</h1>
<!--Before this section - Day 1: Starting w/ Seurat, Initial QC, & Batch correction (SCtransform)-->
<!--Instruction Note: Using integrated data here and will compare to unintegrated results in next section, similar to  [here](https://bmcbioinformatics.biomedcentral.com/articles/10.1186/s12859-021-03957-4)?-->
<p>In addition to the “high dimensionality” expected for single-cell
data, we expect the data to be both “sparse” and “noisy”, as many genes
will either not be expressed or not measured in many of the cells - due
to biological variability and practical limitations of both capture and
sequencing depth <a
href="https://ouyanglab.com/singlecell/basic.html#a-gentle-introduction-to-dr">source</a>.</p>
<p>Single-cell data is sometimes described as “zero inflated”, however
work by <a
href="https://www.biorxiv.org/content/10.1101/582064v1.full">Svensson</a>
has challenged that characterization and argued that the higher number
of zeros observed in scRNA-seq compared to bulk RNA-seq is more likely
due to biological variance and lower sequencing saturation than a
technical artifact.</p>
<p>For sparse high dimensional <em>biological</em> data, we expect many
genes to have correlated expression as they would be impacted by the
same biological process and for many genes with either low (more noisy)
or similar expression across the cell population.</p>
<p>So how do we determine what and how many genes to use before
classifying cells in our samples into cell-types/subtypes based on their
expression?</p>
<blockquote>
<h4 id="more-on-dimensionality-reduction"
class="unlisted unnumbered">More on dimensionality reduction</h4>
<p>The <a
href="https://ouyanglab.com/singlecell/basic.html#a-gentle-introduction-to-dr">Ouyang
Lab has a “gentle introduction” section of their materials that</a> that
goes into greater details on dimensionality reduction including how
similar strategies are used in deep learning models.</p>
</blockquote>
<div id="what-is-pca" class="section level2">
<h2>What is PCA?</h2>
<!--- Give some context and background on PCA
- What is PCA
- Why do we use it 
--->
<p>Principal component analysis (PCA) determines axes in
high-dimensional space that capture the most variation.</p>
<div id="pca-example" class="section level3">
<h3>PCA example</h3>
<p>To understand how PCA works, we can consider a smaller dataset
measuring the expression of four genes measured in just two cells, we
could plot the expression of those four genes, with data from one cell
plotted on the x-axis and the second cell plotted on the y-axis.</p>
<p><img
src="images/curriculum/04-PCAandDimReduction/HBC-PCA_2sample_genes.png"
alt="A simple PCA example (from HBC)" />
<!--- Update plot to replace "Sample" with cell in figure and table ---></p>
<p>If we wanted to represent the most variation across the data, we
would draw a diagonal line between gene B and gene C - this would
represent the first principal component. However, this line doesn’t
capture all the variance in this data as the genes also vary above and
below the line, so we could draw another line (at a right angle to the
first) representing the second most variation int the data - which would
represent the second principal component (PC).</p>
<p>However, as we can see in the example below, genes near the end of
each line are those with the greatest influence on the direction and
length of the PC.</p>
<p><img
src="images/curriculum/04-PCAandDimReduction/HBC-PCA_2sample_variation3.png"
alt="PCA gene loadings (from HBC)" />
<!--- Update plot to replace "Sample" with cell in figure ---></p>
<p>We’ll skip the process of calculating the score for each cell per PC,
but after running PCA on our data we expect each cell to have a score
based on the expression of the genes contributing to each PC and each
gene will have a weight or loading. By definition, the PCs capture the
greatest factors of heterogeneity in the data set.</p>
</div>
<div id="why-pca-is-useful-for-single-cell-data" class="section level3">
<h3>Why PCA is useful for single-cell data</h3>
<p>In contrast to bulk RNA-seq, where the majority of the variance in a
given dataset are usually explained by the first and second PC, we
expect that many more PCs are contributing to the overall variance in a
single-cell data set.</p>
<p>However, we can assume that biological processes affect multiple
genes in a coordinated way. Therefore the top PCs likely represent
biological structure rather than random technical or biological noise,
which affects each gene independently and and we can use the top several
PCs to approximate the full data set in downstream analysis[(source)[<a
href="https://bioconductor.org/books/3.12/OSCA/dimensionality-reduction.html#principal-components-analysis"
class="uri">https://bioconductor.org/books/3.12/OSCA/dimensionality-reduction.html#principal-components-analysis</a>]].</p>
<p>This reduction in dimensionality, from 10,000 cells x 19,000 genes
for example, to several order magnitudes less of PCs (often 30-60 PCs
will be considered) and allows us to select PCs that are more likely to
distinguish between biological variation related to the expected cell
types/subtypes and (uninteresting) technical variation.</p>
<blockquote>
<h4 id="more-context-using-pca-for-single-cell-data"
class="unlisted unnumbered">More context using PCA for single-cell
data</h4>
<p>To read more on PCA, please refer to the <a
href="https://hbctraining.github.io/scRNA-seq_online/lessons/05_theory_of_PCA.html">HBC
- Theory of PCA content</a>, from which this section is adapted and the
original source material for that content, specifically- <a
href="https://www.youtube.com/watch?v=_UVHneBUBW0">Josh Starmer’s
StatQuest video</a>. For additional detail, the OSCA chapter on <a
href="https://bioconductor.org/books/3.15/OSCA.basic/dimensionality-reduction.html">Principle
components analysis</a> includes a more descriptive overview of PCA in
the context of single-cell data.</p>
</blockquote>
</div>
<div id="run-pca-on-our-dataset" class="section level3">
<h3>Run PCA on our dataset</h3>
<!--- Add introduction to function, including link to documentation --->
<p>Let’s run PCA on the SCT normalized assay. We will name the reduction
in an informative way to keep them clear for us in the future. In this
case, our data has not yet been integrated, but it has been SCT
normalized. So we will name the reduction
<code>unintegrated.sct.pca</code>. Note that <code>SCTTransform()</code>
returned a set of highly variable genes, and the <code>RunPCA()</code>
function will use this subset to determine the PCs and genes associated
with those PCs.</p>
<pre class="r"><code>geo_so = RunPCA(geo_so, reduction.name = &#39;unintegrated.sct.pca&#39;)</code></pre>
<p>In the console, the first 5 PCs are listed along with the associated
genes. If we look at the <code>geo_so</code> object, we’ll see our
dimension reduction:</p>
<pre class="r"><code>geo_so</code></pre>
<pre><code>An object of class Seurat 
38503 features across 22320 samples within 2 assays 
Active assay: SCT (17129 features, 3000 variable features)
 3 layers present: counts, data, scale.data
 1 other assay present: RNA
 1 dimensional reductions calculated: unintegrated.sct.pca</code></pre>
<!--- Should this dim loading section drawn from [Ho Lab](https://holab-hku.github.io/Fundamental-scRNA/downstream.html#standard-pre-processing-workflow) be included in the main content? Note - would need to run command in script & capture output  --->
<p>We can also visualize both the cell and gene features that define
each principle components using Seurat provided functions.</p>
<p>First, we can print out the top 5 (gene) features per dimension by
accessing that part of the Seurat object:</p>
<pre class="r"><code>print(geo_so[[&quot;pca&quot;]], dims = 1:5, nfeatures = 5) ## update reduction name</code></pre>
<p>We can also highlight genes loaded for each dimension using a
visualization:</p>
<pre class="r"><code>VizDimLoadings(geo_so, dims = 1:2, reduction = &quot;pca&quot;)</code></pre>
<p>We can look at how cells load on the first two principle components,
similarly to how we often look at samples for bulk RNA-seq, with the
<code>DimPlot</code> function:</p>
<pre class="r"><code>DimPlot(geo_so, reduction=&quot;pca&quot;)</code></pre>
<p>Lastly, a common visualization used in Seurat tutorials is
<code>DimHeatmap()</code>, which orders both cells and features
according to their PCA scores and allows us to see some general patterns
in the data:</p>
<pre class="r"><code>DimHeatmap(geo_so, dims=1:3, cells=500, balanced=TRUE)</code></pre>
<blockquote>
<h4 id="how-does-seurat-use-pca-scores" class="unlisted unnumbered">How
does Seurat use PCA scores?</h4>
<p>Per the <a
href="https://holab-hku.github.io/Fundamental-scRNA/downstream.html#perform-linear-dimensional-reduction">Ho
Lab’s materials</a> - “To overcome the extensive technical noise in any
single feature for scRNA-seq data, Seurat clusters cells based on their
PCA scores, with each PC essentially representing a ‘metafeature’ that
combines information across a correlated feature set. The top principal
components therefore represent a robust compression of the dataset.”
<!-- Note, may want to edit/remove above section --></p>
</blockquote>
</div>
</div>
<div
id="choosing-the-number-of-significant-pcs-for-dimensionality-reduction"
class="section level2">
<h2>Choosing the number of significant PCs for dimensionality
reduction</h2>
<!-- Section still needs to be edited more & consider adding a figure -->
<p>Why not use every principal component generated for this dataset?
Using too many PCs risks including uninteresting technical variation and
a single cell itself is not a reasonable functional unit for most
biology.</p>
<p>A good starting point for determining how many PCs to select for your
single-cell analysis is to understand the “resolution” of your
biological question. Is answering your biological question dependent on
identifying rarer cell types or specific subtypes? Or are broader
cell-types more relevant?</p>
<p>For this dataset, we are expected a diversity of cell types and the
cell populations that mediate wound healing or that are part of the
aberrant transition to bone might be more rare in the population so
after evaluating the relative contributions across the first set of PCs,
we might want to consider selecting too many rather than too few PCs to
start.</p>
<!-- Section still needs to be edited
Related - how important is that decision to the downstream impact (e.g. how much does changing the number of PCs change the clustering)?
-->
<div id="visualizing-relative-contributions-of-each-pc"
class="section level3">
<h3>Visualizing relative contributions of each PC</h3>
<p>One way to determine how many PCs to include is by looking at an
elbow plot, which shows the percent variance explained by successive
PCs.</p>
<pre class="r"><code>elbow_plot = ElbowPlot(geo_so, ndims = 50, reduction = &#39;unintegrated.sct.pca&#39;)</code></pre>
<p><img
src="images/curriculum/04-PCAandDimReduction/qc_sct_elbow_plot.png" /></p>
<p>In this plot, we could arbitrarily choose a number along the x-axis
that looks like a sharp change in the variance from one PC to the next,
that is, an elbow. Of course, the choice is not always obvious, and this
plot is no different. We could also try to quantify our choice.</p>
</div>
<div id="using-a-crude-optimization-to-select-a-starting-point"
class="section level3">
<h3>Using a (crude) optimization to select a starting point</h3>
<p>Function to return minimum PCs based on two possible metrics:</p>
<pre class="r"><code># Function to determine optimal PCs after RunPCA()
optimal_pcs = function(so, reduction) {
    # quantitative check for number of PCs to include
    pct = so@reductions[[reduction]]@stdev / sum(so@reductions[[reduction]]@stdev) * 100
    cum = cumsum(pct)
    co1 = which(cum &gt; 90 &amp; pct &lt; 5)[1]
    co2 = sort(which((pct[1:length(pct)-1] - pct[2:length(pct)]) &gt; .1), decreasing = T)[1] + 1
    pcs = min(co1, co2) 
    
    return(pcs)
}</code></pre>
<p>If we apply this function to our data, then we expect a number of PCs
somewhere within the bend of the elbow plot to be returned:</p>
<pre class="r"><code>optimal_pcs(geo_so, &#39;unintegrated.sct.pca&#39;)</code></pre>
<pre><code>23</code></pre>
<p>Again, this number is likely a starting point and may need to revise
depending on the outcome of the downstream steps.</p>
<p>While outside the scope of this workshop, there are community efforts
to develop more sophisticated methods to select like the [chooseR
package] or the option to use clustering trees to evaluate the stability
of this parameter choice.</p>
</div>
</div>
</div>
<div id="integrate-layers" class="section level1">
<h1>Integrate Layers</h1>
<p>Having normalized the data with <code>SCTransform()</code> and
performed the dimension reduction with <code>RunPCA()</code>, we are now
able to run <code>IntegrateLayers()</code>:</p>
<pre class="r"><code>geo_so = IntegrateLayers(
    object = geo_so, 
    method = CCAIntegration, 
    orig.reduction = &#39;unintegrated.sct.pca&#39;,
    normalization.method = &#39;SCT&#39;,
    new.reduction = &#39;integrated.sct.cca&#39;)</code></pre>
<p>Note we have specified the unintegrated reduction
<code>unintegrated.sct.pca</code>, which is what
<code>IntegrateLayers()</code> operates on, along with the
<code>SCT</code> assay. Let’s take a look to see what’s different about
the <code>Seurat</code> object:</p>
<pre class="r"><code>geo_so</code></pre>
<pre><code>An object of class Seurat 
38503 features across 22320 samples within 2 assays 
Active assay: SCT (17129 features, 3000 variable features)
 3 layers present: counts, data, scale.data
 1 other assay present: RNA
 2 dimensional reductions calculated: unintegrated.sct.pca, integrated.sct.cca</code></pre>
<p>Observe that we now have a new reduction,
<code>integrated.sct.cca</code>, which we will use downstream.</p>
<div id="alternate-integration-methods" class="section level2">
<h2>Alternate integration methods</h2>
<p>New to Seurat v5 is the improved <code>IntegrateLayers()</code>
function which makes selecting different integration methods much
easier. The results of each alternative integration method are stored
within the same <code>Seurat</code> object, which makes comparing
downstream effects much easier. <a
href="https://satijalab.org/seurat/articles/seurat5_integration#perform-streamlined-one-line-integrative-analysis">This
Seurat vignette on integrative analysis in v5</a> provides examples of
each integration method. Note, that the code in this vignette assumes
the
<code>NormalizeData() &gt; ScaleData() &gt; FindVariableFeatures()</code>
pipeline was used. So, for example, if we wanted to run the slightly
faster <code>RPCAIntegration</code> method, we would run:</p>
<pre class="r"><code>geo_so = IntegrateLayers(
  object = geo_so, 
  method = RPCAIntegration, 
  orig.reduction = &#39;unintegrated.sct.pca&#39;,
  normalization.method = &#39;SCT&#39;,
  new.reduction = &#39;integrated.sct.rpca&#39;)</code></pre>
<p>Note the specification of <code>normalization.method = 'SCT'</code>
is different than the vignette.</p>
</div>
</div>
<div id="summary" class="section level1">
<h1>Summary</h1>
<p>Now that we have selected a number of principal components that
likely correspond to biological variation of interest, we will use that
number as a parameter in the next section, along with our integrated
results, to generate clusters.</p>
<p>Next steps: Clustering and projection</p>
<hr />
<p>These materials have been adapted and extended from materials listed
above. These are open access materials distributed under the terms of
the <a href="http://creativecommons.org/licenses/by/4.0/">Creative
Commons Attribution license (CC BY 4.0)</a>, which permits unrestricted
use, distribution, and reproduction in any medium, provided the original
author and source are credited.</p>
<br/> <br/>
<hr/>
<table style="width:100%;">
<colgroup>
<col width="28%" />
<col width="42%" />
<col width="28%" />
</colgroup>
<thead>
<tr class="header">
<th align="left"><a href="03-Integration.html">Previous lesson</a></th>
<th align="center"><a href="#top">Top of this lesson</a></th>
<th align="right"><a href="05-ProjectionAndClustering.html">Next
lesson</a></th>
</tr>
</thead>
<tbody>
</tbody>
</table>
</div>

<div id="rmd-source-code">---
title: "PCA and dimensionality reduction"
author: "UM Bioinformatics Core"
date: "`r Sys.Date()`"
output:
        html_document:
            includes:
                in_header: header.html
            theme: paper
            toc: true
            toc_depth: 4
            toc_float: true
            number_sections: false
            fig_caption: true
            markdown: GFM
            code_download: true
---

<style type="text/css">
body, td {
   font-size: 18px;
}
code.r{
  font-size: 12px;
}
pre {
  font-size: 12px
}
</style>

```{r, include = FALSE}
source("../bin/chunk-options.R")
knitr_fig_path("XX")
```
# Workflow Overview {.unlisted .unnumbered}

<br/>
<img src="images/wayfinder/wayfinder.png" alt="wayfinder" style="height: 400px;"/>
<br/>
<br/>



# Introduction 

One of our ultimate goals is to cluster the cells according to their expression profiles, and determine what cell types correspond to these clusters for the purposes of differential expression between clusters and/or between conditions within clusters.  However, for our data we saw that our 12 samples contain 1K - 3K cells each with measurements across 21K genes, and single cell experiments can generate up to 10K -12K cells per sample.

This means that even after filtering the data, we are working very "high dimensional" data. High dimensional data presents some major challenges and requires something called dimensionality reduction. Dimension reduction helps us reduce the number of separate dimensions by grouping correlated features into a single dimension. This makes downstream analysis steps more computationally feasible.

Similar to the previous sections, the process of selecting informative dimensions in a dataset is often iterative only a single value is likely to be reported, even when multiple parameter choices may have lead to similar results/conclusions.  

In this section, we will demonstrate the steps to perform dimensionality reduction on our data using principal component analysis (PCA).


## Objectives

- Understand why we use PCA for dimensionality reduction
- Choose an appropriate number of principal components to cluster our data

---


# Dimensionality reduction

<!--Before this section - Day 1: Starting w/ Seurat, Initial QC, & Batch correction (SCtransform)-->

<!--Instruction Note: Using integrated data here and will compare to unintegrated results in next section, similar to  [here](https://bmcbioinformatics.biomedcentral.com/articles/10.1186/s12859-021-03957-4)?-->

In addition to the "high dimensionality" expected for single-cell data, we expect the data to be both "sparse" and "noisy", as many genes will either not be expressed or not measured in many of the cells -  due to biological variability and practical limitations of both capture and sequencing depth [source](https://ouyanglab.com/singlecell/basic.html#a-gentle-introduction-to-dr). 

Single-cell data is sometimes described as "zero inflated", however work by [Svensson](https://www.biorxiv.org/content/10.1101/582064v1.full) has challenged that characterization and argued that the higher number of zeros observed in scRNA-seq compared to bulk RNA-seq is more likely due to biological variance and lower sequencing saturation than a technical artifact. 

For sparse high dimensional _biological_ data, we expect many genes to have correlated expression as they would be impacted by the same biological process and for many genes with either low (more noisy) or similar expression across the cell population. 

So how do we determine what and how many genes to use before classifying cells in our samples into cell-types/subtypes based on their expression?

> #### More on dimensionality reduction {.unlisted .unnumbered}
> The [Ouyang Lab has a "gentle introduction" section of their materials that ](https://ouyanglab.com/singlecell/basic.html#a-gentle-introduction-to-dr) that goes into greater details on dimensionality reduction including how similar strategies are used in deep learning models. 


## What is PCA?

<!--- Give some context and background on PCA
- What is PCA
- Why do we use it 
--->

Principal component analysis (PCA) determines axes in high-dimensional space that capture the most variation. 

### PCA example

To understand how PCA works, we can consider a smaller dataset measuring the expression of four genes measured in just two cells, we could plot the expression of those four genes, with data from one cell plotted on the x-axis and the second cell plotted on the y-axis.

![A simple PCA example (from HBC)](./images/curriculum/04-PCAandDimReduction/HBC-PCA_2sample_genes.png)
<!--- Update plot to replace "Sample" with cell in figure and table --->

If we wanted to represent the most variation across the data, we would draw a diagonal line between gene B and gene C - this would represent the first principal component. However, this line doesn't capture all the variance in this data as the genes also vary above and below the line, so we could draw another line (at a right angle to the first) representing the second most variation int the data - which would represent the second principal component (PC). 

However, as we can see in the example below, genes near the end of each line are those with the greatest influence on the direction and length of the PC.

![PCA gene loadings (from HBC)](./images/curriculum/04-PCAandDimReduction/HBC-PCA_2sample_variation3.png)
<!--- Update plot to replace "Sample" with cell in figure --->

We'll skip the process of calculating the score for each cell per PC, but after running PCA on our data we expect each cell to have a score based on the expression of the genes contributing to each PC and each gene will have a weight or loading. By definition, the PCs capture the greatest factors of heterogeneity in the data set. 

### Why PCA is useful for single-cell data

In contrast to bulk RNA-seq, where the majority of the variance in a given dataset are usually explained by the first and second PC, we expect that many more PCs are contributing to the overall variance in a single-cell data set. 

However, we can assume that biological processes affect multiple genes in a coordinated way. Therefore the top PCs likely represent biological structure rather than random technical or biological noise, which affects each gene independently and and we can use the top several PCs to approximate the full data set in downstream analysis[(source)[https://bioconductor.org/books/3.12/OSCA/dimensionality-reduction.html#principal-components-analysis]].

This reduction in dimensionality, from 10,000 cells x 19,000 genes for example, to several order magnitudes less of PCs (often 30-60 PCs will be considered) and allows us to select PCs that are more likely to distinguish between biological variation related to the expected cell types/subtypes and (uninteresting) technical variation.

> #### More context using PCA for single-cell data {.unlisted .unnumbered}
>
> To read more on PCA, please refer to the [HBC - Theory of PCA content](https://hbctraining.github.io/scRNA-seq_online/lessons/05_theory_of_PCA.html), from which this section is adapted and the original source material for that content, specifically- [Josh Starmer's StatQuest video](https://www.youtube.com/watch?v=_UVHneBUBW0). For additional detail, the OSCA chapter on [Principle components analysis](https://bioconductor.org/books/3.15/OSCA.basic/dimensionality-reduction.html) includes a more descriptive overview of PCA in the context of single-cell data.

### Run PCA on our dataset

<!--- Add introduction to function, including link to documentation --->

Let's run PCA on the SCT normalized assay. We will name the reduction in an informative way to keep them clear for us in the future. In this case, our data has not yet been integrated, but it has been SCT normalized. So we will name the reduction `unintegrated.sct.pca`. Note that `SCTTransform()` returned a set of highly variable genes, and the `RunPCA()` function will use this subset to determine the PCs and genes associated with those PCs.

```{r, eval=FALSE}
geo_so = RunPCA(geo_so, reduction.name = 'unintegrated.sct.pca')
```

In the console, the first 5 PCs are listed along with the associated genes. If we look at the `geo_so` object, we'll see our dimension reduction:
```{r, eval = FALSE}
geo_so
```
~~~
An object of class Seurat 
38503 features across 22320 samples within 2 assays 
Active assay: SCT (17129 features, 3000 variable features)
 3 layers present: counts, data, scale.data
 1 other assay present: RNA
 1 dimensional reductions calculated: unintegrated.sct.pca
~~~


<!--- Should this dim loading section drawn from [Ho Lab](https://holab-hku.github.io/Fundamental-scRNA/downstream.html#standard-pre-processing-workflow) be included in the main content? Note - would need to run command in script & capture output  --->

We can also visualize both the cell and gene features that define each principle components using Seurat provided functions.

First, we can print out the top 5 (gene) features per dimension by accessing that part of the Seurat object:
```{r, eval=FALSE}
print(geo_so[["pca"]], dims = 1:5, nfeatures = 5) ## update reduction name
```

We can also highlight genes loaded for each dimension using a visualization:
```{r, eval=FALSE}
VizDimLoadings(geo_so, dims = 1:2, reduction = "pca")
```

We can look at how cells load on the first two principle components, similarly to how we often look at samples for bulk RNA-seq, with the `DimPlot` function:
```{r, eval=FALSE}
DimPlot(geo_so, reduction="pca")
```

Lastly, a common visualization used in Seurat tutorials is `DimHeatmap()`, which orders both cells and features according to their PCA scores and allows us to see some general patterns in the data:
```{r, eval=FALSE}
DimHeatmap(geo_so, dims=1:3, cells=500, balanced=TRUE)
```

> #### How does Seurat use PCA scores? {.unlisted .unnumbered}
> 
> Per the [Ho Lab's materials](https://holab-hku.github.io/Fundamental-scRNA/downstream.html#perform-linear-dimensional-reduction) - "To overcome the extensive technical noise in any single feature for scRNA-seq data, Seurat clusters cells based on their PCA scores, with each PC essentially representing a ‘metafeature’ that combines information across a correlated feature set. The top principal components therefore represent a robust compression of the dataset." 
<!-- Note, may want to edit/remove above section -->

## Choosing the number of significant PCs for dimensionality reduction

<!-- Section still needs to be edited more & consider adding a figure --> 
Why not use every principal component generated for this dataset? Using too many PCs risks including uninteresting technical variation and a single cell itself is not a reasonable functional unit for most biology. 

A good starting point for determining how many PCs to select for your single-cell analysis is to understand the "resolution" of your biological question. Is answering your biological question dependent on identifying rarer cell types or specific subtypes? Or are broader cell-types more relevant?

For this dataset, we are expected a diversity of cell types and the cell populations that mediate wound healing or that are part of the aberrant transition to bone might be more rare in the population so after evaluating the relative contributions across the first set of PCs, we might want to consider selecting too many rather than too few PCs to start.


<!-- Section still needs to be edited
Related - how important is that decision to the downstream impact (e.g. how much does changing the number of PCs change the clustering)?
--> 


### Visualizing relative contributions of each PC


One way to determine how many PCs to include is by looking at an elbow plot, which shows the percent variance explained by successive PCs.

```{r, eval=FALSE}
elbow_plot = ElbowPlot(geo_so, ndims = 50, reduction = 'unintegrated.sct.pca')
```

![](./images/curriculum/04-PCAandDimReduction/qc_sct_elbow_plot.png)

In this plot, we could arbitrarily choose a number along the x-axis that looks like a sharp change in the variance from one PC to the next, that is, an elbow. Of course, the choice is not always obvious, and this plot is no different. We could also try to quantify our choice.


### Using a (crude) optimization to select a starting point

Function to return minimum PCs based on two possible metrics:
```{r, eval=FALSE}
# Function to determine optimal PCs after RunPCA()
optimal_pcs = function(so, reduction) {
    # quantitative check for number of PCs to include
    pct = so@reductions[[reduction]]@stdev / sum(so@reductions[[reduction]]@stdev) * 100
    cum = cumsum(pct)
    co1 = which(cum > 90 & pct < 5)[1]
    co2 = sort(which((pct[1:length(pct)-1] - pct[2:length(pct)]) > .1), decreasing = T)[1] + 1
    pcs = min(co1, co2) 
    
    return(pcs)
}
```

If we apply this function to our data, then we expect a number of PCs somewhere within the bend of the elbow plot to be returned:
```{r, eval=FALSE}
optimal_pcs(geo_so, 'unintegrated.sct.pca')
```
~~~~
23
~~~~

Again, this number is likely a starting point and may need to revise depending on the outcome of the downstream steps. 

While outside the scope of this workshop, there are community efforts to develop more sophisticated methods to select like the [chooseR package] or the option to use clustering trees to evaluate the stability of this parameter choice.


# Integrate Layers


Having normalized the data with `SCTransform()` and performed the dimension reduction with `RunPCA()`, we are now able to run `IntegrateLayers()`:


```{r, eval = FALSE}
geo_so = IntegrateLayers(
    object = geo_so, 
    method = CCAIntegration, 
    orig.reduction = 'unintegrated.sct.pca',
    normalization.method = 'SCT',
    new.reduction = 'integrated.sct.cca')
```

Note we have specified the unintegrated reduction `unintegrated.sct.pca`, which is what `IntegrateLayers()` operates on, along with the `SCT` assay. Let's take a look to see what's different about the `Seurat` object:

```{r, eval = FALSE}
geo_so
```
~~~
An object of class Seurat 
38503 features across 22320 samples within 2 assays 
Active assay: SCT (17129 features, 3000 variable features)
 3 layers present: counts, data, scale.data
 1 other assay present: RNA
 2 dimensional reductions calculated: unintegrated.sct.pca, integrated.sct.cca
~~~

Observe that we now have a new reduction, `integrated.sct.cca`, which we will use downstream.

## Alternate integration methods

New to Seurat v5 is the improved `IntegrateLayers()` function which makes selecting different integration methods much easier. The results of each alternative integration method are stored within the same `Seurat` object, which makes comparing downstream effects much easier. [This Seurat vignette on integrative analysis in v5](https://satijalab.org/seurat/articles/seurat5_integration#perform-streamlined-one-line-integrative-analysis) provides examples of each integration method. Note, that the code in this vignette assumes the `NormalizeData() > ScaleData() > FindVariableFeatures()` pipeline was used. So, for example, if we wanted to run the slightly faster `RPCAIntegration` method, we would run:

```{r, eval = FALSE}
geo_so = IntegrateLayers(
  object = geo_so, 
  method = RPCAIntegration, 
  orig.reduction = 'unintegrated.sct.pca',
  normalization.method = 'SCT',
  new.reduction = 'integrated.sct.rpca')
```

Note the specification of `normalization.method = 'SCT'` is different than the vignette.


# Summary

Now that we have selected a number of principal components that likely correspond to biological variation of interest, we will use that number as a parameter in the next section, along with our integrated results, to generate clusters.

Next steps: Clustering and projection




----

These materials have been adapted and extended from materials listed above. These are open access materials distributed under the terms of the [Creative Commons Attribution license (CC BY 4.0)](http://creativecommons.org/licenses/by/4.0/), which permits unrestricted use, distribution, and reproduction in any medium, provided the original author and source are credited.

<br/>
<br/>
<hr/>
| [Previous lesson](03-Integration.html) | [Top of this lesson](#top) | [Next lesson](05-ProjectionAndClustering.html) |
| :--- | :----: | ---: |



</div>


</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->
<script>
$(document).ready(function () {
  window.initializeSourceEmbed("04-PCAandDimReduction.Rmd");
});
</script>

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3,h4",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
