<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="UM Bioinformatics Core" />

<meta name="date" content="2024-10-14" />

<title>Clustering and Projection</title>

<script src="site_libs/header-attrs-2.25/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/paper.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.13.2/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<script src="site_libs/navigation-1.1/codefolding.js"></script>
<script src="site_libs/navigation-1.1/sourceembed.js"></script>
<link href="site_libs/highlightjs-9.12.0/default.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<script src="site_libs/clipboard-1.7.1/clipboard.min.js"></script>
<link href="site_libs/primer-tooltips-1.4.0/build.css" rel="stylesheet" />
<link href="site_libs/klippy-0.0.0.9500/css/klippy.min.css" rel="stylesheet" />
<script src="site_libs/klippy-0.0.0.9500/js/klippy.min.js"></script>
<!--
Favicon dervied from
https://twemoji.twitter.com/
https://favicon.io/emoji-favicons/dna/
-->
<link rel="apple-touch-icon" sizes="180x180" href="apple-touch-icon.png">
<link rel="icon" type="image/png" sizes="32x32" href="favicon-32x32.png">
<link rel="icon" type="image/png" sizes="16x16" href="favicon-16x16.png">
<link rel="shortcut icon" href="favicon-16x16.png" />
<link rel="manifest" href="site.webmanifest">
<link rel="mask-icon" href="safari-pinned-tab.svg" color="#5bbad5">
<meta name="msapplication-TileColor" content="#da532c">
<meta name="theme-color" content="#ffffff">

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>




<style type="text/css">
#rmd-source-code {
  display: none;
}
</style>





<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->
<style type="text/css">
.code-folding-btn { margin-bottom: 4px; }
</style>



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-inverse  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">Intro to scRNA-Seq</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="workshop_intro.html">Intro</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Day 1
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="00A-OrientingOnScRNASeq.html">Orienting on scRNA-Seq</a>
    </li>
    <li>
      <a href="01-GettingStarted.html">Getting Started with Seurat</a>
    </li>
    <li>
      <a href="00B-CellRangerInAction.html">Cell Ranger in Action</a>
    </li>
    <li>
      <a href="02-QCandFiltering.html">Secondary QC &amp; Filtering</a>
    </li>
    <li>
      <a href="03-Normalization.html">Normalization</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Day 2
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="04-PCAandIntegration.html">PCA &amp; Integration</a>
    </li>
    <li>
      <a href="05-ProjectionAndClustering.html">Projection &amp; Clustering</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Day 3
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="06-MarkerVisualization.html">Marker identification and visualization</a>
    </li>
    <li>
      <a href="07-CellTypeAnnos.html">Cell type annotation</a>
    </li>
    <li>
      <a href="08-DifferentialExpression.html">Differential expression analysis</a>
    </li>
  </ul>
</li>
<li>
  <a href="workshop_wrap_up.html">Wrap up</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">

<div class="btn-group pull-right float-right">
<button type="button" class="btn btn-default btn-xs btn-secondary btn-sm dropdown-toggle" data-toggle="dropdown" data-bs-toggle="dropdown" aria-haspopup="true" aria-expanded="false"><span>Code</span> <span class="caret"></span></button>
<ul class="dropdown-menu dropdown-menu-right" style="min-width: 50px;">
<li><a id="rmd-download-source" href="#">Download Rmd</a></li>
</ul>
</div>



<h1 class="title toc-ignore">Clustering and Projection</h1>
<h4 class="author">UM Bioinformatics Core</h4>
<h4 class="date">2024-10-14</h4>

</div>


<style type="text/css">
body, td {
   font-size: 18px;
}
code.r{
  font-size: 12px;
}
pre {
  font-size: 12px
}

table.fig, th.fig, td.fig {
  border: 1px solid black;
  border-collapse: collapse;
  padding: 15px;
}

table.fig, th.fig, td.fig {
  border: 1px solid black;
  border-collapse: collapse;
  padding: 15px;
}
</style>
<script>
  addClassKlippyTo("pre.r, pre.markdown, pre.bash");
  addKlippy('right', 'top', 'auto', '1', 'Copy code', 'Copied!');
</script>
<div id="workflow-overview" class="section level1 unlisted unnumbered">
<h1 class="unlisted unnumbered">Workflow Overview</h1>
<p><br/>
<img src="images/wayfinder/wayfinder.png" alt="wayfinder" style="height: 400px;"/>
<br/> <br/></p>
</div>
<div id="introduction" class="section level1">
<h1>Introduction</h1>
<!--- General goal: to generate clusters that reasonably approximate cell-types or sub-types of interest --->
<p>One of our goals in a single-cell analysis is to generate clusters
that reasonably approximate cell-types or sub-types of interest in our
samples before determining if there are differences in the proportions
of these populations or differences in gene expression within these
populations between experimental conditions.</p>
<table class="fig">
<tr class="fig">
<td class="fig">
<img
src="images/graphical_abstracts/graphical_abstract_cluster_projection.png" />
</td>
</tr>
<tr class="fig">
<td class="fig">
A. The prior step used the filtered, normalized counts as input to a
prinicpal component analysis. The top principle components (PCs) are the
input to the clustering step. <br/> B. Cells are plotted in a
high-dimensional space (where the num dimesnions = num of included PCs,
e.g. 15 dimensions); cells with similar gene expression profiles are
closer to each other and are clustered together. <br/> C. The cells
clusters are projected from the high-dimensional space (e.g. 15
dimensions) down to two-dimensional space for visualization.
</td>
</tr>
</table>
<p><br/></p>
<p>In this section, we will demonstrate how to generate clusters using
Seurat’s graph based clustering approach and visualize those clustering
assignments via a lower-dimensional projection of the full dataset.</p>
<p>Like other steps in our analysis, multiple parameters may need to be
tested and evaluated while we would expect that only the final would be
reported. Clustering is considered part of data exploration so an
iterative approach is reasonable, and often expected (<a
href="https://bioconductor.org/books/3.15/OSCA.basic/clustering.html">source</a>).</p>
<div id="objectives" class="section level2">
<h2>Objectives</h2>
<ul>
<li>Understand the clustering process and input parameters</li>
<li>Generate initial clusters using ``</li>
<li>Visualize our clustering results</li>
</ul>
<hr />
</div>
</div>
<div id="clustering-and-projection" class="section level1">
<h1>Clustering and projection</h1>
<p>Now that we selected a number of PCs that we think are likely to
represent biological variation and integrated our data across
samples/batches, our next task is clustering.</p>
<p>An important aspect of parameter selection for clustering is to
understand the “resolution” of the underlying biology and your
experimental design:</p>
<ul>
<li>Is answering your biological question dependent on identifying rarer
cell types or specific subtypes?<br />
</li>
<li>Or are broader cell-types more relevant to address your biological
question?</li>
</ul>
<p>The OSCA book has a <a
href="https://bioconductor.org/books/3.15/OSCA.basic/clustering.html#overview-1">helpful
analogy comparing clustering to microscopy</a> and points out that
“asking for an unqualified ‘best’ clustering is akin to asking for the
best magnification on a microscope without any context”.</p>
<p>To generate clusters, we will generate “communities” of cells using
the PCs we selected, before choosing a resolution parameter to divide
those communities into discrete clusters.</p>
<!--- Contrast the previous dimensionality reduction versus nearest neighbors clustering and plotting the cells in lower dimensionality with the cluster labels? --->
<div id="clustering" class="section level2">
<h2>Clustering</h2>
<p>Seurat uses a graph-based clustering approach to assign cells to
clusters using a distance metric based on the previously generated PCs,
with improvements based on work by (<a
href="https://academic.oup.com/bioinformatics/article/31/12/1974/214505">Xu
and Su 2015</a>) and CyTOF data (<a
href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4508757/">Levine et
al. 2015</a>) implemented in Seurat v3 and v5 and building on the
initial strategies for droplet-based single-cell technology (<a
href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4481139/">Macosko et
al. 2015</a>) (<a
href="https://satijalab.org/seurat/articles/pbmc3k_tutorial.html">source</a>).
A key aspect of this process is that while the clusters are based on
similarity of expression between the cells, the clustering is based on
the selected PCs and not the full data set.</p>
<center>
<div class="float">
<img
src="images/curriculum/05-ProjectionAndClustering/BioCellGene-NearestNeighborNetworks.png"
alt="Image: kNN example - section on graph based clustering (from Cambridge Bioinformatics course)" />
<div class="figcaption">Image: kNN example - section on graph based
clustering (from Cambridge Bioinformatics course)</div>
</div>
</center>
<p>To briefly summarize, cells are embedded in a k-nearest neighbors
(kNN) graph (illustrated above) based on “the euclidean distance in PCA
space” between the cells and the edge weights between any two cells
(e.g. their “closeness”) is refined based on Jaccard similarity (<a
href="https://hbctraining.github.io/scRNA-seq_online/lessons/07_SC_clustering_cells_SCT.html">source</a>).</p>
<details>
<summary>
<em>Additional context and sources for graph-based clustering</em>
</summary>
<a
href="https://biocellgen-public.svi.edu.au/mig_2019_scrnaseq-workshop/clustering-and-cell-annotation.html">Cambridge
Bioinformatics’ Analysis of single cell RNA-seq data course
materials</a>, the source of the image above, delves into kNN and other
graph based clustering methods in much greater detail, including
outlining possible downsides for these methods. To described kNN, we
have also drawn from the <a
href="https://holab-hku.github.io/Fundamental-scRNA/downstream.html#perform-linear-dimensional-reduction">Ho
Lab’s description of this process for Seurat v3</a> as well as the <a
href="https://hbctraining.github.io/scRNA-seq_online/lessons/07_SC_clustering_cells_SCT.html">HBC
materials on clustering</a> and the <a
href="https://bioconductor.org/books/3.15/OSCA.basic/clustering.html#clustering-graph">OSCA
book’s more general overview of graph based clustering</a>, which also
describes the drawbacks for these methods.
</details>
<p><br></p>
<p>This process is performed with the <code>FindNeighbors()</code> <a
href="https://satijalab.org/seurat/reference/findneighbors">command</a>,
using the number of principal components we selected in the previous
section.</p>
<p>The second step is to iteratively partition the kNN graph into
“cliques” or clusters using the Louvain modularity optimization
algorithm (for the default parameters), with the “granularity” of the
clusters set by a <code>resolution</code> parameter (<a
href="https://satijalab.org/seurat/articles/pbmc3k_tutorial.html">source</a>).</p>
<div class="float">
<img
src="images/curriculum/05-ProjectionAndClustering/HBC-07-k-meansFigure.png"
alt="Image: K-means clustering example (from Cambridge Bioinformatics course)" />
<div class="figcaption">Image: K-means clustering example (from
Cambridge Bioinformatics course)</div>
</div>
<p>We’ll use the <code>FindClusters()</code> <a
href="https://satijalab.org/seurat/reference/findclusters">function</a>,
selecting a resolution of <code>0.4</code> to start, although we could
also add other resolutions at this stage to look at in later steps. See
<a
href="https://link.springer.com/article/10.1140/epjb/e2013-40829-0">Waltman
and Jan van Eck (2013)</a> for the underlying algorithms.</p>
<p>Again, how a “cell type” or “subtype” should be defined for your data
is important to consider in selecting a resolution - we’d start with a
higher resolution for smaller/more rare clusters and a lower resolution
for larger/more general clusters.</p>
<p>Then, when we look at the meta data we should see that cluster labels
have been added for each cell:</p>
<pre class="r"><code># Code block - clustering steps
# Create KNN graph with `FindNeighbors()`
geo_so = FindNeighbors(geo_so, dims = 1:pcs, reduction = &#39;integrated.sct.rpca&#39;)

# generate clusters
geo_so = FindClusters(geo_so, resolution = 0.4, cluster.name = &#39;integrated.sct.rpca.clusters&#39;)</code></pre>
<pre><code>Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck

Number of nodes: 31560
Number of edges: 1032683

Running Louvain algorithm...
Maximum modularity in 10 random starts: 0.9451
Number of communities: 20
Elapsed time: 5 seconds</code></pre>
<pre class="r"><code># look at meta.data to see cluster labels
head(geo_so@meta.data)</code></pre>
<pre><code>                                            orig.ident nCount_RNA nFeature_RNA
HODay0replicate1_AAACCTGAGAGAACAG-1 HO.Day0.replicate1      10234         3226
HODay0replicate1_AAACCTGGTCATGCAT-1 HO.Day0.replicate1       3158         1499
HODay0replicate1_AAACCTGTCAGAGCTT-1 HO.Day0.replicate1      13464         4102
HODay0replicate1_AAACGGGAGAGACTTA-1 HO.Day0.replicate1        577          346
HODay0replicate1_AAACGGGAGGCCCGTT-1 HO.Day0.replicate1       1189          629
HODay0replicate1_AAACGGGCAACTGGCC-1 HO.Day0.replicate1       7726         2602
                                     day  replicate percent.mt nCount_SCT
HODay0replicate1_AAACCTGAGAGAACAG-1 Day0 replicate1   1.240962       6061
HODay0replicate1_AAACCTGGTCATGCAT-1 Day0 replicate1   7.536415       4609
HODay0replicate1_AAACCTGTCAGAGCTT-1 Day0 replicate1   3.112002       5319
HODay0replicate1_AAACGGGAGAGACTTA-1 Day0 replicate1   1.559792       3880
HODay0replicate1_AAACGGGAGGCCCGTT-1 Day0 replicate1   3.700589       4186
HODay0replicate1_AAACGGGCAACTGGCC-1 Day0 replicate1   2.938131       5855
                                    nFeature_SCT integrated.sct.rpca.clusters
HODay0replicate1_AAACCTGAGAGAACAG-1         2867                            1
HODay0replicate1_AAACCTGGTCATGCAT-1         1509                            1
HODay0replicate1_AAACCTGTCAGAGCTT-1         2386                            1
HODay0replicate1_AAACGGGAGAGACTTA-1         1032                            9
HODay0replicate1_AAACGGGAGGCCCGTT-1          915                            0
HODay0replicate1_AAACGGGCAACTGGCC-1         2587                            1
                                    seurat_clusters
HODay0replicate1_AAACCTGAGAGAACAG-1               1
HODay0replicate1_AAACCTGGTCATGCAT-1               1
HODay0replicate1_AAACCTGTCAGAGCTT-1               1
HODay0replicate1_AAACGGGAGAGACTTA-1               9
HODay0replicate1_AAACGGGAGGCCCGTT-1               0
HODay0replicate1_AAACGGGCAACTGGCC-1               1</code></pre>
<p>The result of <code>FindNeighbors()</code> adds graph information to
the <code>graphs</code> slot:</p>
<div class="float">
<img src="images/seurat_schematic/Slide8.png"
alt="Image: Schematic after FindNeighbors()." />
<div class="figcaption">Image: Schematic after FindNeighbors().</div>
</div>
<p>The result of <code>FindClusters()</code> adds two columns to the
<code>meta.data</code> table and changes the <code>active.ident</code>
to the “seurat_clusters” column. In other words, the cells now belong to
clusters rather than to their <code>orig.ident</code>.</p>
<div class="float">
<img src="images/seurat_schematic/Slide9.png"
alt="Image: Schematic after FindClusters()." />
<div class="figcaption">Image: Schematic after FindClusters().</div>
</div>
<p>Generally it’s preferable to err on the side of too many clusters, as
they can be combined manually in later steps. In our experience, this is
another parameter that often needs to be iteratively revised and
reviewed.</p>
<details>
<summary>
<em>Resolution parameter recommendations</em>
</summary>
The <a
href="https://holab-hku.github.io/Fundamental-scRNA/downstream.html#perform-linear-dimensional-reduction">Seurat
clustering tutorial</a> recommends selecting a resolution between 0.4 -
1.2 for datasets of approximately 3k cells, while the <a
href="https://hbctraining.github.io/scRNA-seq_online/lessons/07_SC_clustering_cells_SCT.html">HBC
course</a> recommends 0.4-1.4 for 3k-5k cells. However, in our
experience reasonable starting resolutions can be very dataset
dependent.
</details>
<p><br></p>
</div>
</div>
<div id="cluster-plots" class="section level1">
<h1>Cluster plots</h1>
<p>To visualize the cell clusters, we can use dimensionality reduction
techniques to visualize and explore our large, high-dimensional dataset.
Two popular methods that are supported by Seurat are t-distributed
stochastic neighbor embedding (t-SNE) and Uniform Manifold Approximation
and Projection (UMAP) techniques. These techniques allow us to visualize
our high-dimensional single-cell data in 2D space and see if cells
grouped together within graph-based clusters co-localize in these
representations (<a
href="https://satijalab.org/seurat/articles/pbmc3k_tutorial.html#run-non-linear-dimensional-reduction-umaptsne">source</a>).</p>
<p>While we unfortunately don’t have time to compare and contrast tSNE,
and UMAP, we would highly recommend <a
href="https://pair-code.github.io/understanding-umap/">this blog post
contrasting tSNE and UMAP</a> for illustrative examples. The Seurat
authors additionally caution that while these methods are useful for
data exploration, to avoid drawing biological conclusions solely based
on these visualizations (<a
href="https://satijalab.org/seurat/articles/pbmc3k_tutorial.html#run-non-linear-dimensional-reduction-umaptsne">source</a>).</p>
<p>To start this process, we’ll use the <code>RunUMAP()</code> <a
href="https://satijalab.org/seurat/reference/runumap">function</a> to
calculate the UMAP reduction for our data. Notice how the previous
dimensionality choices carry through the downstream analysis and that
the number of PCs selected in the previous steps are included as an
argument.</p>
<pre class="r"><code># Code block - create UMAP reduction
geo_so = RunUMAP(geo_so, dims = 1:pcs, reduction = &#39;integrated.sct.rpca&#39;, reduction.name = &#39;umap.integrated.sct.rpca&#39;)
geo_so # Notice a third reduction has been added: `umap.integrated.sct.rpca`</code></pre>
<pre><code>An object of class Seurat 
46955 features across 31560 samples within 2 assays 
Active assay: SCT (20466 features, 3000 variable features)
 3 layers present: counts, data, scale.data
 1 other assay present: RNA
 3 dimensional reductions calculated: unintegrated.sct.pca, integrated.sct.rpca, umap.integrated.sct.rpca</code></pre>
<p>The resulting Seurat object now has an additional
<code>umap.integrated.sct.rpca</code> in the <code>reduction</code>
slot:</p>
<div class="float">
<img src="images/seurat_schematic/Slide10.png"
alt="Image: Schematic after RunUMAP()." />
<div class="figcaption">Image: Schematic after RunUMAP().</div>
</div>
</div>
<div id="visualizing-and-evaluating-clustering" class="section level1">
<h1>Visualizing and evaluating clustering</h1>
<!--- How many clusters should I get and how do I adjust the number? --->
<!--- Add example of changing resolution?--->
<p>After we generate the UMAP reduction, we can then visualize the
results using the <code>DimPlot()</code> <a
href="https://satijalab.org/seurat/reference/dimplot">function</a>,
labeling our plot by the auto generated <code>seurat_clusters</code>
that correspond to the most recent clustering results generated.</p>
<p>At this stage, we want to determine if the clusters look fairly well
separated, if they seem to correspond to how cells are grouped in the
UMAP, and if the number of clusters is generally aligned with the
resolution of our biological question. Again, if there are “too many”
clusters that’s not necessarily a problem.</p>
<p>We can also look at the same UMAP labeled by <code>day</code> to
visually inspect if the UMAP structure corresponds to the
<code>day</code>.</p>
<pre class="r"><code># Code block - UMAP cluster visualization 
# cluster ID labels
post_integration_umap_plot_clusters = DimPlot(geo_so, group.by = &#39;seurat_clusters&#39;, label = TRUE, reduction = &#39;umap.integrated.sct.rpca&#39;) + NoLegend()
post_integration_umap_plot_clusters

ggsave(filename = &#39;results/figures/umap_integrated_sct_clusters.png&#39;, plot = post_integration_umap_plot_clusters, width = 6, height = 6, units = &#39;in&#39;)

# clusters with labels, split by condition
post_integration_umap_plot_split_clusters = DimPlot(geo_so, group.by = &#39;seurat_clusters&#39;, split.by = &#39;day&#39;, label = TRUE, reduction = &#39;umap.integrated.sct.rpca&#39;) + NoLegend()
post_integration_umap_plot_split_clusters

ggsave(filename = &#39;results/figures/umap_integrated_sct_split_clusters.png&#39;, plot = post_integration_umap_plot_clusters, width = 14, height = 6, units = &#39;in&#39;)

# UMAP with day labels (note - we added this column to the meta-data yesterday)
post_integration_umap_plot_day = DimPlot(geo_so, group.by = &#39;day&#39;, label = FALSE, reduction = &#39;umap.integrated.sct.rpca&#39;)
post_integration_umap_plot_day

ggsave(filename = &#39;results/figures/umap_integrated_sct_day.png&#39;, plot = post_integration_umap_plot_day, width = 8, height = 6, units = &#39;in&#39;)</code></pre>
<p><img src="images/curriculum/05-ProjectionAndClustering/05-umap_by_clusters-1.png" width="816" style="display: block; margin: auto;" /><img src="images/curriculum/05-ProjectionAndClustering/05-umap_by_clusters-2.png" width="816" style="display: block; margin: auto;" /><img src="images/curriculum/05-ProjectionAndClustering/05-umap_by_clusters-3.png" width="816" style="display: block; margin: auto;" /></p>
<p>Similar to the PCA plots, the <code>day</code> labeled UMAP can tell
us if technical sources of variation might be driving or stratifying the
clusters, which would suggest that the normalization and integration
steps should be revisted before proceeding.</p>
<p>Another approach is to evaluate the number of cells per cluster using
the <code>table()</code> function, split by <code>day</code> or split by
<code>orig.ident</code> to see if the individual samples are driving any
of the UMAP structure:</p>
<pre class="r"><code># Code block - cell count tables
# cells per cluster, split by condition
table(geo_so@meta.data$day, geo_so@meta.data$integrated.sct.rpca.clusters)</code></pre>
<pre><code>       
           0    1    2    3    4    5    6    7    8    9   10   11   12   13
  Day0   131 1176  923   54   68   23  166  169  108  105  265   56   42  136
  Day7  3903 2161  668 2314 2032 2263 1593  924  555  526  393  627  660  403
  Day21 1581 1458  898  110  242   34  461  142  454  363  211  104   65  158
       
          14   15   16   17   18   19
  Day0     3    0  151  137  132   30
  Day7   608  621  301   96   87  120
  Day21   35    2   42  258  190   22</code></pre>
<pre class="r"><code># cells per cluster per sample
table(geo_so@meta.data$orig.ident, geo_so@meta.data$integrated.sct.rpca.clusters)</code></pre>
<pre><code>                     
                         0    1    2    3    4    5    6    7    8    9   10
  HO.Day0.replicate1    37  338  234   22   23    8   51   57   23   24   70
  HO.Day0.replicate2    19  152  182    7    9    8   30   28   24   20   51
  HO.Day0.replicate3    46  423  273    8   22    2   49   51   31   30   73
  HO.Day0.replicate4    29  263  234   17   14    5   36   33   30   31   71
  HO.Day7.replicate1   613  558  242  794  507  819  368  256  106    0   28
  HO.Day7.replicate2  1331  458  149  524  528  477  462  214  165    0  277
  HO.Day7.replicate3  1092  831  182  555  522  607  408  301  187    0   72
  HO.Day7.replicate4   867  314   95  441  475  360  355  153   97  526   16
  HO.Day21.replicate1  418  477  263   37   63   12  141   40  133  103   65
  HO.Day21.replicate2  213  250  221   13   47    6   83   31   67   72   30
  HO.Day21.replicate3  223  223  190   18   49    4   94   26   75   65   32
  HO.Day21.replicate4  727  508  224   42   83   12  143   45  179  123   84
                     
                        11   12   13   14   15   16   17   18   19
  HO.Day0.replicate1    17   11   23    2    0   48   22   33   10
  HO.Day0.replicate2     8    8   30    1    0   12   24   13    3
  HO.Day0.replicate3    16   12   38    0    0   52   41   52    3
  HO.Day0.replicate4    15   11   45    0    0   39   50   34   14
  HO.Day7.replicate1   172  206   93    0  272   70   41   28   42
  HO.Day7.replicate2   187  165  123    0  126   89   18   13   18
  HO.Day7.replicate3   176  163  155    0  160   99   26   36   54
  HO.Day7.replicate4    92  126   32  608   63   43   11   10    6
  HO.Day21.replicate1   37   14   49   13    0   14   65   57    3
  HO.Day21.replicate2   16   14   26    4    0    5   53   28    7
  HO.Day21.replicate3   15   18   36    3    0    9   54   38    6
  HO.Day21.replicate4   36   19   47   15    2   14   86   67    6</code></pre>
<!-- consider adding later 
Add DimPlot for % mitochondria to see if high % are across dataset or limited to one/few clusters (related to what Olivia talked about during Day 1)

# Other approaches for visualizing scRNA-sq data
e.g. code for tSNE visualization 
-->
</div>
<div id="comparing-to-unintegrated-data" class="section level1">
<h1>Comparing to unintegrated data</h1>
<p>If we had proceeded with our filtered data and only normalized our
data without doing any integration, including through the dimensionality
reduction and clustering steps and then labeled the cells with their
sample of origin, then we would see the following for our data:</p>
<!--Add updated example of UMAP / sample labels for unintegrated data - need to check with Raymond regarding how best to do this in place -->
<pre><code>Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck

Number of nodes: 31560
Number of edges: 998753

Running Louvain algorithm...
Maximum modularity in 10 random starts: 0.9506
Number of communities: 22
Elapsed time: 5 seconds</code></pre>
<p><img src="images/curriculum/05-ProjectionAndClustering/05-run_umap_unintegrated-1.png" width="816" height="80%" style="display: block; margin: auto;" /></p>
<p>In the plot at left, we see that while there are distinct clusters,
those clusters seem to stratified by day. This suggests that without
integration, these batch effects could skew the biological variability
in our data. While on the right, we see little stratification within our
clusters which means the integration seems to have removed those batch
effects.</p>
<!-- Edit section to incorporate --->
<details>
<summary>
<strong>Rewind: Pre-integration evaluation clustering and visualization
(code)</strong>
</summary>
<p>Prior to integration, we could follow the same steps we’ve just run
for the integrated to see if the resulting clusters tend to be
determined by sample or condition (in this case, the day):</p>
<pre class="r"><code>geo_so = FindNeighbors(geo_so, dims = 1:pcs, assay = &#39;RNA&#39;, reduction = &#39;unintegrated.sct.pca&#39;, graph.name = c(&#39;RNA_nn&#39;, &#39;RNA_snn&#39;))
geo_so = FindClusters(geo_so, resolution = 0.4, graph.name = &#39;RNA_snn&#39;, cluster.name = &#39;unintegrated.sct.clusters&#39;)
geo_so = RunUMAP(geo_so, dims = 1:pcs, reduction = &#39;unintegrated.sct.pca&#39;, reduction.name = &#39;umap.unintegrated.sct.pca&#39;)</code></pre>
<p>The plots above were generated with:</p>
<pre class="r"><code># Code block - show unintegrated
pre_integration_umap_plot_orig.ident = DimPlot(geo_so, group.by = &#39;orig.ident&#39;, label = FALSE, reduction = &#39;umap.unintegrated.sct.pca&#39;)
ggsave(filename = &#39;results/figures/umap_unintegrated_sct_orig.ident.png&#39;, plot = pre_integration_umap_plot_orig.ident, width = 8, height = 6, units = &#39;in&#39;)

pre_integration_umap_plot_day = DimPlot(geo_so, group.by = &#39;day&#39;, label = FALSE, reduction = &#39;umap.unintegrated.sct.pca&#39;)
ggsave(filename = &#39;results/figures/umap_unintegrated_sct_day.png&#39;, plot = pre_integration_umap_plot_day, width = 8, height = 6, units = &#39;in&#39;)</code></pre>
</details>
<p><br> <br></p>
<details>
<summary>
<strong>Alternative clustering resolutions</strong>
</summary>
<p>While we show a single resolution, we can generate and plot multiple
resolutions iteratively and compare between them before selecting a
clustering result for the next steps:</p>
<pre class="r"><code>resolutions = c(0.4, 0.8)

for(res in resolutions) {
    message(res)

    cluster_column = sprintf(&#39;SCT_snn_res.%s&#39;, res)
    umap_file = sprintf(&#39;results/figures/umap_integrated_sct_%s.png&#39;, res)

    geo_so = FindClusters(geo_so, resolution = res)

    DimPlot(geo_so, group.by = cluster_column, label = TRUE, reduction = &#39;umap.integrated.sct.rpca&#39;) + NoLegend()
    ggsave(filename = umap_file, width = 8, height = 7, units = &#39;in&#39;)
}</code></pre>
<p>In the results, we’ll see multiple resolutions should now be added to
the metadata slot.</p>
<pre class="r"><code>head(geo_so@meta.data)</code></pre>
</details>
<p><br> <br></p>
</div>
<div id="save-our-progress" class="section level1">
<h1>Save our progress</h1>
<p>Before moving on to our next section, we will output our updated
Seurat object to file:</p>
<pre class="r"><code>saveRDS(geo_so, file = &#39;results/rdata/geo_so_sct_clustered.rds&#39;)</code></pre>
</div>
<div id="summary" class="section level1">
<h1>Summary</h1>
<p>In this section we:</p>
<ul>
<li>Generated cluster assignments for our cells using
<code>FindNeighbors()</code> and <code>FindClusters()</code></li>
<li>Evaluated our initial clusters using <code>RunUMAP</code>
dimensional reduction and visualization</li>
</ul>
<p>Next steps: Marker genes</p>
<hr />
<p>These materials have been adapted and extended from materials listed
above. These are open access materials distributed under the terms of
the <a href="http://creativecommons.org/licenses/by/4.0/">Creative
Commons Attribution license (CC BY 4.0)</a>, which permits unrestricted
use, distribution, and reproduction in any medium, provided the original
author and source are credited.</p>
<br/> <br/>
<hr/>
<table style="width:100%;">
<colgroup>
<col width="28%" />
<col width="42%" />
<col width="28%" />
</colgroup>
<thead>
<tr class="header">
<th align="left"><a href="04-PCAandIntegration.html">Previous
lesson</a></th>
<th align="center"><a href="#top">Top of this lesson</a></th>
<th align="right"><a href="06-MarkerVisualization.html">Next
lesson</a></th>
</tr>
</thead>
<tbody>
</tbody>
</table>
</div>

<div id="rmd-source-code">---
title: "Clustering and Projection"
author: "UM Bioinformatics Core"
date: "`r Sys.Date()`"
output:
        html_document:
            includes:
                in_header: header.html
            theme: paper
            toc: true
            toc_depth: 4
            toc_float: true
            number_sections: false
            fig_caption: true
            markdown: GFM
            code_download: true
---

<style type="text/css">
body, td {
   font-size: 18px;
}
code.r{
  font-size: 12px;
}
pre {
  font-size: 12px
}

table.fig, th.fig, td.fig {
  border: 1px solid black;
  border-collapse: collapse;
  padding: 15px;
}

table.fig, th.fig, td.fig {
  border: 1px solid black;
  border-collapse: collapse;
  padding: 15px;
}
</style>

```{r klippy, echo=FALSE, include=TRUE}
klippy::klippy(lang = c("r", "markdown", "bash"), position = c("top", "right"))
```

```{r, include = FALSE}
source("../bin/chunk-options.R")
knitr_fig_path("05-ProjectionAndClustering/05-")
```

# Workflow Overview {.unlisted .unnumbered}

<br/>
<img src="images/wayfinder/wayfinder.png" alt="wayfinder" style="height: 400px;"/>
<br/>
<br/>

# Introduction

<!--- General goal: to generate clusters that reasonably approximate cell-types or sub-types of interest --->
One of our goals in a single-cell analysis is to generate clusters that reasonably approximate cell-types or sub-types of interest in our samples before determining if there are differences in the proportions of these populations or differences in gene expression within these populations between experimental conditions.

<table class='fig'>
<tr class='fig'><td class='fig'>![](images/graphical_abstracts/graphical_abstract_cluster_projection.png)</td></tr>
<tr class='fig'><td class='fig'>A. The prior step used the filtered, normalized counts as input to a prinicpal component analysis. The top principle components (PCs) are the input to the clustering step. <br/>
B. Cells are plotted in a high-dimensional space (where the num dimesnions = num of included PCs, e.g. 15 dimensions); cells with similar gene expression profiles are closer to each other and are clustered together. <br/>
C. The cells clusters are projected from the high-dimensional space (e.g. 15 dimensions) down to two-dimensional space for visualization.
</td></tr>
</table>
<br/>


In this section, we will demonstrate how to generate clusters using Seurat's graph based clustering approach and visualize those clustering assignments via a lower-dimensional projection of the full dataset.

Like other steps in our analysis, multiple parameters may need to be tested and evaluated while we would expect that only the final would be reported. Clustering is considered part of data exploration so an iterative approach is reasonable, and often expected ([source](https://bioconductor.org/books/3.15/OSCA.basic/clustering.html)). 


## Objectives

- Understand the clustering process and input parameters
- Generate initial clusters using `` 
- Visualize our clustering results

---

```{r, read_rds_hidden, echo = FALSE, warning = FALSE, message = FALSE}
if(!exists('geo_so')) {
  library(Seurat)
  library(BPCells)
  library(tidyverse)

  options(future.globals.maxSize = 1e9)

  geo_so = readRDS('results/rdata/geo_so_sct_integrated.rds')
}
```

# Clustering and projection

Now that we selected a number of PCs that we think are likely to represent biological variation and integrated our data across samples/batches, our next task is clustering. 


An important aspect of parameter selection for clustering is to understand the "resolution" of the underlying biology and your experimental design:   

- Is answering your biological question dependent on identifying rarer cell types or specific subtypes?   
- Or are broader cell-types more relevant to address your biological question?   

The OSCA book has a [helpful analogy comparing clustering to microscopy](https://bioconductor.org/books/3.15/OSCA.basic/clustering.html#overview-1) and points out that "asking for an unqualified 'best' clustering is akin to asking for the best magnification on a microscope without any context". 

To generate clusters, we will generate "communities" of cells using the PCs we selected, before choosing a resolution parameter to divide those communities into discrete clusters.

<!--- Contrast the previous dimensionality reduction versus nearest neighbors clustering and plotting the cells in lower dimensionality with the cluster labels? --->

## Clustering

Seurat uses a graph-based clustering approach to assign cells to clusters using a distance metric based on the previously generated PCs, with improvements based on work by ([Xu and Su 2015](https://academic.oup.com/bioinformatics/article/31/12/1974/214505)) and CyTOF data ([Levine et al. 2015](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4508757/)) implemented in Seurat v3 and v5 and building on the initial strategies for droplet-based single-cell technology ([Macosko et al. 2015](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4481139/))  ([source](https://satijalab.org/seurat/articles/pbmc3k_tutorial.html)). A key aspect of this process is that while the clusters are based on similarity of expression between the cells, the clustering is based on the selected PCs and not the full data set. 

<center>
![Image: kNN example - section on graph based clustering (from Cambridge Bioinformatics course)](./images/curriculum/05-ProjectionAndClustering/BioCellGene-NearestNeighborNetworks.png)
</center>

To briefly summarize, cells are embedded in a k-nearest neighbors (kNN) graph (illustrated above) based on "the euclidean distance in PCA space" between the cells and the edge weights between any two cells (e.g. their "closeness") is refined based on Jaccard similarity ([source](https://hbctraining.github.io/scRNA-seq_online/lessons/07_SC_clustering_cells_SCT.html)).

<details>
    <summary>*Additional context and sources for graph-based clustering*</summary>
    [Cambridge Bioinformatics' Analysis of single cell RNA-seq data course materials](https://biocellgen-public.svi.edu.au/mig_2019_scrnaseq-workshop/clustering-and-cell-annotation.html), the source of the image above, delves into kNN and other graph based clustering methods in much greater detail, including outlining possible downsides for these methods. 
     To described kNN, we have also drawn from the [Ho Lab's description of this process for Seurat v3](https://holab-hku.github.io/Fundamental-scRNA/downstream.html#perform-linear-dimensional-reduction) as well as the [HBC materials on clustering](https://hbctraining.github.io/scRNA-seq_online/lessons/07_SC_clustering_cells_SCT.html) and the [OSCA book's more general overview of graph based clustering](https://bioconductor.org/books/3.15/OSCA.basic/clustering.html#clustering-graph), which also describes the drawbacks for these methods.
</details>
<br>

This process is performed with the `FindNeighbors()` [command](https://satijalab.org/seurat/reference/findneighbors), using the number of principal components we selected in the previous section.

The second step is to iteratively partition the kNN graph into "cliques" or clusters using the Louvain modularity optimization algorithm (for the default parameters), with the "granularity" of the clusters set by a `resolution` parameter ([source](https://satijalab.org/seurat/articles/pbmc3k_tutorial.html)).

![Image: K-means clustering example (from Cambridge Bioinformatics course)](./images/curriculum/05-ProjectionAndClustering/HBC-07-k-meansFigure.png)

We'll use the `FindClusters()` [function](https://satijalab.org/seurat/reference/findclusters), selecting a resolution of `0.4` to start, although we could also add other resolutions at this stage to look at in later steps. See [Waltman and Jan van Eck (2013)](https://link.springer.com/article/10.1140/epjb/e2013-40829-0) for the underlying algorithms.

Again, how a “cell type” or “subtype” should be defined for your data is important to consider in selecting a resolution - we'd start with a higher resolution for smaller/more rare clusters and a lower resolution for larger/more general clusters.  

Then, when we look at the meta data we should see that cluster labels have been added for each cell:

```{r, find_neighbors, cache = TRUE, cache.lazy = FALSE, warning = FALSE, message = FALSE}
# Code block - clustering steps
# Create KNN graph with `FindNeighbors()`
geo_so = FindNeighbors(geo_so, dims = 1:pcs, reduction = 'integrated.sct.rpca')

# generate clusters
geo_so = FindClusters(geo_so, resolution = 0.4, cluster.name = 'integrated.sct.rpca.clusters')

# look at meta.data to see cluster labels
head(geo_so@meta.data)
```

The result of `FindNeighbors()` adds graph information to the `graphs` slot:

![Image: Schematic after FindNeighbors().](images/seurat_schematic/Slide8.png)

The result of `FindClusters()` adds two columns to the `meta.data` table and changes the `active.ident` to the "seurat_clusters" column. In other words, the cells now belong to clusters rather than to their `orig.ident`.

![Image: Schematic after FindClusters().](images/seurat_schematic/Slide9.png)

Generally it's preferable to err on the side of too many clusters, as they can be combined manually in later steps. In our experience, this is another parameter that often needs to be iteratively revised and reviewed. 

<details>
    <summary>*Resolution parameter recommendations*</summary>
    The [Seurat clustering tutorial](https://holab-hku.github.io/Fundamental-scRNA/downstream.html#perform-linear-dimensional-reduction) recommends selecting a resolution between 0.4 - 1.2 for datasets of approximately 3k cells, while the [HBC course](https://hbctraining.github.io/scRNA-seq_online/lessons/07_SC_clustering_cells_SCT.html) recommends 0.4-1.4 for 3k-5k cells. However, in our experience reasonable starting resolutions can be very dataset dependent.
</details>
<br>

# Cluster plots 

To visualize the cell clusters, we can use dimensionality reduction techniques to visualize and explore our large, high-dimensional dataset. Two popular methods that are supported by Seurat are t-distributed stochastic neighbor embedding (t-SNE) and Uniform Manifold Approximation and Projection (UMAP) techniques. These techniques allow us to visualize our high-dimensional single-cell data in 2D space and see if cells grouped together within graph-based clusters co-localize in these representations ([source](https://satijalab.org/seurat/articles/pbmc3k_tutorial.html#run-non-linear-dimensional-reduction-umaptsne)).

While we unfortunately don't have time to compare and contrast tSNE, and UMAP, we would highly recommend [this blog post contrasting tSNE and UMAP](https://pair-code.github.io/understanding-umap/) for illustrative examples. The Seurat authors additionally caution that while these methods are useful for data exploration, to avoid drawing biological conclusions solely based on these visualizations ([source](https://satijalab.org/seurat/articles/pbmc3k_tutorial.html#run-non-linear-dimensional-reduction-umaptsne)).

To start this process, we'll use the `RunUMAP()` [function](https://satijalab.org/seurat/reference/runumap) to calculate the UMAP reduction for our data. Notice how the previous dimensionality choices carry through the downstream analysis and that the number of PCs selected in the previous steps are included as an argument.

```{r, run_umap, cache = TRUE, cache.lazy = FALSE, warning = FALSE, message = FALSE}
# Code block - create UMAP reduction
geo_so = RunUMAP(geo_so, dims = 1:pcs, reduction = 'integrated.sct.rpca', reduction.name = 'umap.integrated.sct.rpca')
geo_so # Notice a third reduction has been added: `umap.integrated.sct.rpca`
```

The resulting Seurat object now has an additional `umap.integrated.sct.rpca` in the `reduction` slot:

![Image: Schematic after RunUMAP().](images/seurat_schematic/Slide10.png)

# Visualizing and evaluating clustering

<!--- How many clusters should I get and how do I adjust the number? --->

<!--- Add example of changing resolution?--->

After we generate the UMAP reduction, we can then visualize the results using the `DimPlot()` [function](https://satijalab.org/seurat/reference/dimplot), labeling our plot by the auto generated `seurat_clusters` that correspond to the most recent clustering results generated.

At this stage, we want to determine if the clusters look fairly well separated, if they seem to correspond to how cells are grouped in the UMAP, and if the number of clusters is generally aligned with the resolution of our biological question. Again, if there are "too many" clusters that's not necessarily a problem.

We can also look at the same UMAP labeled by `day` to visually inspect if the UMAP structure corresponds to the `day`.

```{r, umap_by_clusters, fig.show='hold'}
# Code block - UMAP cluster visualization 
# cluster ID labels
post_integration_umap_plot_clusters = DimPlot(geo_so, group.by = 'seurat_clusters', label = TRUE, reduction = 'umap.integrated.sct.rpca') + NoLegend()
post_integration_umap_plot_clusters

ggsave(filename = 'results/figures/umap_integrated_sct_clusters.png', plot = post_integration_umap_plot_clusters, width = 6, height = 6, units = 'in')

# clusters with labels, split by condition
post_integration_umap_plot_split_clusters = DimPlot(geo_so, group.by = 'seurat_clusters', split.by = 'day', label = TRUE, reduction = 'umap.integrated.sct.rpca') + NoLegend()
post_integration_umap_plot_split_clusters

ggsave(filename = 'results/figures/umap_integrated_sct_split_clusters.png', plot = post_integration_umap_plot_clusters, width = 14, height = 6, units = 'in')

# UMAP with day labels (note - we added this column to the meta-data yesterday)
post_integration_umap_plot_day = DimPlot(geo_so, group.by = 'day', label = FALSE, reduction = 'umap.integrated.sct.rpca')
post_integration_umap_plot_day

ggsave(filename = 'results/figures/umap_integrated_sct_day.png', plot = post_integration_umap_plot_day, width = 8, height = 6, units = 'in')
```

Similar to the PCA plots, the `day` labeled UMAP can tell us if technical sources of variation might be driving or stratifying the clusters, which would suggest that the normalization and integration steps should be revisted before proceeding.

Another approach is to evaluate the number of cells per cluster using the `table()` function, split by `day` or split by `orig.ident` to see if the individual samples are driving any of the UMAP structure:

```{r, table_by_day}
# Code block - cell count tables
# cells per cluster, split by condition
table(geo_so@meta.data$day, geo_so@meta.data$integrated.sct.rpca.clusters)

# cells per cluster per sample
table(geo_so@meta.data$orig.ident, geo_so@meta.data$integrated.sct.rpca.clusters)
```

<!-- consider adding later 
Add DimPlot for % mitochondria to see if high % are across dataset or limited to one/few clusters (related to what Olivia talked about during Day 1)

# Other approaches for visualizing scRNA-sq data
e.g. code for tSNE visualization 
-->

# Comparing to unintegrated data

If we had proceeded with our filtered data and only normalized our data without doing any integration, including through the dimensionality reduction and clustering steps and then labeled the cells with their sample of origin, then we would see the following for our data:

<!--Add updated example of UMAP / sample labels for unintegrated data - need to check with Raymond regarding how best to do this in place --> 

```{r, run_umap_unintegrated, echo = FALSE, cache = TRUE, cache.lazy = FALSE, warning = FALSE, message = FALSE, out.height='80%'}
# Create KNN graph with `FindNeighbors()` for unintegrated data
geo_so = FindNeighbors(geo_so, dims = 1:pcs, reduction = 'unintegrated.sct.pca')

# generate clusters for unintegrated data
geo_so = FindClusters(geo_so, resolution = 0.4, cluster.name = 'unintegrated.sct.pca.clusters')

geo_so = RunUMAP(geo_so, dims = 1:pcs, reduction = 'unintegrated.sct.pca', reduction.name = 'umap.unintegrated.sct.pca')

pre_integration_umap_plot_day = DimPlot(geo_so, group.by = 'day', label = FALSE, reduction = 'umap.unintegrated.sct.pca')

print(pre_integration_umap_plot_day + post_integration_umap_plot_day)
```

In the plot at left, we see that while there are distinct clusters, those clusters seem to stratified by day. This suggests that without integration, these batch effects could skew the biological variability in our data. While on the right, we see little stratification within our clusters which means the integration seems to have removed those batch effects.

<!-- Edit section to incorporate --->

<details>
<summary>**Rewind: Pre-integration evaluation clustering and visualization (code)**</summary>
Prior to integration, we could follow the same steps we've just run for the integrated to see if the resulting clusters tend to be determined by sample or condition (in this case, the day):

```{r, eval=FALSE}
geo_so = FindNeighbors(geo_so, dims = 1:pcs, assay = 'RNA', reduction = 'unintegrated.sct.pca', graph.name = c('RNA_nn', 'RNA_snn'))
geo_so = FindClusters(geo_so, resolution = 0.4, graph.name = 'RNA_snn', cluster.name = 'unintegrated.sct.clusters')
geo_so = RunUMAP(geo_so, dims = 1:pcs, reduction = 'unintegrated.sct.pca', reduction.name = 'umap.unintegrated.sct.pca')
```
	
The plots above were generated with:

```{r, eval=FALSE}
# Code block - show unintegrated
pre_integration_umap_plot_orig.ident = DimPlot(geo_so, group.by = 'orig.ident', label = FALSE, reduction = 'umap.unintegrated.sct.pca')
ggsave(filename = 'results/figures/umap_unintegrated_sct_orig.ident.png', plot = pre_integration_umap_plot_orig.ident, width = 8, height = 6, units = 'in')

pre_integration_umap_plot_day = DimPlot(geo_so, group.by = 'day', label = FALSE, reduction = 'umap.unintegrated.sct.pca')
ggsave(filename = 'results/figures/umap_unintegrated_sct_day.png', plot = pre_integration_umap_plot_day, width = 8, height = 6, units = 'in')
```
</details>
<br>
<br>

<details>
<summary>**Alternative clustering resolutions**</summary>
While we show a single resolution, we can generate and plot multiple resolutions iteratively and compare between them before selecting a clustering result for the next steps:

```{r, eval=FALSE}
resolutions = c(0.4, 0.8)

for(res in resolutions) {
    message(res)

    cluster_column = sprintf('SCT_snn_res.%s', res)
    umap_file = sprintf('results/figures/umap_integrated_sct_%s.png', res)

    geo_so = FindClusters(geo_so, resolution = res)

    DimPlot(geo_so, group.by = cluster_column, label = TRUE, reduction = 'umap.integrated.sct.rpca') + NoLegend()
    ggsave(filename = umap_file, width = 8, height = 7, units = 'in')
}
```

In the results, we'll see multiple resolutions should now be added to the metadata slot.

```{r, eval=FALSE}
head(geo_so@meta.data)
```
</details>
<br>
<br>

# Save our progress

Before moving on to our next section, we will output our updated Seurat object to file:

```{r, save_rds_hidden, echo = FALSE}
if(!file.exists('results/rdata/geo_so_sct_clustered.rds')) {
  saveRDS(geo_so, file = 'results/rdata/geo_so_sct_clustered.rds')
}
```

```{r, eval=FALSE}
saveRDS(geo_so, file = 'results/rdata/geo_so_sct_clustered.rds')
```

# Summary

In this section we:

- Generated cluster assignments for our cells using `FindNeighbors()` and `FindClusters()`
- Evaluated our initial clusters using `RunUMAP` dimensional reduction and visualization

Next steps: Marker genes

----

These materials have been adapted and extended from materials listed above. These are open access materials distributed under the terms of the [Creative Commons Attribution license (CC BY 4.0)](http://creativecommons.org/licenses/by/4.0/), which permits unrestricted use, distribution, and reproduction in any medium, provided the original author and source are credited.

<br/>
<br/>
<hr/>
| [Previous lesson](04-PCAandIntegration.html) | [Top of this lesson](#top) | [Next lesson](06-MarkerVisualization.html) |
| :--- | :----: | ---: |
</div>


</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->
<script>
$(document).ready(function () {
  window.initializeSourceEmbed("05-ProjectionAndClustering.Rmd");
});
</script>

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3,h4",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
