<!DOCTYPE html>

<html lang="en" xml:lang="en">

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="UM Bioinformatics Core Workshop Team" />

<meta name="date" content="2025-04-09" />

<title>Clustering and Projection</title>

<script src="site_libs/header-attrs-2.28/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/paper.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.13.2/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/default.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<script src="site_libs/clipboard-1.7.1/clipboard.min.js"></script>
<link href="site_libs/primer-tooltips-1.4.0/build.css" rel="stylesheet" />
<link href="site_libs/klippy-0.0.0.9500/css/klippy.min.css" rel="stylesheet" />
<script src="site_libs/klippy-0.0.0.9500/js/klippy.min.js"></script>
<script src="site_libs/kePrint-0.0.1/kePrint.js"></script>
<link href="site_libs/lightable-0.0.1/lightable.css" rel="stylesheet" />
<!--
Favicon dervied from
https://twemoji.twitter.com/
https://favicon.io/emoji-favicons/dna/
-->
<link rel="apple-touch-icon" sizes="180x180" href="apple-touch-icon.png">
<link rel="icon" type="image/png" sizes="32x32" href="favicon-32x32.png">
<link rel="icon" type="image/png" sizes="16x16" href="favicon-16x16.png">
<link rel="shortcut icon" href="favicon-16x16.png" />
<link rel="manifest" href="site.webmanifest">
<link rel="mask-icon" href="safari-pinned-tab.svg" color="#5bbad5">
<meta name="msapplication-TileColor" content="#da532c">
<meta name="theme-color" content="#ffffff">

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>









<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-inverse  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">Intro to scRNA-Seq</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="workshop_intro.html">Intro</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Day 1
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="00A-OrientingOnScRNASeq.html">Orienting on scRNA-Seq</a>
    </li>
    <li>
      <a href="01-GettingStarted.html">Getting Started with Seurat</a>
    </li>
    <li>
      <a href="00B-CellRangerInAction.html">Cell Ranger in Action</a>
    </li>
    <li>
      <a href="02-QCandFiltering.html">Secondary QC &amp; Filtering</a>
    </li>
    <li>
      <a href="03-Normalization.html">Normalization</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Day 2
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="04-PCAandIntegration.html">PCA &amp; Integration</a>
    </li>
    <li>
      <a href="05-ProjectionAndClustering.html">Projection &amp; Clustering</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Day 3
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="06-MarkerVisualization.html">Marker identification and visualization</a>
    </li>
    <li>
      <a href="07-CellTypeAnnos.html">Cell type annotation</a>
    </li>
    <li>
      <a href="08-DifferentialExpression.html">Differential expression analysis</a>
    </li>
  </ul>
</li>
<li>
  <a href="workshop_wrap_up.html">Wrap up</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Clustering and Projection</h1>
<h4 class="author">UM Bioinformatics Core Workshop Team</h4>
<h4 class="date">2025-04-09</h4>

</div>


<style type="text/css">

body {
   font-size: 18px;
}

code.r{
  font-size: 12px;
}
pre {
  font-size: 12px
}

table.fig, th.fig, td.fig {
  border: 1px solid lightgray;
  border-collapse: collapse;
  padding: 12px;
}

</style>
<script>
  addClassKlippyTo("pre.r, pre.markdown, pre.bash");
  addKlippy('right', 'top', 'auto', '1', 'Copy code', 'Copied!');
</script>
<div id="workflow-overview" class="section level1 unlisted unnumbered">
<h1 class="unlisted unnumbered">Workflow Overview</h1>
<p><br/> <img
src="images/wayfinder/05-ProjectionAndClustering-Wayfinder.png" />
<br/></p>
</div>
<div id="introduction" class="section level1">
<h1>Introduction</h1>
<br/>
<table class="fig">
<tr>
<td class="fig">
<img
src="images/graphical_abstracts/05-ProjectionAndClustering-Abstract.png" />
</td>
</tr>
<tr>
<td class="fig">
Starting with reduced dimensionality data [PCs x cells] for all samples
- cells are organized into networks and then split up to into clusters
with similar expression programs, regardless of experimental condition.
</td>
</tr>
</table>
<p><br/></p>
<!--- General goal: to generate clusters that reasonably approximate cell-types or sub-types of interest --->
<p>Before making any comparisons between experimental conditions, its
important to identify cell-types or sub-types present across all samples
and unsupervised clustering is a reasonable starting point to accomplish
this. <br/></p>
<div id="objectives" class="section level2">
<h2>Objectives</h2>
<ul>
<li>Understand the clustering process and input parameters</li>
<li>Generate initial clusters using <code>FindNeighbors()</code> and
<code>FindClusters()</code></li>
<li>Visualize our clustering results with <code>DimPlot()</code></li>
</ul>
<p><br/> Like other steps in our analysis, multiple parameters may need
to be tested and evaluated while we would expect that only the final
would be reported. Clustering is considered part of data exploration so
an iterative approach is common (<a
href="https://bioconductor.org/books/3.15/OSCA.basic/clustering.html">source</a>).
<br/></p>
<hr />
</div>
</div>
<div id="clustering-and-projection" class="section level1">
<h1>Clustering and projection</h1>
<p>Now that we selected a number of PCs that we think are likely to
represent biological variation and integrated our data across
samples/batches, our next task is clustering.</p>
<p>An important aspect of parameter selection for clustering is to
understand the “resolution” of the underlying biology and your
experimental design:</p>
<ul>
<li>Is answering your biological question dependent on identifying rarer
cell types or specific subtypes?<br />
</li>
<li>Or are broader cell-types more relevant to address your biological
question?</li>
</ul>
<p>The OSCA book has a <a
href="https://bioconductor.org/books/3.15/OSCA.basic/clustering.html#overview-1">helpful
analogy comparing clustering to microscopy</a> and points out that
“asking for an unqualified ‘best’ clustering is akin to asking for the
best magnification on a microscope without any context”.</p>
<p>To generate clusters, we will generate “communities” of cells using
the PCs we selected, before choosing a resolution parameter to divide
those communities into discrete clusters.</p>
<!--- Contrast the previous dimensionality reduction versus nearest neighbors clustering and plotting the cells in lower dimensionality with the cluster labels? --->
<div id="clustering" class="section level2">
<h2>Clustering</h2>
<p>Seurat uses a graph-based clustering approach to assign cells to
clusters using a distance metric based on the previously generated PCs,
with improvements based on work by (<a
href="https://academic.oup.com/bioinformatics/article/31/12/1974/214505">Xu
and Su 2015</a>) and CyTOF data (<a
href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4508757/">Levine et
al. 2015</a>) implemented in Seurat v3 and v5 and building on the
initial strategies for droplet-based single-cell technology (<a
href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4481139/">Macosko et
al. 2015</a>) (<a
href="https://satijalab.org/seurat/articles/pbmc3k_tutorial.html">source</a>).
A key aspect of this process is that while the clusters are based on
similarity of expression between the cells, the clustering is based on
the selected PCs and not the full data set.</p>
<center>
<div class="float">
<img
src="images/curriculum/05-ProjectionAndClustering/BioCellGene-NearestNeighborNetworks.png"
alt="Image: kNN example - section on graph based clustering (from Cambridge Bioinformatics course)" />
<div class="figcaption">Image: kNN example - section on graph based
clustering (from Cambridge Bioinformatics course)</div>
</div>
</center>
<p>To briefly summarize, cells are embedded in a k-nearest neighbors
(kNN) graph (illustrated above) based on “the euclidean distance in PCA
space” between the cells and the edge weights between any two cells
(e.g. their “closeness”) is refined based on Jaccard similarity (<a
href="https://hbctraining.github.io/scRNA-seq_online/lessons/07_SC_clustering_cells_SCT.html">source</a>).</p>
<details>
<summary>
<em>Additional context and sources for graph-based clustering</em>
</summary>
<a
href="https://biocellgen-public.svi.edu.au/mig_2019_scrnaseq-workshop/clustering-and-cell-annotation.html">Cambridge
Bioinformatics’ Analysis of single cell RNA-seq data course
materials</a>, the source of the image above, delves into kNN and other
graph based clustering methods in much greater detail, including
outlining possible downsides for these methods. To described kNN, we
have also drawn from the <a
href="https://holab-hku.github.io/Fundamental-scRNA/downstream.html#perform-linear-dimensional-reduction">Ho
Lab’s description of this process for Seurat v3</a> as well as the <a
href="https://hbctraining.github.io/scRNA-seq_online/lessons/07_SC_clustering_cells_SCT.html">HBC
materials on clustering</a> and the <a
href="https://bioconductor.org/books/3.15/OSCA.basic/clustering.html#clustering-graph">OSCA
book’s more general overview of graph based clustering</a>, which also
describes the drawbacks for these methods.
</details>
<p><br></p>
<p>This process is performed with the <code>FindNeighbors()</code> <a
href="https://satijalab.org/seurat/reference/findneighbors">command</a>,
using the number of principal components we selected in the previous
section.</p>
<p>The second step is to iteratively partition the kNN graph into
“cliques” or clusters using the Louvain modularity optimization
algorithm (for the default parameters), with the “granularity” of the
clusters set by a <code>resolution</code> parameter (<a
href="https://satijalab.org/seurat/articles/pbmc3k_tutorial.html">source</a>).</p>
<div class="float">
<img
src="images/curriculum/05-ProjectionAndClustering/HBC-07-k-meansFigure.png"
alt="Image: K-means clustering example (from Cambridge Bioinformatics course)" />
<div class="figcaption">Image: K-means clustering example (from
Cambridge Bioinformatics course)</div>
</div>
<p>We’ll use the <code>FindClusters()</code> <a
href="https://satijalab.org/seurat/reference/findclusters">function</a>,
selecting a resolution of <code>0.4</code> to start, although we could
also add other resolutions at this stage to look at in later steps. See
<a
href="https://link.springer.com/article/10.1140/epjb/e2013-40829-0">Waltman
and Jan van Eck (2013)</a> for the underlying algorithms.</p>
<p>Again, how a “cell type” or “subtype” should be defined for your data
is important to consider in selecting a resolution - we’d start with a
higher resolution for smaller/more rare clusters and a lower resolution
for larger/more general clusters.</p>
<p>Then, when we look at the meta data we should see that cluster labels
have been added for each cell:</p>
<pre class="r"><code># Cluster PCAs ------------------------------------------------------------
# Create KNN graph with `FindNeighbors()`
geo_so = FindNeighbors(geo_so, dims = 1:pcs, reduction = &#39;integrated.sct.rpca&#39;)

# generate clusters
geo_so = FindClusters(geo_so,
                      resolution = 0.4,
                      cluster.name = &#39;integrated.sct.rpca.clusters&#39;)

# look at meta.data to see cluster labels
head(geo_so@meta.data)</code></pre>
<pre><code>Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck

Number of nodes: 31559
Number of edges: 1024599

Running Louvain algorithm...
Maximum modularity in 10 random starts: 0.9441
Number of communities: 18
Elapsed time: 5 seconds</code></pre>
<div
style="border: 1px solid #ddd; padding: 0px; overflow-y: scroll; height:100%; overflow-x: scroll; width:1000px; ">
<table class="table table-condensed" style="font-size: 12px; font-family: &quot;Arial Narrow&quot;, arial, helvetica, sans-serif; margin-left: auto; margin-right: auto;">
<thead>
<tr>
<th style="text-align:left;position: sticky; top:0; background-color: #FFFFFF;">
</th>
<th style="text-align:left;position: sticky; top:0; background-color: #FFFFFF;">
orig.ident
</th>
<th style="text-align:right;position: sticky; top:0; background-color: #FFFFFF;">
nCount_RNA
</th>
<th style="text-align:right;position: sticky; top:0; background-color: #FFFFFF;">
nFeature_RNA
</th>
<th style="text-align:left;position: sticky; top:0; background-color: #FFFFFF;">
condition
</th>
<th style="text-align:left;position: sticky; top:0; background-color: #FFFFFF;">
time
</th>
<th style="text-align:left;position: sticky; top:0; background-color: #FFFFFF;">
replicate
</th>
<th style="text-align:left;position: sticky; top:0; background-color: #FFFFFF;">
day
</th>
<th style="text-align:right;position: sticky; top:0; background-color: #FFFFFF;">
percent.mt
</th>
<th style="text-align:right;position: sticky; top:0; background-color: #FFFFFF;">
nCount_SCT
</th>
<th style="text-align:right;position: sticky; top:0; background-color: #FFFFFF;">
nFeature_SCT
</th>
<th style="text-align:left;position: sticky; top:0; background-color: #FFFFFF;">
integrated.sct.rpca.clusters
</th>
<th style="text-align:left;position: sticky; top:0; background-color: #FFFFFF;">
seurat_clusters
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
HODay0replicate1_AAACCTGAGAGAACAG-1
</td>
<td style="text-align:left;">
HODay0replicate1
</td>
<td style="text-align:right;">
10234
</td>
<td style="text-align:right;">
3226
</td>
<td style="text-align:left;">
HO
</td>
<td style="text-align:left;">
Day0
</td>
<td style="text-align:left;">
replicate1
</td>
<td style="text-align:left;">
Day0
</td>
<td style="text-align:right;">
1.240962
</td>
<td style="text-align:right;">
6062
</td>
<td style="text-align:right;">
2867
</td>
<td style="text-align:left;">
0
</td>
<td style="text-align:left;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
HODay0replicate1_AAACCTGGTCATGCAT-1
</td>
<td style="text-align:left;">
HODay0replicate1
</td>
<td style="text-align:right;">
3158
</td>
<td style="text-align:right;">
1499
</td>
<td style="text-align:left;">
HO
</td>
<td style="text-align:left;">
Day0
</td>
<td style="text-align:left;">
replicate1
</td>
<td style="text-align:left;">
Day0
</td>
<td style="text-align:right;">
7.536416
</td>
<td style="text-align:right;">
4607
</td>
<td style="text-align:right;">
1509
</td>
<td style="text-align:left;">
0
</td>
<td style="text-align:left;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
HODay0replicate1_AAACCTGTCAGAGCTT-1
</td>
<td style="text-align:left;">
HODay0replicate1
</td>
<td style="text-align:right;">
13464
</td>
<td style="text-align:right;">
4102
</td>
<td style="text-align:left;">
HO
</td>
<td style="text-align:left;">
Day0
</td>
<td style="text-align:left;">
replicate1
</td>
<td style="text-align:left;">
Day0
</td>
<td style="text-align:right;">
3.112002
</td>
<td style="text-align:right;">
5314
</td>
<td style="text-align:right;">
2370
</td>
<td style="text-align:left;">
0
</td>
<td style="text-align:left;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
HODay0replicate1_AAACGGGAGAGACTTA-1
</td>
<td style="text-align:left;">
HODay0replicate1
</td>
<td style="text-align:right;">
577
</td>
<td style="text-align:right;">
346
</td>
<td style="text-align:left;">
HO
</td>
<td style="text-align:left;">
Day0
</td>
<td style="text-align:left;">
replicate1
</td>
<td style="text-align:left;">
Day0
</td>
<td style="text-align:right;">
1.559792
</td>
<td style="text-align:right;">
3877
</td>
<td style="text-align:right;">
1031
</td>
<td style="text-align:left;">
11
</td>
<td style="text-align:left;">
11
</td>
</tr>
<tr>
<td style="text-align:left;">
HODay0replicate1_AAACGGGAGGCCCGTT-1
</td>
<td style="text-align:left;">
HODay0replicate1
</td>
<td style="text-align:right;">
1189
</td>
<td style="text-align:right;">
629
</td>
<td style="text-align:left;">
HO
</td>
<td style="text-align:left;">
Day0
</td>
<td style="text-align:left;">
replicate1
</td>
<td style="text-align:left;">
Day0
</td>
<td style="text-align:right;">
3.700589
</td>
<td style="text-align:right;">
4166
</td>
<td style="text-align:right;">
915
</td>
<td style="text-align:left;">
0
</td>
<td style="text-align:left;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
HODay0replicate1_AAACGGGCAACTGGCC-1
</td>
<td style="text-align:left;">
HODay0replicate1
</td>
<td style="text-align:right;">
7726
</td>
<td style="text-align:right;">
2602
</td>
<td style="text-align:left;">
HO
</td>
<td style="text-align:left;">
Day0
</td>
<td style="text-align:left;">
replicate1
</td>
<td style="text-align:left;">
Day0
</td>
<td style="text-align:right;">
2.938131
</td>
<td style="text-align:right;">
5865
</td>
<td style="text-align:right;">
2588
</td>
<td style="text-align:left;">
0
</td>
<td style="text-align:left;">
0
</td>
</tr>
</tbody>
</table>
</div>
<p>The result of <code>FindNeighbors()</code> adds graph information to
the <code>graphs</code> slot:</p>
<div class="float">
<img src="images/seurat_schematic/Slide8.png"
alt="Image: Schematic after FindNeighbors()." />
<div class="figcaption">Image: Schematic after FindNeighbors().</div>
</div>
<p><br></p>
<p>The result of <code>FindClusters()</code> adds two columns to the
<code>meta.data</code> table and changes the <code>active.ident</code>
to the “seurat_clusters” column. In other words, the cells now belong to
clusters rather than to their <code>orig.ident</code>.</p>
<p><img src="images/seurat_schematic/Slide9.png"
alt="Image: Schematic after FindClusters()." /> <br></p>
<p>Generally it’s preferable to err on the side of too many clusters, as
they can be combined manually in later steps. In our experience, this is
another parameter that often needs to be iteratively revised and
reviewed.</p>
<div id="resolution-parameters-recommendations" class="section level3">
<h3>Resolution parameters recommendations</h3>
<details>
<summary>
<em>More details on choosing a clustering resolution</em>
</summary>
The <a
href="https://holab-hku.github.io/Fundamental-scRNA/downstream.html#perform-linear-dimensional-reduction">Seurat
clustering tutorial</a> recommends selecting a resolution between 0.4 -
1.2 for datasets of approximately 3k cells, while the <a
href="https://hbctraining.github.io/scRNA-seq_online/lessons/07_SC_clustering_cells_SCT.html">HBC
course</a> recommends 0.4-1.4 for 3k-5k cells. However, in our
experience reasonable starting resolutions can be very dataset
dependent.
</details>
<p><br></p>
</div>
</div>
</div>
<div id="cluster-plots" class="section level1">
<h1>Cluster plots</h1>
<p>To visualize the cell clusters, we can use dimensionality reduction
techniques to visualize and explore our large, high-dimensional dataset.
Two popular methods that are supported by Seurat are t-distributed
stochastic neighbor embedding (t-SNE) and Uniform Manifold Approximation
and Projection (UMAP) techniques. These techniques allow us to visualize
our high-dimensional single-cell data in 2D space and see if cells
grouped together within graph-based clusters co-localize in these
representations (<a
href="https://satijalab.org/seurat/articles/pbmc3k_tutorial.html#run-non-linear-dimensional-reduction-umaptsne">source</a>).</p>
<p>While we unfortunately don’t have time to compare and contrast tSNE,
and UMAP, we would highly recommend <a
href="https://pair-code.github.io/understanding-umap/">this blog post
contrasting tSNE and UMAP</a> for illustrative examples. The Seurat
authors additionally caution that while these methods are useful for
data exploration, to avoid drawing biological conclusions solely based
on these visualizations (<a
href="https://satijalab.org/seurat/articles/pbmc3k_tutorial.html#run-non-linear-dimensional-reduction-umaptsne">source</a>).</p>
<p>To start this process, we’ll use the <code>RunUMAP()</code> <a
href="https://satijalab.org/seurat/reference/runumap">function</a> to
calculate the UMAP reduction for our data. Notice how the previous
dimensionality choices carry through the downstream analysis and that
the number of PCs selected in the previous steps are included as an
argument.</p>
<pre class="r"><code># Create UMAP reduction ---------------------------------------------------
geo_so = RunUMAP(geo_so, dims = 1:pcs, reduction = &#39;integrated.sct.rpca&#39;, reduction.name = &#39;umap.integrated.sct.rpca&#39;)

# Note a third reduction has been added: `umap.integrated.sct.rpca`
geo_so </code></pre>
<pre><code>An object of class Seurat 
46957 features across 31559 samples within 2 assays 
Active assay: SCT (20468 features, 3000 variable features)
 3 layers present: counts, data, scale.data
 1 other assay present: RNA
 3 dimensional reductions calculated: unintegrated.sct.pca, integrated.sct.rpca, umap.integrated.sct.rpca</code></pre>
<p>Refering back to the schematic, the resulting Seurat object now has
an additional <code>umap.integrated.sct.rpca</code> in the
<code>reduction</code> slot:</p>
<div class="float">
<img src="images/seurat_schematic/Slide10.png"
alt="Image: Schematic after RunUMAP()." />
<div class="figcaption">Image: Schematic after RunUMAP().</div>
</div>
<p><br/> <br/></p>
</div>
<div id="visualizing-and-evaluating-clustering" class="section level1">
<h1>Visualizing and evaluating clustering</h1>
<!--- How many clusters should I get and how do I adjust the number? --->
<!--- Add example of changing resolution?--->
<p>After we generate the UMAP reduction, we can then visualize the
results using the <code>DimPlot()</code> <a
href="https://satijalab.org/seurat/reference/dimplot">function</a>,
labeling our plot by the auto generated <code>seurat_clusters</code>
that correspond to the most recent clustering results generated.</p>
<p>At this stage, we want to determine if the clusters look fairly well
separated, if they seem to correspond to how cells are grouped in the
UMAP, and if the number of clusters is generally aligned with the
resolution of our biological question. Again, if there are “too many”
clusters that’s not necessarily a problem.</p>
<p>We can also look at the same UMAP labeled by <code>day</code> to
visually inspect if the UMAP structure corresponds to the
<code>day</code>.</p>
<pre class="r"><code># Visualize UMAP cluster 1 --------------------------------------------------
# cluster ID labels
post_integration_umap_plot_clusters = DimPlot(geo_so, group.by = &#39;seurat_clusters&#39;, label = TRUE, reduction = &#39;umap.integrated.sct.rpca&#39;) + NoLegend()
post_integration_umap_plot_clusters

ggsave(filename = &#39;results/figures/umap_integrated_sct_clusters.png&#39;, plot = post_integration_umap_plot_clusters, width = 6, height = 6, units = &#39;in&#39;)</code></pre>
<p><img src="images/curriculum/05-ProjectionAndClustering/05-umap_by_clusters-1.png" width="576" style="display: block; margin: auto;" /></p>
<pre class="r"><code># Visualize UMAP cluster 2 --------------------------------------------------
# clusters with labels, split by condition
post_integration_umap_plot_split_clusters = DimPlot(geo_so, group.by = &#39;seurat_clusters&#39;, split.by = &#39;time&#39;, label = TRUE, reduction = &#39;umap.integrated.sct.rpca&#39;) + NoLegend()
post_integration_umap_plot_split_clusters

ggsave(filename = &#39;results/figures/umap_integrated_sct_split_clusters.png&#39;, plot = post_integration_umap_plot_clusters, width = 14, height = 6, units = &#39;in&#39;)</code></pre>
<p><img src="images/curriculum/05-ProjectionAndClustering/05-umap_by_splitclusters-1.png" width="816" style="display: block; margin: auto;" /></p>
<pre class="r"><code># Visualize UMAP cluster 3 --------------------------------------------------
# UMAP with day labels (note - we added this column to the meta-data yesterday)
post_integration_umap_plot_day = DimPlot(geo_so, group.by = &#39;day&#39;, label = FALSE, reduction = &#39;umap.integrated.sct.rpca&#39;)
post_integration_umap_plot_day

ggsave(filename = &#39;results/figures/umap_integrated_sct_day.png&#39;, plot = post_integration_umap_plot_day, width = 8, height = 6, units = &#39;in&#39;)</code></pre>
<p><img src="images/curriculum/05-ProjectionAndClustering/05-umap_by_dayclusters-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>Similar to the PCA plots, the <code>day</code> labeled UMAP can tell
us if technical sources of variation might be driving or stratifying the
clusters, which would suggest that the normalization and integration
steps should be revisted before proceeding.</p>
<p>Another approach is to evaluate the number of cells per cluster using
the <code>table()</code> function, split by <code>day</code> or split by
<code>orig.ident</code> to see if the individual samples are driving any
of the UMAP structure:</p>
<pre class="r"><code>clusters_counts_condition &lt;- geo_so@meta.data %&gt;% 
  select(time, integrated.sct.rpca.clusters) %&gt;%
  summarise(counts = n(), .by = c(time, integrated.sct.rpca.clusters)) %&gt;% 
  pivot_wider(names_from = integrated.sct.rpca.clusters, values_from = counts, names_sort=TRUE) %&gt;%
  bind_rows(summarise_all(., ~if(is.numeric(.)) sum(.) else &quot;TOTAL&quot;))
clusters_counts_condition 


# cells per cluster per sample
clusters_counts_sample &lt;- geo_so@meta.data %&gt;% 
  select(orig.ident, integrated.sct.rpca.clusters) %&gt;%
  summarise(counts = n(), .by = c(orig.ident, integrated.sct.rpca.clusters)) %&gt;% 
  pivot_wider(names_from = integrated.sct.rpca.clusters, values_from = counts, names_sort=TRUE) %&gt;%
  bind_rows(summarise_all(., ~if(is.numeric(.)) sum(., na.rm = TRUE) else &quot;TOTAL&quot;))
clusters_counts_sample</code></pre>
<div
style="border: 1px solid #ddd; padding: 0px; overflow-y: scroll; height:100%; overflow-x: scroll; width:825px; ">
<table class="table table-condensed" style="font-size: 12px; font-family: &quot;Arial Narrow&quot;, arial, helvetica, sans-serif; margin-left: auto; margin-right: auto;">
<thead>
<tr>
<th style="text-align:left;position: sticky; top:0; background-color: #FFFFFF;">
time
</th>
<th style="text-align:right;position: sticky; top:0; background-color: #FFFFFF;">
0
</th>
<th style="text-align:right;position: sticky; top:0; background-color: #FFFFFF;">
1
</th>
<th style="text-align:right;position: sticky; top:0; background-color: #FFFFFF;">
2
</th>
<th style="text-align:right;position: sticky; top:0; background-color: #FFFFFF;">
3
</th>
<th style="text-align:right;position: sticky; top:0; background-color: #FFFFFF;">
4
</th>
<th style="text-align:right;position: sticky; top:0; background-color: #FFFFFF;">
5
</th>
<th style="text-align:right;position: sticky; top:0; background-color: #FFFFFF;">
6
</th>
<th style="text-align:right;position: sticky; top:0; background-color: #FFFFFF;">
7
</th>
<th style="text-align:right;position: sticky; top:0; background-color: #FFFFFF;">
8
</th>
<th style="text-align:right;position: sticky; top:0; background-color: #FFFFFF;">
9
</th>
<th style="text-align:right;position: sticky; top:0; background-color: #FFFFFF;">
10
</th>
<th style="text-align:right;position: sticky; top:0; background-color: #FFFFFF;">
11
</th>
<th style="text-align:right;position: sticky; top:0; background-color: #FFFFFF;">
12
</th>
<th style="text-align:right;position: sticky; top:0; background-color: #FFFFFF;">
13
</th>
<th style="text-align:right;position: sticky; top:0; background-color: #FFFFFF;">
14
</th>
<th style="text-align:right;position: sticky; top:0; background-color: #FFFFFF;">
15
</th>
<th style="text-align:right;position: sticky; top:0; background-color: #FFFFFF;">
16
</th>
<th style="text-align:right;position: sticky; top:0; background-color: #FFFFFF;">
17
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
Day0
</td>
<td style="text-align:right;">
1229
</td>
<td style="text-align:right;">
69
</td>
<td style="text-align:right;">
206
</td>
<td style="text-align:right;">
201
</td>
<td style="text-align:right;">
40
</td>
<td style="text-align:right;">
820
</td>
<td style="text-align:right;">
21
</td>
<td style="text-align:right;">
223
</td>
<td style="text-align:right;">
17
</td>
<td style="text-align:right;">
279
</td>
<td style="text-align:right;">
316
</td>
<td style="text-align:right;">
37
</td>
<td style="text-align:right;">
56
</td>
<td style="text-align:right;">
30
</td>
<td style="text-align:right;">
2
</td>
<td style="text-align:right;">
162
</td>
<td style="text-align:right;">
137
</td>
<td style="text-align:right;">
29
</td>
</tr>
<tr>
<td style="text-align:left;">
Day7
</td>
<td style="text-align:right;">
2725
</td>
<td style="text-align:right;">
3079
</td>
<td style="text-align:right;">
2336
</td>
<td style="text-align:right;">
2067
</td>
<td style="text-align:right;">
2278
</td>
<td style="text-align:right;">
716
</td>
<td style="text-align:right;">
1506
</td>
<td style="text-align:right;">
638
</td>
<td style="text-align:right;">
1396
</td>
<td style="text-align:right;">
761
</td>
<td style="text-align:right;">
370
</td>
<td style="text-align:right;">
577
</td>
<td style="text-align:right;">
604
</td>
<td style="text-align:right;">
658
</td>
<td style="text-align:right;">
629
</td>
<td style="text-align:right;">
314
</td>
<td style="text-align:right;">
99
</td>
<td style="text-align:right;">
101
</td>
</tr>
<tr>
<td style="text-align:left;">
Day21
</td>
<td style="text-align:right;">
1682
</td>
<td style="text-align:right;">
1316
</td>
<td style="text-align:right;">
312
</td>
<td style="text-align:right;">
537
</td>
<td style="text-align:right;">
96
</td>
<td style="text-align:right;">
854
</td>
<td style="text-align:right;">
31
</td>
<td style="text-align:right;">
654
</td>
<td style="text-align:right;">
17
</td>
<td style="text-align:right;">
276
</td>
<td style="text-align:right;">
238
</td>
<td style="text-align:right;">
279
</td>
<td style="text-align:right;">
111
</td>
<td style="text-align:right;">
67
</td>
<td style="text-align:right;">
33
</td>
<td style="text-align:right;">
49
</td>
<td style="text-align:right;">
263
</td>
<td style="text-align:right;">
16
</td>
</tr>
<tr>
<td style="text-align:left;">
TOTAL
</td>
<td style="text-align:right;">
5636
</td>
<td style="text-align:right;">
4464
</td>
<td style="text-align:right;">
2854
</td>
<td style="text-align:right;">
2805
</td>
<td style="text-align:right;">
2414
</td>
<td style="text-align:right;">
2390
</td>
<td style="text-align:right;">
1558
</td>
<td style="text-align:right;">
1515
</td>
<td style="text-align:right;">
1430
</td>
<td style="text-align:right;">
1316
</td>
<td style="text-align:right;">
924
</td>
<td style="text-align:right;">
893
</td>
<td style="text-align:right;">
771
</td>
<td style="text-align:right;">
755
</td>
<td style="text-align:right;">
664
</td>
<td style="text-align:right;">
525
</td>
<td style="text-align:right;">
499
</td>
<td style="text-align:right;">
146
</td>
</tr>
</tbody>
</table>
</div>
<p><br/></p>
<div
style="border: 1px solid #ddd; padding: 0px; overflow-y: scroll; height:100%; overflow-x: scroll; width:825px; ">
<table class="table table-condensed" style="font-size: 12px; font-family: &quot;Arial Narrow&quot;, arial, helvetica, sans-serif; margin-left: auto; margin-right: auto;">
<thead>
<tr>
<th style="text-align:left;position: sticky; top:0; background-color: #FFFFFF;">
orig.ident
</th>
<th style="text-align:right;position: sticky; top:0; background-color: #FFFFFF;">
0
</th>
<th style="text-align:right;position: sticky; top:0; background-color: #FFFFFF;">
1
</th>
<th style="text-align:right;position: sticky; top:0; background-color: #FFFFFF;">
2
</th>
<th style="text-align:right;position: sticky; top:0; background-color: #FFFFFF;">
3
</th>
<th style="text-align:right;position: sticky; top:0; background-color: #FFFFFF;">
4
</th>
<th style="text-align:right;position: sticky; top:0; background-color: #FFFFFF;">
5
</th>
<th style="text-align:right;position: sticky; top:0; background-color: #FFFFFF;">
6
</th>
<th style="text-align:right;position: sticky; top:0; background-color: #FFFFFF;">
7
</th>
<th style="text-align:right;position: sticky; top:0; background-color: #FFFFFF;">
8
</th>
<th style="text-align:right;position: sticky; top:0; background-color: #FFFFFF;">
9
</th>
<th style="text-align:right;position: sticky; top:0; background-color: #FFFFFF;">
10
</th>
<th style="text-align:right;position: sticky; top:0; background-color: #FFFFFF;">
11
</th>
<th style="text-align:right;position: sticky; top:0; background-color: #FFFFFF;">
12
</th>
<th style="text-align:right;position: sticky; top:0; background-color: #FFFFFF;">
13
</th>
<th style="text-align:right;position: sticky; top:0; background-color: #FFFFFF;">
14
</th>
<th style="text-align:right;position: sticky; top:0; background-color: #FFFFFF;">
15
</th>
<th style="text-align:right;position: sticky; top:0; background-color: #FFFFFF;">
16
</th>
<th style="text-align:right;position: sticky; top:0; background-color: #FFFFFF;">
17
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
HODay0replicate1
</td>
<td style="text-align:right;">
358
</td>
<td style="text-align:right;">
15
</td>
<td style="text-align:right;">
68
</td>
<td style="text-align:right;">
63
</td>
<td style="text-align:right;">
20
</td>
<td style="text-align:right;">
229
</td>
<td style="text-align:right;">
7
</td>
<td style="text-align:right;">
52
</td>
<td style="text-align:right;">
7
</td>
<td style="text-align:right;">
72
</td>
<td style="text-align:right;">
43
</td>
<td style="text-align:right;">
10
</td>
<td style="text-align:right;">
17
</td>
<td style="text-align:right;">
11
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
48
</td>
<td style="text-align:right;">
22
</td>
<td style="text-align:right;">
10
</td>
</tr>
<tr>
<td style="text-align:left;">
HODay0replicate2
</td>
<td style="text-align:right;">
158
</td>
<td style="text-align:right;">
10
</td>
<td style="text-align:right;">
34
</td>
<td style="text-align:right;">
36
</td>
<td style="text-align:right;">
5
</td>
<td style="text-align:right;">
146
</td>
<td style="text-align:right;">
8
</td>
<td style="text-align:right;">
35
</td>
<td style="text-align:right;">
NA
</td>
<td style="text-align:right;">
55
</td>
<td style="text-align:right;">
81
</td>
<td style="text-align:right;">
5
</td>
<td style="text-align:right;">
8
</td>
<td style="text-align:right;">
7
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
13
</td>
<td style="text-align:right;">
24
</td>
<td style="text-align:right;">
3
</td>
</tr>
<tr>
<td style="text-align:left;">
HODay0replicate3
</td>
<td style="text-align:right;">
438
</td>
<td style="text-align:right;">
28
</td>
<td style="text-align:right;">
64
</td>
<td style="text-align:right;">
58
</td>
<td style="text-align:right;">
4
</td>
<td style="text-align:right;">
246
</td>
<td style="text-align:right;">
2
</td>
<td style="text-align:right;">
78
</td>
<td style="text-align:right;">
4
</td>
<td style="text-align:right;">
76
</td>
<td style="text-align:right;">
89
</td>
<td style="text-align:right;">
11
</td>
<td style="text-align:right;">
16
</td>
<td style="text-align:right;">
5
</td>
<td style="text-align:right;">
NA
</td>
<td style="text-align:right;">
59
</td>
<td style="text-align:right;">
41
</td>
<td style="text-align:right;">
3
</td>
</tr>
<tr>
<td style="text-align:left;">
HODay0replicate4
</td>
<td style="text-align:right;">
275
</td>
<td style="text-align:right;">
16
</td>
<td style="text-align:right;">
40
</td>
<td style="text-align:right;">
44
</td>
<td style="text-align:right;">
11
</td>
<td style="text-align:right;">
199
</td>
<td style="text-align:right;">
4
</td>
<td style="text-align:right;">
58
</td>
<td style="text-align:right;">
6
</td>
<td style="text-align:right;">
76
</td>
<td style="text-align:right;">
103
</td>
<td style="text-align:right;">
11
</td>
<td style="text-align:right;">
15
</td>
<td style="text-align:right;">
7
</td>
<td style="text-align:right;">
NA
</td>
<td style="text-align:right;">
42
</td>
<td style="text-align:right;">
50
</td>
<td style="text-align:right;">
13
</td>
</tr>
<tr>
<td style="text-align:left;">
HODay7replicate1
</td>
<td style="text-align:right;">
688
</td>
<td style="text-align:right;">
468
</td>
<td style="text-align:right;">
572
</td>
<td style="text-align:right;">
555
</td>
<td style="text-align:right;">
755
</td>
<td style="text-align:right;">
250
</td>
<td style="text-align:right;">
566
</td>
<td style="text-align:right;">
134
</td>
<td style="text-align:right;">
538
</td>
<td style="text-align:right;">
71
</td>
<td style="text-align:right;">
83
</td>
<td style="text-align:right;">
NA
</td>
<td style="text-align:right;">
174
</td>
<td style="text-align:right;">
212
</td>
<td style="text-align:right;">
NA
</td>
<td style="text-align:right;">
68
</td>
<td style="text-align:right;">
41
</td>
<td style="text-align:right;">
38
</td>
</tr>
<tr>
<td style="text-align:left;">
HODay7replicate2
</td>
<td style="text-align:right;">
607
</td>
<td style="text-align:right;">
1071
</td>
<td style="text-align:right;">
616
</td>
<td style="text-align:right;">
634
</td>
<td style="text-align:right;">
460
</td>
<td style="text-align:right;">
161
</td>
<td style="text-align:right;">
337
</td>
<td style="text-align:right;">
177
</td>
<td style="text-align:right;">
270
</td>
<td style="text-align:right;">
412
</td>
<td style="text-align:right;">
106
</td>
<td style="text-align:right;">
NA
</td>
<td style="text-align:right;">
183
</td>
<td style="text-align:right;">
164
</td>
<td style="text-align:right;">
NA
</td>
<td style="text-align:right;">
93
</td>
<td style="text-align:right;">
18
</td>
<td style="text-align:right;">
15
</td>
</tr>
<tr>
<td style="text-align:left;">
HODay7replicate3
</td>
<td style="text-align:right;">
998
</td>
<td style="text-align:right;">
861
</td>
<td style="text-align:right;">
700
</td>
<td style="text-align:right;">
561
</td>
<td style="text-align:right;">
486
</td>
<td style="text-align:right;">
199
</td>
<td style="text-align:right;">
396
</td>
<td style="text-align:right;">
223
</td>
<td style="text-align:right;">
386
</td>
<td style="text-align:right;">
160
</td>
<td style="text-align:right;">
141
</td>
<td style="text-align:right;">
NA
</td>
<td style="text-align:right;">
174
</td>
<td style="text-align:right;">
165
</td>
<td style="text-align:right;">
NA
</td>
<td style="text-align:right;">
104
</td>
<td style="text-align:right;">
28
</td>
<td style="text-align:right;">
44
</td>
</tr>
<tr>
<td style="text-align:left;">
HODay7replicate4
</td>
<td style="text-align:right;">
432
</td>
<td style="text-align:right;">
679
</td>
<td style="text-align:right;">
448
</td>
<td style="text-align:right;">
317
</td>
<td style="text-align:right;">
577
</td>
<td style="text-align:right;">
106
</td>
<td style="text-align:right;">
207
</td>
<td style="text-align:right;">
104
</td>
<td style="text-align:right;">
202
</td>
<td style="text-align:right;">
118
</td>
<td style="text-align:right;">
40
</td>
<td style="text-align:right;">
577
</td>
<td style="text-align:right;">
73
</td>
<td style="text-align:right;">
117
</td>
<td style="text-align:right;">
629
</td>
<td style="text-align:right;">
49
</td>
<td style="text-align:right;">
12
</td>
<td style="text-align:right;">
4
</td>
</tr>
<tr>
<td style="text-align:left;">
HODay21replicate1
</td>
<td style="text-align:right;">
552
</td>
<td style="text-align:right;">
331
</td>
<td style="text-align:right;">
83
</td>
<td style="text-align:right;">
163
</td>
<td style="text-align:right;">
38
</td>
<td style="text-align:right;">
238
</td>
<td style="text-align:right;">
10
</td>
<td style="text-align:right;">
190
</td>
<td style="text-align:right;">
6
</td>
<td style="text-align:right;">
98
</td>
<td style="text-align:right;">
91
</td>
<td style="text-align:right;">
56
</td>
<td style="text-align:right;">
35
</td>
<td style="text-align:right;">
14
</td>
<td style="text-align:right;">
12
</td>
<td style="text-align:right;">
16
</td>
<td style="text-align:right;">
69
</td>
<td style="text-align:right;">
2
</td>
</tr>
<tr>
<td style="text-align:left;">
HODay21replicate2
</td>
<td style="text-align:right;">
285
</td>
<td style="text-align:right;">
174
</td>
<td style="text-align:right;">
66
</td>
<td style="text-align:right;">
94
</td>
<td style="text-align:right;">
13
</td>
<td style="text-align:right;">
196
</td>
<td style="text-align:right;">
6
</td>
<td style="text-align:right;">
96
</td>
<td style="text-align:right;">
2
</td>
<td style="text-align:right;">
45
</td>
<td style="text-align:right;">
53
</td>
<td style="text-align:right;">
55
</td>
<td style="text-align:right;">
19
</td>
<td style="text-align:right;">
14
</td>
<td style="text-align:right;">
4
</td>
<td style="text-align:right;">
6
</td>
<td style="text-align:right;">
53
</td>
<td style="text-align:right;">
5
</td>
</tr>
<tr>
<td style="text-align:left;">
HODay21replicate3
</td>
<td style="text-align:right;">
255
</td>
<td style="text-align:right;">
199
</td>
<td style="text-align:right;">
61
</td>
<td style="text-align:right;">
106
</td>
<td style="text-align:right;">
13
</td>
<td style="text-align:right;">
182
</td>
<td style="text-align:right;">
4
</td>
<td style="text-align:right;">
118
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
36
</td>
<td style="text-align:right;">
52
</td>
<td style="text-align:right;">
43
</td>
<td style="text-align:right;">
17
</td>
<td style="text-align:right;">
18
</td>
<td style="text-align:right;">
3
</td>
<td style="text-align:right;">
10
</td>
<td style="text-align:right;">
54
</td>
<td style="text-align:right;">
6
</td>
</tr>
<tr>
<td style="text-align:left;">
HODay21replicate4
</td>
<td style="text-align:right;">
590
</td>
<td style="text-align:right;">
612
</td>
<td style="text-align:right;">
102
</td>
<td style="text-align:right;">
174
</td>
<td style="text-align:right;">
32
</td>
<td style="text-align:right;">
238
</td>
<td style="text-align:right;">
11
</td>
<td style="text-align:right;">
250
</td>
<td style="text-align:right;">
8
</td>
<td style="text-align:right;">
97
</td>
<td style="text-align:right;">
42
</td>
<td style="text-align:right;">
125
</td>
<td style="text-align:right;">
40
</td>
<td style="text-align:right;">
21
</td>
<td style="text-align:right;">
14
</td>
<td style="text-align:right;">
17
</td>
<td style="text-align:right;">
87
</td>
<td style="text-align:right;">
3
</td>
</tr>
<tr>
<td style="text-align:left;">
TOTAL
</td>
<td style="text-align:right;">
5636
</td>
<td style="text-align:right;">
4464
</td>
<td style="text-align:right;">
2854
</td>
<td style="text-align:right;">
2805
</td>
<td style="text-align:right;">
2414
</td>
<td style="text-align:right;">
2390
</td>
<td style="text-align:right;">
1558
</td>
<td style="text-align:right;">
1515
</td>
<td style="text-align:right;">
1430
</td>
<td style="text-align:right;">
1316
</td>
<td style="text-align:right;">
924
</td>
<td style="text-align:right;">
893
</td>
<td style="text-align:right;">
771
</td>
<td style="text-align:right;">
755
</td>
<td style="text-align:right;">
664
</td>
<td style="text-align:right;">
525
</td>
<td style="text-align:right;">
499
</td>
<td style="text-align:right;">
146
</td>
</tr>
</tbody>
</table>
</div>
<p><br/> <br/></p>
<!-- consider adding later 
Add DimPlot for % mitochondria to see if high % are across dataset or limited to one/few clusters (related to what Olivia talked about during Day 1)

# Other approaches for visualizing scRNA-sq data
e.g. code for tSNE visualization 
-->
</div>
<div id="comparing-to-unintegrated-data" class="section level1">
<h1>Comparing to unintegrated data</h1>
<p>If we had proceeded with our filtered data and only normalized our
data without doing any integration, including through the dimensionality
reduction and clustering steps and then labeled the cells with their
sample of origin, then we would see the following for our data:</p>
<!--Add updated example of UMAP / sample labels for unintegrated data - need to check with Raymond regarding how best to do this in place -->
<pre><code>Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck

Number of nodes: 31559
Number of edges: 977875

Running Louvain algorithm...
Maximum modularity in 10 random starts: 0.9522
Number of communities: 22
Elapsed time: 5 seconds</code></pre>
<p><img src="images/curriculum/05-ProjectionAndClustering/05-run_umap_unintegrated-1.png" width="816" style="display: block; margin: auto;" /></p>
<p>In the plot at left, we see that while there are distinct clusters,
those clusters seem to stratified by day. This suggests that without
integration, these batch effects could skew the biological variability
in our data. While on the right, we see little stratification within our
clusters which means the integration seems to have removed those batch
effects.</p>
<details>
<summary>
<strong>Rewind: Pre-integration evaluation clustering and visualization
(code)</strong>
</summary>
<p>Prior to integration, we could follow the same steps we’ve just run
for the integrated to see if the resulting clusters tend to be
determined by sample or condition (in this case, the day):</p>
<pre class="r"><code>geo_so = FindNeighbors(geo_so, dims = 1:pcs, assay = &#39;RNA&#39;, reduction = &#39;unintegrated.sct.pca&#39;, graph.name = c(&#39;RNA_nn&#39;, &#39;RNA_snn&#39;))
geo_so = FindClusters(geo_so, resolution = 0.4, graph.name = &#39;RNA_snn&#39;, cluster.name = &#39;unintegrated.sct.clusters&#39;)
geo_so = RunUMAP(geo_so, dims = 1:pcs, reduction = &#39;unintegrated.sct.pca&#39;, reduction.name = &#39;umap.unintegrated.sct.pca&#39;)</code></pre>
<p>The plots above were generated with:</p>
<pre class="r"><code># Code block - show unintegrated
pre_integration_umap_plot_orig.ident = DimPlot(geo_so, group.by = &#39;orig.ident&#39;, label = FALSE, reduction = &#39;umap.unintegrated.sct.pca&#39;)
ggsave(filename = &#39;results/figures/umap_unintegrated_sct_orig.ident.png&#39;, plot = pre_integration_umap_plot_orig.ident, width = 8, height = 6, units = &#39;in&#39;)

pre_integration_umap_plot_day = DimPlot(geo_so, group.by = &#39;time&#39;, label = FALSE, reduction = &#39;umap.unintegrated.sct.pca&#39;)
ggsave(filename = &#39;results/figures/umap_unintegrated_sct_day.png&#39;, plot = pre_integration_umap_plot_day, width = 8, height = 6, units = &#39;in&#39;)</code></pre>
<pre class="r"><code># Remove the plot variables from the environment to avoid excessive memory usage
rm(pre_integration_umap_plot_orig.ident) 
rm(pre_integration_umap_plot_day)
gc()</code></pre>
</details>
<p><br> <br></p>
<details>
<summary>
<strong>Alternative clustering resolutions</strong>
</summary>
<p>While we show a single resolution, we can generate and plot multiple
resolutions iteratively and compare between them before selecting a
clustering result for the next steps:</p>
<pre class="r"><code>resolutions = c(0.4, 0.8)

for(res in resolutions) {
    message(res)

    cluster_column = sprintf(&#39;SCT_snn_res.%s&#39;, res)
    umap_file = sprintf(&#39;results/figures/umap_integrated_sct_%s.png&#39;, res)

    geo_so = FindClusters(geo_so, resolution = res)

    DimPlot(geo_so, group.by = cluster_column, label = TRUE, reduction = &#39;umap.integrated.sct.rpca&#39;) + NoLegend()
    ggsave(filename = umap_file, width = 8, height = 7, units = &#39;in&#39;)
}</code></pre>
<p>In the results, we’ll see multiple resolutions should now be added to
the metadata slot.</p>
<pre class="r"><code>head(geo_so@meta.data)</code></pre>
</details>
<p><br> <br></p>
<div id="remove-plot-variables-once-they-are-saved"
class="section level3">
<h3>Remove plot variables once they are saved</h3>
<p>ggplots are terrific for visualizing data and we use them liberally
in our analysis and the workshop. Interestingly, ggplots save a
reference to the source dataset - in our case a rather large Seurat
object. That’s ok in most cases, BUT if you save your RStudio
environment (or RStudio temporarily suspends your session), RStudio will
save a <strong>separate copy</strong> of the Seurat object for each
ggplot object; RStudio’s simplistic approach avoids a lot of complexity,
but can drastically increase your load time and memory usage.</p>
<p>Since (a) we have the script to recreate the plots and (b) we
explicitly saved them as graphic objects - we can simply remove them
from memory.</p>
<pre class="r"><code># Remove plot variables from the environment to avoid excessive memory usage

plots = c(&quot;pre_integration_umap_plot_day&quot;, 
          &quot;post_integration_umap_plot_clusters&quot;, 
          &quot;post_integration_umap_plot_split_clusters&quot;, 
          &quot;post_integration_umap_plot_day&quot;)

# Only remove plots that actually exist in the environment
rm(list=Filter(exists, plots))
gc()</code></pre>
<pre><code>            used   (Mb) gc trigger    (Mb)   max used    (Mb)
Ncells   7200805  384.6   12035948   642.8   12035948   642.8
Vcells 745939034 5691.1 1331493210 10158.5 1330520953 10151.1</code></pre>
</div>
</div>
<div id="save-our-progress" class="section level1">
<h1>Save our progress</h1>
<p>Before moving on to our next section, we will output our updated
Seurat object to file:</p>
<pre class="r"><code># Save Seurat object ------------------------------------------------------
saveRDS(geo_so, file = &#39;results/rdata/geo_so_sct_clustered.rds&#39;)</code></pre>
<p><br/> <br/></p>
</div>
<div id="summary" class="section level1">
<h1>Summary</h1>
<br/>
<table class="fig">
<tr>
<td class="fig">
<img
src="images/graphical_abstracts/05-ProjectionAndClustering-Abstract.png" />
</td>
</tr>
<tr>
<td class="fig">
Starting with reduced dimensionality data [PCs x cells] for all samples
- cells are organized into networks and then split up to into clusters
with similar expression programs, regardless of experimental condition.
</td>
</tr>
</table>
<p><br/></p>
<p>In this section we:</p>
<ul>
<li>Generated cluster assignments for our cells using
<code>FindNeighbors()</code> and <code>FindClusters()</code></li>
<li>Evaluated our initial clusters using <code>RunUMAP</code>
dimensional reduction and visualization</li>
</ul>
<p>Next steps: Marker genes</p>
<hr />
<p>These materials have been adapted and extended from materials listed
above. These are open access materials distributed under the terms of
the <a href="http://creativecommons.org/licenses/by/4.0/">Creative
Commons Attribution license (CC BY 4.0)</a>, which permits unrestricted
use, distribution, and reproduction in any medium, provided the original
author and source are credited.</p>
<br/> <br/>
<hr/>
<table style="width:100%;">
<colgroup>
<col width="28%" />
<col width="42%" />
<col width="28%" />
</colgroup>
<thead>
<tr class="header">
<th align="left"><a href="04-PCAandIntegration.html">Previous
lesson</a></th>
<th align="center"><a href="#top">Top of this lesson</a></th>
<th align="right"><a href="06-MarkerVisualization.html">Next
lesson</a></th>
</tr>
</thead>
<tbody>
</tbody>
</table>
</div>



</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3,h4",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
