<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="UM Bioinformatics Core" />

<meta name="date" content="2024-02-23" />

<title>Clustering and Projection</title>

<script src="site_libs/header-attrs-2.25/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/paper.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.13.2/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<script src="site_libs/navigation-1.1/codefolding.js"></script>
<script src="site_libs/navigation-1.1/sourceembed.js"></script>
<link href="site_libs/highlightjs-9.12.0/default.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<!--
Favicon dervied from
https://twemoji.twitter.com/
https://favicon.io/emoji-favicons/dna/
-->
<link rel="apple-touch-icon" sizes="180x180" href="/apple-touch-icon.png">
<link rel="icon" type="image/png" sizes="32x32" href="/favicon-32x32.png">
<link rel="icon" type="image/png" sizes="16x16" href="/favicon-16x16.png">
<link rel="shortcut icon" href="favicon-16x16.png" />
<link rel="manifest" href="/site.webmanifest">
<link rel="mask-icon" href="/safari-pinned-tab.svg" color="#5bbad5">
<meta name="msapplication-TileColor" content="#da532c">
<meta name="theme-color" content="#ffffff">

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>




<style type="text/css">
#rmd-source-code {
  display: none;
}
</style>





<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->
<style type="text/css">
.code-folding-btn { margin-bottom: 4px; }
</style>



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-inverse  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">Intro to scRNA-Seq</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="workshop_intro.html">Intro</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Day 1
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="00A-OrientingOnScRNASeq.html">Orienting on scRNA-Seq</a>
    </li>
    <li>
      <a href="00B-CellRangerInAction.html">Cell Ranger in Action</a>
    </li>
    <li>
      <a href="01-GettingStarted.html">Seurat</a>
    </li>
    <li>
      <a href="02-QCandFiltering.html">Secondary QC &amp; Filtering</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Day 2
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="03-Integration.html">Normalization</a>
    </li>
    <li>
      <a href="04-PCAandDimReduction.html">PCA &amp; Integration</a>
    </li>
    <li>
      <a href="05-ProjectionAndClustering.html">Projection &amp; Clustering</a>
    </li>
    <li>
      <a href="06-MarkerVisualization.html">Marker identification and visualization</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Day 3
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="07-CellTypeAnnos.html">Cell type annotation</a>
    </li>
    <li>
      <a href="08-DifferentialExpression.html">Differential expression analysis</a>
    </li>
    <li>
      <a href="09-IndependentExercise.html">Independent exercise</a>
    </li>
  </ul>
</li>
<li>
  <a href="workshop_wrap_up.html">Wrap up</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">

<div class="btn-group pull-right float-right">
<button type="button" class="btn btn-default btn-xs btn-secondary btn-sm dropdown-toggle" data-toggle="dropdown" data-bs-toggle="dropdown" aria-haspopup="true" aria-expanded="false"><span>Code</span> <span class="caret"></span></button>
<ul class="dropdown-menu dropdown-menu-right" style="min-width: 50px;">
<li><a id="rmd-download-source" href="#">Download Rmd</a></li>
</ul>
</div>



<h1 class="title toc-ignore">Clustering and Projection</h1>
<h4 class="author">UM Bioinformatics Core</h4>
<h4 class="date">2024-02-23</h4>

</div>


<style type="text/css">
body, td {
   font-size: 18px;
}
code.r{
  font-size: 12px;
}
pre {
  font-size: 12px
}
</style>
<div id="workflow-overview" class="section level1 unlisted unnumbered">
<h1 class="unlisted unnumbered">Workflow Overview</h1>
<p><br/>
<img src="images/wayfinder/wayfinder.png" alt="wayfinder" style="height: 400px;"/>
<br/> <br/></p>
</div>
<div id="introduction" class="section level1">
<h1>Introduction</h1>
<!--Add specific goals for section-->
<p>Point out possible gaps between how single-cell data is presented
versus the reality of running an analysis E.g. Papers might report final
filtering thresholds, but our process for choosing those is likely to be
more iterative</p>
<p>General goal: Filtering out the “cells” that either were poorly
measured or are either not single-cells or not cells at all</p>
<div id="objectives" class="section level2">
<h2>Objectives</h2>
<ul>
<li>XX</li>
<li>XX</li>
</ul>
<hr />
</div>
</div>
<div id="projection-and-clustering" class="section level1">
<h1>Projection and clustering</h1>
<p>Contrast the previous dimensionality reduction/nearest neighbors
clustering and plotting the cells in lower dimensionality with the
cluster labels.</p>
<div id="clustering" class="section level2">
<h2>Clustering</h2>
<p>Nearest neighbors algorithm - based on similarity of expression
pattern (just for genes/PCs selected)</p>
<p>Again, how are you defining a “cell type” or “subtype” should help
with ambiguity not only with “How do I decide how many PCs to include?”
but also “How do I choose a resolution for finding clusters?” (as well
as evaluating in the downstream if your initial choices should be
revised).</p>
<!-- per [Ho Lab](https://holab-hku.github.io/Fundamental-scRNA/downstream.html#perform-linear-dimensional-reduction) - The previous version of Seurat (v3) applied a graph-based clustering approach, building upon initial strategies in the development of the initial droplet based single cell technology([Macosko et al. 2015](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4481139/)). "Importantly, the distance metric which drives the clustering analysis (based on previously identified PCs) remains the same."

"However, approach in Seurat to partitioning the cellular distance matrix into clusters has dramatically improved. This approach was heavily inspired by recent publications which applied graph-based clustering approaches to scRNA-seq data ([Xu and Su 2015](https://academic.oup.com/bioinformatics/article/31/12/1974/214505)) and CyTOF data ([Levine et al. 2015](https://www.cell.com/cell/pdfExtended/S0092-8674(15)00637-6)). Briefly, these methods embed cells in a graph structure - for example a K-nearest neighbor (KNN) graph, with edges drawn between cells with similar feature expression patterns, and then attempt to partition this graph into highly interconnected ‘quasi-cliques’ or ‘communities’.

As in PhenoGraph, they first construct a KNN graph based on the euclidean distance in PCA space, and refine the edge weights between any two cells based on the shared overlap in their local neighborhoods (Jaccard similarity). This step is performed using the FindNeighbors() function, and takes as input the previously defined dimensionality of the dataset (first 20 PCs)." -->
<p>Our first step is to generate the k-Nearest Neighbor (KNN) graph
using the number of principal components chosen in the previous
section.</p>
<pre class="r"><code>geo_so = FindNeighbors(geo_so, dims = 1:pcs, reduction = &#39;integrated.sct.rpca.clusters&#39;)</code></pre>
<!--- per [Ho Lab](https://holab-hku.github.io/Fundamental-scRNA/downstream.html#perform-linear-dimensional-reduction) - "To cluster the cells, we next apply modularity optimization techniques such as the Louvain algorithm (default) or SLM (Blondel et al. 2008), to iteratively group cells together, with the goal of optimizing the standard modularity function. The FindClusters() function implements this procedure, and contains a resolution parameter that sets the ‘granularity’ of the downstream clustering, with increased values leading to a greater number of clusters. We find that setting this parameter between 0.4-1.2 typically returns good results for single-cell datasets of around 3K cells. Optimal resolution often increases for larger datasets. The clusters can be found using the Idents() function."" --->
<p>The second step is to partition the kNN graph into “cliques” or
clusters using the Louvain modularity optimization algorithm, specifying
the cluster resolution.</p>
<pre class="r"><code>geo_so = FindClusters(geo_so, resolution = 0.4, cluster.name = &#39;integrated.sct.rpca.clusterss&#39;)</code></pre>
<p>A larger cluster resolution would generate more clusters; generally
it can be preferable to err on the side of too many clusters, as they
can be combined manually in later steps. However, this is another
parameter that may need to be revised after evaluating this initial
clustering in later steps. Recommended resolution ranges are often
dependent on cell number but the <a
href="https://holab-hku.github.io/Fundamental-scRNA/downstream.html#perform-linear-dimensional-reduction">Ho
Lab recommends</a> resolutions between 0.4-1.2 for datasets of
approximately 3k cells.</p>
<pre class="r"><code>geo_so</code></pre>
<pre><code># An object of class Seurat 
# 47037 features across 29615 samples within 2 assays 
# Active assay: SCT (20548 features, 3000 variable features)
#  3 layers present: counts, data, scale.data
#  1 other assay present: RNA
#  3 dimensional reductions calculated: unintegrated.sct.pca, integrated.sct.rpca, umap.unintegrated.sct.pca</code></pre>
</div>
</div>
<div id="cluster-plots" class="section level1">
<h1>Cluster plots</h1>
<p>Note how dimensionality choices are carried through downstream
functions.</p>
<p>What’s different beteween PCA, tSNE, and UMAP?</p>
<p>Note: tSNE vs UMAP example: <a
href="https://pair-code.github.io/understanding-umap/"
class="uri">https://pair-code.github.io/understanding-umap/</a></p>
<!-- per [Ho Lab](https://holab-hku.github.io/Fundamental-scRNA/downstream.html#run-non-linear-dimensional-reduction-umaptsne) "Seurat offers several non-linear dimensional reduction techniques, such as tSNE and UMAP, to visualize and explore these datasets. The goal of these algorithms is to learn the underlying manifold of the data in order to place similar cells together in low-dimensional space. Cells within the graph-based clusters determined above should co-localize on these dimension reduction plots. As input to the UMAP and tSNE, we suggest using the same PCs as input to the clustering analysis."" --->
<pre class="r"><code>geo_so = RunUMAP(geo_so, dims = 1:25, reduction = &#39;unintegrated.sct.pca&#39;, reduction.name = &#39;umap.unintegrated.sct.pca&#39;)</code></pre>
</div>
<div id="visualizing-and-evaluating-clustering" class="section level1">
<h1>Visualizing and evaluating clustering</h1>
<p>How many clusters should I get and how do I adjust the number?</p>
<!--- Show example of changing resolution?--->
<p>We can check the how well the normalization and integration .</p>
<pre class="r"><code>post_integration_umap_plot_day = DimPlot(geo_so, group.by = &#39;day&#39;, label = FALSE, reduction = &#39;umap.integrated.sct.rpca&#39;)

ggsave(filename = &#39;results/figures/umap_integrated_sct_day.png&#39;, plot = post_integration_umap_plot_day)</code></pre>
<p><img src="images/curriculum/umap_integrated_sct_day.png" /></p>
<pre class="r"><code>post_integration_umap_plot_clusters = DimPlot(geo_so, group.by = &#39;seurat_clusters&#39;, label = FALSE, reduction = &#39;umap.integrated.sct.rpca&#39;)

ggsave(filename = &#39;results/figures/umap_integrated_sct_clusters.png&#39;, plot = post_integration_umap_plot_clusters, width = 8, height = 6, units = &#39;in&#39;)</code></pre>
<p><img
src="images/curriculum/umap_integrated_sct_orig.ident.png" /></p>
<p>We can also evaluate the number of cells per cluster using the
<code>table()</code> function, split by <code>day</code>:</p>
<pre class="r"><code>table(geo_so@meta.data$day, geo_so@meta.data$integrated.sct.rpca.clusters)</code></pre>
<pre><code>#       0    1    2    3    4    5    6    7    8    9   10   11   12   13   14   15   16   17   18   19
Day0    65  240  843  317  167   36   36  125  778  132  125  305   55   29  159  133   56  141   37   19
Day7  3184 2497  972 1689 2037 2374 1596  929  108  548  566  382  627  686  309   97  262   86   14   36
Day21 1329  355  964  663  461  147   81  452  383  499  379  227  108   72   42  252  117  195   80   12
</code></pre>
<p>… or split by <code>orig.ident</code> e.g. by individual sample:</p>
<pre class="r"><code>table(geo_so@meta.data$orig.ident, geo_so@meta.data$integrated.sct.rpca.clusters)</code></pre>
<pre><code>#                   0    1    2    3    4    5    6    7    8    9   10   11   12   13   14   15   16   17   18   19
HO.Day0.replicate1    14   82  250   96   52   17   15   24  203   33   32   84   17   11   47   20    5   35    5    7
HO.Day0.replicate2    10   38  110   33   33    4    6   34  152   27   20   57    8    6   13   25   13   16    9    1
HO.Day0.replicate3    25   72  300  117   46    3    5   25  227   40   42   86   16    5   58   40   16   54   12    2
HO.Day0.replicate4    16   48  183   71   36   12   10   42  196   32   31   78   14    7   41   48   22   36   11    9
HO.Day7.replicate1   410  587  285  431  506  766  755  157   45  113  190   79  167  213   63   40   72   28    4   17
HO.Day7.replicate2  1156  616  184  361  580  468  220  372    7  158  135  126  180  160   86   19   88   12    1    3
HO.Day7.replicate3   887  722  377  592  534  395   90  241   48  178  146  110  163  164   93   27   72   36    8   14
HO.Day7.replicate4   731  572  126  305  417  745  531  159    8   99   95   67  117  149   67   11   30   10    1    2
HO.Day21.replicate1  338   96  317  221  144   40   30  139  115  145  110   73   35   14   13   65   26   56   20    3
HO.Day21.replicate2  168   69  189   86   82   25   12  106  113   78   48   40   15   14    6   51   29   29   17    5
HO.Day21.replicate3  184   72  168   93   91   26    8   62  100   90   69   34   16   20    7   55   20   39    9    3
HO.Day21.replicate4  639  118  290  263  144   56   31  145   55  186  152   80   42   24   16   81   42   71   34    1</code></pre>
</div>
<div id="other-approaches-for-visualizing-scrna-sq-data"
class="section level1">
<h1>Other approaches for visualizing scRNA-sq data</h1>
<p>tSNE, PCA</p>
</div>
<div id="comparing-to-unintegrated-data" class="section level1">
<h1>Comparing to unintegrated data</h1>
<p>If we had proceeded with our filtered data and only normalized our
data without doing any integration, including through the dimensionality
reduction and clustering steps and then labeled the cells with their
sample of origin, then we would see the following for our data:</p>
<!--Add example of UMAP / sample labels for unintegrated data from analysis folder to images-->
<p><img src="umap_comparison_sct.png" /></p>
<p>In this plot, we … see that while there are distinct clusters, those
clusters seem to separate primarily by sample id. This suggests that
without integration, these batch effects could skew the biological
variability in our data.
<!--move this section to end of Projection module to contrast with integrated results--></p>
<!-- Edit section to incorporate --->
</div>
<div id="rewind-pre-integration-evaluation" class="section level1">
<h1>Rewind: Pre-integration evaluation</h1>
<p>Prior to integration, we would like to know if there are any batch
effects that carry through to the cell clusters. Most importantly, we
would like to know if clusters tend to be determined by sample or
condition (in this case, the day). In order to figure this out we have
already run the following three commands, which we will describe in more
detail later.</p>
<pre class="r"><code># DefaultAssay(geo_so) = &#39;RNA&#39;

geo_so = RunPCA(geo_so, reduction.name = &#39;unintegrated.sct.pca&#39;)
geo_so = FindNeighbors(geo_so, dims = 1:25, reduction = &#39;unintegrated.sct.pca&#39;)
geo_so = FindClusters(geo_so, resolution = 0.4, cluster.name = &#39;unintegrated.sct.clusters&#39;)
geo_so = RunUMAP(geo_so, dims = 1:25, reduction = &#39;unintegrated.sct.pca&#39;, reduction.name = &#39;umap.unintegrated.sct.pca&#39;)</code></pre>
<p>From this image, we can qualitatively see that some clusters seem to
be driven by the original identity. We can see this even more clearly if
we use <code>group.by = 'day'</code>:</p>
<pre class="r"><code>pre_integration_umap_plot = DimPlot(geo_so, group.by = &#39;day&#39;, label = FALSE, reduction = &#39;umap.unintegrated.sct.pca&#39;) + NoLegend()
ggsave(filename = &#39;ISC_R/results/figures/umap_unintegrated_sct_day.png&#39;, plot = pre_integration_umap_plot)</code></pre>
<p><img src="images/curriculum/umap_unintegrated_sct_day.png" /></p>
<pre class="r"><code>pre_post_integration_plot = pre_integration_umap_plot_day | post_integration_umap_plot_day

ggsave(filename = &#39;results/figures/umap_comparison_sct.png&#39;, plot = pre_post_integration_plot, width = 12, height = 6, units = &#39;in&#39;)</code></pre>
<!-- Make sure above section is edited -->
<div id="alternative-clustering-resolutions" class="section level2">
<h2>Alternative clustering resolutions</h2>
<pre class="r"><code>resolutions = c(0.4, 0.8)
# seq(from = 0.1, to = 1, by = 0.1)

for(res in resolutions) {
    message(res)
    
    cluster_column = sprintf(&#39;SCT_snn_res.%s&#39;, res)
    umap_file = sprintf(&#39;results/figures/umap_integrated_sct_%s.png&#39;, res)
    
    geo_so = FindClusters(geo_so, resolution = res)
    
    DimPlot(geo_so, group.by = cluster_column, label = FALSE, reduction = &#39;umap.integrated.sct.rpca&#39;)
    ggsave(filename = umap_file, width = 8, height = 7, units = &#39;in&#39;)
}</code></pre>
<p>Look at our results</p>
<pre class="r"><code>head(geo_so@meta.data)</code></pre>
<pre><code>#                                   orig.ident nCount_RNA nFeature_RNA  day  replicate percent.mt nCount_SCT nFeature_SCT
 HODay0replicate1_AAACCTGAGAGAACAG-1 HO.Day0.replicate1      10234         3226 Day0 replicate1   1.240962       6061         2865
 HODay0replicate1_AAACCTGGTCATGCAT-1 HO.Day0.replicate1       3158         1499 Day0 replicate1   7.536415       4625         1510
 HODay0replicate1_AAACCTGTCAGAGCTT-1 HO.Day0.replicate1      13464         4102 Day0 replicate1   3.112002       5426         2485
 HODay0replicate1_AAACGGGAGGCCCGTT-1 HO.Day0.replicate1       1189          629 Day0 replicate1   3.700589       4165          919
 HODay0replicate1_AAACGGGCAACTGGCC-1 HO.Day0.replicate1       7726         2602 Day0 replicate1   2.938131       5858         2585
 HODay0replicate1_AAACGGGGTCCGAATT-1 HO.Day0.replicate1       5165         2362 Day0 replicate1   9.196515       5162         2348
#                                     unintegrated.sct.clusters seurat_clusters integrated.sct.rpca.clusters SCT_snn_res.0.4 SCT_snn_res.0.8
 HODay0replicate1_AAACCTGAGAGAACAG-1                         4               8                            2               2               8
 HODay0replicate1_AAACCTGGTCATGCAT-1                         4               8                            2               2               8
 HODay0replicate1_AAACCTGTCAGAGCTT-1                         4               6                            3               3               6
 HODay0replicate1_AAACGGGAGGCCCGTT-1                         4               2                            2               2               2
 HODay0replicate1_AAACGGGCAACTGGCC-1                         4               8                            2               2               8
 HODay0replicate1_AAACGGGGTCCGAATT-1                        11               9                            8               8               9 </code></pre>
</div>
<div id="save-our-progress" class="section level2">
<h2>Save our progress</h2>
<pre class="r"><code>saveRDS(object = geo_so, file = &#39;results/rdata/geo_so_sct_clustered.rda&#39;)</code></pre>
</div>
</div>
<div id="summary" class="section level1">
<h1>Summary</h1>
<p>In this section we:</p>
<ul>
<li>XX</li>
<li>XX</li>
<li>XX</li>
</ul>
<p>Next steps: Marker genes</p>
<hr />
<p>These materials have been adapted and extended from materials listed
above. These are open access materials distributed under the terms of
the <a href="http://creativecommons.org/licenses/by/4.0/">Creative
Commons Attribution license (CC BY 4.0)</a>, which permits unrestricted
use, distribution, and reproduction in any medium, provided the original
author and source are credited.</p>
<br/> <br/>
<hr/>
<table style="width:100%;">
<colgroup>
<col width="28%" />
<col width="42%" />
<col width="28%" />
</colgroup>
<thead>
<tr class="header">
<th align="left"><a href="04-PCAandDimReduction.html">Previous
lesson</a></th>
<th align="center"><a href="#top">Top of this lesson</a></th>
<th align="right"><a href="06-MarkerVisualization.html">Next
lesson</a></th>
</tr>
</thead>
<tbody>
</tbody>
</table>
</div>

<div id="rmd-source-code">---
title: "Clustering and Projection"
author: "UM Bioinformatics Core"
date: "`r Sys.Date()`"
output:
        html_document:
            includes:
                in_header: header.html
            theme: paper
            toc: true
            toc_depth: 4
            toc_float: true
            number_sections: false
            fig_caption: true
            markdown: GFM
            code_download: true
---

<style type="text/css">
body, td {
   font-size: 18px;
}
code.r{
  font-size: 12px;
}
pre {
  font-size: 12px
}
</style>

```{r, include = FALSE}
source("../bin/chunk-options.R")
knitr_fig_path("XX")
```

# Workflow Overview {.unlisted .unnumbered}

<br/>
<img src="images/wayfinder/wayfinder.png" alt="wayfinder" style="height: 400px;"/>
<br/>
<br/>

# Introduction

<!--Add specific goals for section-->

Point out possible gaps between how single-cell data is presented versus the reality of running an analysis
E.g. Papers might report final filtering thresholds, but our process for choosing those is likely to be more iterative

General goal: Filtering out the "cells" that either were poorly measured or are either not single-cells or not cells at all

## Objectives

- XX
- XX

---


# Projection and clustering

Contrast the previous dimensionality reduction/nearest neighbors clustering and plotting the cells in lower dimensionality with the cluster labels.


## Clustering

Nearest neighbors algorithm - based on similarity of expression pattern (just for genes/PCs selected)

Again, how are you defining a “cell type” or “subtype” should help with ambiguity not only with "How do I decide how many PCs to include?" but also "How do I choose a resolution for finding clusters?" (as well as evaluating in the downstream if your initial choices should be revised).


<!-- per [Ho Lab](https://holab-hku.github.io/Fundamental-scRNA/downstream.html#perform-linear-dimensional-reduction) - The previous version of Seurat (v3) applied a graph-based clustering approach, building upon initial strategies in the development of the initial droplet based single cell technology([Macosko et al. 2015](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4481139/)). "Importantly, the distance metric which drives the clustering analysis (based on previously identified PCs) remains the same."

"However, approach in Seurat to partitioning the cellular distance matrix into clusters has dramatically improved. This approach was heavily inspired by recent publications which applied graph-based clustering approaches to scRNA-seq data ([Xu and Su 2015](https://academic.oup.com/bioinformatics/article/31/12/1974/214505)) and CyTOF data ([Levine et al. 2015](https://www.cell.com/cell/pdfExtended/S0092-8674(15)00637-6)). Briefly, these methods embed cells in a graph structure - for example a K-nearest neighbor (KNN) graph, with edges drawn between cells with similar feature expression patterns, and then attempt to partition this graph into highly interconnected ‘quasi-cliques’ or ‘communities’.

As in PhenoGraph, they first construct a KNN graph based on the euclidean distance in PCA space, and refine the edge weights between any two cells based on the shared overlap in their local neighborhoods (Jaccard similarity). This step is performed using the FindNeighbors() function, and takes as input the previously defined dimensionality of the dataset (first 20 PCs)." -->

Our first step is to generate the k-Nearest Neighbor (KNN) graph using the number of principal components chosen in the previous section.

```{r, eval = FALSE}
geo_so = FindNeighbors(geo_so, dims = 1:pcs, reduction = 'integrated.sct.rpca.clusters')
```

<!--- per [Ho Lab](https://holab-hku.github.io/Fundamental-scRNA/downstream.html#perform-linear-dimensional-reduction) - "To cluster the cells, we next apply modularity optimization techniques such as the Louvain algorithm (default) or SLM (Blondel et al. 2008), to iteratively group cells together, with the goal of optimizing the standard modularity function. The FindClusters() function implements this procedure, and contains a resolution parameter that sets the ‘granularity’ of the downstream clustering, with increased values leading to a greater number of clusters. We find that setting this parameter between 0.4-1.2 typically returns good results for single-cell datasets of around 3K cells. Optimal resolution often increases for larger datasets. The clusters can be found using the Idents() function."" --->

The second step is to partition the kNN graph into "cliques" or clusters using the Louvain modularity optimization algorithm, specifying the cluster resolution. 

```{r, eval=FALSE}
geo_so = FindClusters(geo_so, resolution = 0.4, cluster.name = 'integrated.sct.rpca.clusterss')
```

A larger cluster resolution would generate more clusters; generally it can be preferable to err on the side of too many clusters, as they can be combined manually in later steps. However, this is another parameter that may need to be revised after evaluating this initial clustering in later steps. Recommended resolution ranges are often dependent on cell number but the [Ho Lab recommends](https://holab-hku.github.io/Fundamental-scRNA/downstream.html#perform-linear-dimensional-reduction) resolutions between 0.4-1.2 for datasets of approximately 3k cells.

```{r, eval=FALSE}
geo_so
```

~~~
# An object of class Seurat 
# 47037 features across 29615 samples within 2 assays 
# Active assay: SCT (20548 features, 3000 variable features)
#  3 layers present: counts, data, scale.data
#  1 other assay present: RNA
#  3 dimensional reductions calculated: unintegrated.sct.pca, integrated.sct.rpca, umap.unintegrated.sct.pca
~~~


# Cluster plots 

Note how dimensionality choices are carried through downstream functions.

What's different beteween PCA, tSNE, and UMAP?

Note: tSNE vs UMAP example: https://pair-code.github.io/understanding-umap/ 

<!-- per [Ho Lab](https://holab-hku.github.io/Fundamental-scRNA/downstream.html#run-non-linear-dimensional-reduction-umaptsne) "Seurat offers several non-linear dimensional reduction techniques, such as tSNE and UMAP, to visualize and explore these datasets. The goal of these algorithms is to learn the underlying manifold of the data in order to place similar cells together in low-dimensional space. Cells within the graph-based clusters determined above should co-localize on these dimension reduction plots. As input to the UMAP and tSNE, we suggest using the same PCs as input to the clustering analysis."" --->

```{r, eval=FALSE}
geo_so = RunUMAP(geo_so, dims = 1:25, reduction = 'unintegrated.sct.pca', reduction.name = 'umap.unintegrated.sct.pca')
```

# Visualizing and evaluating clustering

How many clusters should I get and how do I adjust the number?

<!--- Show example of changing resolution?--->


We can check the how well the normalization and integration .

```{r, eval = FALSE}
post_integration_umap_plot_day = DimPlot(geo_so, group.by = 'day', label = FALSE, reduction = 'umap.integrated.sct.rpca')

ggsave(filename = 'results/figures/umap_integrated_sct_day.png', plot = post_integration_umap_plot_day)
```

![](./images/curriculum/umap_integrated_sct_day.png)



```{r, eval = FALSE}
post_integration_umap_plot_clusters = DimPlot(geo_so, group.by = 'seurat_clusters', label = FALSE, reduction = 'umap.integrated.sct.rpca')

ggsave(filename = 'results/figures/umap_integrated_sct_clusters.png', plot = post_integration_umap_plot_clusters, width = 8, height = 6, units = 'in')
```

![](./images/curriculum/umap_integrated_sct_orig.ident.png)



We can also evaluate the number of cells per cluster using the `table()` function, split by `day`:

```{r, eval=FALSE}
table(geo_so@meta.data$day, geo_so@meta.data$integrated.sct.rpca.clusters)
```

~~~
# 		0    1    2    3    4    5    6    7    8    9   10   11   12   13   14   15   16   17   18   19
Day0    65  240  843  317  167   36   36  125  778  132  125  305   55   29  159  133   56  141   37   19
Day7  3184 2497  972 1689 2037 2374 1596  929  108  548  566  382  627  686  309   97  262   86   14   36
Day21 1329  355  964  663  461  147   81  452  383  499  379  227  108   72   42  252  117  195   80   12

~~~


... or split by `orig.ident` e.g. by individual sample:

```{r, eval=FALSE}
table(geo_so@meta.data$orig.ident, geo_so@meta.data$integrated.sct.rpca.clusters)
```

~~~
# 					0    1    2    3    4    5    6    7    8    9   10   11   12   13   14   15   16   17   18   19
HO.Day0.replicate1    14   82  250   96   52   17   15   24  203   33   32   84   17   11   47   20    5   35    5    7
HO.Day0.replicate2    10   38  110   33   33    4    6   34  152   27   20   57    8    6   13   25   13   16    9    1
HO.Day0.replicate3    25   72  300  117   46    3    5   25  227   40   42   86   16    5   58   40   16   54   12    2
HO.Day0.replicate4    16   48  183   71   36   12   10   42  196   32   31   78   14    7   41   48   22   36   11    9
HO.Day7.replicate1   410  587  285  431  506  766  755  157   45  113  190   79  167  213   63   40   72   28    4   17
HO.Day7.replicate2  1156  616  184  361  580  468  220  372    7  158  135  126  180  160   86   19   88   12    1    3
HO.Day7.replicate3   887  722  377  592  534  395   90  241   48  178  146  110  163  164   93   27   72   36    8   14
HO.Day7.replicate4   731  572  126  305  417  745  531  159    8   99   95   67  117  149   67   11   30   10    1    2
HO.Day21.replicate1  338   96  317  221  144   40   30  139  115  145  110   73   35   14   13   65   26   56   20    3
HO.Day21.replicate2  168   69  189   86   82   25   12  106  113   78   48   40   15   14    6   51   29   29   17    5
HO.Day21.replicate3  184   72  168   93   91   26    8   62  100   90   69   34   16   20    7   55   20   39    9    3
HO.Day21.replicate4  639  118  290  263  144   56   31  145   55  186  152   80   42   24   16   81   42   71   34    1
~~~


# Other approaches for visualizing scRNA-sq data

tSNE, PCA


# Comparing to unintegrated data

If we had proceeded with our filtered data and only normalized our data without doing any integration, including through the dimensionality reduction and clustering steps and then labeled the cells with their sample of origin, then we would see the following for our data:

<!--Add example of UMAP / sample labels for unintegrated data from analysis folder to images-->

![](umap_comparison_sct.png)



In this plot, we ... see that while there are distinct clusters, those clusters seem to separate primarily by sample id. This suggests that without integration, these batch effects could skew the biological variability in our data. <!--move this section to end of Projection module to contrast with integrated results-->


<!-- Edit section to incorporate --->

# Rewind: Pre-integration evaluation

Prior to integration, we would like to know if there are any batch effects that carry through to the cell clusters. Most importantly, we would like to know if clusters tend to be determined by sample or condition (in this case, the day). In order to figure this out we have already run the following three commands, which we will describe in more detail later.

```{r, eval=FALSE}
# DefaultAssay(geo_so) = 'RNA'

geo_so = RunPCA(geo_so, reduction.name = 'unintegrated.sct.pca')
geo_so = FindNeighbors(geo_so, dims = 1:25, reduction = 'unintegrated.sct.pca')
geo_so = FindClusters(geo_so, resolution = 0.4, cluster.name = 'unintegrated.sct.clusters')
geo_so = RunUMAP(geo_so, dims = 1:25, reduction = 'unintegrated.sct.pca', reduction.name = 'umap.unintegrated.sct.pca')
```

From this image, we can qualitatively see that some clusters seem to be driven by the original identity. We can see this even more clearly if we use `group.by = 'day'`:

```{r, eval=FALSE}
pre_integration_umap_plot = DimPlot(geo_so, group.by = 'day', label = FALSE, reduction = 'umap.unintegrated.sct.pca') + NoLegend()
ggsave(filename = 'ISC_R/results/figures/umap_unintegrated_sct_day.png', plot = pre_integration_umap_plot)
```

![](./images/curriculum/umap_unintegrated_sct_day.png)


```{r, eval=FALSE}
pre_post_integration_plot = pre_integration_umap_plot_day | post_integration_umap_plot_day

ggsave(filename = 'results/figures/umap_comparison_sct.png', plot = pre_post_integration_plot, width = 12, height = 6, units = 'in')
```

<!-- Make sure above section is edited -->

## Alternative clustering resolutions

```{r, eval=FALSE}
resolutions = c(0.4, 0.8)
# seq(from = 0.1, to = 1, by = 0.1)

for(res in resolutions) {
    message(res)
    
    cluster_column = sprintf('SCT_snn_res.%s', res)
    umap_file = sprintf('results/figures/umap_integrated_sct_%s.png', res)
    
    geo_so = FindClusters(geo_so, resolution = res)
    
    DimPlot(geo_so, group.by = cluster_column, label = FALSE, reduction = 'umap.integrated.sct.rpca')
    ggsave(filename = umap_file, width = 8, height = 7, units = 'in')
}
```


Look at our results
```{r, eval=FALSE}
head(geo_so@meta.data)
```

~~~
#									orig.ident nCount_RNA nFeature_RNA  day  replicate percent.mt nCount_SCT nFeature_SCT
 HODay0replicate1_AAACCTGAGAGAACAG-1 HO.Day0.replicate1      10234         3226 Day0 replicate1   1.240962       6061         2865
 HODay0replicate1_AAACCTGGTCATGCAT-1 HO.Day0.replicate1       3158         1499 Day0 replicate1   7.536415       4625         1510
 HODay0replicate1_AAACCTGTCAGAGCTT-1 HO.Day0.replicate1      13464         4102 Day0 replicate1   3.112002       5426         2485
 HODay0replicate1_AAACGGGAGGCCCGTT-1 HO.Day0.replicate1       1189          629 Day0 replicate1   3.700589       4165          919
 HODay0replicate1_AAACGGGCAACTGGCC-1 HO.Day0.replicate1       7726         2602 Day0 replicate1   2.938131       5858         2585
 HODay0replicate1_AAACGGGGTCCGAATT-1 HO.Day0.replicate1       5165         2362 Day0 replicate1   9.196515       5162         2348
#                                     unintegrated.sct.clusters seurat_clusters integrated.sct.rpca.clusters SCT_snn_res.0.4 SCT_snn_res.0.8
 HODay0replicate1_AAACCTGAGAGAACAG-1                         4               8                            2               2               8
 HODay0replicate1_AAACCTGGTCATGCAT-1                         4               8                            2               2               8
 HODay0replicate1_AAACCTGTCAGAGCTT-1                         4               6                            3               3               6
 HODay0replicate1_AAACGGGAGGCCCGTT-1                         4               2                            2               2               2
 HODay0replicate1_AAACGGGCAACTGGCC-1                         4               8                            2               2               8
 HODay0replicate1_AAACGGGGTCCGAATT-1                        11               9                            8               8               9 
~~~


## Save our progress

```{r, eval=FALSE}
saveRDS(object = geo_so, file = 'results/rdata/geo_so_sct_clustered.rda')
```

# Summary

In this section we:

- XX
- XX
- XX

Next steps: Marker genes


----

These materials have been adapted and extended from materials listed above. These are open access materials distributed under the terms of the [Creative Commons Attribution license (CC BY 4.0)](http://creativecommons.org/licenses/by/4.0/), which permits unrestricted use, distribution, and reproduction in any medium, provided the original author and source are credited.

<br/>
<br/>
<hr/>
| [Previous lesson](04-PCAandDimReduction.html) | [Top of this lesson](#top) | [Next lesson](06-MarkerVisualization.html) |
| :--- | :----: | ---: |

</div>


</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->
<script>
$(document).ready(function () {
  window.initializeSourceEmbed("05-ProjectionAndClustering.Rmd");
});
</script>

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3,h4",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
