<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="UM Bioinformatics Core" />

<meta name="date" content="2024-02-23" />

<title>Marker identification and visualization</title>

<script src="site_libs/header-attrs-2.25/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/paper.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.13.2/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<script src="site_libs/navigation-1.1/codefolding.js"></script>
<script src="site_libs/navigation-1.1/sourceembed.js"></script>
<link href="site_libs/highlightjs-9.12.0/default.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<!--
Favicon dervied from
https://twemoji.twitter.com/
https://favicon.io/emoji-favicons/dna/
-->
<link rel="apple-touch-icon" sizes="180x180" href="/apple-touch-icon.png">
<link rel="icon" type="image/png" sizes="32x32" href="/favicon-32x32.png">
<link rel="icon" type="image/png" sizes="16x16" href="/favicon-16x16.png">
<link rel="shortcut icon" href="favicon-16x16.png" />
<link rel="manifest" href="/site.webmanifest">
<link rel="mask-icon" href="/safari-pinned-tab.svg" color="#5bbad5">
<meta name="msapplication-TileColor" content="#da532c">
<meta name="theme-color" content="#ffffff">

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>




<style type="text/css">
#rmd-source-code {
  display: none;
}
</style>





<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->
<style type="text/css">
.code-folding-btn { margin-bottom: 4px; }
</style>



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-inverse  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">Intro to scRNA-Seq</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="workshop_intro.html">Intro</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Day 1
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="00A-OrientingOnScRNASeq.html">Orienting on scRNA-Seq</a>
    </li>
    <li>
      <a href="00B-CellRangerInAction.html">Cell Ranger in Action</a>
    </li>
    <li>
      <a href="01-GettingStarted.html">Seurat</a>
    </li>
    <li>
      <a href="02-QCandFiltering.html">Secondary QC &amp; Filtering</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Day 2
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="03-Integration.html">Normalization</a>
    </li>
    <li>
      <a href="04-PCAandDimReduction.html">PCA &amp; Integration</a>
    </li>
    <li>
      <a href="05-ProjectionAndClustering.html">Projection &amp; Clustering</a>
    </li>
    <li>
      <a href="06-MarkerVisualization.html">Marker identification and visualization</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Day 3
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="07-CellTypeAnnos.html">Cell type annotation</a>
    </li>
    <li>
      <a href="08-DifferentialExpression.html">Differential expression analysis</a>
    </li>
    <li>
      <a href="09-IndependentExercise.html">Independent exercise</a>
    </li>
  </ul>
</li>
<li>
  <a href="workshop_wrap_up.html">Wrap up</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">

<div class="btn-group pull-right float-right">
<button type="button" class="btn btn-default btn-xs btn-secondary btn-sm dropdown-toggle" data-toggle="dropdown" data-bs-toggle="dropdown" aria-haspopup="true" aria-expanded="false"><span>Code</span> <span class="caret"></span></button>
<ul class="dropdown-menu dropdown-menu-right" style="min-width: 50px;">
<li><a id="rmd-download-source" href="#">Download Rmd</a></li>
</ul>
</div>



<h1 class="title toc-ignore">Marker identification and
visualization</h1>
<h4 class="author">UM Bioinformatics Core</h4>
<h4 class="date">2024-02-23</h4>

</div>


<style type="text/css">
body, td {
   font-size: 18px;
}
code.r{
  font-size: 12px;
}
pre {
  font-size: 12px
}
</style>
<div id="workflow-overview" class="section level1 unlisted unnumbered">
<h1 class="unlisted unnumbered">Workflow Overview</h1>
<p><br/>
<img src="images/wayfinder/wayfinder.png" alt="wayfinder" style="height: 400px;"/>
<br/> <br/></p>
</div>
<div id="introduction" class="section level1">
<h1>Introduction</h1>
<p>Point out possible gaps between how single-cell data is presented
versus the reality of running an analysis E.g. Papers might report final
filtering thresholds, but our process for choosing those is likely to be
more iterative</p>
<p>General goal: Filtering out the “cells” that either were poorly
measured or are either not single-cells or not cells at all</p>
<!--Add specific goals for section-->
</div>
<div id="marker-identification" class="section level1">
<h1>Marker identification</h1>
<p>Ensure that correct resolution is selected</p>
<pre class="r"><code>geo_so = SetIdent(geo_so, value = &#39;SCT_snn_res.0.4&#39;)</code></pre>
<p>We can use <code>FindAllMarkers</code> to run comparisons between
each cluster and all other cells, regardless of the experimental
group.</p>
<pre class="r"><code>geo_so = PrepSCTFindMarkers(geo_so)
geo_markers = FindAllMarkers(geo_so, only.pos = TRUE)</code></pre>
<!--- Consider adding figure showing example of what's being compared (circle cluster 1 and then circle all other cells for example) -->
<p>Then we can write out the full results to file.</p>
<pre class="r"><code>write_csv(geo_markers, file = &#39;ISC_R/results/tables/marker_genes_0.4.csv&#39;)</code></pre>
<p>In addition to generating the full set of marker genes, it can be
useful to filter the results to highlight the top positive markers
(since a positive fold-change would mean that gene is more highly
expressed in the cluster compared to all other cells):</p>
<pre class="r"><code># add code to create table of top 5 markers per cluster
top_5_by_log2FC = geo_markers %&gt;% group_by(cluster) %&gt;% arrange(p_val_adj, desc(avg_log2FC)) %&gt;% slice_head(n = 5)</code></pre>
<p>If we look at the marker table, we see several columns:</p>
<pre class="r"><code>head(top_5_by_log2FC)</code></pre>
<pre><code># add expected output</code></pre>
<div id="marker-visualization" class="section level2">
<h2>Marker visualization</h2>
<p>Use SCT values, create a dotplot</p>
<pre class="r"><code>top_5_sct_dot_plot = DotPlot(geo_so, features = unique(top_5_by_log2FC$gene)) + 
    theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5)) +
    labs(title = &#39;Top 5 Cluster Genes by FDR and avg_log2FC&#39;)</code></pre>
<p>And we can save the figure to file</p>
<pre class="r"><code>ggsave(filename = &#39;ISC_R/results/figures/markers_top_5_sct_dot_plot.png&#39;, plot = top_5_sct_dot_plot, width = 12, height = 6, units = &#39;in&#39;) </code></pre>
<p>Can also use raw RNA values <!--- move to dropdown ---></p>
<pre class="r"><code>top_5_rna_dot_plot = DotPlot(geo_so, features = unique(top_5_by_log2FC$gene), assay = &#39;RNA&#39;) + 
    theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5)) +
    labs(title = &#39;Top 5 Cluster Genes by FDR and avg_log2FC&#39;)

ggsave(filename = &#39;ISC_R/results/figures/markers_top_5_sct_dot_plot.png&#39;, plot = top_5_rna_dot_plot, width = 12, height = 6, units = &#39;in&#39;) </code></pre>
</div>
</div>
<div id="utilizing-genes-of-interest" class="section level1">
<h1>Utilizing genes of interest</h1>
<p>Often we have</p>
<pre class="r"><code># Using genes from the paper - https://www.ncbi.nlm.nih.gov/pmc/articles/PMC7002453/

fig1g_markers = c(&#39;Cxcl1&#39;, &#39;Cxcl2&#39;, &#39;Ccl2&#39;, &#39;Ccl3&#39;, &#39;Ccl4&#39;, &#39;Il1b&#39;, &#39;Il6b&#39;, &#39;Tnf&#39;, &#39;Tgfb1&#39;, &#39;Tgfb2&#39;, &#39;Tgfb3&#39;, &#39;Cxcl5&#39;)
fig1h_markers = c(&#39;Cxcr2&#39;, &#39;Csf1r&#39;, &#39;Csf3r&#39;, &#39;Tgfbr1&#39;, &#39;Tgfbr3&#39;, &#39;Il1r1&#39;, &#39;Il6ra&#39;, &#39;Lifr&#39;, &#39;Tgfbr2&#39;)</code></pre>
</div>
<div id="save-our-progress" class="section level1">
<h1>Save our progress</h1>
<p>Create an output for our updated Seurat object</p>
<pre class="r"><code>saveRDS(object = geo_so, file = &#39;ISC_R/results/rdata/geo_so_sct_integrated_with_markers.rda&#39;)</code></pre>
</div>
<div id="summary" class="section level1">
<h1>Summary</h1>
<p>Now that we have …</p>
<p>However, marker genes alone might not be sufficient to determine
cell-type or sub-type labels for our clusters so it can be helpful to
apply other approaches, which we will discuss in the next section.</p>
<p>Next steps: Cell type prediction tools</p>
<hr />
<p>These materials have been adapted and extended from materials listed
above. These are open access materials distributed under the terms of
the <a href="http://creativecommons.org/licenses/by/4.0/">Creative
Commons Attribution license (CC BY 4.0)</a>, which permits unrestricted
use, distribution, and reproduction in any medium, provided the original
author and source are credited.</p>
<br/> <br/>
<hr/>
<table style="width:100%;">
<colgroup>
<col width="28%" />
<col width="42%" />
<col width="28%" />
</colgroup>
<thead>
<tr class="header">
<th align="left"><a href="05-ProjectionAndClustering.html">Previous
lesson</a></th>
<th align="center"><a href="#top">Top of this lesson</a></th>
<th align="right"><a href="07-CellTypeAnnos.html">Next lesson</a></th>
</tr>
</thead>
<tbody>
</tbody>
</table>
</div>

<div id="rmd-source-code">LS0tCnRpdGxlOiAiTWFya2VyIGlkZW50aWZpY2F0aW9uIGFuZCB2aXN1YWxpemF0aW9uIgphdXRob3I6ICJVTSBCaW9pbmZvcm1hdGljcyBDb3JlIgpkYXRlOiAiYHIgU3lzLkRhdGUoKWAiCm91dHB1dDoKICAgICAgICBodG1sX2RvY3VtZW50OgogICAgICAgICAgICBpbmNsdWRlczoKICAgICAgICAgICAgICAgIGluX2hlYWRlcjogaGVhZGVyLmh0bWwKICAgICAgICAgICAgdGhlbWU6IHBhcGVyCiAgICAgICAgICAgIHRvYzogdHJ1ZQogICAgICAgICAgICB0b2NfZGVwdGg6IDQKICAgICAgICAgICAgdG9jX2Zsb2F0OiB0cnVlCiAgICAgICAgICAgIG51bWJlcl9zZWN0aW9uczogZmFsc2UKICAgICAgICAgICAgZmlnX2NhcHRpb246IHRydWUKICAgICAgICAgICAgbWFya2Rvd246IEdGTQogICAgICAgICAgICBjb2RlX2Rvd25sb2FkOiB0cnVlCi0tLQoKPHN0eWxlIHR5cGU9InRleHQvY3NzIj4KYm9keSwgdGQgewogICBmb250LXNpemU6IDE4cHg7Cn0KY29kZS5yewogIGZvbnQtc2l6ZTogMTJweDsKfQpwcmUgewogIGZvbnQtc2l6ZTogMTJweAp9Cjwvc3R5bGU+CgpgYGB7ciwgaW5jbHVkZSA9IEZBTFNFfQpzb3VyY2UoIi4uL2Jpbi9jaHVuay1vcHRpb25zLlIiKQprbml0cl9maWdfcGF0aCgiWFgiKQpgYGAKCiMgV29ya2Zsb3cgT3ZlcnZpZXcgey51bmxpc3RlZCAudW5udW1iZXJlZH0KCjxici8+CjxpbWcgc3JjPSJpbWFnZXMvd2F5ZmluZGVyL3dheWZpbmRlci5wbmciIGFsdD0id2F5ZmluZGVyIiBzdHlsZT0iaGVpZ2h0OiA0MDBweDsiLz4KPGJyLz4KPGJyLz4KCiMgSW50cm9kdWN0aW9uCgpQb2ludCBvdXQgcG9zc2libGUgZ2FwcyBiZXR3ZWVuIGhvdyBzaW5nbGUtY2VsbCBkYXRhIGlzIHByZXNlbnRlZCB2ZXJzdXMgdGhlIHJlYWxpdHkgb2YgcnVubmluZyBhbiBhbmFseXNpcwpFLmcuIFBhcGVycyBtaWdodCByZXBvcnQgZmluYWwgZmlsdGVyaW5nIHRocmVzaG9sZHMsIGJ1dCBvdXIgcHJvY2VzcyBmb3IgY2hvb3NpbmcgdGhvc2UgaXMgbGlrZWx5IHRvIGJlIG1vcmUgaXRlcmF0aXZlCgpHZW5lcmFsIGdvYWw6IEZpbHRlcmluZyBvdXQgdGhlICJjZWxscyIgdGhhdCBlaXRoZXIgd2VyZSBwb29ybHkgbWVhc3VyZWQgb3IgYXJlIGVpdGhlciBub3Qgc2luZ2xlLWNlbGxzIG9yIG5vdCBjZWxscyBhdCBhbGwKCjwhLS1BZGQgc3BlY2lmaWMgZ29hbHMgZm9yIHNlY3Rpb24tLT4KCiMgTWFya2VyIGlkZW50aWZpY2F0aW9uCgpFbnN1cmUgdGhhdCBjb3JyZWN0IHJlc29sdXRpb24gaXMgc2VsZWN0ZWQKYGBge3IsIGV2YWwgPSBGQUxTRX0KZ2VvX3NvID0gU2V0SWRlbnQoZ2VvX3NvLCB2YWx1ZSA9ICdTQ1Rfc25uX3Jlcy4wLjQnKQpgYGAKCldlIGNhbiB1c2UgYEZpbmRBbGxNYXJrZXJzYCB0byBydW4gY29tcGFyaXNvbnMgYmV0d2VlbiBlYWNoIGNsdXN0ZXIgYW5kIGFsbCBvdGhlciBjZWxscywgcmVnYXJkbGVzcyBvZiB0aGUgZXhwZXJpbWVudGFsIGdyb3VwLiAKCmBgYHtyLCBldmFsPUZBTFNFfQpnZW9fc28gPSBQcmVwU0NURmluZE1hcmtlcnMoZ2VvX3NvKQpnZW9fbWFya2VycyA9IEZpbmRBbGxNYXJrZXJzKGdlb19zbywgb25seS5wb3MgPSBUUlVFKQpgYGAKCjwhLS0tIENvbnNpZGVyIGFkZGluZyBmaWd1cmUgc2hvd2luZyBleGFtcGxlIG9mIHdoYXQncyBiZWluZyBjb21wYXJlZCAoY2lyY2xlIGNsdXN0ZXIgMSBhbmQgdGhlbiBjaXJjbGUgYWxsIG90aGVyIGNlbGxzIGZvciBleGFtcGxlKSAtLT4KClRoZW4gd2UgY2FuIHdyaXRlIG91dCB0aGUgZnVsbCByZXN1bHRzIHRvIGZpbGUuIApgYGB7ciwgZXZhbD1GQUxTRX0Kd3JpdGVfY3N2KGdlb19tYXJrZXJzLCBmaWxlID0gJ0lTQ19SL3Jlc3VsdHMvdGFibGVzL21hcmtlcl9nZW5lc18wLjQuY3N2JykKYGBgCgpJbiBhZGRpdGlvbiB0byBnZW5lcmF0aW5nIHRoZSBmdWxsIHNldCBvZiBtYXJrZXIgZ2VuZXMsIGl0IGNhbiBiZSB1c2VmdWwgdG8gZmlsdGVyIHRoZSByZXN1bHRzIHRvIGhpZ2hsaWdodCB0aGUgdG9wIHBvc2l0aXZlIG1hcmtlcnMgKHNpbmNlIGEgcG9zaXRpdmUgZm9sZC1jaGFuZ2Ugd291bGQgbWVhbiB0aGF0IGdlbmUgaXMgbW9yZSBoaWdobHkgZXhwcmVzc2VkIGluIHRoZSBjbHVzdGVyIGNvbXBhcmVkIHRvIGFsbCBvdGhlciBjZWxscyk6CmBgYHtyLCBldmFsPUZBTFNFfQojIGFkZCBjb2RlIHRvIGNyZWF0ZSB0YWJsZSBvZiB0b3AgNSBtYXJrZXJzIHBlciBjbHVzdGVyCnRvcF81X2J5X2xvZzJGQyA9IGdlb19tYXJrZXJzICU+JSBncm91cF9ieShjbHVzdGVyKSAlPiUgYXJyYW5nZShwX3ZhbF9hZGosIGRlc2MoYXZnX2xvZzJGQykpICU+JSBzbGljZV9oZWFkKG4gPSA1KQpgYGAKCklmIHdlIGxvb2sgYXQgdGhlIG1hcmtlciB0YWJsZSwgd2Ugc2VlIHNldmVyYWwgY29sdW1uczoKYGBge3IsIGV2YWw9RkFMU0V9CmhlYWQodG9wXzVfYnlfbG9nMkZDKQpgYGAKfn5+fgojIGFkZCBleHBlY3RlZCBvdXRwdXQKfn5+fgoKCiMjIE1hcmtlciB2aXN1YWxpemF0aW9uCgpVc2UgU0NUIHZhbHVlcywgY3JlYXRlIGEgZG90cGxvdApgYGB7ciwgZXZhbD1GQUxTRX0KdG9wXzVfc2N0X2RvdF9wbG90ID0gRG90UGxvdChnZW9fc28sIGZlYXR1cmVzID0gdW5pcXVlKHRvcF81X2J5X2xvZzJGQyRnZW5lKSkgKyAKICAgIHRoZW1lKGF4aXMudGV4dC54ID0gZWxlbWVudF90ZXh0KGFuZ2xlID0gOTAsIGhqdXN0ID0gMSwgdmp1c3QgPSAwLjUpKSArCiAgICBsYWJzKHRpdGxlID0gJ1RvcCA1IENsdXN0ZXIgR2VuZXMgYnkgRkRSIGFuZCBhdmdfbG9nMkZDJykKYGBgCgpBbmQgd2UgY2FuIHNhdmUgdGhlIGZpZ3VyZSB0byBmaWxlCmBgYHtyLCBldmFsPUZBTFNFfQpnZ3NhdmUoZmlsZW5hbWUgPSAnSVNDX1IvcmVzdWx0cy9maWd1cmVzL21hcmtlcnNfdG9wXzVfc2N0X2RvdF9wbG90LnBuZycsIHBsb3QgPSB0b3BfNV9zY3RfZG90X3Bsb3QsIHdpZHRoID0gMTIsIGhlaWdodCA9IDYsIHVuaXRzID0gJ2luJykgCmBgYAoKCkNhbiBhbHNvIHVzZSByYXcgUk5BIHZhbHVlcyA8IS0tLSBtb3ZlIHRvIGRyb3Bkb3duIC0tLT4KCmBgYHtyLCBldmFsPUZBTFNFfQp0b3BfNV9ybmFfZG90X3Bsb3QgPSBEb3RQbG90KGdlb19zbywgZmVhdHVyZXMgPSB1bmlxdWUodG9wXzVfYnlfbG9nMkZDJGdlbmUpLCBhc3NheSA9ICdSTkEnKSArIAogICAgdGhlbWUoYXhpcy50ZXh0LnggPSBlbGVtZW50X3RleHQoYW5nbGUgPSA5MCwgaGp1c3QgPSAxLCB2anVzdCA9IDAuNSkpICsKICAgIGxhYnModGl0bGUgPSAnVG9wIDUgQ2x1c3RlciBHZW5lcyBieSBGRFIgYW5kIGF2Z19sb2cyRkMnKQoKZ2dzYXZlKGZpbGVuYW1lID0gJ0lTQ19SL3Jlc3VsdHMvZmlndXJlcy9tYXJrZXJzX3RvcF81X3NjdF9kb3RfcGxvdC5wbmcnLCBwbG90ID0gdG9wXzVfcm5hX2RvdF9wbG90LCB3aWR0aCA9IDEyLCBoZWlnaHQgPSA2LCB1bml0cyA9ICdpbicpIApgYGAKCgojIFV0aWxpemluZyBnZW5lcyBvZiBpbnRlcmVzdAoKT2Z0ZW4gd2UgaGF2ZSAKCmBgYHtyLCBldmFsPUZBTFNFfQojIFVzaW5nIGdlbmVzIGZyb20gdGhlIHBhcGVyIC0gaHR0cHM6Ly93d3cubmNiaS5ubG0ubmloLmdvdi9wbWMvYXJ0aWNsZXMvUE1DNzAwMjQ1My8KCmZpZzFnX21hcmtlcnMgPSBjKCdDeGNsMScsICdDeGNsMicsICdDY2wyJywgJ0NjbDMnLCAnQ2NsNCcsICdJbDFiJywgJ0lsNmInLCAnVG5mJywgJ1RnZmIxJywgJ1RnZmIyJywgJ1RnZmIzJywgJ0N4Y2w1JykKZmlnMWhfbWFya2VycyA9IGMoJ0N4Y3IyJywgJ0NzZjFyJywgJ0NzZjNyJywgJ1RnZmJyMScsICdUZ2ZicjMnLCAnSWwxcjEnLCAnSWw2cmEnLCAnTGlmcicsICdUZ2ZicjInKQpgYGAKCiMgU2F2ZSBvdXIgcHJvZ3Jlc3MKCkNyZWF0ZSBhbiBvdXRwdXQgZm9yIG91ciB1cGRhdGVkIFNldXJhdCBvYmplY3QKYGBge3IsIGV2YWw9RkFMU0V9CnNhdmVSRFMob2JqZWN0ID0gZ2VvX3NvLCBmaWxlID0gJ0lTQ19SL3Jlc3VsdHMvcmRhdGEvZ2VvX3NvX3NjdF9pbnRlZ3JhdGVkX3dpdGhfbWFya2Vycy5yZGEnKQpgYGAKCgojIFN1bW1hcnkKCk5vdyB0aGF0IHdlIGhhdmUgLi4uIAoKSG93ZXZlciwgbWFya2VyIGdlbmVzIGFsb25lIG1pZ2h0IG5vdCBiZSBzdWZmaWNpZW50IHRvIGRldGVybWluZSBjZWxsLXR5cGUgb3Igc3ViLXR5cGUgbGFiZWxzIGZvciBvdXIgY2x1c3RlcnMgc28gaXQgY2FuIGJlIGhlbHBmdWwgdG8gYXBwbHkgb3RoZXIgYXBwcm9hY2hlcywgd2hpY2ggd2Ugd2lsbCBkaXNjdXNzIGluIHRoZSBuZXh0IHNlY3Rpb24uCgpOZXh0IHN0ZXBzOiBDZWxsIHR5cGUgcHJlZGljdGlvbiB0b29scwoKCgotLS0tCgpUaGVzZSBtYXRlcmlhbHMgaGF2ZSBiZWVuIGFkYXB0ZWQgYW5kIGV4dGVuZGVkIGZyb20gbWF0ZXJpYWxzIGxpc3RlZCBhYm92ZS4gVGhlc2UgYXJlIG9wZW4gYWNjZXNzIG1hdGVyaWFscyBkaXN0cmlidXRlZCB1bmRlciB0aGUgdGVybXMgb2YgdGhlIFtDcmVhdGl2ZSBDb21tb25zIEF0dHJpYnV0aW9uIGxpY2Vuc2UgKENDIEJZIDQuMCldKGh0dHA6Ly9jcmVhdGl2ZWNvbW1vbnMub3JnL2xpY2Vuc2VzL2J5LzQuMC8pLCB3aGljaCBwZXJtaXRzIHVucmVzdHJpY3RlZCB1c2UsIGRpc3RyaWJ1dGlvbiwgYW5kIHJlcHJvZHVjdGlvbiBpbiBhbnkgbWVkaXVtLCBwcm92aWRlZCB0aGUgb3JpZ2luYWwgYXV0aG9yIGFuZCBzb3VyY2UgYXJlIGNyZWRpdGVkLgoKPGJyLz4KPGJyLz4KPGhyLz4KfCBbUHJldmlvdXMgbGVzc29uXSgwNS1Qcm9qZWN0aW9uQW5kQ2x1c3RlcmluZy5odG1sKSB8IFtUb3Agb2YgdGhpcyBsZXNzb25dKCN0b3ApIHwgW05leHQgbGVzc29uXSgwNy1DZWxsVHlwZUFubm9zLmh0bWwpIHwKfCA6LS0tIHwgOi0tLS06IHwgLS0tOiB8Cgo=</div>


</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->
<script>
$(document).ready(function () {
  window.initializeSourceEmbed("06-MarkerVisualization.Rmd");
});
</script>

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3,h4",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
