<!DOCTYPE html>

<html lang="en" xml:lang="en">

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="UM Bioinformatics Core Workshop Team" />


<title>Exercises - Intro to Single-Cell Analysis</title>

<script src="site_libs/header-attrs-2.29/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/paper.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.13.2/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/default.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<script src="site_libs/clipboard-1.7.1/clipboard.min.js"></script>
<link href="site_libs/primer-tooltips-1.4.0/build.css" rel="stylesheet" />
<link href="site_libs/klippy-0.0.0.9500/css/klippy.min.css" rel="stylesheet" />
<script src="site_libs/klippy-0.0.0.9500/js/klippy.min.js"></script>
<!--
Favicon dervied from
https://twemoji.twitter.com/
https://favicon.io/
-->
<link rel="apple-touch-icon" sizes="180x180" href="/apple-touch-icon.png">
<link rel="icon" type="image/png" sizes="32x32" href="/favicon-32x32.png">
<link rel="icon" type="image/png" sizes="16x16" href="/favicon-16x16.png">
<link rel="shortcut icon" href="favicon-16x16.png" />
<link rel="manifest" href="/site.webmanifest">
<link rel="mask-icon" href="/safari-pinned-tab.svg" color="#5bbad5">
<meta name="msapplication-TileColor" content="#da532c">
<meta name="theme-color" content="#ffffff">


<meta property="og:url" content="https://michmed.org/re39e"/>
<meta property="og:title" content="Introduction to Single-Cell Analysis"/>
<meta property="og:description" content="A virtual workshop for researchers that details the analysis steps of single-cell RNA-Seq data using RStudio and Seurat." />

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>









<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-inverse  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">Intro to scRNA-Seq</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="workshop_intro.html">Intro</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Day 1
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="00A-OrientingOnScRNASeq.html">Orienting on scRNA-Seq</a>
    </li>
    <li>
      <a href="01-GettingStarted.html">Getting Started with Seurat</a>
    </li>
    <li>
      <a href="00B-CellRangerInAction.html">Cell Ranger in Action</a>
    </li>
    <li>
      <a href="02-QCandFiltering.html">Secondary QC &amp; Filtering</a>
    </li>
    <li>
      <a href="03-Normalization.html">Normalization</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Day 2
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="04-PCAandIntegration.html">PCA &amp; Integration</a>
    </li>
    <li>
      <a href="05-ProjectionAndClustering.html">Clustering &amp; Projection</a>
    </li>
    <li>
      <a href="clusters_faq.html">Clusters, PCs, and Resolutions: FAQ</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Day 3
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="06-MarkerVisualization.html">Marker identification and visualization</a>
    </li>
    <li>
      <a href="07-CellTypeAnnos.html">Cell type annotation</a>
    </li>
    <li>
      <a href="08-DifferentialExpression.html">Differential expression analysis</a>
    </li>
    <li>
      <a href="08A-AnalysisFinale.html">Analysis summary and next steps</a>
    </li>
  </ul>
</li>
<li>
  <a href="exercises.html">Exercises</a>
</li>
<li>
  <a href="workshop_wrap_up.html">Wrap up</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Exercises - Intro to Single-Cell
Analysis</h1>
<h4 class="author">UM Bioinformatics Core Workshop Team</h4>

</div>


<script>
  addClassKlippyTo("pre.r, pre.markdown, pre.bash");
  addKlippy('right', 'top', 'auto', '1', 'Copy code', 'Copied!');
</script>
<style type="text/css">

body, td {
   font-size: 18px;
}

.doNotRun {
background-color: CornSilk;
}

.pro_tip {
  border-radius: 10px;
  padding: 5px 5px 5px 95px;
  background: #EBF7FF left 10px top 10px / 85px no-repeat;
}

.pro_tip_icon {
  background-image: url("images/curriculum/pro-tip.png");
}

</style>
<div id="workflow-overview" class="section level1 unlisted unnumbered">
<h1 class="unlisted unnumbered">Workflow Overview</h1>
<p><br/> <img src="images/wayfinder/00-FullWayfinder.png" /> <br/>
<br/></p>
</div>
<div id="workshop-exercises" class="section level1">
<h1>Workshop exercises</h1>
<p>Build on the content and analysis steps covered in the workshop
sessions by working through these independent exercises. <strong>Note -
if you work on the exercises make sure to restart R session to clear out
environment before closing the window (like we have at the end of each
session) to avoid lags when logging in the next day</strong> <br/></p>
<hr />
<div id="reminder-rstudio-seurat-and-memory-management"
class="section level2 pro_tip pro_tip_icon">
<h2>Reminder: RStudio, Seurat, and Memory Management</h2>
<ul>
<li><p>R and RStudio are designed to grab computer memory (RAM) when
necessary and release that memory when it’s no longer needed. R does
this with a fancy computer science technique called <strong>garbage
collection</strong>.</p></li>
<li><p>The R <strong>garbage collector</strong> is very good for small
objects and simple analyses, but complex single-cell analysis can
overwhelm it. So <strong>don’t rely on R/RStudio’s built in garbage
collection and session management for single-cell analysis</strong>.
Instead, do the following:</p></li>
</ul>
<hr />
<ol style="list-style-type: decimal">
<li>Save intermediate versions of the Seurat object as you go along
using <code>saveRDS()</code>. This gives you “savepoints” so if you need
to backtrack, you don’t have to go back to the beginning. <br/></li>
<li>Save ggplots into image files as you go along. Periodically clear
ggplot objects from the R environment. <br/></li>
<li>At the end of each day, wherever you are in the analysis, explicitly
save the Seurat object and “power down” your session. <br/></li>
<li>The next day, you can load the Seruat object using readRDS().
<br/></li>
</ol>
<hr />
<p>This simple pattern will make your code more reproducible and your
RStudio session more reliable.</p>
</div>
<p><br/></p>
<div id="day-1" class="section level2">
<h2>Day 1</h2>
<!-- Add Rmd version of exercises with smaller code blocks -->
<p>You don’t need to have a fresh environment for these exercises. All
variable names will be different from the lessons. Load the libraries
(it’s okay if they’re already loaded) and the unfiltered Seurat object
with the following code:</p>
<pre class="r"><code># =========================================================================
# Independent Exercise - Day 1 Startup
# =========================================================================

# After restarting our R session, load the required libraries &amp; input data
library(Seurat)
library(BPCells)
library(tidyverse)

# Use provided copy of integrated data
exso = readRDS(&#39;./inputs/prepared_data/rdata/geo_so_unfiltered.rds&#39;)

# Add percent.mt column to meta.data
exso$percent.mt = PercentageFeatureSet(exso, pattern = &#39;^mt-&#39;)

## NOTE - BEFORE STOPPING WORK ON THE EXERCISES REMEMBER TO POWER DOWN AND RESTART R SESSION !!!!</code></pre>
<p>Let’s practice filtering cells based on different criteria than we’re
using in the lessons.</p>
<div id="day-1-exercise-1" class="section level3">
<h3>Day 1 Exercise 1</h3>
<p>Subset <code>exso</code> so that all cells have
<code>nFeature_RNA</code> greater than or equal to 1000. How many cells
remain?</p>
<p>Hint 1: For a reminder of how to subset a Seurat object, see the
“Removing low-quality cells” section of the <a
href="https://umich-brcf-bioinf.github.io/workshop-intro-single-cell/main/html/02-QCandFiltering.html">Secondary
QC and Filtering module</a>. Hint 2: For a reminder, of how to count the
number cells per sample, see the “Cell counts” section of the <a
href="https://umich-brcf-bioinf.github.io/workshop-intro-single-cell/main/html/02-QCandFiltering.html">Secondary
QC and Filtering module</a>.</p>
</div>
<div id="day-1-exercise-2" class="section level3">
<h3>Day 1 Exercise 2</h3>
<p>Subset <code>exso</code> so that all cells have
<code>nCount_RNA</code> greater than or equal to 5000. How many cells
remain?</p>
</div>
<div id="day-1-exercise-3" class="section level3">
<h3>Day 1 Exercise 3</h3>
<p>Subset <code>exso</code> so that all cells have
<code>percent.mt</code> less than 10%. How many cells remain?</p>
</div>
<div id="day-1-exercise-4" class="section level3">
<h3>Day 1 Exercise 4</h3>
<p>Subset <code>exso</code> so that each of the previous three
conditions are satisfied. How many cells remain?</p>
<!-- Use R script as key in hidden dropdown -->
<details>
<summary>
<em>Example Solutions - Day 1</em>
</summary>
<pre class="r"><code># =========================================================================
# Independent Exercise - Day 1 Startup
# =========================================================================

# After restarting our R session, load the required libraries &amp; input data
library(Seurat)
library(BPCells)
library(tidyverse)

# Load the unfiltered version and give it a new variable name
exso = readRDS(&#39;./inputs/prepared_data/rdata/geo_so_unfiltered.rds&#39;)

# Add percent.mt column to meta.data
exso$percent.mt = PercentageFeatureSet(exso, pattern = &#39;^mt-&#39;)

## NOTE - BEFORE STOPPING WORK ON THE EXERCISES REMEMBER TO POWER DOWN AND RESTART R SESSION !!!!

# Exercise 1
exso = subset(exso, nFeature_RNA &gt;= 1000)

ex1 = exso@meta.data %&gt;% count(orig.ident, name = &#39;postfilter_cells&#39;)
ex1

# Exercise 2
exso = subset(exso, nCount_RNA &gt;= 1000)

ex2 = exso@meta.data %&gt;% count(orig.ident, name = &#39;postfilter_cells&#39;)
ex2

# Exercise 3
exso = subset(exso, percent.mt &lt; 10)

ex3 = exso@meta.data %&gt;% count(orig.ident, name = &#39;postfilter_cells&#39;)
ex3

# Exercise 4
exso = subset(exso, nFeature_RNA &gt;= 1000 &amp; nCount_RNA &gt;= 1000 &amp; percent.mt &lt; 10)

ex4 = exso@meta.data %&gt;% count(orig.ident, name = &#39;postfilter_cells&#39;)
ex4</code></pre>
</details>
<p><br></p>
<hr />
</div>
</div>
<div id="day-2" class="section level2">
<h2>Day 2</h2>
<!-- link out to `clusters_faq.Rmd` to show examples? -->
<p>Since we are working with larger object sizes, it’s best to start
with a fresh session. All variable names will be different from the
lessons. Load the libraries (it’s okay if they’re already loaded) and
the <strong>integrated</strong> Seurat object with the following
code:</p>
<!-- Exploring clustering thresholds -->
<pre class="r"><code># =========================================================================
# Independent Exercise - Day 2 Startup
# =========================================================================

# After restarting our R session, load the required libraries &amp; input data
library(Seurat)
library(BPCells)
library(tidyverse)

# Use provided copy of integrated data
exso2 = readRDS(&#39;inputs/prepared_data/rdata/geo_so_sct_integrated.rds&#39;)
exso2 # check that object loaded

## NOTE - BEFORE STOPPING WORK ON THE EXERCISES REMEMBER TO POWER DOWN AND RESTART R SESSION !!!!</code></pre>
<p><br/></p>
<div id="day-2-exercise-1---clustering-with-reduced-number-of-pcs"
class="section level3">
<h3>Day 2 Exercise 1 - Clustering with reduced number of PCs</h3>
<p>Test how the clustering results would change if you used
<em>fewer</em> or <em>more</em> PCs when clustering these data.</p>
<pre class="r"><code># -------------------------------------------------------------------------
# Testing fewer PCs for clustering 

# look at elbow plot to check PCs and consider alternatives to the number selected for main materials
ElbowPlot(exso2, ndims = 50, reduction = &#39;unintegrated.sct.pca&#39;)

# select alternative value to try (choose a number &lt;10 or &gt;10 PCs)
pcs = # Your value here #

# generate nearest neighbor (graph), using selected number of PCs
exso2 = FindNeighbors(exso2, dims = 1:pcs, reduction = &#39;integrated.sct.rpca&#39;)</code></pre>
<p><br/></p>
<p>After generating clustering, set resolution parameter - remember this
only impacts the how the boundaries of the neighbors are drawn, not the
underlying NN graph/structure.</p>
<pre class="r"><code># -------------------------------------------------------------------------
# Testing resolution options to see impact

# start with one resolution
res = # Your value here #

# generate clusters, using `pcs` and `res` to make a custom cluster name that will be added to the metadata
exso2 = FindClusters(exso2, resolution = res, 
                     cluster.name = paste0(&#39;int.sct.rpca.clusters&#39;, res))

# look at meta.data to see cluster labels
head(exso2@meta.data)

# run UMAP reduction to prepare for plotting
exso2 = RunUMAP(exso2, dims = 1:pcs, 
                  reduction = &#39;integrated.sct.rpca&#39;, 
                  reduction.name = paste0(&#39;umap.integrated.sct.rpca_alt&#39;, res))

# check object to see if named reduction was added
exso2</code></pre>
<p><br/></p>
<p><strong>Challenge 1</strong>: How could we generate clusters for
multiple alternative resolutions (e.g. 0.4, 0.8, &amp; 1.0)?
<em>Hint</em> - <a
href="https://r4ds.had.co.nz/iteration.html"><code>for</code> loops</a>
can be useful for iteration in many programming languages, including
R.</p>
<p><br/></p>
</div>
<div id="day-2-exercise-2---plotting-alternative-clustering-results"
class="section level3">
<h3>Day 2 Exercise 2 - Plotting alternative clustering results</h3>
<p>After generating our clustering results it’s time to visualize them -
create a UMAP plot, ensuring that you generate a UMAP reduction for the
appropriate reduction</p>
<p><em>Hint 1</em> - Remember the previous code chunk used
<code>paste0('umap.integrated.sct.rpca_alt', res))</code> for the
cluster name parameter.</p>
<p><em>Hint 2</em> - For a reminder of how to generate a UMAP plot, look
at the <a
href="https://umich-brcf-bioinf.github.io/workshop-intro-single-cell/main/html/05-ProjectionAndClustering.html#Visualizing_and_evaluating_clustering">Visualizing
and evaluating clustering</a> from the Clustering and Projection
module.</p>
<p><br/></p>
<p><strong>Challenge 2</strong>: Create UMAP plots for the alternative
resolutions (e.g. 0.4, 0.8, &amp; 1.0) generated in the previous
challenge.</p>
<p><br/></p>
<p><strong>Check-in Questions - UMAP for alternative PCs</strong></p>
<blockquote>
<p>How does the UMAP plot look when fewer or more PCs are used? Do you
notice any relationship between the PC parameter and resolution
parameter?</p>
</blockquote>
<p><br/></p>
</div>
<div id="clean-up-session" class="section level3">
<h3>Clean up session</h3>
<p>Before closing out the window, <strong>make sure to clear environment
and restart R session to manage the memory usage</strong></p>
<pre class="r"><code>## ----------------------------------------------------------
## Clean up session, including any plot objects
rm(list=names(which(unlist(eapply(.GlobalEnv, is.ggplot)))));
gc()

## Save copy of Seurat object in current state to file
saveRDS(exso2, file = paste0(&#39;results/rdata/geo_so_sct_integrated_exercise.rds&#39;))
rm(exso2)
gc()</code></pre>
<p><br/> <br/></p>
<!-- Use R script as key in hidden dropdown -->
<details>
<summary>
<em>Example Solutions - Day 2</em>
</summary>
<pre class="r"><code># =========================================================================
# Independent Exercise - Day 2 Startup
# =========================================================================

# After restarting our R session, load the required libraries &amp; input data
library(Seurat)
library(BPCells)
library(tidyverse)

# Use provided copy of integrated data
exso2 = readRDS(&#39;inputs/prepared_data/rdata/geo_so_sct_integrated.rds&#39;)
exso2 # check that object loaded


### Day 2 Exercise 1 - Clustering with reduced number of PCs
# -------------------------------------------------------------------------
# Testing fewer PCs for clustering 

# look at elbow plot to check PCs and consider alternatives to the number selected for main materials
ElbowPlot(exso2, ndims = 50, reduction = &#39;unintegrated.sct.pca&#39;)

# select alternative value to try (can choose a number &lt;10 or &gt;10 PCs)
pcs = 6

# generate nearest neighbor (graph), using selected number of PCs
exso2 = FindNeighbors(exso2, dims = 1:pcs, reduction = &#39;integrated.sct.rpca&#39;)

# -------------------------------------------------------------------------
# start with one resolution
res = 0.2

# generate clusters, using `pcs` and `res` to make a custom cluster name that will be added to the metadata
exso2 = FindClusters(exso2, resolution = res, 
                     cluster.name = paste0(&#39;int.sct.rpca.clusters&#39;, res))

# look at meta.data to see cluster labels
head(exso2@meta.data)

# run UMAP reduction to prepare for plotting
exso2 = RunUMAP(exso2, dims = 1:pcs, 
                reduction = &#39;integrated.sct.rpca&#39;, 
                reduction.name = paste0(&#39;umap.integrated.sct.rpca_alt&#39;, res))

# check object to see if named reduction was added
exso2


## Challenge 1 - solution option
# -------------------------------------------------------------------------
# use for loop to generate clustering with alternative resolutions
for(i in c(0.4, 0.8, 1.0)){
  exso2 = FindClusters(exso2, resolution = i, 
                       cluster.name = paste0(&#39;int.sct.rpca.clusters&#39;, i))
  
  # look at meta.data to see cluster labels
  head(exso2@meta.data)
  
  # run UMAP reduction to prepare for plotting
  exso2 = RunUMAP(exso2, dims = 1:pcs, 
                  reduction = &#39;integrated.sct.rpca&#39;, 
                  reduction.name = paste0(&#39;umap.integrated.sct.rpca_alt&#39;, i))
  
  # check object to see if multiple cluster resolutions are added
  head(exso2@meta.data)
}



### Day 2 Exercise 2 - Plotting alternative clustering results
# -------------------------------------------------------------------------
# plot clustering results
post_integration_umap_clusters_testing = 
  DimPlot(exso2, group.by = paste0(&#39;int.sct.rpca.clusters&#39;, res), label = TRUE, 
          reduction = paste0(&#39;umap.integrated.sct.rpca_alt&#39;, res)) + NoLegend()
post_integration_umap_clusters_testing # look at plot

# output to file, including the number of PCs and resolution used to generate results
ggsave(filename = paste0(&#39;results/figures/umap_int_sct_clusters_exercise_&#39;, 
                         pcs,&#39;PC.&#39;,res,&#39;res&#39;,&#39;.png&#39;),
       plot = post_integration_umap_clusters_testing, 
       width = 8, height = 6, units = &#39;in&#39;)


## Challenge 2 - solution option
# -------------------------------------------------------------------------
# use for loop to visualize clustering across tested resolutions
post_integration_umap_plots &lt;- c()
for(i in c(0.4, 0.8, 1.0)){
  res_type = paste0(&quot;res_&quot;, i)
  post_integration_umap_plots[[res_type]] = 
    DimPlot(exso2, group.by = paste0(&#39;int.sct.rpca.clusters&#39;, i), label = TRUE, 
            reduction = paste0(&#39;umap.integrated.sct.rpca_alt&#39;, i)) + NoLegend()
}

# look at plots for each resolution stored in list
post_integration_umap_plots
# remove plot list to clean up session 
rm(post_integration_umap_plots)


## ----------------------------------------------------------
## Clean up session, including any plot objects
rm(list=names(which(unlist(eapply(.GlobalEnv, is.ggplot)))));
gc()

## Save copy of Seurat object in current state to file
saveRDS(exso2, file = paste0(&#39;results/rdata/geo_so_sct_clustered_exercise.rds&#39;))
rm(exso2)
gc()</code></pre>
</details>
<p><br></p>
<hr />
</div>
</div>
<div id="day-3" class="section level2">
<h2>Day 3</h2>
<!-- Add Rmd version of exercises with smaller code blocks -->
<p>Since we are working with larger object sizes, it’s best to start
with a fresh session. All variable names will be different from the
lessons. Load the libraries (it’s okay if they’re already loaded) and a
Seurat object with alternative clustering using the following code:
<!-- add copy of alt clustering to share with learners --></p>
<pre class="r"><code># =========================================================================
# Independent Exercise - Day 3 Startup
# =========================================================================

# After restarting our R session, load the required libraries
library(Seurat)
library(BPCells)
library(tidyverse)
library(scCATCH)

# Load in seurat object with alternative clustering results from yesterday&#39;s exercises
exso3 = readRDS(&#39;results/rdata/geo_so_sct_integrated_exercise.rds&#39;)
exso3 # check that object loaded

## NOTE - BEFORE STOPPING WORK ON THE EXERCISES REMEMBER TO POWER DOWN AND RESTART R SESSION !!!!</code></pre>
<p><br/></p>
<p>Before testing what marker genes and cell-type predictions are
generated for an alternative clustering, start by ensuring the Seurat
object is set up and has an expected identity set.</p>
<pre class="r"><code>## ----------------------------------------------------------
# Check what identities are set
Idents(exso3) %&gt;% head()

pcs = ## use same values as yesterday
res = ## use same values as yesterday

## Set identities to clustering for selected resolution
exso3 = SetIdent(exso3, value = paste0(&#39;int.sct.rpca.clusters&#39;, res))
Idents(exso3) %&gt;% head()</code></pre>
<div
id="day-3-exercise-1---examine-marker-genes-for-alternative-clusters"
class="section level3">
<h3>Day 3 Exercise 1 - examine marker genes for alternative
clusters</h3>
<p>Using the approach from the <a
href="https://umich-brcf-bioinf.github.io/workshop-intro-single-cell/main/html/06-MarkerVisualization.html#Marker_identification">marker
identification</a> module, run <code>PrepSCTFindMarkers</code> and
generate markers for the alternative clustering results. Bonus to
generate a table of the top 5 markers and outputing that to file.</p>
<blockquote>
<p>Check-in Question - clustering impact on marker genes After you
generate markers for the “fewer” <code>pcs</code> option, how do the
results differ from the markers found for <code>pcs=10</code> in the
workshop session? What do you think would happen if you tried this with
the “more” results?</p>
</blockquote>
<p><br/></p>
</div>
<div
id="day-3-exercise-2---generate-predictions-for-alternative-clustering"
class="section level3">
<h3>Day 3 Exercise 2 - Generate predictions for alternative
clustering</h3>
<p>Using <a
href="https://umich-brcf-bioinf.github.io/workshop-intro-single-cell/main/html/07-CellTypeAnnos.html#Refine_annotations_and_label_clusters">scCATCH
like we did for our annotation process</a>, generate predictions for the
alternative clustering results.</p>
<p><br/></p>
</div>
<div id="day-3-exercise-3---add-predictions-to-seurat-object"
class="section level3">
<h3>Day 3 Exercise 3 - Add predictions to Seurat object</h3>
<p>Skipping the refinement step, add the cell-type predictions to the
Seurat object using the numeric labels as a key, similarly to the our <a
href="file:///Users/damki/Desktop/ActiveProjectsMount/damki/repos/workshop-intro-single-cell/html/07-CellTypeAnnos.html#Refine_annotations_and_label_clusters">annotation
step</a>.</p>
<p><br/></p>
</div>
<div id="day-3-exercise-4---plot-umap-with-new-cluster-labels"
class="section level3">
<h3>Day 3 Exercise 4 - Plot UMAP with new cluster labels</h3>
<p>Using a similar approach used to <a
href="https://umich-brcf-bioinf.github.io/workshop-intro-single-cell/main/html/07-CellTypeAnnos.html">visualize
the annotations in the workshop</a>, generate a UMAP plot</p>
<p><em>Hint</em> - use the same reduction name as yesterday’s exercise
and make sure the <code>group_by</code> parameter matches the column
name added to the metadata that stores the predicted cell-types.</p>
<p><br/></p>
<p><strong>Check-in Question</strong></p>
<blockquote>
<p>How did the number of pcs and/or resolution change the predictions?
Do you think the predictions correspond better or worse to the cluster
structure we see in the UMAP?</p>
</blockquote>
<p></br></p>
</div>
<div id="clean-up-session-1" class="section level3">
<h3>Clean up session</h3>
<p>Before closing out the window, <strong>make sure to clear environment
and restart R session to manage the memory usage</strong></p>
<pre class="r"><code>## ----------------------------------------------------------
## Clean up session 
rm(list=names(which(unlist(eapply(.GlobalEnv, is.ggplot))))); 
rm(catch_celltypes, catch_markers, geo2_catch, geo2_markers, new_metadata, top_5); 
gc()

## (Optional) - Save copy of exso3
saveRDS(exso3, file = paste0(&#39;results/rdata/geo_so_sct_integrated_with_markers_exercise.rds&#39;))

## BEFORE CLOSING WINDOW - POWER DOWN AND RESTART R SESSION</code></pre>
<p><br/></p>
<!--- hidden for now - additional DE comparisons
### Day 3 Challenge - Run additional DE comparisons using session results


``` r
## BEFORE PROCEEDING, MAKE SURE SESSION HAS BEEN restarted and environment is clear

# After restarting our R session, load the required libraries
library(Seurat)
library(BPCells)
library(tidyverse)

# load a copy of the final Seurat object (from end of Day 3)
exso3 = readRDS('/home/workshop/damki/ISC_R/results/rdata/exso3_sct_integrated_final.rds')

# check what identities are set for the loaded Seurat object
Idents(exso3) # expect to see Day + Cluster labels; if different identities are set, how would you change them?
## add code to change identities if needed
```

Once the cluster identities are set to the combined Day + Cluster labels, generate additional DE comparisons using the provided "psuedocode" which are comments that outline what steps are expected/needed to guide adding code to complete those outlined steps.

``` r
## ---------------------------------------------------
## Further extension - how might you generate DE comparisons between D21 and D7 for all annotated clusters?

# check the unique cluster labels - how many clusters are present?
unique(exso3$cell_type)

## Below is comments to help guide you though the process of looping through all the comparisons of interest, for all labeled clusters
## This section from the HBC materials could be a helpful reference to aid in filling in the code: https://hbctraining.github.io/Intro-to-scRNAseq/lessons/pseudobulk_DESeq2_scrnaseq.html

# Create list(s) of pair-wise comparisons of interest between conditions, ideally specifying Case and Control 
# (e.g. Day 7 vs Day 0, Day 21 vs Day 7, Day 21 vs Day 0)

# Use list of comparisons and cluster identities to loop through comparisons of interest (for standard DE comparisons) and save results to files
# Try to use an outer `for` loop step though the comparisons listed in previous step and an inner `for` loop to cycle through all the clusters

# Bonus - summarize the DE results for the comparison, choosing a reasonable set of thresholds to call the number of DE genes, & save summary to file

# Clean up session. Caution - this next command clears your environment of all objects
rm(list = ls())
gc()

## BEFORE CLOSING WINDOW - POWER DOWN AND RESTART R SESSION !!!!
```
-->
<p></br> <br/></p>
<!-- Use R script as key in hidden dropdown -->
<details>
<summary>
<em>Example Solutions - Day 3</em>
</summary>
<pre class="r"><code># =========================================================================
# Independent Exercise - Day 3 Startup
# =========================================================================

# After restarting our R session, load the required libraries
library(Seurat)
library(BPCells)
library(tidyverse)
library(scCATCH)

# Load in seurat object with alternative clustering results from yesterday&#39;s exercises
exso3 = readRDS(&#39;results/rdata/geo_so_sct_integrated_exercise.rds&#39;)
exso3 # check that object loaded

## NOTE - BEFORE STOPPING WORK ON THE EXERCISES REMEMBER TO POWER DOWN AND RESTART R SESSION !!!!


### Day 3 Exercise 1 - examine marker genes and cell type predictions for alternative clustering
## ----------------------------------------------------------
# Check what identities are set
Idents(exso3) %&gt;% head()

## use same values as previous exercise
pcs = 6
res = 0.2 

## Set identities to clustering for selected resolution
exso3 = SetIdent(exso3, value = paste0(&#39;int.sct.rpca.clusters&#39;, res))
Idents(exso3) %&gt;% head()


## ----------------------------------------------------------
## Generate cluster markers to see how that changes with new parameters 
exso3 = PrepSCTFindMarkers(exso3, assay = &quot;SCT&quot;) # NOTE - this step will take some time to run
exso3_markers = FindAllMarkers(exso3, only.pos = TRUE)
head(exso3)

# Create table of top 5 markers per cluster (using default ranking)
top_5 = exso3_markers %&gt;% filter(p_val_adj &lt; 0.01) %&gt;% group_by(cluster) %&gt;% slice_head(n = 5)
head(top_5, n = 10) # look at results
write_csv(top_5, file = paste0(&#39;results/tables/top5_marker_genes_exercise.csv&#39;))


#### Day 3 Exercise 2 - Generate predictions for alternative clustering 
## ----------------------------------------------------
## Next - run scCATCH predictions for alternative clustering results
exso3_catch = createscCATCH(data = exso3@assays$SCT@counts, cluster = as.character(Idents(exso3)))
catch_markers = exso3_markers %&gt;% rename(&#39;logfc&#39; = &#39;avg_log2FC&#39;)
exso3_catch@markergene = exso3_markers
exso3_catch@marker = cellmatch[cellmatch$species == &#39;Mouse&#39; &amp; cellmatch$tissue %in% c(&#39;Blood&#39;, &#39;Peripheral Blood&#39;, &#39;Muscle&#39;, &#39;Skeletal muscle&#39;, &#39;Epidermis&#39;, &#39;Skin&#39;), ]
exso3_catch = findcelltype(exso3_catch)

# Check predictions
exso3_catch@celltype %&gt;% select(cluster, cell_type, celltype_score)


#### Day 3 Exercise 3 - Add predictions to Seurat object
## ------------------------------------------------------
## Use predictions to label clusters and UMAP plot
catch_celltypes = exso3_catch@celltype %&gt;% select(cluster, cell_type)
colnames(catch_celltypes)[2] = paste0(&#39;cell_type.&#39;,pcs,&#39;PC.&#39;,res,&#39;res&#39;)
new_metadata = exso3@meta.data %&gt;% 
  left_join(catch_celltypes, 
            by = c(&#39;seurat_clusters&#39; = &#39;cluster&#39;)) # using `seurat_clusters`, which will store the most recently generated cluster labels for each cell
rownames(new_metadata) = rownames(exso3@meta.data) #  We are implicitly relying on the same row order!
exso3@meta.data = new_metadata # Replace the meta.data
head(exso3@meta.data)


#### Day 3 Exercise 4 - Plot UMAP with new cluster labels
## ------------------------------------------------------
catch_umap_plot = DimPlot(exso3, group.by = paste0(&#39;cell_type.&#39;,pcs,&#39;PC.&#39;,res,&#39;res&#39;), 
                          label = TRUE, reduction = paste0(&#39;umap.integrated.sct.rpca_alt&#39;, res))
catch_umap_plot


# Save the plot to file
# output to file, including the number of PCs and resolution used to generate results
ggsave(filename = paste0(&#39;results/figures/umap_int_catch-labeled_&#39;, 
                         pcs,&#39;PC.&#39;,res,&#39;res&#39;,&#39;.png&#39;),
       plot = catch_umap_plot, 
       width = 8, height = 6, units = &#39;in&#39;)


## ----------------------------------------------------------
## Clean up session 
rm(list=names(which(unlist(eapply(.GlobalEnv, is.ggplot))))); 
rm(catch_celltypes, catch_markers, exso3_catch, exso3_markers, new_metadata, top_5); 
gc()

## (Optional) - Save copy of exso3
saveRDS(exso3, file = paste0(&#39;results/rdata/geo_so_sct_integrated_with_markers_exercise.rds&#39;))
rm(exso3)
gc()

## BEFORE PROCEEDING TO THE NEXT SECTION or closing window - POWER DOWN AND RESTART R SESSION</code></pre>
</details>
<p><br></p>
</div>
</div>
</div>



</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3,h4",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
