<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="UM Bioinformatics Core" />


<title>Intro to Single Cell RNA-Seq Workshop</title>

<script src="site_libs/header-attrs-2.25/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/paper.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.13.2/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/default.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<!--
Favicon dervied from
https://twemoji.twitter.com/
https://favicon.io/emoji-favicons/dna/
-->
<link rel="apple-touch-icon" sizes="180x180" href="/apple-touch-icon.png">
<link rel="icon" type="image/png" sizes="32x32" href="/favicon-32x32.png">
<link rel="icon" type="image/png" sizes="16x16" href="/favicon-16x16.png">
<link rel="shortcut icon" href="favicon-16x16.png" />
<link rel="manifest" href="/site.webmanifest">
<link rel="mask-icon" href="/safari-pinned-tab.svg" color="#5bbad5">
<meta name="msapplication-TileColor" content="#da532c">
<meta name="theme-color" content="#ffffff">

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>









<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-inverse  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">Intro to scRNA-Seq</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="workshop_intro.html">Intro</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Day 1
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="00A-OrientingOnScRNASeq.html">Orienting on scRNA-Seq</a>
    </li>
    <li>
      <a href="01-GettingStarted.html">Getting Started with Seurat</a>
    </li>
    <li>
      <a href="00B-CellRangerInAction.html">Cell Ranger in Action</a>
    </li>
    <li>
      <a href="02-QCandFiltering.html">Secondary QC &amp; Filtering</a>
    </li>
    <li>
      <a href="03-Normalization.html">Normalization</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Day 2
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="04-PCAandIntegration.html">PCA &amp; Integration</a>
    </li>
    <li>
      <a href="05-ProjectionAndClustering.html">Projection &amp; Clustering</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Day 3
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="06-MarkerVisualization.html">Marker identification and visualization</a>
    </li>
    <li>
      <a href="07-CellTypeAnnos.html">Cell type annotation</a>
    </li>
    <li>
      <a href="08-DifferentialExpression.html">Differential expression analysis</a>
    </li>
  </ul>
</li>
<li>
  <a href="workshop_wrap_up.html">Wrap up</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Intro to Single Cell RNA-Seq Workshop</h1>
<h4 class="author">UM Bioinformatics Core</h4>

</div>


<style type="text/css">

body, td {
   font-size: 18px;
}
code.r{
  font-size: 12px;
}
pre {
  font-size: 12px
}

a.external {
    background: url(images/external-link.png) center right no-repeat;
    padding-right: 13px;
}
</style>
<div id="wrapping-up" class="section level2">
<h2>Wrapping up</h2>
<p>We hope you now have more familiarity with key concepts, data types,
tools, and how they all connect to enable single-cell gene expression
analysis from RNA-Seq data.</p>
<p><img src="images/wayfinder/wayfinder_workshop_wrap_up.png" /></p>
<hr />
</div>
<div id="housekeeping" class="section level2">
<h2>Housekeeping</h2>
<ul>
<li><p>Please take our optional <a
href="https://forms.gle/mafY39iccNRoEAVZ8" target="_blank">post-workshop
survey</a> (5-10 minutes).</p></li>
<li><p>Sign up here to attend <a
href="https://forms.gle/9ZxC3zj8EuFYXeos5" target="_blank">virtual
office hours</a> on 4/22 1-2 ET.</p></li>
<li><p>We will email you a link to the final session recordings by the
end of next week.</p></li>
<li><p>The website/notes for this workshop will be available.</p></li>
<li><p>The <a href="https://umbioinfcoreworkshops.slack.com"
target="_blank">UM Bioinformatics Core Workshop Slack channel</a>
content will be available for 90 days.</p></li>
</ul>
<hr />
</div>
<div id="looking-ahead" class="section level2">
<h2>Looking ahead</h2>
<div id="workshop-environment" class="section level4">
<h4>Workshop environment</h4>
<ul>
<li><p>RStudio workshop compute environment will be available until
<strong>5/1/2024</strong>.</p>
<ul>
<li><em>Please save all your R scripts now</em> so that we can
“right-size” the compute environment immediately following today’s
workshop session.</li>
</ul></li>
<li><p>You can download files from the workshop environment from your
terminal/command line window as below. (You will need to substitute your
actual workshop username and type workshop password when prompted.)</p>
<pre><code>mkdir intro_scrnaseq-workshop
cd intro_scrnaseq-workshop
scp -r YOUR_USERNAME@bfx-workshop01.med.umich.edu:&quot;ISC_R*&quot; .</code></pre>
<ul>
<li>Note that the full download of the R data is about 8Gb, so depending
on your internet speeds it could take a while. (We do not recommend you
download the full set of Cell Ranger outputs.)</li>
</ul></li>
</ul>
</div>
<div id="installing-software-locally" class="section level4">
<h4>Installing software locally</h4>
<ul>
<li>You can install necessary programs to run programs locally. Note
that for typical data, Cell Ranger steps (reviewed Day 1) assume your
computer has powerful compute (many CPUs and lots of RAM) and sizable
storage capacity. (i.e. it’s impractical to run these on your
laptop.)</li>
<li>Installing bioinformatics software is non-trivial and comprehensive
instructions to setup a complete compute environment are outside the
scope of this workshop. (We are considering offering a Computational
Reproducibility workshop that would cover this.) For the intrepid, see
relevant links below:
<ul>
<li><a href="workshop_setup/setup_instructions_advanced.html">Setting up
R/R-Studio</a></li>
</ul></li>
</ul>
<details>
<summary>
<em>University of Michigan Resources</em>
</summary>
<ul>
<li><a href="https://datascience.isr.umich.edu/events/coderspaces/"
target="_blank">UM CoderSpaces “office hours”</a> and UM CoderSpaces
Slack workspace. <em>(See “Useful Resources” section of the CoderSpaces
page for instructions on how to join the CoderSpaces Slack
workspace.)</em></li>
<li>Upcoming <a href="https://arc.umich.edu/events/" target="_blank">UM
Advanced Research Computing workshops</a>.</li>
<li>Advanced Research Computing (ARC) at University of Michigan hosts a
high-performance computing (HPC) platform called <em>Great Lakes</em>
which combines high-end computers, fast/resilient storage, and
pre-installed software. Great Lakes may be a good resource for folks who
need to run the more compute intensive steps and a substantial block of
compute and storage is subsidized by ARC making it essentially free to
many UM researchers.
<ul>
<li>About <a href="https://arc.umich.edu/greatlakes/"
target="_blank">Great Lakes</a>.</li>
<li>About the ARC <a href="https://arc.umich.edu/umrcp/"
target="_blank">Research Computing Package</a>.</li>
<li>Videos on <a
href="https://www.mivideo.it.umich.edu/channel/ARC-TS%2BTraining/181860561/"
target="_blank">getting started with Great Lakes</a>. (Videos are
available to UM folks.)</li>
</ul></li>
</ul>
</div>
</div>
<div id="resources-for-continued-learning" class="section level2">
<h2>Resources for continued learning</h2>
<ul>
<li>Learning bioinformatic analyses is more like a process than a task.
Resources to consider:
<ul>
<li>A CZI NDCN curated list of resources for <a
href="https://ndcn.github.io/self-guided-learning/">self-guided
learning</a></li>
<li>UM Bioinformatics Core links to <a
href="https://brcf.medicine.umich.edu/cores/bioinformatics-core/training/"
target="_blank">training resources/workshops</a>.</li>
<li><strong>Techniques in Bash and R</strong>
<ul>
<li>Lesson plans from the UM Bioinformatics Core’s recent <a
href="https://umich-brcf-bioinf.github.io/workshop-computational-foundations/main/html/"
target="_blank">Computational Foundations Workshop</a>.</li>
<li>Lessons and workshops in Bash / Git / R / Python :
<a href="https://software-carpentry.org/lessons/" target="_blank">
Software Carpentry</a></li>
</ul></li>
<li><strong>More info on NGS analysis</strong> (A decidedly
non-exhaustive list in no discernible order.)
<ul>
<li>Lesson plans from UM Bioinformatics Core’s recent <a
href="https://umich-brcf-bioinf.github.io/workshop-rnaseq-demystified/main/html/"
target="_blank">RNA-Seq Demystified Workshop</a>. (Note this is for bulk
RNA-Seq.)</li>
<li>Training materials from <a
href="https://hbctraining.github.io/main/#advanced-topics-analysis-of-high-throughput-sequencing-ngs-data"
target="_blank">Harvard Chan Bioinformatics Core</a>.</li>
<li><a href="https://bioconductor.org/books/3.12/OSCA/index.html"
target="_blank">Orchestrating Single-Cell Analysis with
Bioconductor</a>.</li>
<li>Welcome Sanger Instutute’s <a
href="https://www.singlecellcourse.org/index.html"
target="_blank">Analysis of single cell RNA-seq data</a>.</li>
<li>In addition to maintaining the Seurat analysis platform, New York
Genome Center’s <a href="https://satijalab.org/" target="_blank">Satija
lab</a> publishes many excellent step-by-step tutorials on single-cell
analysis, e.g. <a
href="https://satijalab.org/seurat/articles/pbmc3k_tutorial"
target="_blank">Seurat - Guided Clustering Tutorial</a>. The Satija lab
also hosts <a href="https://satijalab.org/scgd24/"
target="_blank">Single Cell Genomics Day</a>, a live-streamed annual
symposium on emerging techniques in single-cell analysis.</li>
<li>Nayak, Richa, and Yasha Hasija. <strong>“A hitchhiker’s guide to
single-cell transcriptomics and data analysis pipelines.”</strong>
Genomics vol. 113,2 (2021): 606-619.<br/> <a
href="https://pubmed.ncbi.nlm.nih.gov/33485955/"
class="uri">https://pubmed.ncbi.nlm.nih.gov/33485955/</a></li>
<li>Luecken, Malte D, and Fabian J Theis. <strong>“Current best
practices in single-cell RNA-seq analysis: a tutorial.”</strong>
Molecular systems biology vol. 15,6 e8746. 19 Jun. 2019.<br/> <a
href="https://pubmed.ncbi.nlm.nih.gov/31217225/"
class="uri">https://pubmed.ncbi.nlm.nih.gov/31217225/</a></li>
<li>He, Jiangping et al. <strong>“Practical bioinformatics pipelines for
single-cell RNA-seq data analysis.”</strong> Biophysics reports vol. 8,3
(2022): 158-169.<br/> <a
href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC10189648/"
class="uri">https://www.ncbi.nlm.nih.gov/pmc/articles/PMC10189648/</a></li>
</ul></li>
<li>For more context on the experiments and data we referenced during
the workshop, see the source paper: <br/> Sorkin, Michael et
al. <strong>“Regulation of heterotopic ossification by monocytes in a
mouse model of aberrant wound healing.”</strong> Nature communications
vol. 11,1 722. 5 Feb. 2020.<br/>
<a class="external" href="https://pubmed.ncbi.nlm.nih.gov/32024825/" target="_blank">https://pubmed.ncbi.nlm.nih.gov/32024825/</a></li>
</ul></li>
</ul>
<hr />
</div>
<div id="thank-you-to-our-sponsors" class="section level2">
<h2>Thank you to our sponsors</h2>
<p><img src="images/workshop_intro/sponsor_logos.png" /></p>
<p><br/></p>
</div>
<div id="thank-you-tofrom-the-workshop-team" class="section level2">
<h2>Thank you to/from the workshop team</h2>
<table>
<colgroup>
<col width="25%" />
<col width="25%" />
<col width="25%" />
<col width="25%" />
</colgroup>
<thead>
<tr class="header">
<th align="center"><img
src="images/headshots/headshot_cgates.jpg" /></th>
<th align="center"><img
src="images/headshots/headshot_mbradenb.jpg" /></th>
<th align="center"><img
src="images/headshots/headshot_rcavalca.jpg" /></th>
<th align="center"><img
src="images/headshots/headshot_damki.jpg" /></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="center"><strong>Chris (Gates)</strong></td>
<td align="center"><strong>Marci</strong></td>
<td align="center"><strong>Raymond</strong></td>
<td align="center"><strong>Dana</strong></td>
</tr>
<tr class="even">
<td align="center"><img
src="images/headshots/headshot_ncarruth.jpg" /></td>
<td align="center"><img
src="images/headshots/headshot_trsaari.jpg" /></td>
<td align="center"><img
src="images/headshots/headshot_csifuentes.jpg" /></td>
<td align="center"></td>
</tr>
<tr class="odd">
<td align="center"><strong>Nick</strong></td>
<td align="center"><strong>Travis</strong></td>
<td align="center"><strong>Chris (Sifuentes)</strong></td>
<td align="center"></td>
</tr>
<tr class="even">
<td align="center"><img
src="images/headshots/headshot_mkdohert.jpg" /></td>
<td align="center"><img
src="images/headshots/headshot_okoues.jpg" /></td>
<td align="center"><img
src="images/headshots/headshot_phillipa.jpg" /></td>
<td align="center"></td>
</tr>
<tr class="odd">
<td align="center"><strong>Matt</strong></td>
<td align="center"><strong>Olivia</strong></td>
<td align="center"><strong>Tricia</strong></td>
<td align="center"></td>
</tr>
</tbody>
</table>
<p><br/></p>
<p>Thank you for participating in our workshop. We welcome your
questions and feedback now and in the future.</p>
<p>Bioinformatics Workshop Team</p>
<p><a
href="mailto:bioinformatics-workshops@umich.edu">bioinformatics-workshops@umich.edu</a>
<br/> <a
href="https://medresearch.umich.edu/office-research/about-office-research/biomedical-research-core-facilities/bioinformatics-core"
target="_blank">UM BRCF Bioinformatics Core</a></p>
</div>



</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3,h4",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
