<!DOCTYPE html>

<html lang="en" xml:lang="en">

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="UM Bioinformatics Core Workshop Team" />


<title>Intro to Single Cell RNA-Seq Workshop</title>

<script src="site_libs/header-attrs-2.28/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/paper.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.13.2/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/default.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<script src="site_libs/clipboard-1.7.1/clipboard.min.js"></script>
<link href="site_libs/primer-tooltips-1.4.0/build.css" rel="stylesheet" />
<link href="site_libs/klippy-0.0.0.9500/css/klippy.min.css" rel="stylesheet" />
<script src="site_libs/klippy-0.0.0.9500/js/klippy.min.js"></script>
<!--
Favicon dervied from
https://twemoji.twitter.com/
https://favicon.io/emoji-favicons/dna/
-->
<link rel="apple-touch-icon" sizes="180x180" href="apple-touch-icon.png">
<link rel="icon" type="image/png" sizes="32x32" href="favicon-32x32.png">
<link rel="icon" type="image/png" sizes="16x16" href="favicon-16x16.png">
<link rel="shortcut icon" href="favicon-16x16.png" />
<link rel="manifest" href="site.webmanifest">
<link rel="mask-icon" href="safari-pinned-tab.svg" color="#5bbad5">
<meta name="msapplication-TileColor" content="#da532c">
<meta name="theme-color" content="#ffffff">

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>









<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-inverse  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">Intro to scRNA-Seq</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="workshop_intro.html">Intro</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Day 1
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="00A-OrientingOnScRNASeq.html">Orienting on scRNA-Seq</a>
    </li>
    <li>
      <a href="01-GettingStarted.html">Getting Started with Seurat</a>
    </li>
    <li>
      <a href="00B-CellRangerInAction.html">Cell Ranger in Action</a>
    </li>
    <li>
      <a href="02-QCandFiltering.html">Secondary QC &amp; Filtering</a>
    </li>
    <li>
      <a href="03-Normalization.html">Normalization</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Day 2
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="04-PCAandIntegration.html">PCA &amp; Integration</a>
    </li>
    <li>
      <a href="05-ProjectionAndClustering.html">Projection &amp; Clustering</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Day 3
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="06-MarkerVisualization.html">Marker identification and visualization</a>
    </li>
    <li>
      <a href="07-CellTypeAnnos.html">Cell type annotation</a>
    </li>
    <li>
      <a href="08-DifferentialExpression.html">Differential expression analysis</a>
    </li>
  </ul>
</li>
<li>
  <a href="workshop_wrap_up.html">Wrap up</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Intro to Single Cell RNA-Seq Workshop</h1>
<h4 class="author">UM Bioinformatics Core Workshop Team</h4>

</div>


<style type="text/css">

body, td {
   font-size: 18px;
}
code.r{
  font-size: 12px;
}
pre {
  font-size: 12px
}

</style>
<script>
  addClassKlippyTo("pre.r, pre.markdown, pre.bash");
  addKlippy('right', 'top', 'auto', '1', 'Copy code', 'Copied!');
</script>
<div id="wrapping-up" class="section level2">
<h2>Wrapping up</h2>
<p>We hope you now have more familiarity with key concepts, data types,
tools, and how they all connect to enable single-cell gene expression
analysis from RNA-Seq data.</p>
<p><img src="images/wayfinder/00-FullWayfinder.png"
title="Wayfinder: Wrap-up" /></p>
<hr />
</div>
<div id="example-methods" class="section level2">
<h2>Example methods</h2>
<details>
<summary>
<strong>Single cell RNA sequencing using 10× genomics</strong>
</summary>
<p>The example dataset used in the workshop was inspired by this Sorkin
et al. and the experimental methods below were excerpted directly from
that paper:</p>
<ul>
<li>Sorkin, Michael et al. <strong>“Regulation of heterotopic
ossification by monocytes in a mouse model of aberrant wound
healing.”</strong> Nature communications vol. 11,1 722. 5
Feb. 2020.<br/>
<a class="external" href="https://pubmed.ncbi.nlm.nih.gov/32024825/" target="_blank">https://pubmed.ncbi.nlm.nih.gov/32024825/</a></li>
</ul>
<p>Tissues harvested from the extremity injury site were digested for
45 min in 0.3% Type 1 Collagenase and 0.4% Dispase II (Gibco) in Roswell
Park Memorial Institute (RPMI) medium at 37 °C under constant agitation
at 120 rpm. Digestions were subsequently quenched with 10% FBS RPMI and
filtered through 40μm sterile strainers. Cells were then washed in PBS
with 0.04% BSA, counted and resuspended at a concentration of ~1000
cells/μl. Cell viability was assessed with Trypan blue exclusion on a
Countess II (Thermo Fisher Scientific) automated counter and only
samples with &gt;85% viability were processed for further
sequencing.</p>
University of Michigan Biomedical Research Core Facilities Advanced
Genomics Core generated single-cell 3’ libraries on the 10× Genomics
Chromium Controller following the manufacturers protocol for the v2
reagent kit (10× Genomics, Pleasanton, CA, USA). Cell suspensions were
loaded onto a Chromium Single-Cell A chip along with reverse
transcription (RT) master mix and single cell 3’ gel beads, aiming for
2000–6000 cells per channel. In this experiment, 8700 cells were
encapsulated into emulsion droplets at a concentration of 700–1200
cells/ul which targets 5000 single cells with an expected multiplet rate
of 3.9%. Following generation of single-cell gel bead-in-emulsions
(GEMs), reverse transcription was performed and the resulting Post
GEM-RT product was cleaned up using DynaBeads MyOne Silane beads (Thermo
Fisher Scientific, Waltham, MA, USA). The cDNA was amplified, SPRIselect
(Beckman Coulter, Brea, CA, USA) cleaned and quantified then
enzymatically fragmented and size selected using SPRIselect beads to
optimize the cDNA amplicon size prior to library construction. An
additional round of double-sided SPRI bead cleanup is performed after
end repair and A-tailing. Another single-sided cleanup is done after
adapter ligation. Indexes were added during PCR amplification and a
final double-sided SPRI cleanup was performed. Libraries were quantified
by Kapa qPCR for Illumina Adapters (Roche) and size was determined by
Agilent tapestation D1000 tapes. Read 1 primer sequence are added to the
molecules during GEM incubation. P5, P7 and sample index and read 2
primer sequence are added during library construction via end repair,
A-tailing, adaptor ligation and PCR. Libraries were generated with
unique sample indices (SI) for each sample. Libraries were sequenced on
a HiSeq 4000, (Illumina, San Diego, CA, USA) using a HiSeq 4000 PE
Cluster Kit (PN PE-410-1001) with HiSeq 4000 SBS Kit (100 cycles, PN
FC-410-1002) reagents, loaded at 200 pM following Illumina’s denaturing
and dilution recommendations. The run configuration was 26 × 8 × 98
cycles for Read 1, Index and Read 2, respectively.
</details>
<details>
<summary>
<strong>Data analysis</strong>
</summary>
<p>University of Michigan Biomedical Research Core Facilities Advanced
Genomics Core executed 10x Genomics Cell Ranger (v9.0.0) to perform
sample de-multiplexing, barcode processing, and single cell gene
counting (Alignment, Barcoding and UMI Count); alignments were against
mm10-2020-A and included intronic sequence. The Cell Ranger filtered
barcode feature matrix was used as input to downstream analysis.</p>
<p>All analysis and graphics were generated in R (v4.4.1) [1]. Analysis
was performed primarily using the Seurat package (v5.0.1) [2]. Cells
with extreme values (which indicate low complexity, doublets, or
apoptotic cells) were excluded by filtering to include only cells where
Genes/cell &gt;300 and % mitochondrial &lt; 15% resulting in
approximately 600-5,600 cells per sample after filtering. Counts were
then normalized using the SCTransform method with default parameters
[3].</p>
<p>Normalized data were integrated using the RPCA method
(IntegrateLayers function with “SCT” as normalization method. Principal
Component Analysis (PCA) was then performed and the first significant
components were used for finding nearest neighbors followed by
graph-based, semi-unsupervised Louvain clustering into distinct
populations (resolution = 0.4). All uniform manifold approximation and
projection (UMAP) plots were generated using default settings [4]. To
identify marker genes, the clusters in the integrated data were compared
pairwise for differential gene expression using Wilcoxon rank-sum test
for single-cell gene expression (FindAllMarkers function; default
parameters) [5]. Additional marker genes and cell-type predictions were
generated with scCATCH (3.2.2) [6].</p>
<p>To identify differentially expressed (DE) genes within the total
population, Case and Control samples were compared pairwise for
differential expression expression (FindAllMarkers function; log2FC =
1.5; test.use = “Wilcoxon”). For each cluster, the results were further
limited to significantly different genes (Benjamini-Hochberg adjusted
p-value &lt;= 0.05). Intra-cluster case-vs-control pseudo bulk
comparisons were analyzed using DESeq2 (v1.44.0) [7].</p>
<p><strong>References</strong></p>
<ol style="list-style-type: decimal">
<li>R Core Team (2019). R: A language and environment for statistical
computing. R Foundation for Statistical Computing, Vienna, Austria. <a
href="https://www.R-project.org/"
class="uri">https://www.R-project.org/</a>.</li>
<li>Stuart T, Butler A, Hoffman P, Hafemeister C, Papalexi E, III WMM,
Stoeckius M, Smibert P, Satija R (2018). “Comprehensive integration of
single cell data.” bioRxiv. doi: 10.1101/460147, <a
href="https://www.biorxiv.org/content/10.1101/460147v1"
class="uri">https://www.biorxiv.org/content/10.1101/460147v1</a></li>
<li>Hafemeister, C., Satija, R (2019). Normalization and variance
stabilization of single-cell RNA-seq data using regularized negative
binomial regression. Genome Biol 20, 296. <a
href="https://doi.org/10.1186/s13059-019-1874-1"
class="uri">https://doi.org/10.1186/s13059-019-1874-1</a></li>
<li>Becht, E. et al (2018). Dimensionality reduction for visualizing
single-cell data using UMAP. Nat. Biotechnol. 37, 38–44 .</li>
<li>Myles Hollander and Douglas A. Wolfe (1973). Nonparametric
Statistical Methods. New York: John Wiley &amp; Sons. Pages 68–75.</li>
<li>Shao et al (2020), scCATCH:Automatic Annotation on Cell Types of
Clusters from Single-Cell RNA Sequencing Data, iScience, Volume 23,
Issue 3. doi: 10.1016/j.isci.2020.100882.</li>
<li>Love MI, Huber W, Anders S (2014). “Moderated estimation of fold
change and dispersion for RNA-seq data with DESeq2.” Genome Biology, 15,
550. <a href="doi:10.1186/s13059-014-0550-8"
class="uri">doi:10.1186/s13059-014-0550-8</a>.</li>
</ol>
</details>
<details>
<summary>
<strong>Session Info</strong>
</summary>
<p><em><code>Session info</code> lists the relevant versions of R and
all the libraries that were loaded in the analysis. This info is
generally not included in main body of a paper, but serious
reproducibility street-cred if it shows up in supplemental.<br/> Hear me
now, believe me later.</em></p>
<pre><code>devtools::session_info()
─ Session info ────────────────────────────────────────────────────────────────
 setting  value
 version  R version 4.4.1 (2024-06-14)
 os       Ubuntu 22.04.5 LTS
 system   x86_64, linux-gnu
 ui       RStudio
 language (EN)
 collate  C.UTF-8
 ctype    C.UTF-8
 tz       America/Detroit
 date     2025-02-04
 rstudio  2023.12.0+369 Ocean Storm (server)
 pandoc   3.4 @ /usr/lib/rstudio-server/bin/quarto/bin/tools/x86_64/ (via rmarkdown)

─ Packages ──────────────────────────────────────────────────────────────────────────────────────────────────────────
 package          * version    date (UTC) lib source
 abind              1.4-8      2024-09-12 [2] CRAN (R 4.4.1)
 assertthat         0.2.1      2019-03-21 [2] CRAN (R 4.4.1)
 BiocGenerics       0.50.0     2024-04-30 [2] Bioconductor 3.19 (R 4.4.1)
 BPCells          * 0.2.0      2024-07-18 [2] Github (bnprks/BPCells@5677cf1)
 cachem             1.1.0      2024-05-16 [2] CRAN (R 4.4.1)
 cli                3.6.3      2024-06-21 [2] CRAN (R 4.4.1)
 cluster            2.1.6      2023-12-01 [2] CRAN (R 4.4.1)
 codetools          0.2-20     2024-03-31 [2] CRAN (R 4.4.1)
 colorspace         2.1-1      2024-07-26 [2] CRAN (R 4.4.1)
 cowplot            1.1.3      2024-01-22 [2] CRAN (R 4.4.1)
 crayon             1.5.3      2024-06-20 [2] CRAN (R 4.4.1)
 data.table         1.16.0     2024-08-27 [2] CRAN (R 4.4.1)
 deldir             2.0-4      2024-02-28 [2] CRAN (R 4.4.1)
 devtools           2.4.5      2022-10-11 [2] CRAN (R 4.4.1)
 digest             0.6.37     2024-08-19 [2] CRAN (R 4.4.1)
 dotCall64          1.1-1      2023-11-28 [2] CRAN (R 4.4.1)
 dplyr            * 1.1.4      2023-11-17 [2] CRAN (R 4.4.1)
 ellipsis           0.3.2      2021-04-29 [2] CRAN (R 4.4.1)
 evaluate           0.24.0     2024-06-10 [2] CRAN (R 4.4.1)
 fansi              1.0.6      2023-12-08 [2] CRAN (R 4.4.1)
 farver             2.1.2      2024-05-13 [2] CRAN (R 4.4.1)
 fastDummies        1.7.4      2024-08-16 [2] CRAN (R 4.4.1)
 fastmap            1.2.0      2024-05-15 [2] CRAN (R 4.4.1)
 fitdistrplus       1.2-1      2024-07-12 [2] CRAN (R 4.4.1)
 forcats          * 1.0.0      2023-01-29 [2] CRAN (R 4.4.1)
 fs                 1.6.4      2024-04-25 [2] CRAN (R 4.4.1)
 future             1.34.0     2024-07-29 [2] CRAN (R 4.4.1)
 future.apply       1.11.2     2024-03-28 [2] CRAN (R 4.4.1)
 generics           0.1.3      2022-07-05 [2] CRAN (R 4.4.1)
 GenomeInfoDb       1.40.1     2024-05-24 [2] Bioconductor 3.19 (R 4.4.1)
 GenomeInfoDbData   1.2.12     2024-07-17 [2] Bioconductor
 GenomicRanges      1.56.1     2024-06-12 [2] Bioconductor 3.19 (R 4.4.1)
 ggplot2          * 3.5.1      2024-04-23 [2] CRAN (R 4.4.1)
 ggrepel            0.9.6      2024-09-07 [2] CRAN (R 4.4.1)
 ggridges           0.5.6      2024-01-23 [2] CRAN (R 4.4.1)
 globals            0.16.3     2024-03-08 [2] CRAN (R 4.4.1)
 glue               1.7.0      2024-01-09 [2] CRAN (R 4.4.1)
 goftest            1.2-3      2021-10-07 [2] CRAN (R 4.4.1)
 gridExtra          2.3        2017-09-09 [2] CRAN (R 4.4.1)
 gtable             0.3.5      2024-04-22 [2] CRAN (R 4.4.1)
 hms                1.1.3      2023-03-21 [2] CRAN (R 4.4.1)
 htmltools          0.5.8.1    2024-04-04 [2] CRAN (R 4.4.1)
 htmlwidgets        1.6.4      2023-12-06 [2] CRAN (R 4.4.1)
 httpuv             1.6.15     2024-03-26 [2] CRAN (R 4.4.1)
 httr               1.4.7      2023-08-15 [2] CRAN (R 4.4.1)
 ica                1.0-3      2022-07-08 [2] CRAN (R 4.4.1)
 igraph             2.0.3      2024-03-13 [2] CRAN (R 4.4.1)
 IRanges            2.38.1     2024-07-03 [2] Bioconductor 3.19 (R 4.4.1)
 irlba              2.3.5.1    2022-10-03 [2] CRAN (R 4.4.1)
 jsonlite           1.8.8      2023-12-04 [2] CRAN (R 4.4.1)
 KernSmooth         2.23-24    2024-05-17 [2] CRAN (R 4.4.1)
 knitr              1.48       2024-07-07 [2] CRAN (R 4.4.1)
 later              1.3.2      2023-12-06 [2] CRAN (R 4.4.1)
 lattice            0.22-6     2024-03-20 [2] CRAN (R 4.4.1)
 lazyeval           0.2.2      2019-03-15 [2] CRAN (R 4.4.1)
 leiden             0.4.3.1    2023-11-17 [2] CRAN (R 4.4.1)
 lifecycle          1.0.4      2023-11-07 [2] CRAN (R 4.4.1)
 listenv            0.9.1      2024-01-29 [2] CRAN (R 4.4.1)
 lmtest             0.9-40     2022-03-21 [2] CRAN (R 4.4.1)
 lubridate        * 1.9.3      2023-09-27 [2] CRAN (R 4.4.1)
 magrittr           2.0.3      2022-03-30 [2] CRAN (R 4.4.1)
 MASS               7.3-61     2024-06-13 [2] CRAN (R 4.4.1)
 Matrix             1.7-0      2024-04-26 [2] CRAN (R 4.4.1)
 MatrixGenerics     1.16.0     2024-04-30 [2] Bioconductor 3.19 (R 4.4.1)
 matrixStats        1.4.1      2024-09-08 [2] CRAN (R 4.4.1)
 memoise            2.0.1      2021-11-26 [2] CRAN (R 4.4.1)
 mime               0.12       2021-09-28 [2] CRAN (R 4.4.1)
 miniUI             0.1.1.1    2018-05-18 [2] CRAN (R 4.4.1)
 munsell            0.5.1      2024-04-01 [2] CRAN (R 4.4.1)
 nlme               3.1-166    2024-08-14 [2] CRAN (R 4.4.1)
 parallelly         1.38.0     2024-07-27 [2] CRAN (R 4.4.1)
 patchwork          1.3.0      2024-09-16 [2] CRAN (R 4.4.1)
 pbapply            1.7-2      2023-06-27 [2] CRAN (R 4.4.1)
 pillar             1.9.0      2023-03-22 [2] CRAN (R 4.4.1)
 pkgbuild           1.4.4      2024-03-17 [2] CRAN (R 4.4.1)
 pkgconfig          2.0.3      2019-09-22 [2] CRAN (R 4.4.1)
 pkgload            1.4.0      2024-06-28 [2] CRAN (R 4.4.1)
 plotly             4.10.4     2024-01-13 [2] CRAN (R 4.4.1)
 plyr               1.8.9      2023-10-02 [2] CRAN (R 4.4.1)
 png                0.1-8      2022-11-29 [2] CRAN (R 4.4.1)
 polyclip           1.10-7     2024-07-23 [2] CRAN (R 4.4.1)
 prettyunits        1.2.0      2023-09-24 [2] CRAN (R 4.4.1)
 profvis            0.3.8      2023-05-02 [2] CRAN (R 4.4.1)
 progress           1.2.3      2023-12-06 [2] CRAN (R 4.4.1)
 progressr          0.14.0     2023-08-10 [2] CRAN (R 4.4.1)
 promises           1.3.0      2024-04-05 [2] CRAN (R 4.4.1)
 purrr            * 1.0.2      2023-08-10 [2] CRAN (R 4.4.1)
 R6                 2.5.1      2021-08-19 [2] CRAN (R 4.4.1)
 RANN               2.6.2      2024-08-25 [2] CRAN (R 4.4.1)
 RColorBrewer       1.1-3      2022-04-03 [2] CRAN (R 4.4.1)
 Rcpp               1.0.13     2024-07-17 [2] CRAN (R 4.4.1)
 RcppAnnoy          0.0.22     2024-01-23 [2] CRAN (R 4.4.1)
 RcppHNSW           0.6.0      2024-02-04 [2] CRAN (R 4.4.1)
 readr            * 2.1.5      2024-01-10 [2] CRAN (R 4.4.1)
 remotes            2.5.0      2024-03-17 [2] CRAN (R 4.4.1)
 reshape2           1.4.4      2020-04-09 [2] CRAN (R 4.4.1)
 reticulate         1.39.0     2024-09-05 [2] CRAN (R 4.4.1)
 rlang              1.1.4      2024-06-04 [2] CRAN (R 4.4.1)
 rmarkdown          2.28       2024-08-17 [2] CRAN (R 4.4.1)
 ROCR               1.0-11     2020-05-02 [2] CRAN (R 4.4.1)
 RSpectra           0.16-2     2024-07-18 [2] CRAN (R 4.4.1)
 rstudioapi         0.16.0     2024-03-24 [2] CRAN (R 4.4.1)
 Rtsne              0.17       2023-12-07 [2] CRAN (R 4.4.1)
 S4Vectors          0.42.1     2024-07-03 [2] Bioconductor 3.19 (R 4.4.1)
 scales             1.3.0      2023-11-28 [2] CRAN (R 4.4.1)
 scattermore        1.2        2023-06-12 [2] CRAN (R 4.4.1)
 scCATCH          * 3.2.2      2023-04-23 [2] CRAN (R 4.4.1)
 sctransform        0.4.1      2023-10-19 [2] CRAN (R 4.4.1)
 sessioninfo        1.2.2      2021-12-06 [2] CRAN (R 4.4.1)
 Seurat           * 5.1.0      2024-05-10 [2] CRAN (R 4.4.1)
 SeuratObject     * 5.0.2      2024-05-08 [2] CRAN (R 4.4.1)
 shiny              1.9.1      2024-08-01 [2] CRAN (R 4.4.1)
 sp               * 2.1-4      2024-04-30 [2] CRAN (R 4.4.1)
 spam               2.10-0     2023-10-23 [2] CRAN (R 4.4.1)
 spatstat.data      3.1-2      2024-06-21 [2] CRAN (R 4.4.1)
 spatstat.explore   3.3-2      2024-08-21 [2] CRAN (R 4.4.1)
 spatstat.geom      3.3-2      2024-07-15 [2] CRAN (R 4.4.1)
 spatstat.random    3.3-1      2024-07-15 [2] CRAN (R 4.4.1)
 spatstat.sparse    3.1-0      2024-06-21 [2] CRAN (R 4.4.1)
 spatstat.univar    3.0-1      2024-09-05 [2] CRAN (R 4.4.1)
 spatstat.utils     3.1-0      2024-08-17 [2] CRAN (R 4.4.1)
 stringi            1.8.4      2024-05-06 [2] CRAN (R 4.4.1)
 stringr          * 1.5.1      2023-11-14 [2] CRAN (R 4.4.1)
 survival           3.7-0      2024-06-05 [2] CRAN (R 4.4.1)
 tensor             1.5        2012-05-05 [2] CRAN (R 4.4.1)
 tibble           * 3.2.1      2023-03-20 [2] CRAN (R 4.4.1)
 tidyr            * 1.3.1      2024-01-24 [2] CRAN (R 4.4.1)
 tidyselect         1.2.1      2024-03-11 [2] CRAN (R 4.4.1)
 tidyverse        * 2.0.0      2023-02-22 [2] CRAN (R 4.4.1)
 timechange         0.3.0      2024-01-18 [2] CRAN (R 4.4.1)
 tzdb               0.4.0      2023-05-12 [2] CRAN (R 4.4.1)
 UCSC.utils         1.0.0      2024-04-30 [2] Bioconductor 3.19 (R 4.4.1)
 urlchecker         1.0.1      2021-11-30 [2] CRAN (R 4.4.1)
 usethis            3.0.0      2024-07-29 [2] CRAN (R 4.4.1)
 utf8               1.2.4      2023-10-22 [2] CRAN (R 4.4.1)
 uwot               0.2.2      2024-04-21 [2] CRAN (R 4.4.1)
 vctrs              0.6.5      2023-12-01 [2] CRAN (R 4.4.1)
 viridisLite        0.4.2      2023-05-02 [2] CRAN (R 4.4.1)
 withr              3.0.1      2024-07-31 [2] CRAN (R 4.4.1)
 xfun               0.47       2024-08-17 [2] CRAN (R 4.4.1)
 xtable             1.8-4      2019-04-21 [2] CRAN (R 4.4.1)
 XVector            0.44.0     2024-04-30 [2] Bioconductor 3.19 (R 4.4.1)
 yaml               2.3.10     2024-07-26 [2] CRAN (R 4.4.1)
 zlibbioc           1.50.0     2024-04-30 [2] Bioconductor 3.19 (R 4.4.1)
 zoo                1.8-12     2023-04-13 [2] CRAN (R 4.4.1)
</code></pre>
</details>
<hr />
</div>
<div id="housekeeping" class="section level2">
<h2>Housekeeping</h2>
<ul>
<li><p>Please take our optional <a href="https://michmed.org/7k9Zr"
target="_blank">post-workshop survey</a> (5-10 minutes).</p></li>
<li><p>We will email you a link to the final session recordings next
week.</p></li>
<li><p>The website/notes for this workshop will be available.</p></li>
<li><p>The <a href="https://umbioinfcoreworkshops.slack.com"
target="_blank">UM Bioinformatics Core Workshop Slack channel</a>
content will be available for 90 days.</p></li>
</ul>
<hr />
</div>
<div id="looking-ahead" class="section level2">
<h2>Looking ahead</h2>
<div id="workshop-environment" class="section level4">
<h4>Workshop environment</h4>
<ul>
<li><p>RStudio workshop compute environment will be available until
<strong>3/7/2025</strong>.</p>
<ul>
<li><em>Please save all your R scripts now</em> so that we can
“right-size” the compute environment immediately following today’s
workshop session.</li>
</ul></li>
<li><p>You can download files from the workshop environment from your
terminal/command line window as below. (You will need to substitute your
actual workshop username and type workshop password when prompted.)</p>
<pre class="bash"><code># download workshop files -------------------------------------------------
mkdir intro_scrnaseq_workshop
cd intro_scrnaseq_workshop
scp -r YOUR_USERNAME@bfx-workshop01.med.umich.edu:&quot;ISC_R*&quot; .</code></pre>
<ul>
<li>Note that the full download of the R data is about 8Gb, so depending
on your internet speeds it could take a while. (We do not recommend you
download the full set of Cell Ranger outputs.)</li>
</ul></li>
</ul>
</div>
<div id="installing-software-locally" class="section level4">
<h4>Installing software locally</h4>
<ul>
<li>You can install necessary programs to run programs locally. Note
that for typical data, Cell Ranger steps (reviewed Day 1) assume your
computer has powerful compute (many CPUs and lots of RAM) and sizable
storage capacity. (i.e. it’s impractical to run these on your laptop.)
<ul>
<li><a href="workshop_setup/setup_instructions_advanced.html">Setting up
R/R-Studio</a></li>
</ul></li>
<li>Note that installing bioinformatics software is often
non-trivial.</li>
</ul>
</div>
<div id="downloading-seurat-inputs" class="section level4">
<h4>Downloading Seurat inputs</h4>
<ul>
<li><p>The Seurat inputs (cellranger triples and dbcells files) used in
this workshop can be installed locally like so:</p>
<pre class="bash"><code># download Seurat inputs --------------------------------------------------
mkdir -p intro_scrnaseq_workshop/ISC_R
cd intro_scrnaseq_workshop/ISC_R
# Curl download a ~500Mb tarball
curl -o workshop_isc_inputs-20250212.tgz https://umich-brcf-bioinf-workshop-intro-single-cell.s3.us-east-1.amazonaws.com/workshop_isc_inputs-20250212.tgz 
# tar unpacks the tarball into directories and then removes the tarball
tar xzvf workshop_isc_inputs-20250212.tgz || rm workshop_isc_inputs-20250212.tgz</code></pre></li>
</ul>
</div>
<div id="an-independent-exercise" class="section level4">
<h4><a href="09-IndependentExercise.html">An independent
exercise</a></h4>
<ul>
<li>A variation on the workshop workflow that considers different
clustering parameters and their downstream effects.</li>
</ul>
<details>
<summary>
<em>University of Michigan Resources</em>
</summary>
<ul>
<li><a href="https://datascience.isr.umich.edu/events/coderspaces/"
target="_blank">UM CoderSpaces “office hours”</a> and UM CoderSpaces
Slack workspace. <em>(See “Useful Resources” section of the CoderSpaces
page for instructions on how to join the CoderSpaces Slack
workspace.)</em></li>
<li>Upcoming <a href="https://arc.umich.edu/events/" target="_blank">UM
Advanced Research Computing workshops</a>.</li>
<li>Advanced Research Computing (ARC) at University of Michigan hosts a
high-performance computing (HPC) platform called <em>Great Lakes</em>
which combines high-end computers, fast/resilient storage, and
pre-installed software. Great Lakes may be a good resource for folks who
need to run the more compute intensive steps and a substantial block of
compute and storage is subsidized by ARC making it essentially free to
many UM researchers.
<ul>
<li>About <a href="https://arc.umich.edu/greatlakes/"
target="_blank">Great Lakes</a>.</li>
<li>About the ARC <a href="https://arc.umich.edu/umrcp/"
target="_blank">Research Computing Package</a>.</li>
<li>Videos on <a
href="https://www.mivideo.it.umich.edu/channel/ARC-TS%2BTraining/181860561/"
target="_blank">getting started with Great Lakes</a>. (Videos are
available to UM folks.)</li>
</ul></li>
</ul>
</div>
</div>
<div id="resources-for-continued-learning" class="section level2">
<h2>Resources for continued learning</h2>
<ul>
<li>Learning bioinformatic analyses is more like a process than a task.
Resources to consider:
<ul>
<li>A CZI NDCN curated list of resources for <a
href="https://ndcn.github.io/self-guided-learning/"
target="_blank">self-guided learning</a></li>
<li>UM Bioinformatics Core links to <a
href="https://brcf.medicine.umich.edu/cores/bioinformatics-core/training/"
target="_blank">training resources/workshops</a>.</li>
<li><strong>Techniques in Bash and R</strong>
<ul>
<li>Lesson plans from the UM Bioinformatics Core’s recent <a
href="https://umich-brcf-bioinf.github.io/workshop-computational-foundations/main/html/"
target="_blank">Computational Foundations Workshop</a>.</li>
<li>Lessons and workshops in Bash / Git / R / Python :
<a href="https://software-carpentry.org/lessons/" target="_blank">
Software Carpentry</a></li>
</ul></li>
<li><strong>More info on NGS analysis</strong> (A decidedly
non-exhaustive list in no discernible order.)
<ul>
<li>Lesson plans from UM Bioinformatics Core’s recent <a
href="https://umich-brcf-bioinf.github.io/workshop-rnaseq-demystified/main/html/"
target="_blank">RNA-Seq Demystified Workshop</a>. (Note this is for bulk
RNA-Seq.)</li>
<li>Training materials from <a
href="https://hbctraining.github.io/main/#advanced-topics-analysis-of-high-throughput-sequencing-ngs-data"
target="_blank">Harvard Chan Bioinformatics Core</a>.</li>
<li><a href="https://bioconductor.org/books/3.12/OSCA/index.html"
target="_blank">Orchestrating Single-Cell Analysis with
Bioconductor</a>.</li>
<li>Welcome Sanger Instutute’s <a
href="https://www.singlecellcourse.org/index.html"
target="_blank">Analysis of single cell RNA-seq data</a>.</li>
<li>In addition to maintaining the Seurat analysis platform, New York
Genome Center’s <a href="https://satijalab.org/" target="_blank">Satija
lab</a> publishes many excellent step-by-step tutorials on single-cell
analysis, e.g. <a
href="https://satijalab.org/seurat/articles/pbmc3k_tutorial"
target="_blank">Seurat - Guided Clustering Tutorial</a>. The Satija lab
also hosts <a href="https://satijalab.org/scgd24/"
target="_blank">Single Cell Genomics Day</a>, a live-streamed annual
symposium on emerging techniques in single-cell analysis.</li>
<li>Nayak, Richa, and Yasha Hasija. <strong>“A hitchhiker’s guide to
single-cell transcriptomics and data analysis pipelines.”</strong>
Genomics vol. 113,2 (2021): 606-619.<br/> <a
href="https://pubmed.ncbi.nlm.nih.gov/33485955/"
class="uri">https://pubmed.ncbi.nlm.nih.gov/33485955/</a></li>
<li>Luecken, Malte D, and Fabian J Theis. <strong>“Current best
practices in single-cell RNA-seq analysis: a tutorial.”</strong>
Molecular systems biology vol. 15,6 e8746. 19 Jun. 2019.<br/> <a
href="https://pubmed.ncbi.nlm.nih.gov/31217225/"
class="uri">https://pubmed.ncbi.nlm.nih.gov/31217225/</a></li>
<li>He, Jiangping et al. <strong>“Practical bioinformatics pipelines for
single-cell RNA-seq data analysis.”</strong> Biophysics reports vol. 8,3
(2022): 158-169.<br/> <a
href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC10189648/"
class="uri">https://www.ncbi.nlm.nih.gov/pmc/articles/PMC10189648/</a></li>
</ul></li>
<li>For more context on the experiments and data we referenced during
the workshop, see the source paper: <br/> Sorkin, Michael et
al. <strong>“Regulation of heterotopic ossification by monocytes in a
mouse model of aberrant wound healing.”</strong> Nature communications
vol. 11,1 722. 5 Feb. 2020.<br/>
<a class="external" href="https://pubmed.ncbi.nlm.nih.gov/32024825/" target="_blank">https://pubmed.ncbi.nlm.nih.gov/32024825/</a></li>
</ul></li>
</ul>
<hr />
</div>
<div id="thank-you-to-our-sponsors" class="section level2">
<h2>Thank you to our sponsors</h2>
<p><img src="images/workshop_intro/sponsor_logos.png" /></p>
<p><br/></p>
</div>
<div id="thank-you-tofrom-the-workshop-team" class="section level2">
<h2>Thank you to/from the workshop team</h2>
<table>
<colgroup>
<col width="33%" />
<col width="33%" />
<col width="33%" />
</colgroup>
<thead>
<tr class="header">
<th align="center"><img
src="images/headshots/headshot_cgates.jpg" /></th>
<th align="center"><img
src="images/headshots/headshot_mbradenb.jpg" /></th>
<th align="center"><img
src="images/headshots/headshot_rcavalca.jpg" /></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="center"><strong>Chris</strong></td>
<td align="center"><strong>Marci</strong></td>
<td align="center"><strong>Raymond</strong></td>
</tr>
<tr class="even">
<td align="center"><img
src="images/headshots/headshot_damki.jpg" /></td>
<td align="center"><img
src="images/headshots/headshot_dfhannum.jpg" /></td>
<td align="center"><img
src="images/headshots/headshot_trsaari.jpg" /></td>
</tr>
<tr class="odd">
<td align="center"><strong>Dana</strong></td>
<td align="center"><strong>Ford</strong></td>
<td align="center"><strong>Travis</strong></td>
</tr>
<tr class="even">
<td align="center"><img
src="images/headshots/headshot_phillipa.jpg" /></td>
<td align="center"><img
src="images/headshots/headshot_ncarruth.jpg" /></td>
<td align="center"></td>
</tr>
<tr class="odd">
<td align="center"><strong>Tricia</strong></td>
<td align="center"><strong>Nick</strong></td>
<td align="center"></td>
</tr>
</tbody>
</table>
<p><br/></p>
<p>Thank you for participating in our workshop. We welcome your
questions and feedback now and in the future.</p>
<p>Bioinformatics Workshop Team</p>
<p><a
href="mailto:bioinformatics-workshops@umich.edu">bioinformatics-workshops@umich.edu</a>
<br/> <a
href="https://medresearch.umich.edu/office-research/about-office-research/biomedical-research-core-facilities/bioinformatics-core"
target="_blank">UM BRCF Bioinformatics Core</a></p>
</div>



</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3,h4",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
