<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="UM Bioinformatics Core Workshop Team" />


<title>Cell Ranger in Action</title>

<script src="site_libs/header-attrs-2.21/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/paper.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/default.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<!--
Favicon dervied from
https://twemoji.twitter.com/
https://favicon.io/emoji-favicons/dna/
-->
<link rel="apple-touch-icon" sizes="180x180" href="/apple-touch-icon.png">
<link rel="icon" type="image/png" sizes="32x32" href="/favicon-32x32.png">
<link rel="icon" type="image/png" sizes="16x16" href="/favicon-16x16.png">
<link rel="shortcut icon" href="favicon-16x16.png" />
<link rel="manifest" href="/site.webmanifest">
<link rel="mask-icon" href="/safari-pinned-tab.svg" color="#5bbad5">
<meta name="msapplication-TileColor" content="#da532c">
<meta name="theme-color" content="#ffffff">

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>









<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-inverse  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">Intro to scRNA-Seq</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="workshop_intro.html">Intro</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Day 1
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="00A-OrientingOnScRNASeq.html">Orienting on scRNA-Seq</a>
    </li>
    <li>
      <a href="00B-CellRangerInAction.html">Cell Ranger in Action</a>
    </li>
    <li>
      <a href="01-GettingStarted.html">Seurat</a>
    </li>
    <li>
      <a href="02-QCandFiltering.html">Secondary QC &amp; Filtering</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Day 2
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="03-Integration.html">Normalization</a>
    </li>
    <li>
      <a href="04-PCAandDimReduction.html">PCA &amp; Integration</a>
    </li>
    <li>
      <a href="05-ProjectionAndClustering.html">Projection &amp; Clustering</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Day 3
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="06-MarkerVisualization.html">Marker identification and visualization</a>
    </li>
    <li>
      <a href="07-CellTypeAnnos.html">Cell type annotation</a>
    </li>
    <li>
      <a href="08-DifferentialExpression.html">Differential expression analysis</a>
    </li>
  </ul>
</li>
<li>
  <a href="workshop_wrap_up.html">Wrap up</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Cell Ranger in Action</h1>
<h4 class="author">UM Bioinformatics Core Workshop Team</h4>

</div>


<style type="text/css">

body, td {
   font-size: 18px;
}
code.r{
  font-size: 12px;
}
pre {
  font-size: 12px
}
</style>
<div id="objective" class="section level2">
<h2>Objective</h2>
<ul>
<li>By the end of the session you will be able to run
<code>cellranger count</code> with data from the AGC to transform FASTQ
files into a feature-barcode matrix.</li>
<li>(borrowed heavily from 10x tutorials, but as if with data from
AGC)</li>
</ul>
</div>
<div id="do-what-10x-does" class="section level2">
<h2>Do what 10x does</h2>
<div id="what-you-get-from-agc" class="section level3">
<h3>What you get from AGC</h3>
<div id="base-data-structure" class="section level4">
<h4>Base Data Structure</h4>
<p>The basic data package from the AGC includes:</p>
<ul>
<li>an <strong>*.md5</strong> file to validate your data transfer.</li>
<li>a <strong>DemuxStats_*.csv</strong> that has some basic metrics
about how your samples performed on the sequencer.</li>
<li>the <strong>fastq_*</strong> folder containing the fastq.gz
files</li>
<li>a <strong>README.txt</strong> including details about how your
samples were processed.</li>
</ul>
<p>It looks like the tree below.</p>
<pre><code>0000-SR
├── 0000-SR.md5
├── DemuxStats_0000-SR.csv
├── fastqs_0000-SR
│   ├── 0000-SR-1-GEX_S25_R1_001.fastq.gz
│   ├── 0000-SR-1-GEX_S25_R2_001.fastq.gz
│   ├── 0000-SR-2-GEX_S26_R1_001.fastq.gz
│   ├── 0000-SR-2-GEX_S26_R2_001.fastq.gz
│   ├── 0000-SR-3-GEX_S27_R1_001.fastq.gz
│   ├── 0000-SR-3-GEX_S27_R2_001.fastq.gz
│   ├── 0000-SR-4-GEX_S28_R1_001.fastq.gz
│   └── 0000-SR-4-GEX_S28_R2_001.fastq.gz
└── README.txt</code></pre>
<p>All you need to run 10x cellranger count are the above 10x sample
fastq files and the correct <a
href="https://www.10xgenomics.com/support/software/cell-ranger/downloads#reference-downloads">reference
genome file</a>.</p>
<p>If the AGC does basic 10x analysis you also receive a
<strong>10x_analysis_*</strong> folder with the following structure and
files.</p>
<pre><code>0000-SR
|...
└── 10x_analysis_0000-SR
    ├── Sample_0000-SR-1
    │   ├── analysis
    │   ├── cloupe.cloupe
    │   ├── filtered_feature_bc_matrix
    │   ├── filtered_feature_bc_matrix.h5
    │   ├── metrics_summary.csv
    │   ├── molecule_info.h5
    │   ├── possorted_genome_bam.bam
    │   ├── possorted_genome_bam.bam.bai
    │   ├── raw_feature_bc_matrix
    │   ├── raw_feature_bc_matrix.h5
    │   └── web_summary.html
    ├── Sample_0000-SR-2
    │   ├── ...
    ├── Sample_0000-SR-3
    │   ├── ...
    └── Sample_0000-SR-4
        └── ...
</code></pre>
</div>
</div>
<div id="how-does-the-agc-generate-the-10x_analysis-folder-and-data"
class="section level3">
<h3>How does the AGC generate the 10x_analysis folder and data</h3>
<p><strong><em>You cannot run the cellranger pipeline easily on a
personal computer/laptop because of the <a
href="https://www.10xgenomics.com/support/software/cell-ranger/downloads/cr-system-requirements">software
requirements</a></em></strong>, briefly:</p>
<blockquote>
<p>Cell Ranger pipelines run on Linux systems that meet these minimum
requirements:</p>
<ul>
<li>8-core Intel or AMD processor (16 cores recommended).</li>
<li>64GB RAM (128GB recommended).<br />
</li>
<li>1TB free disk space.<br />
</li>
<li>64-bit CentOS/RedHat 7.0 or Ubuntu 14.04 - See the 10x Genomics OS
Support page for details.</li>
</ul>
<p>The pipelines also run on clusters that meet these additional minimum
requirements:</p>
<ul>
<li>Shared filesystem (e.g. NFS)</li>
<li>Slurm batch scheduling system</li>
</ul>
</blockquote>
<p>The AGC uses the Great Lake Computing cluster which meets the above
requirements.</p>
<ul>
<li><strong>If your lab has signed up for the <a
href="https://arc.umich.edu/umrcp/">RCP</a>, you can also get access to
the cluster and the available software.</strong></li>
<li>Using the cluster requires an <strong>*.sbat</strong> file, or slurm
batch file. This file allows the slurm scheduler to get the correct
resources needed to process your data using cellranger and charge the
correct computing account. <em>If you have questions about this, please
ask them in office hours.</em></li>
<li>At the AGC we construct the cellranger command that goes into the
*.sbat using one node per sample with 16 cores and 128G of RAM, as
recommeded by the software requirements.</li>
</ul>
<div id="constructing-the-cellranger-command-from-the-10x-tutorial"
class="section level4">
<h4>Constructing the cellranger command (From the 10x Tutorial)</h4>
<p>If running interactively on the cluster the first step is to load the
cell ranger module with <code>module load cellranger</code>.</p>
<p>Print the usage statement to see what is needed to build the command.
<code>cellranger count --help</code></p>
<p>The output is similar to the following:</p>
<pre><code>cellranger-count
Count gene expression (targeted or whole-transcriptome) and/or feature barcode reads from a single sample and GEM well
 
USAGE:
    cellranger count [FLAGS] [OPTIONS] --id &amp;lt;ID&amp;gt; --transcriptome &amp;lt;PATH&amp;gt;
     
FLAGS:
          --no-bam                  Do not generate a bam file
          --nosecondary             Disable secondary analysis, e.g. clustering. Optional
          --include-introns         Include intronic reads in count
          --no-libraries            Proceed with processing using a --feature-ref but no Feature Barcode libraries
                                    specified with the &amp;#x27;libraries&amp;#x27; flag
          --no-target-umi-filter    Turn off the target UMI filtering subpipeline. Only applies when --target-panel is
                            used
          --dry                     Do not execute the pipeline. Generate a pipeline invocation (.mro) file and stop
          --disable-ui              Do not serve the web UI
          --noexit                  Keep web UI running after pipestance completes or fails
          --nopreflight             Skip preflight checks
      -h, --help                    Prints help information
  ...
  </code></pre>
<p>To run <code>cellranger count</code>, you need to specify an
<strong><code>--id</code></strong>. This can be any string, which is a
sequence of alpha-numeric characters, underscores, or dashes and no
spaces, that is less than 64 characters. Cell Ranger creates an output
directory that is named using this id. This directory is called a
"pipeline instance" or pipestance for short.</p>
<p>The <code>--fastqs</code> should be a path to the directory
containing the FASTQ files. If there is more than one sample in the
FASTQ directory, use the <strong><code>--sample</code></strong> argument
to specify which samples to use. This
<strong><code>--sample</code></strong> argument works off of the sample
id at the beginning of the FASTQ file name. The last argument needed is
the path to the <strong><code>--transcriptome</code></strong> reference
package. Be sure to edit the file paths in the command below.</p>
<pre><code>cellranger count --id=0000-SR-1 \
   --fastqs=/nfs/turbo/path/to/0000-SR/0000-SR_fastqs \
   --sample=0000-SR-1\
   --transcriptome=path/to/refereces/cellranger_count/refdata-gex-GRCh38-2020-A`</code></pre>
<p>A full-sized dataset can take several hours to complete, but we’re
using a smaller data set that should process quickly.</p>
<p>The <code>cellranger count</code> screen output is similar to the
following:</p>
<pre><code>/mnt/yard/user.name/yard/apps/cellranger-7.2.0/bin
cellranger count (7.2.0)

cellranger count (7.2.0)
Copyright (c) 2021 10x Genomics, Inc.  All rights reserved.
-------------------------------------------------------------------------------

Martian Runtime - v4.0.6
...
2021-10-15 17:12:42 [perform] Serializing pipestance performance data.
Waiting 6 seconds for UI to do final refresh.
Pipestance completed successfully!</code></pre>
<p>When the output of the <code>cellranger count</code> command says,
“Pipestance completed successfully!”, this means the job is done. The
<code>cellranger count</code> pipeline outputs are in the pipestance
directory in the outs folder. List the contents of this directory with
<code>ls -1</code>.</p>
<p><code>ls -1 0000-SR-1/outs</code></p>
<p>The output is similar to the following:</p>
<pre><code>├── analysis 
├── cloupe.cloupe
├── filtered_feature_bc_matrix
├── filtered_feature_bc_matrix.h5
├── metrics_summary.csv
├── molecule_info.h5
├── possorted_genome_bam.bam
├── possorted_genome_bam.bam.bai
├── raw_feature_bc_matrix
├── raw_feature_bc_matrix.h5
└── web_summary.html</code></pre>
</div>
</div>
</div>
<div id="qc---key-outputs-of-the-quality-control-report-to-review"
class="section level2">
<h2>QC - key outputs of the quality control report to review</h2>
<p><strong>QC of small smaple/realistic samples</strong></p>
<p><strong><em><em>While cellranger processes we can take a break or
discuss what cellranger is doing under the hood if we haven’t covered it
elsewhere.</em></em></strong></p>
<p>The important outputs we look at to determine the quility of the 10x
analaysis include:</p>
<ul>
<li><strong>web_summary.html</strong>
<ul>
<li><strong>Barcode knee plot</strong></li>
<li>Cell count, UMI count, gene count</li>
<li><strong>Seq saturation</strong></li>
<li>Cell count, UMI count, gene count</li>
</ul></li>
<li><strong>metrics_summary.csv</strong></li>
</ul>
<div id="web-summary-html-report." class="section level3">
<h3>Web Summary html report.</h3>
<p>The <code>cellranger count</code> web summary has Summary and Gene
Expression tabs, Similar web summaries are also output from the other
cellranger pipelines.</p>
<p>The run summary from <code>cellranger count</code> can be viewed by
clicking Summary in the top left tab of the HTML file. The summary
metrics describe sequencing quality and various characteristics of the
detected cells.</p>
<style data-emotion="css 92repi">.css-92repi{display:block;max-width:100%;}</style>
<p><img alt="" class=" css-92repi" src="https://cdn.10xgenomics.com/image/upload/v1646938726/software-support/3p-Single-Cell-GEX/count-web-summary/cr7-count-gexAb-summary1.png"/>
<br/></p>
<div id="summary-tab" class="section level4">
<h4>Summary tab</h4>
The estimated number of cells detected, mean reads per cell, and median
genes detected per cell are prominently displayed near the top of the
page.
</p>
<p>Estimated <a
href="https://kb.10xgenomics.com/hc/en-us/articles/115005062366-What-is-sequencing-saturation">Sequecing
Saturation</a> is also displayed in the sequencing table.</p>
<ul>
<li>The recommended saturation is &gt;60%</li>
</ul>
Click the <code>?</code> icons next to the Sequencing, Mapping, and
Cells sections to display information about each metric in the
dashboard.
</p>
<div id="the-barcode-rank-plot" class="section level5">
<h5>The Barcode Rank Plot</h5>
<p>The <a
href="/support/software/cell-ranger/latest/advanced/cr-barcode-rank-plot">GEX
Barcode Rank Plot</a> under the Cells dashboard shows the distribution
of barcode counts and which barcodes were inferred to be associated with
cells. The y-axis is the number of UMI counts mapped to each barcode and
the x-axis is the number of barcodes below that value. A steep drop-off
is indicative of good separation between the cell-associated barcodes
and the barcodes associated with empty partitions. Since barcodes can be
associated with cells based on their UMI count or by their RNA profiles,
some regions of the graph can contain both cell-associated and
background-associated barcodes. The color of the graph represents the
local density of barcodes that are cell-associated.</p>
<p>See this <a
href="/support/software/cell-ranger/latest/advanced/cr-barcode-rank-plot">Guided
Tour of the Barcode Rank Plot</a> for more details.</p>
</div>
</div>
<div id="gene-expression-tab" class="section level4">
<h4>Gene Expression tab</h4>
The automated secondary analysis results can be viewed by clicking the
Gene Expression tab in the top left corner. Click the <code>?</code>
icons next to each section title to display information about the
secondary analyses shown in the dashboard.
</p>
The t-SNE Projection section shows the data reduced to two dimensions,
colored by UMI count (left) or clustering (right). It is a good starting
point to explore structure in the data. The projection colored by UMI
counts is indicative of the RNA content of the cells and often
correlates with cell size - redder points are cells with more RNA in
them. For the projection colored by clustering results, select the type
of clustering analysis to display from the drop-down button on the upper
right (Graph-based by default) - change the category to vary the type of
clustering and/or number of clusters (K=2-10) that are assigned to the
data.
</p>
The Top Features By Cluster table shows which genes are differentially
expressed in each cluster relative to all other clusters (Graph-based by
default). To find the genes associated with a particular cluster, click
the L2FC or p-value column headers associated with a given cluster
number to sort the table by a specific cluster
</p>
<p><img alt="" class=" css-92repi" src="https://cdn.10xgenomics.com/image/upload/v1646938726/software-support/3p-Single-Cell-GEX/count-web-summary/cr7-count-gexAb-GEX1.png"/>
<br/></p>
<div id="the-sequencing-saturation-plot" class="section level5">
<h5>The Sequencing Saturation plot</h5>
The Sequencing Saturation plot shows the effect of decreased sequencing
depth on sequencing saturation, which is a measure of the fraction of
library complexity that was observed. The right-most point on the line
is the full sequencing depth obtained in this run.
</p>
Similarly, the Median Genes per Cell plot shows the effect of decreased
sequencing depth on median genes per cell, which is a way of measuring
data yield as a function of depth. The right-most point on the line is
the full sequencing depth obtained in this run.
</p>
<p><img alt="" class=" css-92repi" src="https://cdn.10xgenomics.com/image/upload/v1646938726/software-support/3p-Single-Cell-GEX/count-web-summary/cr7-count-gexAb-GEX2.png"/>
<br/></p>
</div>
</div>
</div>
<div id="common-errorswarnings-remediations" class="section level3">
<h3>Common errors/warnings &amp; remediations</h3>
<ul>
<li>Saturation &amp; resequencing</li>
<li>Library prep quality issues</li>
</ul>
</div>
</div>
<div id="outputs-used-for-downstream-analysis" class="section level2">
<h2>Outputs used for downstream analysis</h2>
<p><a
href="https://www.10xgenomics.com/support/software/cell-ranger/latest/analysis/outputs/cr-outputs-gex-overview">10x
output documentation</a></p>
<div id="the-metrics_summary.csv" class="section level3">
<h3>The metrics_summary.csv</h3>
<p>Includes much of the same data as the web_summary.html but in csv
format.</p>
<table class="css-1pyqnef">
<style data-emotion="css eckm6m">.css-eckm6m >tr>*{background-color:#F7F9FC;height:32px;padding:0 16px;}.css-eckm6m >tr>:first-of-type{border-top-left-radius:0.375rem;}.css-eckm6m >tr>:last-of-type{border-top-right-radius:0.375rem;}</style>
<thead class="css-eckm6m">
<tr>
<style data-emotion="css vmhyu4">.css-vmhyu4{padding:1rem;text-align:left;vertical-align:bottom;}</style>
<style data-emotion="css yz5ans">.css-yz5ans{-moz-osx-font-smoothing:grayscale;-webkit-font-smoothing:antialiased;box-sizing:border-box;color:#0A2347;font-family:Calibre,sans-serif;font-style:normal;font-weight:500;margin:0;padding:0;font-size:1.125rem;letter-spacing:0.15px;line-height:1.333;padding:1rem;text-align:left;vertical-align:bottom;}</style>
<th class="css-yz5ans" style="padding:1rem">
Column Name
</th>
<th class="css-yz5ans" style="padding:1rem">
Description
</th>
</tr>
</thead>
<style data-emotion="css 8b62zi">.css-8b62zi >tr>*{background-color:#ffffff;height:32px;padding:0 16px;}.css-8b62zi >:last-of-type>:first-of-type{border-bottom-left-radius:0.375rem;}.css-8b62zi >:last-of-type>:last-of-type{border-bottom-right-radius:0.375rem;}</style>
<tbody class="css-8b62zi">
<tr>
<style data-emotion="css s9mn2y">.css-s9mn2y{text-align:left;vertical-align:top;}</style>
<style data-emotion="css 4fk7u9">.css-4fk7u9{-moz-osx-font-smoothing:grayscale;-webkit-font-smoothing:antialiased;box-sizing:border-box;color:#445979;font-family:Calibre,sans-serif;font-style:normal;font-weight:400;margin:0;padding:0;font-size:1.125rem;letter-spacing:0.15px;line-height:1.333;text-align:left;vertical-align:top;}</style>
<td class="css-4fk7u9" style="padding:1rem">
<code>Estimated Number of Cells</code>
</td>
<td class="css-4fk7u9" style="padding:1rem">
The number of barcodes associated with cell-containing partitions,
estimated from the barcode UMI count distribution.
</td>
</tr>
<tr>
<td class="css-4fk7u9" style="padding:1rem">
<code>Mean Reads per Cell</code>
</td>
<td class="css-4fk7u9" style="padding:1rem">
The total number of reads divided by the estimated number of cells.
</td>
</tr>
<tr>
<td class="css-4fk7u9" style="padding:1rem">
<code>Median Genes per Cell</code>
</td>
<td class="css-4fk7u9" style="padding:1rem">
Median number of read pairs sequenced from the cells assigned to this
sample. In case of multiplexing, only cell-associated barcodes assigned
exactly one CMO can be assigned to a sample.
</td>
</tr>
<tr>
<td class="css-4fk7u9" style="padding:1rem">
<code>Number of Reads</code>
</td>
<td class="css-4fk7u9" style="padding:1rem">
Total number of sequenced reads.
</td>
</tr>
<tr>
<td class="css-4fk7u9" style="padding:1rem">
<code>Valid Barcodes</code>
</td>
<td class="css-4fk7u9" style="padding:1rem">
Fraction of reads with cell-barcodes that match the whitelist.
</td>
</tr>
<tr>
<td class="css-4fk7u9" style="padding:1rem">
<code>Sequencing Saturation</code>
</td>
<td class="css-4fk7u9" style="padding:1rem">
Fraction of reads originating from an already-observed UMI. This is a
function of library complexity and sequencing depth. More specifically,
this is a ratio where: the denominator is the number of
confidently-mapped reads with a valid cell-barcode and valid UMI, and
the numerator is the subset of those reads that had a non-unique
combination of (cell-barcode, UMI, gene). This metric was called "cDNA
PCR Duplication" in versions of Cell Ranger prior to 1.2.
</td>
</tr>
<tr>
<td class="css-4fk7u9" style="padding:1rem">
<code>Q30 Bases in Barcode</code>
</td>
<td class="css-4fk7u9" style="padding:1rem">
Fraction of bases with Q-score at least 30 in the cell barcode
sequences. This is the i7 index (I1) read for the Single Cell 3' v1
chemistry and the R1 read for the Single Cell 3' v2 chemistry.
</td>
</tr>
<tr>
<td class="css-4fk7u9" style="padding:1rem">
<code>Q30 Bases in RNA</code>
</td>
<td class="css-4fk7u9" style="padding:1rem">
Fraction of bases with Q-score at least 30 in the RNA read sequences.
This is Illumina R1 for the Single Cell 3' v1 chemistry and Illumina R2
for the Single Cell 3' v2 chemistry.
</td>
</tr>
<tr>
<td class="css-4fk7u9" style="padding:1rem">
<code>Q30 Bases in UMI</code>
</td>
<td class="css-4fk7u9" style="padding:1rem">
Fraction of bases with Q-score at least 30 in the UMI sequences. This is
the R2 read for the Single Cell 3' v1 chemistry and the R1 read for the
Single Cell 3' v2 chemistry.
</td>
</tr>
<tr>
<td class="css-4fk7u9" style="padding:1rem">
<code>Reads Mapped to Genome</code>
</td>
<td class="css-4fk7u9" style="padding:1rem">
Fraction of reads that mapped to the genome.
</td>
</tr>
<tr>
<td class="css-4fk7u9" style="padding:1rem">
<code>Reads Mapped Confidently to Genome</code>
</td>
<td class="css-4fk7u9" style="padding:1rem">
Fraction of reads that mapped uniquely to the genome. If a read mapped
to exonic loci from a single gene and also to non-exonic loci, it is
considered uniquely mapped to one of the exonic loci.
</td>
</tr>
<tr>
<td class="css-4fk7u9" style="padding:1rem">
<code>Reads Mapped Confidently to Intergenic Regions</code>
</td>
<td class="css-4fk7u9" style="padding:1rem">
Fraction of reads that mapped to the intergenic regions of the genome
with a high mapping quality score as reported by the aligner.
</td>
</tr>
<tr>
<td class="css-4fk7u9" style="padding:1rem">
<code>Reads Mapped Confidently to Intronic Regions</code>
</td>
<td class="css-4fk7u9" style="padding:1rem">
Fraction of reads that mapped to the intronic regions of the genome with
a high mapping quality score as reported by the aligner.
</td>
</tr>
<tr>
<td class="css-4fk7u9" style="padding:1rem">
<code>Reads Mapped Confidently to Exonic Regions</code>
</td>
<td class="css-4fk7u9" style="padding:1rem">
Fraction of reads that mapped to the exonic regions of the genome with a
high mapping quality score as reported by the aligner.
</td>
</tr>
<tr>
<td class="css-4fk7u9" style="padding:1rem">
<code>Reads Mapped Confidently to Transcriptome</code>
</td>
<td class="css-4fk7u9" style="padding:1rem">
Fraction of reads that mapped to a unique gene in the transcriptome with
a high mapping quality score as reported by the aligner. The read must
be consistent with annotated splice junctions when
include-introns=false. These reads are considered for UMI counting.
</td>
</tr>
<tr>
<td class="css-4fk7u9" style="padding:1rem">
<code>Reads Confidently Mapped Antisense</code>
</td>
<td class="css-4fk7u9" style="padding:1rem">
Fraction of reads confidently mapped to the transcriptome, but on the
opposite strand of their annotated gene. A read is counted as antisense
if it has any alignments that are consistent with an exon of a
transcript but antisense to it, and has no sense alignments.
</td>
</tr>
<tr>
<td class="css-4fk7u9" style="padding:1rem">
<code>Total Genes Detected Median UMI Counts per Cell</code>
</td>
<td class="css-4fk7u9" style="padding:1rem">
The number of genes with at least one UMI count in any cell.
</td>
</tr>
</tbody>
</table>
</div>
<div id="raw_feature_bc_matrix-filtered_feature_bc_matrix"
class="section level3">
<h3>Raw_feature_bc_matrix / filtered_feature_bc_matrix</h3>
<p>The <code>cellranger</code> pipeline outputs unfiltered (raw) and
filtered feature-barcode matrices in two file formats: the Market
Exchange Format (MEX), which is described on this page, and Hierarchical
Data Format (HDF5), which is described in detail here.</p>
Each element of the feature-barcode matrix is the number of UMIs
associated with a feature (row) and a barcode (column):
</p>
<style data-emotion="css ayshjd">.css-ayshjd{overflow-x:auto;}</style>
<div class="css-ayshjd">
<style data-emotion="css 1pyqnef">.css-1pyqnef{background-color:#DCE0E6;border-spacing:1px;border-radius:0.375rem;margin-bottom:1.5rem;table-layout:auto;width:100%;}</style>
<table class="css-1pyqnef">
<style data-emotion="css eckm6m">.css-eckm6m >tr>*{background-color:#F7F9FC;height:32px;padding:0 16px;}.css-eckm6m >tr>:first-of-type{border-top-left-radius:0.375rem;}.css-eckm6m >tr>:last-of-type{border-top-right-radius:0.375rem;}</style>
<thead class="css-eckm6m">
<tr>
<style data-emotion="css vmhyu4">.css-vmhyu4{padding:1rem;text-align:left;vertical-align:bottom;}</style>
<style data-emotion="css yz5ans">.css-yz5ans{-moz-osx-font-smoothing:grayscale;-webkit-font-smoothing:antialiased;box-sizing:border-box;color:#0A2347;font-family:Calibre,sans-serif;font-style:normal;font-weight:500;margin:0;padding:0;font-size:1.125rem;letter-spacing:0.15px;line-height:1.333;padding:1rem;text-align:left;vertical-align:bottom;}</style>
<th class="css-yz5ans" style="padding:1rem">
Type
</th>
<th class="css-yz5ans" style="padding:1rem">
Description
</th>
</tr>
</thead>
<style data-emotion="css 8b62zi">.css-8b62zi >tr>*{background-color:#ffffff;height:32px;padding:0 16px;}.css-8b62zi >:last-of-type>:first-of-type{border-bottom-left-radius:0.375rem;}.css-8b62zi >:last-of-type>:last-of-type{border-bottom-right-radius:0.375rem;}</style>
<tbody class="css-8b62zi">
<tr>
<style data-emotion="css s9mn2y">.css-s9mn2y{text-align:left;vertical-align:top;}</style>
<style data-emotion="css 4fk7u9">.css-4fk7u9{-moz-osx-font-smoothing:grayscale;-webkit-font-smoothing:antialiased;box-sizing:border-box;color:#445979;font-family:Calibre,sans-serif;font-style:normal;font-weight:400;margin:0;padding:0;font-size:1.125rem;letter-spacing:0.15px;line-height:1.333;text-align:left;vertical-align:top;}</style>
<td class="css-4fk7u9" style="padding:1rem">
Unfiltered feature-barcode matrix
</td>
<td class="css-4fk7u9" style="padding:1rem">
Contains every barcode from the fixed list of known-good barcode
sequences that has at least one read. This includes background and
cell-associated barcodes. <br/>count:
<code>outs/raw_feature_bc_matrix/</code><br/> multi:
<code>outs/multi/count/raw_feature_bc_matrix/</code>
</td>
</tr>
<tr>
<td class="css-4fk7u9" style="padding:1rem">
Filtered feature-barcode matrix
</td>
<td class="css-4fk7u9" style="padding:1rem">
Contains only detected cell-associated barcodes. <br/>count:
<code>outs/filtered_feature_bc_matrix/</code><br/> multi:
<code>outs/per_sample_outs/count/sample_filtered_feature_bc_matrix/</code>
</td>
</tr>
</tbody>
</table>
</div>
For sparse matrices, the matrix is stored in the <a
href="https://math.nist.gov/MatrixMarket/formats.html">Market Exchange
Format (MEX)</a>. It contains gzipped TSV files with feature and barcode
sequences corresponding to row and column indices respectively. For
example, the matrices output may look like:
</p>
<code>cd /home/jdoe/runs/sample345/outs tree filtered_feature_bc_matrix filtered_feature_bc_matrix     ├── barcodes.tsv.gz     ├── features.tsv.gz     └── matrix.mtx.gz 0 directories, 3 files</code>
Features correspond to row indices. For each feature, the feature ID and
name are stored in the first and second column of the (unzipped)
<code>features.tsv.gz</code> file, respectively. The third column
identifies the type of feature, which will be one of
<code>Gene Expression</code>, <code>Antibody Capture</code>,
<code>CRISPR Guide Capture</code>, <code>Multiplexing Capture</code>, or
<code>CUSTOM</code>, depending on the feature type. Below is a minimal
example <code>features.tsv.gz</code> file showing data collected for
three genes.
</p>
<pre><code>ENSG00000141510       TP53         Gene Expression
ENSG00000012048       BRCA1        Gene Expression
ENSG00000139687       RB1          Gene Expression</code></pre>
Gene Expression</code> data, the ID corresponds to
<code class="css-q0dnij">gene_id</code> in the annotation field of the
reference GTF. Similarly, the name corresponds to
<code class="css-q0dnij">gene_name</code> in the annotation field of the
reference GTF. If no <code class="css-q0dnij">gene_name</code> field is
present in the reference GTF, gene name is equivalent to gene ID.
Similarly, for <code class="css-q0dnij">Antibody Capture</code> and
<code class="css-q0dnij">CRISPR Guide Capture</code> data, the
<code class="css-q0dnij">id</code> and
<code class="css-q0dnij">name</code> are taken from the first two
columns of the
<a href="/support/software/cell-ranger/latest/analysis/running-pipelines/cr-feature-bc-analysis#feature-ref" class="css-k0f5v edy1dh10">Feature
Reference CSV file</a>.
</p>
For multi-species experiments, gene IDs and names are prefixed with the
genome name to avoid name collisions between genes of different species
e.g., GAPDH becomes <code class="css-q0dnij">hg19_GAPDH</code> and
Gm15816 becomes <code class="css-q0dnij">mm10_Gm15816</code>.
</p>
Barcode sequences correspond to column indices:
</p>
<pre><code>head filtered_feature_bc_matrices/barcodes.tsv.gz

AAACCCAAGGAGAGTA-1      
AAACGCTTCAGCCCAG-1
AAAGAACAGACGACTG-1
AAAGAACCAATGGCAG-1    
AAAGAACGTCTGCAAT-1    
AAAGGATAGTAGACAT-1    
AAAGGATCACCGGCTA-1    
AAAGGATTCAGCTTGA-1    
AAAGGATTCCGTTTCG-1    
AAAGGGCTCATGCCCT-1</code></pre>
Each barcode sequence includes a suffix with a dash separator followed
by a number:
</p>
<p class="css-d17snu">
More details on the barcode sequence format are available in the
<a href="/support/software/cell-ranger/latest/analysis/outputs/cr-outputs-bam" class="css-k0f5v edy1dh10">barcoded
BAM section</a>.
</p>
<p class="css-d17snu">
R and Python support the MEX format and sparse matrices can be used for
more efficient manipulation.
</p>
<p class="css-d17snu">
For suggestions on downstream analysis with 3rd party R and Python
tools, see the 10x Genomics
<a href="/analysis-guides/continuing-your-journey-after-running-cell-ranger" class="css-k0f5v edy1dh10">Analysis
Guides</a> resource.
</p>
<p class="css-d17snu">
The R package Matrix supports loading MEX format data, and can be easily
used to load the sparse feature-barcode matrix.
</p>
<p>###Converting matrix files to CSV format</p>
Cell Ranger represents the feature-barcode matrix using sparse formats
(only the nonzero entries are stored) in order to minimize file size.
All of our programs, and many other programs for gene expression
analysis, support sparse formats.
</p>
However, certain programs (e.g. Excel) only support dense formats (where
every row-column entry is explicitly stored, even if it's a zero). Here
are a few methods for converting feature-barcode matrices to CSV:
</p>
###Load matrices into Python
<p class="css-d17snu">
The <code class="css-q0dnij">csv</code>,
<code class="css-q0dnij">os</code>,
<code class="css-q0dnij">gzip</code>, and
<code class="css-q0dnij">scipy.io</code> modules can be used to load a
feature-barcode matrix into Python.
</p>
###mat2csv
<p class="css-d17snu">
You can convert a feature-barcode matrix to dense CSV format using the
<code class="css-q0dnij">cellranger mat2csv</code> command.
</p>
<p><strong>Important</strong></p>
<p><strong>WARNING</strong>: Dense files can be very large and may cause
Excel to crash, or even fail in mat2csv if your computer doesn't have
enough memory. For example, a feature-barcode matrix from a human
reference (~33k genes) with ~3k barcodes uses at least 200MB of disk
space. Our 1.3 million mouse neuron dataset, if converted to this
format, would use more than 60GB of disk space. Thus, while you can use
mat2csv for small datasets, we strongly recommend using R or Python (as
shown in the sections above) to examine these matrix files.</p>
This command takes two arguments - an input matrix generated by Cell
Ranger (either an HDF5 file or a MEX directory), and an output path for
the dense CSV. For example, to convert a matrix from a pipestance named
<code class="css-q0dnij">sample123</code> in the current directory,
either of the following commands would work:
</p>
<pre><code># Convert from MEX
cellranger mat2csv sample123/outs/filtered_feature_bc_matrix sample123.csv

# Or, convert from HDF5
cellranger mat2csv sample123/outs/filtered_feature_bc_matrix.h5 sample123.csv</code></pre>
<ul>
<li>H5 vs files</li>
</ul>
<p><a href="https://portal.hdfgroup.org/documentation/index.html">intro
to HD5 format</a></p>
<p><a
href="https://www.10xgenomics.com/support/software/cell-ranger/latest/analysis/outputs/cr-outputs-molecule-info">10x
HDF docs</a></p>
<p>The cellranger pipeline outputs an HDF5 file containing per-molecule
information for all molecules that contain a valid barcode, a valid UMI,
and were assigned with high confidence to a gene or Feature Barcode.
This HDF5 file contains data corresponding to the observed molecules, as
well as data about the libraries and feature set(s) used (general
information about the HDF5 file format available here). This file is
called molecule_info.h5 in cellranger count and sample_molecule_info.h5
in cellranger multi outputs.</p>
<pre><code>(root)
    ├─ barcode_idx
    ├─ barcode_info [HDF5 group]
    │   ├─ genomes
    │   └─ pass_filter
    ├─ barcodes
    ├─ count
    ├─ feature_idx
    ├─ features [HDF5 group]
    │   ├─ _all_tag_keys
    │   ├─ target_sets [for Targeted Gene Expression or Fixed RNA Profiling]
    │   │    └─ [target set name]
    │   ├─ feature_type
    │   ├─ genome
    │   ├─ id
    │   ├─ name
    │   ├─ pattern [Feature Barcode only]
    │   ├─ read [Feature Barcode only]
    │   └─ sequence [Feature Barcode only]
    ├─ gem_group
    ├─ library_idx
    ├─ library_info
    ├─ metrics_json [HDF5 dataset; see below]
    ├── probe_idx          ---------------------|
    ├── probes [HDF5 group]                     |
    │   ├── feature_id                          | [For Fixed RNA Profiling, Cell Ranger v7.1+]
    │   ├── feature_name                        |
    │   ├── probe_id                            |
    │   └── region         ---------------------|
    ├─ umi
    └─ umi_type</code></pre>
<p>You can examine the contents of the H5 file using software such as
HDFView or the h5dump command, as demonstrated below to show the file
contents of the entire H5 object:</p>
<p>```h5dump -n molecule_info.h5</p>
<pre><code>HDF5 &quot;molecule_info.h5&quot; {
FILE_CONTENTS {
group      /
dataset    /barcode_idx
group      /barcode_info
dataset    /barcode_info/genomes
dataset    /barcode_info/pass_filter
dataset    /barcodes
dataset    /count
dataset    /feature_idx
group      /features
dataset    /features/_all_tag_keys
dataset    /features/feature_type
dataset    /features/genome
dataset    /features/id
dataset    /features/name
dataset    /gem_group
dataset    /library_idx
dataset    /library_info
dataset    /metrics_json
dataset    /umi
dataset    /umi_type
}
}</code></pre>
<p>```</p>
<p>More details can be found at the following links:</p>
<p><a href="https://portal.hdfgroup.org/documentation/index.html">intro
to HD5 format</a></p>
<p><a
href="https://www.10xgenomics.com/support/software/cell-ranger/latest/analysis/outputs/cr-outputs-molecule-info">10x
HDF docs</a></p>
<br/> <br/>
<hr/>
<div class="line-block"><a href="00A-OrientingOnScRNASeq.html">Previous
lesson</a> | <a href="#top">Top of this lesson</a> | <a
href="01-GettingStarted.html">Next lesson</a> |</div>
</div>
</div>



</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3,h4",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
