---
title: "Integration and batch correction"
author: "UM Bioinformatics Core"
date: "`r Sys.Date()`"
output:
        html_document:
            includes:
                in_header: header.html
            theme: paper
            toc: true
            toc_depth: 4
            toc_float: true
            number_sections: false
            fig_caption: true
            markdown: GFM
            code_download: true
---

<style type="text/css">
body, td {
   font-size: 18px;
}
code.r{
  font-size: 12px;
}
pre {
  font-size: 12px
}
</style>

```{r, include = FALSE}
source("../bin/chunk-options.R")
knitr_fig_path("XX")
```

# Workflow Overview {.unlisted .unnumbered}

<br/>
<img src="images/wayfinder/wayfinder.png" alt="wayfinder" style="height: 400px;"/>
<br/>
<br/>

# Introduction 

In this section, our goal is to use an integration technique to better control technical variablity between our samples. 

Like in the previous section, one integration method might be reported in a research paper but several options might have been tested before determining which approach sufficiently corrected batch effects without removing expected biological variability.

<!--Add specific goals for section-->


# Normalization & Integration

<!--Is PCA and Find Neighbors required prior to SCTransform?-->

For any single-cell experiment there are several steps that introduce technical variability, including sample isolation, cell dissociation, and running each sample on the instrument <!--although... there are some library approaches that allow samples to be pooled before running, which our colleagues from the Advanced Genomics Core will likely be discussing at the end of the day.--> 

<!--add figure from Seurat showing batch effects-->

To combat this, we will need to integrate our samples into a scaled dataset prior to performing dimensionality reduction and clustering.

# Integration methods

SCTransform, RPCA, 

#



Next steps: PCA and Dim reduction

Note how dimensionality choices are carried through downstream functions.
tSNE vs UMAP
DK- find thread with UMAP vs TSNE
https://pair-code.github.io/understanding-umap/ 

----

These materials have been adapted and extended from materials listed above. These are open access materials distributed under the terms of the [Creative Commons Attribution license (CC BY 4.0)](http://creativecommons.org/licenses/by/4.0/), which permits unrestricted use, distribution, and reproduction in any medium, provided the original author and source are credited.

<br/>
<br/>
<hr/>
| [Previous lesson](02-QCandFiltering.html) | [Top of this lesson](#top) | [Next lesson](PCAandDimReduction.html) |
| :--- | :----: | ---: |




