---
title: "Normalization"
author: "UM Bioinformatics Core"
date: "`r Sys.Date()`"
output:
        html_document:
            includes:
                in_header: header.html
            theme: paper
            toc: true
            toc_depth: 4
            toc_float: true
            number_sections: false
            fig_caption: true
            markdown: GFM
            code_download: true
---

<style type="text/css">
body, td {
   font-size: 18px;
}
code.r{
  font-size: 12px;
}
pre {
  font-size: 12px
}
</style>

```{r, include = FALSE}
source("../bin/chunk-options.R")
knitr_fig_path("XX")
```

# Workflow Overview {.unlisted .unnumbered}

<br/>
<img src="images/wayfinder/wayfinder.png" alt="wayfinder" style="height: 400px;"/>
<br/>
<br/> 

# Introduction 

After removing low-quality cells from the data, the next task is the normalization and variance stabiliziation of the counts for downstream analysis. Variation in scRNA-seq data comes from both biological and technical sources. It is important to control for the technical sources of variation as much as possible, without removing the biological sources of variation.

Biological variation is related to differences in cell type and state, and can help answer biologically meaningful questions. Whereas technical variation may arise from fluctuations in cellular RNA content, efficiency in lysis and reverse transcription, and stochastic sampling during sequencing.

A key driver of technical variation is cellular sequencing depth (that is, the number of UMIs sequenced per cell). The known relationship between gene abundance and gene variance is also an important factor to account for. While methods exist to address these issues in bulk RNA-seq, the sparsity of scRNA-seq data presents challenges to existing methods.

## Objectives

- Understand why normalization is needed.
- Understand normalization options.
- Normalize the counts with `SCTransform()`.

---

# Normalization

We will use the `SCTransform()` function, which uses a generalized linear model (GLM) framework to account for cell-level sequencing depth while also stabilizing the variance of the counts. Let's get the normalization started, as it takes a little time, and then we can explain what it's doing and why it is an improvement on alternate methods.

First, we'll run the following command to separate the data from each sample into layers, a requirement of Seurat v5:

```{r, eval = FALSE}
geo_so[['RNA']] = split(geo_so[['RNA']], f = geo_so$orig.ident)
geo_so
```
~~~
An object of class Seurat 
47037 features across 29615 samples within 1 assay
Active assay: RNA (26489 features, 0 variable features)
 12 layers present: counts.HO.Day0.replicate1, counts.HO.Day0.replicate2, counts.HO.Day0.replicate3, counts.HO.Day0.replicate4, counts.HO.Day21.replicate1, counts.HO.Day21.replicate2, counts.HO.Day21.replicate3, counts.HO.Day21.replicate4, counts.HO.Day7.replicate1, counts.HO.Day7.replicate2, counts.HO.Day7.replicate3, counts.HO.Day7.replicate4
~~~

We see the 12 layers containing the count data for each of the samples. Next, run `SCTransform()`:

```{r, eval = FALSE}
geo_so = SCTransform(geo_so)
```

## Details

The full description and justification of the `SCTransform()` function are provided in two excellent papers:

- Hafemeister & Satija, *Normalization and variance stabilization of single-cell RNA-seq data using regularized negative binomial regression*, 2019, Genome Biology ([link](https://genomebiology.biomedcentral.com/articles/10.1186/s13059-019-1874-1#availability-of-data-and-materials))

- Choudhary & Satija, *Comparison and evaluation of statistical error models for scRNA-seq*, 2022, Genome Biology ([link](https://genomebiology.biomedcentral.com/articles/10.1186/s13059-021-02584-9))

Contained in the title of the first paper is the kernel of the approach: "regularized negative binomial regression". Let's explain the parts of that phrase at a high-level.

- "Regression" means that for each gene in the scRNA-seq data, a generalized linear model is fit whereby the vector of UMI counts for the gene is estimated by a linear equation containing an intercept and the vector of sequencing depth of the cells. As a term in the model, the cell sequencing depth can be accounted for.

- "Negative binomial" refers to the counts being modeled with a negative binomial distribution. It has been observed in both bulk and single-cell RNA-seq that the mean and variance of the count data are related to one another. Put another way, the variance is non-constant. The negative binomial distribution allows for the variance of the counts to be a function of the mean of the counts.

- "Regularized" refers to a method to constrain parameter estimation. When computing the parameters of the model, the authors observed that genes with low to moderate expression levels had widely ranging model parameter estimates, even when the genes were similar to one another. They concluded this was a result of model overfitting--not of biological origin--and decided to constrain the model parameters by combining information across similar genes. That is, regularization. In this way, the variance of the gene counts is stabilized.

Finally, the Choudhary & Satija paper introduced a modified version of the regularized negative binomial regression (`sctransform` v2). The modifications include:

- The slope of the NB GLM is fixed at an analytically derived `ln(10)`. Consequently, only the intercept and overdispersion parameters remain to be estimated for each gene.

- Genes with very low expression or with variance less than the mean have their dispersion parameters set to infinity, essentially reverting to a Poisson count distribution.

- A lower bound on the minimum variance is used to prevent genes with very low UMI counts from being given very high normalized expression values.

Additionally, the benefit of the GLM framework for normalization and variance stabilization is that the GLM can include other terms which can account for unwanted technical variation. One example of this is the percent mitochondrial reads (`percent.mt`). The `vars.to.regress` parameter of `SCTransform()` is used for this purpose. See the [documentation](https://satijalab.org/seurat/reference/sctransform) for details.

## Other normalizations

From the Seurat documentation, the `SCTransform()` function "replaces `NormalizeData()`, `ScaleData()`, and `FindVariableFeatures()`". This chain of functions is referred to as the "log-normalization procedure". You may see these three commands in other vignettes, and even in other Seurat vignettes ([source](https://satijalab.org/seurat/articles/sctransform_vignette)). In the two papers referenced above, the authors show how the log-normalization procedure does not always fully account for cell sequencing depth and overdispersion. Therefore, we urge you to use this alternative pipeline with caution.

# Normalization, continued

By now, `SCTransform()` should have finished running, so let's take a look at the result:

```{r, eval = FALSE}
geo_so
```
~~~
An object of class Seurat 
38503 features across 22320 samples within 2 assays 
Active assay: SCT (17129 features, 3000 variable features)
 3 layers present: counts, data, scale.data
 1 other assay present: RNA
~~~

We observe that the active assay is now "SCT", but that the "RNA" assay is still present. `SCTransform()` has also determined the common variable features across the cells to be used in our downstream analysis.

# Save the Seurat object

Let's save this normalized form of our Seurat object.

```{r, eval = FALSE}
saveRDS(object = geo_so, file = 'results/rdata/geo_so_sct_normalized.rda')
```

# Summary

In this section we have run the `SCTransform()` function to account for variation in cell sequencing depth and to stabilize the variance of the counts. 

Next steps: PCA and integration

----

These materials have been adapted and extended from materials listed above. These are open access materials distributed under the terms of the [Creative Commons Attribution license (CC BY 4.0)](http://creativecommons.org/licenses/by/4.0/), which permits unrestricted use, distribution, and reproduction in any medium, provided the original author and source are credited.

<br/>
<br/>
<hr/>
| [Previous lesson](02-QCandFiltering.html) | [Top of this lesson](#top) | [Next lesson](04-PCAandIntegration.html) |
| :--- | :----: | ---: |




