---
title: "Marker identification and visualization"
author: "UM Bioinformatics Core"
date: "`r Sys.Date()`"
output:
        html_document:
            includes:
                in_header: header.html
            theme: paper
            toc: true
            toc_depth: 4
            toc_float: true
            number_sections: false
            fig_caption: true
            markdown: GFM
            code_download: true
---

<style type="text/css">
body, td {
   font-size: 18px;
}
code.r{
  font-size: 12px;
}
pre {
  font-size: 12px
}
</style>

```{r, include = FALSE}
source("../bin/chunk-options.R")
knitr_fig_path("XX")
```

# Workflow Overview {.unlisted .unnumbered}

<br/>
<img src="images/wayfinder/wayfinder.png" alt="wayfinder" style="height: 400px;"/>
<br/>
<br/>

# Introduction

To interpret our initial clustering results for all cells in our experiment, we'll need to characterized the gene expression driving separation between the clusters. 

In this section, we will demonstrate how to generate analytical marker genes for each cluster and how to visualize the expression of genes of interest to aid in cluster identification and labeling in subsequent steps. 

This process can be a highly variable process, from seeing well-characterized marker genes as top markers to needing to dig through the literature and plot the expression for multiple sets of genes of interest. Like the previous sections (and like other areas of research), when this process is more involved or iterative the final results then often only the final results are reported.

## Objectives
<!--Add specific goals for section-->
- XX
- XX

---

# Cluster markers and characterization

<!-- [OSCA](https://bioconductor.org/books/3.15/OSCA.basic/marker-detection.html) "o interpret our clustering results from Chapter 5, we identify the genes that drive separation between clusters. These marker genes allow us to assign biological meaning to each cluster based on their functional annotation. In the simplest case, we have a priori knowledge of the marker genes associated with particular cell types, allowing us to treat the clustering as a proxy for cell type identity. The same principle can be applied to discover more subtle differences between clusters (e.g., changes in activation or differentiation state) based on the behavior of genes in the affected pathways.

The most straightforward approach to marker gene detection involves testing for differential expression between clusters. If a gene is strongly DE between clusters, it is likely to have driven the separation of cells in the clustering algorithm. Several methods are available to quantify the differences in expression profiles between clusters and obtain a single ranking of genes for each cluster. We will demonstrate some of these choices in this chapter using the 10X PBMC dataset:" --> 

<!-- [Ouyang](https://ouyanglab.com/singlecell/clust.html#identifying-marker-genes) "After identifying the cell clusters in an unbiased manner, we next seek to find the genes that distinguishes these cell clusters. To achieve this, we need to perform differential expression (DE) analysis to identify differentially expressed genes (DEGs) between cell clusters. These marker genes can then be used to characterize and annotate the cell clusters (Section 3.3)." 
"Before the prominence of scRNA-seq, there already exist a plethora of DE tools designed for comparative transcriptomic analysis between bulk RNA-seq samples. The large number of DE tools arises because of the difficulty in estimating gene dispersion using a small number of bulk RNA-seq samples (typically  
n
  = 2 or 3). Accurate estimates of gene dispersion is important in quantifying the DE significance. Thus, different DE methods differ in how they estimate gene dispersion, which consequently affect their accuracy in identifying DEGs.

In the case of scRNA-seq, gene dispersion is a lot easier to estimate since there are a lot more single cells ( 
n
  > 100) in DE comparisons. As a result, “simpler” statistical tests such as the Wilcoxon test (which is the default in Seurat) and the t-test perform well, as shown in a recent review by Soneson and Robinson (2018) (Figure 3.7). In fact, these simple statistical tests seem to outperform DE tools that are dedicated for scRNA-seq. These dedicated DE tools often use a zero-inflated model to account for the large number of zeros in single-cell gene expression. However, a recent report by Svensson (2020) suggests that single-cell expression is not zero-inflated and the large number of zeros are expected under a conventional negative binomial distribution. This could explain the poor performance of some of the single-cell dedicated DE tools. Another consideration is that some of the DE tools initially designed for bulk RNA-seq may not scale well computationally with the increasing number of single cells. From personal experience, the wilcoxon test suffices for simple pairwise DE while edgeR (Robinson, McCarthy, and Smyth 2010) can be used for more complex designs."
-->

After generating clusters, to identify the genes that distinguish them we need to perform differential expression analysis to identify differentially expressed genes for each cluster. [source](https://ouyanglab.com/singlecell/clust.html#identifying-marker-genes).

<details>
    <summary>*Additional considerations for differential expression*</summary>
    The Ouyang Lab has a [section of their tutorial](https://ouyanglab.com/singlecell/clust.html#sec:diffexpr) that discusses the methods available for differential expression as well as [threshold considerations](https://ouyanglab.com/singlecell/clust.html#sec:diffexpr).
</details>
<br>

## Marker identification

Ensure that correct resolution is selected from Seurat object
```{r, eval = FALSE}
geo_so = SetIdent(geo_so, value = 'SCT_snn_res.0.4')
```

We can use `FindAllMarkers` to run comparisons between each cluster and all other cells, regardless of the experimental group. Note that the default option for this function uses a Wilcoxon test to generate the resulting statistics.

```{r, eval=FALSE}
geo_so = PrepSCTFindMarkers(geo_so)
geo_markers = FindAllMarkers(geo_so, only.pos = TRUE)
```

<!--- Consider adding figure showing example of what's being compared (circle cluster 1 and then circle all other cells for example) -->

Then we can write out the full results to file. 
```{r, eval=FALSE}
write_csv(geo_markers, file = 'ISC_R/results/tables/marker_genes_0.4.csv')
```

In addition to generating the full set of marker genes, it can be useful to filter the results to highlight the top positive markers (since a positive fold-change would mean that gene is more highly expressed in the cluster compared to all other cells):
```{r, eval=FALSE}
# add code to create table of top 5 markers per cluster
top_5_by_log2FC = geo_markers %>% group_by(cluster) %>% arrange(p_val_adj, desc(avg_log2FC)) %>% slice_head(n = 5)
```

If we look at the marker table, we see several columns:
```{r, eval=FALSE}
head(top_5_by_log2FC)
```
~~~~
# add expected output
~~~~


## Marker visualization

Use SCT values, create a dotplot
```{r, eval=FALSE}
top_5_sct_dot_plot = DotPlot(geo_so, features = unique(top_5_by_log2FC$gene)) + 
    theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5)) +
    labs(title = 'Top 5 Cluster Genes by FDR and avg_log2FC')
```

And we can save the figure to file
```{r, eval=FALSE}
ggsave(filename = 'ISC_R/results/figures/markers_top_5_sct_dot_plot.png', plot = top_5_sct_dot_plot, width = 12, height = 6, units = 'in') 
```


Can also use raw RNA values <!--- move to dropdown --->

```{r, eval=FALSE}
top_5_rna_dot_plot = DotPlot(geo_so, features = unique(top_5_by_log2FC$gene), assay = 'RNA') + 
    theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5)) +
    labs(title = 'Top 5 Cluster Genes by FDR and avg_log2FC')

ggsave(filename = 'ISC_R/results/figures/markers_top_5_sct_dot_plot.png', plot = top_5_rna_dot_plot, width = 12, height = 6, units = 'in') 
```


# Utilizing genes of interest

Often we have 

```{r, eval=FALSE}
# Using genes from the paper - https://www.ncbi.nlm.nih.gov/pmc/articles/PMC7002453/

fig1g_markers = c('Cxcl1', 'Cxcl2', 'Ccl2', 'Ccl3', 'Ccl4', 'Il1b', 'Il6b', 'Tnf', 'Tgfb1', 'Tgfb2', 'Tgfb3', 'Cxcl5')
fig1h_markers = c('Cxcr2', 'Csf1r', 'Csf3r', 'Tgfbr1', 'Tgfbr3', 'Il1r1', 'Il6ra', 'Lifr', 'Tgfbr2')
```

# Save our progress

Create an output for our updated Seurat object
```{r, eval=FALSE}
saveRDS(object = geo_so, file = 'ISC_R/results/rdata/geo_so_sct_integrated_with_markers.rda')
```


# Summary

Now that we have ... 

However, marker genes alone might not be sufficient to determine cell-type or sub-type labels for our clusters so it can be helpful to apply other approaches, which we will discuss in the next section.

<!-- While there are some automated cluster annotation tools, one of which we'll discuss in the next section, often manual cluster annotation is part of characterizing clusters in single cell data.  -->

Next steps: Cell type prediction tools



----

These materials have been adapted and extended from materials listed above. These are open access materials distributed under the terms of the [Creative Commons Attribution license (CC BY 4.0)](http://creativecommons.org/licenses/by/4.0/), which permits unrestricted use, distribution, and reproduction in any medium, provided the original author and source are credited.

<br/>
<br/>
<hr/>
| [Previous lesson](05-ProjectionAndClustering.html) | [Top of this lesson](#top) | [Next lesson](07-CellTypeAnnos.html) |
| :--- | :----: | ---: |

