---
title: "Marker identification and visualization"
author: "UM Bioinformatics Core"
date: "`r Sys.Date()`"
output:
        html_document:
            includes:
                in_header: header.html
            theme: paper
            toc: true
            toc_depth: 4
            toc_float: true
            number_sections: false
            fig_caption: true
            markdown: GFM
            code_download: true
---

<style type="text/css">
body, td {
   font-size: 18px;
}
code.r{
  font-size: 12px;
}
pre {
  font-size: 12px
}
</style>

```{r, include = FALSE}
source("../bin/chunk-options.R")
knitr_fig_path("XX")
```

# Workflow Overview {.unlisted .unnumbered}

<br/>
<img src="images/wayfinder/wayfinder.png" alt="wayfinder" style="height: 400px;"/>
<br/>
<br/>

# Introduction

Point out possible gaps between how single-cell data is presented versus the reality of running an analysis
E.g. Papers might report final filtering thresholds, but our process for choosing those is likely to be more iterative

General goal: Filtering out the "cells" that either were poorly measured or are either not single-cells or not cells at all

<!--Add specific goals for section-->

# Marker identification

Ensure that correct resolution is selected
```{r, eval = FALSE}
geo_so = SetIdent(geo_so, value = 'SCT_snn_res.0.4')
```

We can use `FindAllMarkers` to run comparisons between each cluster and all other cells, regardless of the experimental group. 

```{r, eval=FALSE}
geo_so = PrepSCTFindMarkers(geo_so)
geo_markers = FindAllMarkers(geo_so, only.pos = TRUE)
```

<!--- Consider adding figure showing example of what's being compared (circle cluster 1 and then circle all other cells for example) -->

Then we can write out the full results to file. 
```{r, eval=FALSE}
write_csv(geo_markers, file = 'ISC_R/results/tables/marker_genes_0.4.csv')
```

In addition to generating the full set of marker genes, it can be useful to filter the results to highlight the top positive markers (since a positive fold-change would mean that gene is more highly expressed in the cluster compared to all other cells):
```{r, eval=FALSE}
# add code to create table of top 5 markers per cluster
top_5_by_log2FC = geo_markers %>% group_by(cluster) %>% arrange(p_val_adj, desc(avg_log2FC)) %>% slice_head(n = 5)
```

If we look at the marker table, we see several columns:
```{r, eval=FALSE}
head(top_5_by_log2FC)
```
~~~~
# add expected output
~~~~


## Marker visualization

Use SCT values, create a dotplot
```{r, eval=FALSE}
top_5_sct_dot_plot = DotPlot(geo_so, features = unique(top_5_by_log2FC$gene)) + 
    theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5)) +
    labs(title = 'Top 5 Cluster Genes by FDR and avg_log2FC')
```

And we can save the figure to file
```{r, eval=FALSE}
ggsave(filename = 'ISC_R/results/figures/markers_top_5_sct_dot_plot.png', plot = top_5_sct_dot_plot, width = 12, height = 6, units = 'in') 
```


Can also use raw RNA values <!--- move to dropdown --->

```{r, eval=FALSE}
top_5_rna_dot_plot = DotPlot(geo_so, features = unique(top_5_by_log2FC$gene), assay = 'RNA') + 
    theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5)) +
    labs(title = 'Top 5 Cluster Genes by FDR and avg_log2FC')

ggsave(filename = 'ISC_R/results/figures/markers_top_5_sct_dot_plot.png', plot = top_5_rna_dot_plot, width = 12, height = 6, units = 'in') 
```


# Utilizing genes of interest

Often we have 

```{r, eval=FALSE}
# Using genes from the paper - https://www.ncbi.nlm.nih.gov/pmc/articles/PMC7002453/

fig1g_markers = c('Cxcl1', 'Cxcl2', 'Ccl2', 'Ccl3', 'Ccl4', 'Il1b', 'Il6b', 'Tnf', 'Tgfb1', 'Tgfb2', 'Tgfb3', 'Cxcl5')
fig1h_markers = c('Cxcr2', 'Csf1r', 'Csf3r', 'Tgfbr1', 'Tgfbr3', 'Il1r1', 'Il6ra', 'Lifr', 'Tgfbr2')
```

# Save our progress

Create an output for our updated Seurat object
```{r, eval=FALSE}
saveRDS(object = geo_so, file = 'ISC_R/results/rdata/geo_so_sct_integrated_with_markers.rda')
```


# Summary

Now that we have ... 

However, marker genes alone might not be sufficient to determine cell-type or sub-type labels for our clusters so it can be helpful to apply other approaches, which we will discuss in the next section.

Next steps: Cell type prediction tools



----

These materials have been adapted and extended from materials listed above. These are open access materials distributed under the terms of the [Creative Commons Attribution license (CC BY 4.0)](http://creativecommons.org/licenses/by/4.0/), which permits unrestricted use, distribution, and reproduction in any medium, provided the original author and source are credited.

<br/>
<br/>
<hr/>
| [Previous lesson](05-ProjectionAndClustering.html) | [Top of this lesson](#top) | [Next lesson](07-CellTypeAnnos.html) |
| :--- | :----: | ---: |

