---
title: "Cell Type Annotation"
author: "UM Bioinformatics Core Workshop Team"
date: "`r Sys.Date()`"
lang: "en"
output:
        html_document:
            includes:
                in_header: header.html
            theme: paper
            toc: true
            toc_depth: 4
            toc_float: true
            number_sections: false
            fig_caption: true
            markdown: GFM
            code_download: false
---

<style type="text/css">
body {
   font-size: 18px;
}

code.r{
  font-size: 12px;
}
pre {
  font-size: 12px
}

table.fig, th.fig, td.fig {
  border: 1px solid lightgray;
  border-collapse: collapse;
  padding: 12px;
}

.prepared_content {
  border-radius: 10px;
  padding: 5px 5px 5px 95px;
  background: #FFF8DC left 10px top 10px / 65px no-repeat;
}

.cooking_show {
  background-image: url("images/curriculum/cooking_show-1.png");
}
</style>

```{r klippy, include=TRUE, echo=FALSE, message=FALSE, warning=FALSE}
klippy::klippy(lang = c("r", "markdown", "bash"), position = c("top", "right"))
library(kableExtra)
```

```{r, include = FALSE}
source("../bin/chunk-options.R")
knitr_fig_path("07-CellTypeAnnotation/07-")
```

# Workflow Overview {.unlisted .unnumbered}

<br/>
![](images/wayfinder/07-CellTypeAnnos-Wayfinder.png)
<br/>
<br/>

# Introduction

<br/>
<table class='fig'><tr><td class='fig'>
![](images/graphical_abstracts/07-CellTypeAnnos-Abstract.png)
</td></tr><tr><tr class='fig'><td class='fig'>Starting with clustered data, we can summarize and compare gene expression for known gene markers within each cluster, irrespective of sample or condition, to help label clusters with the appropriate cell-type or subtype.
</td></tr>
</table>
<br/>

A frequent bottleneck in single-cell analysis is annotating the algorithmicly generated clusters, as it requires bridging the gap between the data and prior knowledge ([source](https://bioconductor.org/books/3.15/OSCA.basic/cell-type-annotation.html)). While generating markers for each cluster it important, it may or may not be sufficient to assign cell-type or sub-type labels without bringing in additional sources. 
<br/>

## Objectives

-  Understand the complexities of cell-type annotation    
-  Use `scCATCH` cell-type predictions to annotate our clusters   

<br/>
Like the previous sections, the steps to assign cell-types to clusters might require some iteration, can be very dataset dependent, and often is more challenging for less characterized tissues. 
<br/>

----

```{r, read_rds_hidden, echo = FALSE, warning = FALSE, message = FALSE}
if(!exists('geo_so')) {
  library(Seurat)
  library(BPCells)
  library(tidyverse)

  options(future.globals.maxSize = 1e9)

  geo_so = readRDS('results/rdata/geo_so_sct_integrated_with_markers.rds')
}

if(!exists('geo_markers')) {
  library(Seurat)
  library(BPCells)
  library(tidyverse)

  options(future.globals.maxSize = 1e9)

  geo_markers = readRDS('results/rdata/geo_markers.rds')
}
```

# Cell type predictions

Automated tools have the advantage of being able to compare between the expression patterns in our dataset and large numbers of reference datasets or databases at a scale that is not feasible to do manually.

As described [in more detail by the Ouyang Lab](https://ouyanglab.com/singlecell/clust.html#annotating-clusters) and summarized in the figure below, there are many computational tools that aim to assign cell type labels for single-cell data. These methods generally fall into three categories: 

1. Marker based approaches that use gene sets drawn from the literature, including previous single-cell studies, 
2. Correlation based approaches that estimate the similarity between the cells or clusters in the input data and some reference data
3. Machine learning approaches that include training on a single-cell reference atlas. 

![Image: Diagram of types of cell annotation approaches (from Oyang materials).](./images/curriculum/07-CellTypeAnnotation/Ouyang_clust-celltype.png){height=50%}     
<br/>

However, across any of these approaches the quality of the reference data (and reliability of the authors labels) and relevancy to your specific tissue/experiment (and the resolution of your biological question) is crucial. Additionally, it's important to consider that rare or novel cell populations may not be present or well-characterized in available references and that even after filtering, some clusters might correspond to stressed or dying cells and not a particular cell-type or subtype. Therefore, any prediction should be reviewed and considered in the context both marker gene expression for the dataset and knowledge of the biological system and broader literature. 

Some tools and references are available solely or primarily for human tissues (and not mouse or rat), particular for tissues other than PBMCs and the brain. For human data, if a relevant reference is available for your experiment, we would recommend trying [Azimuth (created by authors of Seurat)](https://azimuth.hubmapconsortium.org/). [10x has a tutorial](https://www.10xgenomics.com/analysis-guides/automated-cell-type-annotation-from-r-to-loupe-using-louper) that includes example of using Azimuth, including a feature of the tool that allows for first pass of cell-type assignment of more common cell-types followed by identifying rarer populations. 

<details>
    <summary>*Additional automated annotation resources*</summary>
    Automated cell-type annotation is an active area of research and development and many other tools and resources are available, including [OSCA's demonstration of the SingleR method](https://bioconductor.org/books/3.15/OSCA.basic/cell-type-annotation.html), a [Tutorial by Clarke et al. for cell-type annotations](https://pubmed.ncbi.nlm.nih.gov/34031612/), and an [entire chapter of the SC best practices book](https://www.sc-best-practices.org/cellular_structure/annotation.html#automated-annotation).
</details>
<br>

# Using scCATCH

A tool we often use for both mouse and human data cell-type predictions is called [scCATCH](https://github.com/ZJUFanLab/scCATCH/wiki) which, per the author's description in [Shao et al (2020)](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC7031312/), annotates cell-types using a "tissue-specific cellular taxonomy reference database (CellMatch) and [an] evidence-based scoring (ES) protocol". The CellMatch reference is compiled from CellMarker ([Zhang et al., 2019b](https://www.cell.com/servlet/linkout?suffix=e_1_5_1_2_33_2&dbid=8&doi=10.1016/j.isci.2020.100882&key=30289549&cf=)), MCA ([Han et al., 2018](https://www.cell.com/servlet/linkout?suffix=e_1_5_1_2_10_2&dbid=8&doi=10.1016/j.isci.2020.100882&key=29775597&cf=)), CancerSEA ([Yuan et al., 2019](https://www.cell.com/servlet/linkout?suffix=e_1_5_1_2_29_2&dbid=8&doi=10.1016/j.isci.2020.100882&key=30329142&cf=)), and the [CD Marker Handbook](https://www.abcam.com/primary-antibodies/human-cd-antigen-guide) and PMIDs for relevant literature are reported in the prediction results. 

<!-- consider adding [scType](https://cran.r-project.org/web/packages/scCATCH/vignettes/tutorial.html) as an alternative -->

![Image: scCATCH summary from Shao et al (2020).](./images/curriculum/07-CellTypeAnnotation/scCATCH-paper-VizAbstract.jpeg){height=60%}   
<br/>

First, we need to load the scCATCH library. Then, we'll double check that we are using the expected resolution cluster results (this is particularly important if we generated multiple resolutions in our clustering steps), before creating a new object from our `counts` data with `createscCATCH()` and adding our marker genes to the scCATCCH object.

To increase the speed and accuracy of our predictions, we'll create query of relevant tissues (which requires some prior knowledge of the experiment and using the [scCATCH wiki](https://github.com/ZJUFanLab/scCATCH/wiki){target='_blank'} to select tissues from the species) before we run the tool:

```{r, sccatch, message = FALSE, warning = FALSE}
# =========================================================================
# Cell Type Annotation
# =========================================================================

# Load scCATCH
library(scCATCH)

# check that cell identities are set to expected resolution 
all(Idents(geo_so) == geo_so$integrated.sct.rpca.clusters)

```

```{r, sccatch_annotate, message = FALSE, warning = FALSE}
# -------------------------------------------------------------------------
# Annotate clusters using scCATCH

# create scCATCH object, using count data
geo_catch = createscCATCH(data = geo_so@assays$SCT@counts, cluster = as.character(Idents(geo_so)))

# add marker genes to use for predictions
geo_catch@markergene = geo_markers
```

```{r, sccatch_tissues, message = FALSE, warning = FALSE}
# -------------------------------------------------------------------------
# specify tissues/cell-types from the scCATCH reference
tissues_of_interest = c('Blood', 'Peripheral Blood', 'Muscle', 'Skeletal muscle', 'Epidermis', 'Skin')
species_tissue_mask = cellmatch$species == 'Mouse' & cellmatch$tissue %in% tissues_of_interest
geo_catch@marker = cellmatch[species_tissue_mask, ]
```

```{r, sccatch_run, message = FALSE, warning = FALSE}
# -------------------------------------------------------------------------
# run scCATCH to generate predictions
geo_catch = findcelltype(geo_catch)

# look at the predictions
View(head(geo_catch@celltype))
```


```{r, sccatch_cleanup, message = FALSE, warning = FALSE, eval=FALSE}
# -------------------------------------------------------------------------
# cleanup predictions
sc_pred_tbl = geo_catch@celltype %>% 
  select(cluster, cell_type, celltype_score) %>%
  rename(cell_type_prediction = cell_type)
sc_pred_tbl
```

```{r, sccatch_cleanup_table, warning = FALSE, message = FALSE, echo=FALSE}
sc_pred_tbl = geo_catch@celltype %>% 
  select(cluster, cell_type, celltype_score) %>%
  rename(cell_type_prediction = cell_type)

sc_pred_tbl %>%
  kable("html") %>%
  kable_styling(font_size = 12,
                bootstrap_options='condensed',
                html_font = '"Arial Narrow", arial, helvetica, sans-serif') %>%
  scroll_box(height='100%', width='1000px')
```


When we look at our results we can see the cell type score, which gives us an idea of the confidence of that prediction. Not shown here but the full `celltype` table also includes marker genes and PMIDs for relevant literature for each prediction.

In our experience, these kinds of results often help guide cluster annotation but scores can vary and the predictions may need to be revised based on researcher's knowledge of the biological system. As these cell-types correspond to the cell-types and subtypes we'd expect to be present in these data and most of the prediction scores are quite high, we can reasonably use these results to help annotate our clusters with some minor adjustments. 

To revise the cluster annotation from those predicted by scCATCH, it can be helpful to output a table that can be edited in excel or a text editor.
```{r, sccatch_write, message = FALSE, warning = FALSE}
# -------------------------------------------------------------------------
# Write scCATCH predictions to file
write_csv(sc_pred_tbl, file = 'results/tables/scCATCH_cluster_predictions.csv')
```

# Known cell-type markers

To confirm and/or refine the scCATCH predictions, we'll spot check some known markers for immune populations. Then we'll look look at some other key marker genes from some other relevant resources like [Chen et al (2021)](https://www.nature.com/articles/s41467-021-27563-3), [Buechler et al (2021)](https://www.nature.com/articles/s41586-021-03549-5) [Roman (2023)](https://pmc.ncbi.nlm.nih.gov/articles/PMC10296409/#sec3-biomolecules-13-00945 ),  [Li et al (2022)](https://bmcbioinformatics.biomedcentral.com/articles/10.1186/s12859-022-04817-5/figures/2) and [Nestorowa et al (2016)](https://ashpublications.org/blood/article/128/8/e20/35749/A-single-cell-resolution-map-of-mouse) to see if other modifications should be made to the scCATCH predictions:

## Immune marker genes

```{r, imm_marker_genes, fig.show='hold'}
# -------------------------------------------------------------------------
# Create a **named list** of immune cells and associated gene markers

immune_markers = list(
  'Inflam. Macrophage' = c('Cd14', 'Cxcl2'),
  'Platelet' = c('Pf4'),
  'Mast cells' = c('Gata2', 'Kit'),
  'NK cells' = c('Nkg7', 'Klrd1'),
  'B-cell'= c( 'Ly6d', 'Cd19', 'Cd79b', 'Ms4a1'),
  'T-cell' = c( 'Cd3d','Cd3e','Cd3g')
  ) 
immune_markers
```


```{r, imm_marker_plot_basic, fig.show='hold', message=FALSE, warning=FALSE}
# -------------------------------------------------------------------------
# Make a dot plot of immune markers to refine cluster identification
# Since this dot plot is based on known markers instead of empirical markers,
# we'll cue that by using red instead of the default blue. 
immune_markers_plot_basic = DotPlot(geo_so,
                              features = immune_markers,
                              assay = 'SCT',
                              cols = c("lightgrey", "darkred"))

immune_markers_plot_basic
```

Dot plots show the gene expression patterns across clusters. This plot
also groups the genes by their immune cell categories. It's very nice - but the 
genes symbols (bottom) are overlapping and hard to read. Also the immune cell 
categories (top) are truncated. Since this is a ggplot object, we can adjust it
to create a better plot:

```{r, imm_marker_plot_better, fig.show='hold', message=FALSE, warning=FALSE}
# -------------------------------------------------------------------------
# Start with the plot above but reduce the font size and rotate the gene labels 

immune_markers_plot_better = immune_markers_plot_basic + 
  theme(text=element_text(size=10),
        axis.text.x = element_text(angle = 45, vjust = 0.5))

immune_markers_plot_better
```

More readable than before - but with a few more ggplot functions, we can make 
it even better:

- [coord_flip()](https://ggplot2.tidyverse.org/reference/coord_flip.html){target='_blank'} rotates the plot
- [facet_grid()](https://ggplot2.tidyverse.org/reference/facet_grid.html){target='_blank'} adds space between the cell types
- [theme()](https://ggplot2.tidyverse.org/reference/theme.html){target='_blank'} paramaters makes the cell type labels horizontal and to right the gene symbols


```{r, imm_marker_plot_best, fig.show='hold', message=FALSE, warning=FALSE}
# -------------------------------------------------------------------------
# Continuing adjustments to plot above

immune_markers_plot_best = immune_markers_plot_better +
  coord_flip() +
  facet_grid(rows = vars(feature.groups),
             scales = 'free',
             space = 'free',
             switch ='y') +
  theme(strip.placement = 'outside',
        strip.text.y.left = element_text(angle = 0))
immune_markers_plot_best
```

```{r, imm_marker_plot_save, fig.show='hold', message=FALSE, warning=FALSE}
# -------------------------------------------------------------------------
# save dot plot to file

ggsave(filename = 'results/figures/immune_markers_sct_dot_plot.png', 
       plot = immune_markers_plot_best,
       width = 10, height = 5, units = 'in')
```

## Other marker genes

For immune markers (above) we created a **named list** of cell types and gene 
markers. Here's another way to create a named list by building it from a
table. It's a few more steps, but it's a handy alternative and can sometimes
make the data clearer.

```{r, marker_genes_as_tibble, fig.show='hold'}
# -------------------------------------------------------------------------
# Create lists of other cells and associated gene markers

other_markers_string = "
cell_type      | gene_markers         | comment
Pericyte       | Acan, Sox9           | 
SMC            | Acta2, Myh11         | mesenchymal smooth-muscle cell/mesenchymal lineage
Keratinocytes  | Thy1, Dlk1           | fibro progenitors
Myofibroblasts | Tmem100, Cd34, Ly6c1 | hematopoetic stem/activated fibroblast
Fibroblast     | Dpt, Fn1, Col3a1     | activated fib
Endothelial    | Pecam1, Cd38         | from wound healing; Pecam1 also exp in endothelial
HSC            | Ltb, Cd74            | less well defined/conflicting definitions
Erythroid      | Hba-a1               |
"
other_marker_tbl = read_delim(other_markers_string,
                              delim="|",
                              trim_ws=TRUE,
                              show_col_types = FALSE)
other_marker_tbl
```

```{r, marker_genes_as_list, fig.show='hold'}
# -------------------------------------------------------------------------
# Make marker tibble into named list

# Split comma seperated gene symbols into actual lists
other_marker_tbl$gene_markers = str_split(other_marker_tbl$gene_markers, ",")

# Remove any stray whitespace before/after gene symbol
other_marker_tbl$gene_markers = lapply(other_marker_tbl$gene_markers, str_trim)

# Make a named list of cell_type -> marker gene symmbols
other_markers = setNames(other_marker_tbl$gene_markers, other_marker_tbl$cell_type)

other_markers
```

```{r, other_markers_dot_plot_basic, fig.show='hold', warning=FALSE}
# -------------------------------------------------------------------------
# Plot known cell-type markers

other_markers_dot_plot_basic = DotPlot(geo_so, 
                                       features = other_markers,
                                       assay = 'SCT',
                                       cols = c("lightgrey", "darkred")) +
  theme(text=element_text(size=10), axis.text.x = element_text(angle = 45, vjust = 0.5))
other_markers_dot_plot_basic
```

```{r, other_markers_dot_plot_enhanced, fig.show='hold', warning=FALSE}
# -------------------------------------------------------------------------
# Enhance plot aesthetics as we did above

other_markers_dot_plot = other_markers_dot_plot_basic + 
  theme(text=element_text(size=10),
        axis.text.x = element_text(angle = 45, vjust = 0.5)) +
  coord_flip() +
  facet_grid(rows = vars(feature.groups),
             scales = 'free',
             space = 'free',
             switch ='y') +
  theme(strip.placement = 'outside',
        strip.text.y.left = element_text(angle = 0))
other_markers_dot_plot
```

```{r, other_markers_dot_plot_save, fig.show='hold', warning=FALSE}
# -------------------------------------------------------------------------
# Save other markers dot plot to file
ggsave(filename = 'results/figures/other_markers_sct_dot_plot.png', 
       plot = other_markers_dot_plot, width = 12, height = 5, units = 'in')
```

In the first plot, B-cell and T-cell markers seem to line up with the predictions and are limited to single clusters. However, macrophage and dendrocyte markers match to multiple clusters including some annotated with different cell types, so we can consider modifying those cluster labels. 


From the other marker genes, the patterns are less clear so we may want to test other clustering parameters and discuss the results with a researcher familiar with the expected cell types. However, we can notice some patterns that we can use to refine our cluster annotations. 

## Markers from source paper 

<details>
    <summary>*Plotting the expression of genes of interest from Sorkin, Huber et al*</summary>
    Often we have prior information about what cell types are expected in our samples and key marker genes for those populations. This can be an important part of evaluating our clusters, since if genes that are known markers for a specific cell type are found in too many or too few clusters as that can suggest that re-clustering is needed or that some of the clusters should be manually combined before annotating. We can create lists of markers used in figures from the [original paper](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC7002453/Fig1){target='_blank'} before using the same `DotPlot()` function to visualize the expression level and frequency of these genes in our current clusters:

```{r, known_dot_plots_sct_markers, fig.show='hold'}
# -------------------------------------------------------------------------
# Create lists of genes from paper
# https://pmc.ncbi.nlm.nih.gov/articles/PMC7002453/#Fig1
fig1g_markers = c('Cxcl1', 'Cxcl2', 'Ccl2', 'Ccl3', 'Ccl4', 'Il1b', 'Il6', 'Tnf', 'Tgfb1', 'Tgfb2', 'Tgfb3', 'Cxcl5')
fig1h_markers = c('Cxcr2', 'Csf1r', 'Csf3r', 'Tgfbr1', 'Tgfbr3', 'Il1r1', 'Il6ra', 'Lifr', 'Tgfbr2')

# It's diligent to verify genes from paper were measured in the data.
markers_from_paper = unique(c(fig1g_markers, fig1h_markers))
measured_genes = Features(geo_so[["SCT"]])

# If everything worked, missing markers *should be* empty. If it's not empty, we
# could examine the marker symbols to see if we mistyped a symbol or if the
# marker was named something other than the canonical gene symbol in the paper.
missing_markers = setdiff(markers_from_paper, measured_genes)
cat(paste0(length(missing_markers),
            ' gene(s) were listed as markers but not found in the data. Excerpt below:'))
head(missing_markers)
```

```{r, known_dot_plots_sct, fig.show='hold'}
# -------------------------------------------------------------------------
# create DotPlots for genes from paper
fig1g_sct_dot_plot = DotPlot(geo_so,
                             features = fig1g_markers,
                             assay = 'SCT',
                             cols = c("lightgrey", "darkred"))

fig1h_sct_dot_plot = DotPlot(geo_so,
                             features = fig1h_markers,
                             assay = 'SCT',
                             cols = c("lightgrey", "darkred"))

fig1g_sct_dot_plot
fig1h_sct_dot_plot
```

```{r, known_dot_plots_sct_save, fig.show='hold'}
# -------------------------------------------------------------------------
# save plots to file
ggsave(filename = 'results/figures/markers_fig1g_sct_dot_plot.png', 
       plot = fig1g_sct_dot_plot, width = 8, height = 6, units = 'in')
ggsave(filename = 'results/figures/markers_fig1h_sct_dot_plot.png', 
       plot = fig1h_sct_dot_plot, width = 8, height = 6, units = 'in')
```

For known marker genes, it's important to note that since scRNA-seq is only measuring transcriptional signals that markers at the protein level (e.g used for approaches like FACS) may be less effective. An alternative or complement to using marker genes could be methods like using gene set enrichment (GSEA) [as demonstrated in the OSCA book](https://bioconductor.org/books/3.15/OSCA.basic/cell-type-annotation.html#computing-gene-set-activities) to aid in annotations. However, the book ["Best practices for single-cell analysis across modalities" by Heumos, Schaar, Lance, et al. ](https://www.sc-best-practices.org/cellular_structure/annotation.html) points out that "it is often useful to work together with experts ... [like a] biologist who has more extensive knowledge of the tissue, the biology, the expected cell types and markers etc.". In our experience, we find that experience and knowledge of the researchers we work with is invaluable. 

</details>
<br>

<details>
    <summary>*Using raw RNA values for genes of interest from Sorkin, Huber et al*</summary>
    Typically, dot plots are built from normalized, integrated data but you can 
    also show dot plots from unintegrated data. The prior integration step 
    (sctransform) overwrites the normalized data values with normalized and 
    integrated values. To get around this, you can re-load the saved data prior 
    to the integration step.
    
```{r, known_dot_plot_rna, warning=FALSE}
# -------------------------------------------------------------------------
# Visualize manually selected marker genes (from unintegrated data)

# Re-load RNA seq count data
unintegrated_geo_so = readRDS('results/rdata/geo_so_filtered.rds')

# Since there's only one assay, you don't need to specify the *assay* parameter 
# in commands below. That said, I prefer making it explicit.
Assays(unintegrated_geo_so)
Layers(unintegrated_geo_so, assay='RNA')

# Dot plots are typically based on normalized data layer. In prior steps, the 
# data were normalized as a part of running the sctransform() function. We'll 
# do the normalization explicitly.
unintegrated_geo_so = NormalizeData(unintegrated_geo_so)

# The new **data** layer is the normalized count data.
Layers(unintegrated_geo_so, assay='RNA')


# We'll emphasize that these are unintegrated by adding a title and also
# choose a distinct color ramp.
fig1g_unintegrated_dot_plot = DotPlot(unintegrated_geo_so, 
                             features = fig1g_markers,
                             assay = 'RNA',
                             cols = c("lightgrey", "darkgreen")) +
  labs(title="Dot plot from normalized, unintegrated counts", 
       subtitle="Fig1G markers Sorkin, Huber et al")

fig1h_unintegrated_dot_plot = DotPlot(unintegrated_geo_so,
                             features = fig1h_markers,
                             assay = 'RNA',
                             cols = c("lightgrey", "darkgreen")) +
  labs(title="Dot plot from normalized, unintegrated counts", 
       subtitle="Fig1H markers Sorkin, Huber et al")

fig1g_unintegrated_dot_plot

fig1h_unintegrated_dot_plot

ggsave(filename = 'results/figures/markers_fig1g_unintegrated_dot_plot.png',
       plot = fig1g_unintegrated_dot_plot,
       width = 8, height = 6, units = 'in')
ggsave(filename = 'results/figures/markers_fig1h_unintegrated_dot_plot.png',
       plot = fig1h_unintegrated_dot_plot,
       width = 8, height = 6, units = 'in')

# Clean up memory
rm(unintegrated_geo_so, fig1g_unintegrated_dot_plot, fig1h_unintegrated_dot_plot)
gc()

```

</details>
<br>


# Refine and label clusters

Next, we'll use a revised version of the cell type predictions from a file that we prepared by refining the scCATCH predictions. How is a table like this created?

<br/>
![](images/curriculum/07-CellTypeAnnotation/RefiningCellTypeAnnos_altLayout.png)
<br/>

Note: To add the cell-type labels to our Seurat object to replace our clusters' numerical identities, we will create a new metadata object to join in the cell type labels. However this will destroy the row names, which will cause a problem in Seurat so we have to add them back.

::: {.prepared_content .cooking_show data-latex="{cooking_show}"}
**Load a pre-prepared file:**

```{r, annotate_clusters_customize, eval=FALSE}
# -------------------------------------------------------------------------
# Load the revised annotations from prepared file
celltype_annos = read_csv('inputs/prepared_data/revised_cluster_annotations.csv') %>%
  mutate(cluster=factor(cluster))
head(celltype_annos)
```

:::

```{r, annotate_clusters_seurat, eval=FALSE}
# -------------------------------------------------------------------------
# Merge cell types in but as a new table to slide into @meta.data
copy_metadata = geo_so@meta.data
new_metadata = copy_metadata %>% 
  left_join(celltype_annos, by = c('integrated.sct.rpca.clusters' = 'cluster'))
#  We are implicitly relying on the same row order!
rownames(new_metadata) = rownames(geo_so@meta.data)

# Replace the meta.data
geo_so@meta.data = new_metadata 

head(geo_so@meta.data)

```

```{r, annotate_clusters_customize_hidden, echo = FALSE, warning = FALSE, message = FALSE}
celltype_annos = read_csv('inputs/prepared_data/revised_cluster_annotations.csv') %>%
  mutate(cluster=factor(cluster))

# copy metadata (and clean up if needed)
copy_metadata = geo_so@meta.data
copy_metadata = copy_metadata[, 1:11]

# Merge cell types in but as a new table to slide into @meta.data
new_metadata = copy_metadata %>% 
  dplyr::left_join(celltype_annos, 
                   by = c('integrated.sct.rpca.clusters' = 'cluster'))
rownames(new_metadata) = rownames(geo_so@meta.data) #  We are implicitly relying on the same row order!

# Replace the meta.data
geo_so@meta.data = new_metadata 

head(geo_so@meta.data) %>%
  kable("html") %>%
  kable_styling(font_size=12,
                bootstrap_options='condensed',
                html_font = '"Arial Narrow", arial, helvetica, sans-serif') %>%
  scroll_box(height='100%', width='1000px')
```

<br/>
**Checkpoint** : Has the metadata for your `geo_so` object been updated?
<br/>

## Visualise annotated clusters

Lastly, we can generate a revised UMAP plot with our descriptive cluster labels by using our updated Seurat object and providing the new `cell_type` label for the `group.by` argument:

```{r, annos_umap_plot, fig.width = 10, fig.height = 8, fig.show='hold'}
# -------------------------------------------------------------------------
# Make a labeled UMAP plot of clusters
annos_umap_plot = 
  DimPlot(geo_so, group.by = 'cell_type', label = TRUE, reduction = 'umap.integrated.sct.rpca')
annos_umap_plot

ggsave(filename = 'results/figures/umap_integrated_annotated.png', 
       plot =  annos_umap_plot,
       width = 8, height = 6, units = 'in')
```

```{r, annos_umap_plot_day, fig.width = 10, fig.height = 4, fig.show='hold'}
# -------------------------------------------------------------------------
# Repeat, splitting by day
annos_umap_condition_plot = 
  DimPlot(geo_so,
          group.by = 'cell_type',
          split.by = 'time',
          label = FALSE,
          reduction = 'umap.integrated.sct.rpca')
annos_umap_condition_plot

ggsave(filename = 'results/figures/umap_integrated_annotated_byCondition.png', 
       plot = annos_umap_condition_plot,
       width = 14, height = 5, units = 'in')

```

It can also be helpful to compare our annotated UMAP to the original clustering results:

```{r, compare_umap_plots, fig.width = 12, fig.height = 5, fig.show='hold', echo = FALSE, warning = FALSE, message = FALSE}
org_umap_plot = 
  DimPlot(geo_so, group.by = 'integrated.sct.rpca.clusters', label = TRUE, reduction = 'umap.integrated.sct.rpca')

org_umap_plot + annos_umap_plot
```
<br/>
<!-- Add comment on what to observe from the comparison?? -->

<details>
<summary>**Why didn't we just use the original scCATCH labels? (with code)**</summary>
<!-- Update to add the scCATCH labels as they were and generate labeled UMAP -->

Predictive tools can be useful but the labels produced may not match the level of specificity that's relevant to *your* biological question or use the same labels for clusters that don't seems to share similar expression programs 
```{r, catch_umap_plots, fig.width = 12, fig.height = 4.5, fig.show='hold', echo = FALSE, warning = FALSE, message = FALSE}
# -------------------------------------------------------------------------
# optional code to create UMAP with initial scCATCH labels
catch_umap_plot = 
  DimPlot(geo_so, group.by = 'cell_type_prediction', label = TRUE, reduction = 'umap.integrated.sct.rpca')

ggsave(filename = 'results/figures/umap_integrated_scCatch-only.png',
       plot = catch_umap_plot,
       width = 8, height = 6, units = 'in')

# show scCATCH labels compared to original clustering results
org_umap_plot + catch_umap_plot

# -------------------------------------------------------------------------
# Remove the plot variables from the environment to manage memory
rm(catch_umap_plot) 
gc()
```
</details>
<br>
<br/>

<!-- to add - number of cells per cluster and condition/replicate after annotation ?--->

# Save our progress

```{r}
# -------------------------------------------------------------------------
# Discard all ggplot objects currently in environment
# (Ok since we saved the plots as we went along)
rm(list=names(which(unlist(eapply(.GlobalEnv, is.ggplot))))); 
gc()

```

We'll save the scCATCH object and our updated Seurat object with cell type annotations.

```{r, save_rds_hidden, echo = FALSE}
if(!file.exists('results/rdata/geo_so_sct_integrated_with_catch.rds')) {
  saveRDS(geo_so, file = 'results/rdata/geo_so_sct_integrated_with_annos.rds')
}

if(!file.exists('results/rdata/geo_catch.rds')) {
  saveRDS(geo_catch, file = 'results/rdata/geo_catch.rds')
}
```


```{r, save_rds, eval=FALSE}
# -------------------------------------------------------------------------
# Save Seurat object and annotations
saveRDS(geo_so, file = 'results/rdata/geo_so_sct_integrated_with_annos.rds')
saveRDS(geo_catch, file = 'results/rdata/geo_catch.rds')
```

<br/>
<br/>

# Summary

<br/>
<table class='fig'><tr><td class='fig'>
![](images/graphical_abstracts/07-CellTypeAnnos-Abstract.png)
</td></tr><tr><tr class='fig'><td class='fig'>Starting with clustered data, we can summarize and compare gene expression for known gene markers within each cluster, irrespective of sample or condition, to help label clusters with the appropriate cell-type or subtype.
</td></tr>
</table>
<br/>

In this section we:  
   
- Used `scCATCH` to generate predicted cell-type annotations for our initial clusters    
- Plotted the expression per cluster of known markers to confirm and refine the predictions   
- Finalized cell type labels for our clusters     


Next steps: Differential Expression

----

These materials have been adapted and extended from materials listed above. These are open access materials distributed under the terms of the [Creative Commons Attribution license (CC BY 4.0)](http://creativecommons.org/licenses/by/4.0/), which permits unrestricted use, distribution, and reproduction in any medium, provided the original author and source are credited.

<br/>
<br/>
<hr/>
| [Previous lesson](06-MarkerVisualization.html) | [Top of this lesson](#top) | [Next lesson](08-DifferentialExpression.html) |
| :--- | :----: | ---: |
