---
title: "Cell Type Annotation"
author: "UM Bioinformatics Core"
date: "`r Sys.Date()`"
output:
        html_document:
            includes:
                in_header: header.html
            theme: paper
            toc: true
            toc_depth: 4
            toc_float: true
            number_sections: false
            fig_caption: true
            markdown: GFM
            code_download: true
---

<style type="text/css">
body, td {
   font-size: 18px;
}
code.r{
  font-size: 12px;
}
pre {
  font-size: 12px
}
</style>

```{r, include = FALSE}
source("../bin/chunk-options.R")
knitr_fig_path("XX")
```

# Workflow Overview {.unlisted .unnumbered}

<br/>
<img src="images/wayfinder/wayfinder.png" alt="wayfinder" style="height: 400px;"/>
<br/>
<br/>

# Introduction

General goal:

<!-- Point out possible gaps between how single-cell data is presented versus the reality of running an analysis
E.g. Papers might report final filtering thresholds, but our process for choosing those is likely to be more iterative -->

Even more so than the previous sections, cell type annotation is ...

## Objectives

<!--Add specific goals for section-->

----

# Cell type predictions

<!-- [OSCA](https://bioconductor.org/books/3.15/OSCA.basic/cell-type-annotation.html) "The most challenging task in scRNA-seq data analysis is arguably the interpretation of the results. Obtaining clusters of cells is fairly straightforward, but it is more difficult to determine what biological state is represented by each of those clusters. Doing so requires us to bridge the gap between the current dataset and prior biological knowledge, and the latter is not always available in a consistent and quantitative manner. Indeed, even the concept of a “cell type” is not clearly defined, with most practitioners possessing a “I’ll know it when I see it” intuition that is not amenable to computational analysis. As such, interpretation of scRNA-seq data is often manual and a common bottleneck in the analysis workflow." -->

<!--For mouse, we often use scCATCH-->

<!--More options for human samples than mouse-->

For human, if a relevant reference is available would recommend using [Azimuth (created by authors of Seurat)](https://azimuth.hubmapconsortium.org/) <!-- and [10x tutorial](https://www.10xgenomics.com/analysis-guides/automated-cell-type-annotation-from-r-to-loupe-using-louper) that includes example of using Azimuth -->

Other tools?

<!-- [sc best practices](https://www.sc-best-practices.org/cellular_structure/annotation.html#automated-annotation) -->

<!-- [OSCA](https://bioconductor.org/books/3.15/OSCA.basic/cell-type-annotation.html#assigning-cell-labels-from-reference-data) "A conceptually straightforward annotation approach is to compare the single-cell expression profiles with previously annotated reference datasets. Labels can then be assigned to each cell in our uncharacterized test dataset based on the most similar reference sample(s), for some definition of “similar”. This is a standard classification challenge that can be tackled by standard machine learning techniques such as random forests and support vector machines. Any published and labelled RNA-seq dataset (bulk or single-cell) can be used as a reference, though its reliability depends greatly on the expertise of the original authors who assigned the labels in the first place.

In this section, we will demonstrate the use of the SingleR method (Aran et al. 2019) for cell type annotation. This method assigns labels to cells based on the reference samples with the highest Spearman rank correlations, using only the marker genes between pairs of labels to focus on the relevant differences between cell types. It also performs a fine-tuning step for each cell where the correlations are recomputed with just the marker genes for the top-scoring labels. This aims to resolve any ambiguity between those labels by removing noise from irrelevant markers for other labels. Further details can be found in the SingleR book from which most of the examples here are derived." -->


<!-- [Ouyang](https://ouyanglab.com/singlecell/clust.html#annotating-clusters)

# Using scCATCH

<!--- add context about scCATCH--->

[scCATCH documentation](https://github.com/ZJUFanLab/scCATCH/wiki) and [paper by Shao et al](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC7031312/)

Load the scCATCH library
```{r, eval=FALSE}
library(scCATCH)
```

Check that we are using the expected resolution cluster results
```{r, eval=FALSE}
all(Idents(geo_so) == geo_so$SCT_snn_res.0.4)
```

Create scCATCH object
```{r, eval=FALSE}
geo_catch = createscCATCH(data = geo_so@assays$SCT@counts, cluster = as.character(Idents(geo_so)))
```

Convert marker genes to scCATCH compatible format and add to scCATCH object
```{r, eval=FALSE}
catch_markers = geo_markers %>% rename('logfc' = 'avg_log2FC')

geo_catch@markergene = geo_markers
```

Create query for database & run tool
```{r, eval=FALSE}
geo_catch@marker = cellmatch[cellmatch$species == 'Mouse' & cellmatch$tissue %in% c('Blood', 'Peripheral Blood', 'Muscle', 'Skeletal muscle', 'Epidermis', 'Skin'), ]

geo_catch = findcelltype(geo_catch)
```


Look at our results, including cell type score
```{r, eval=FALSE}
geo_catch@celltype %>% select(cluster, cell_type, celltype_score)
```

Add celltype predictions to our Seurat object. Note that we will create a new metadata object where we join cell types, this destroys rownames which will cause a problem in Seurat, so we have to add them back. We are implicitly relying on the same row order!
```{r, eval=FALSE}
catch_celltypes = geo_catch@celltype %>% select(cluster, cell_type)


new_metadata = geo_so@meta.data %>% left_join(catch_celltypes, by = c('SCT_snn_res.0.4' = 'cluster'))
rownames(new_metadata) = rownames(geo_so@meta.data)

geo_so@meta.data = new_metadata 
```

Create a UMAP plot with our new labels
```{r, eval=FALSE}
catch_umap_plot = DimPlot(geo_so, group.by = 'cell_type', label = TRUE, reduction = 'umap.sct.integrated.rpca')

ggsave(filename = 'ISC_R/results/figures/umap_integrated_catch.png', plot = catch_umap_plot, width = 8, height = 7, units = 'in')
```



# Summary

Now that we have ... 

Next steps: Differential Expression



----

These materials have been adapted and extended from materials listed above. These are open access materials distributed under the terms of the [Creative Commons Attribution license (CC BY 4.0)](http://creativecommons.org/licenses/by/4.0/), which permits unrestricted use, distribution, and reproduction in any medium, provided the original author and source are credited.

<br/>
<br/>
<hr/>
| [Previous lesson](06-MarkerVisualization.html) | [Top of this lesson](#top) | [Next lesson](08-DifferentialExpression.html) |
| :--- | :----: | ---: |

