---
title: "Cell Type Annotation"
author: "UM Bioinformatics Core"
date: "`r Sys.Date()`"
output:
        html_document:
            includes:
                in_header: header.html
            theme: paper
            toc: true
            toc_depth: 4
            toc_float: true
            number_sections: false
            fig_caption: true
            markdown: GFM
            code_download: true
---

<style type="text/css">
body, td {
   font-size: 18px;
}
code.r{
  font-size: 12px;
}
pre {
  font-size: 12px
}
</style>

```{r, include = FALSE}
source("../bin/chunk-options.R")
knitr_fig_path("XX")
```

# Workflow Overview {.unlisted .unnumbered}

<br/>
<img src="images/wayfinder/wayfinder.png" alt="wayfinder" style="height: 400px;"/>
<br/>
<br/>

# Introduction

Point out possible gaps between how single-cell data is presented versus the reality of running an analysis
E.g. Papers might report final filtering thresholds, but our process for choosing those is likely to be more iterative

General goal: Filtering out the "cells" that either were poorly measured or are either not single-cells or not cells at all

## Objectives

<!--Add specific goals for section-->

----

# Cell type predictions

More options for human samples than mouse

For mouse, we often use scCATCH

For human, if a relevant reference is available, would recommend XX tools (created by authors of Seurat)

Other tools?

# Using scCATCH

<!--- add context about scCATCH--->
Load the scCATCH library
```{r, eval=FALSE}
library(scCATCH)
```

Check that we are using the expected resolution cluster results
```{r, eval=FALSE}
all(Idents(geo_so) == geo_so$SCT_snn_res.0.4)
```

Create scCATCH object
```{r, eval=FALSE}
geo_catch = createscCATCH(data = geo_so@assays$SCT@counts, cluster = as.character(Idents(geo_so)))
```

Convert marker genes to scCATCH compatible format and add to scCATCH object
```{r, eval=FALSE}
catch_markers = geo_markers %>% rename('logfc' = 'avg_log2FC')

geo_catch@markergene = geo_markers
```

Create query for database & run tool
```{r, eval=FALSE}
geo_catch@marker = cellmatch[cellmatch$species == 'Mouse' & cellmatch$tissue %in% c('Blood', 'Peripheral Blood', 'Muscle', 'Skeletal muscle', 'Epidermis', 'Skin'), ]

geo_catch = findcelltype(geo_catch)
```


Look at our results, including cell type score
```{r, eval=FALSE}
geo_catch@celltype %>% select(cluster, cell_type, celltype_score)
```

Add celltype predictions to our Seurat object. Note that we will create a new metadata object where we join cell types, this destroys rownames which will cause a problem in Seurat, so we have to add them back. We are implicitly relying on the same row order!
```{r, eval=FALSE}
catch_celltypes = geo_catch@celltype %>% select(cluster, cell_type)


new_metadata = geo_so@meta.data %>% left_join(catch_celltypes, by = c('SCT_snn_res.0.4' = 'cluster'))
rownames(new_metadata) = rownames(geo_so@meta.data)

geo_so@meta.data = new_metadata 
```

Create a UMAP plot with our new labels
```{r, eval=FALSE}
catch_umap_plot = DimPlot(geo_so, group.by = 'cell_type', label = TRUE, reduction = 'umap.sct.integrated.rpca')

ggsave(filename = 'ISC_R/results/figures/umap_integrated_catch.png', plot = catch_umap_plot, width = 8, height = 7, units = 'in')
```



# Summary

Now that we have ... 

Next steps: Differential Expression



----

These materials have been adapted and extended from materials listed above. These are open access materials distributed under the terms of the [Creative Commons Attribution license (CC BY 4.0)](http://creativecommons.org/licenses/by/4.0/), which permits unrestricted use, distribution, and reproduction in any medium, provided the original author and source are credited.

<br/>
<br/>
<hr/>
| [Previous lesson](06-MarkerVisualization.html) | [Top of this lesson](#top) | [Next lesson](08-DifferentialExpression.html) |
| :--- | :----: | ---: |

