---
title: "Differential Expression Analysis"
author: "UM Bioinformatics Core"
date: "`r Sys.Date()`"
output:
        html_document:
            includes:
                in_header: header.html
            theme: paper
            toc: true
            toc_depth: 4
            toc_float: true
            number_sections: false
            fig_caption: true
            markdown: GFM
            code_download: true
---

<style type="text/css">
body, td {
   font-size: 18px;
}
code.r{
  font-size: 12px;
}
pre {
  font-size: 12px
}
</style>

```{r, include = FALSE}
source("../bin/chunk-options.R")
knitr_fig_path("XX")
```

# Workflow Overview {.unlisted .unnumbered}

<br/>
<img src="images/wayfinder/wayfinder.png" alt="wayfinder" style="height: 400px;"/>
<br/>
<br/>

# Introduction

Differential expression comparison is a key step to understanding what genes in what cell-types might be contributing to the misregulation in wound healing that lead to bone formation in this experiment.

As a reminder, our data includes cells isolated from issue from day 0 (prior to injury) as controls, and days 7 and 21 post-injury as experimental conditions.

![](./images/curriculum/experimental_design.jpg)

Point out possible gaps between how single-cell data is presented versus the reality of running an analysis
E.g. Papers might report final filtering thresholds, but our process for choosing those is likely to be more iterative

## Objectives

<!--Add specific goals for section-->

----

# Differential Expression

## Model options

Wilcoxin (default), DESeq2, MAST, etc.

## Standard comparisons

![](./images/curriculum/experimental_design.jpg)

<!-- add UMAP with facets by groups to show what's being compared? -->

```{r, eval=FALSE}
### Comparison 1, D21 vs D0 not pseudo-bulk

geo_so$day.celltype = paste(geo_so$day, geo_so$cell_type, sep = '_')

Idents(geo_so) = 'day.celltype'
```


```{r, eval=FALSE}
de_cell_monocytes_D21_vs_D0 = FindMarkers(
    object = geo_so,
    slot = 'data', test = 'wilcox',
    ident.1 = 'Day21_Monocyte', ident.2 = 'Day0_Monocyte')
```

```{r, eval=FALSE}
table(de_cell_monocytes_D21_vs_D0$p_val_adj < 0.05 & abs(de_cell_monocytes_D21_vs_D0$avg_log2FC) > 1.5)
```
~~~
FALSE  TRUE 
10375    15 
~~~

Run additional comparison - Day 7 versus Day 0
```{r, eval=FALSE}
### Comparison 2, D7 vs D0 not pseudo-bulk

de_cell_monocytes_D7_vs_D0 = FindMarkers(
    object = geo_so,
    slot = 'data', test = 'wilcox',
    ident.1 = 'Day7_Monocyte', ident.2 = 'Day0_Monocyte')
```

Look at results - note that thresholds here are more stringent than Seurat defaults
```{r, eval=FALSE}
table(de_cell_monocytes_D7_vs_D0$p_val_adj < 0.05 & abs(de_cell_monocytes_D7_vs_D0$avg_log2FC) > 1.5)
```

~~~
FALSE  TRUE 
10360   313 
~~~

## Pseudobulk comparisons

Standard approach versus pseudobulk

<!-- add UMAP with facets by sample to show what's being compared? -->

<!-- [10x analysis guide for differential expression with biological replicates](https://www.10xgenomics.com/analysis-guides/differential-gene-expression-analysis-in-scrna-seq-data-between-conditions-with-biological-replicates) -->

<!-- [Ouyang](https://ouyanglab.com/singlecell/clust.html#sec:diffexpr) For (ii), single-cell studies are now more complex, often including samples from multiple donors. It is possible that there are more cells being profiled from a specific donor than others and this can skew the DE results. For example, consider a study where there are three diseased samples (D1,D2,D3) and three healthy samples (H1,H2,H3_ and there are a lot more cells from sample D1. In this scenario, a gene that is specifically expressed in D1 may be identified as differentially expressed when comparing diseased and healthy single-cells. In fact, Squair et al. have shown that ignoring biological replicates can often result in false discoveries in single-cell DE (Squair et al. 2021). To circumvent this, Squair et al. suggested the use of pseudo-bulk profiles where the single-cell profiles from each individual is being collected and then subjected to bulk RNA-seq based DE methods. Another possible approach is to downsample the number of single cells such that each individual have roughly a similar number of cells. -->

```{r, eval=FALSE}
pseudo_catch_so = AggregateExpression(geo_so, assays = 'RNA', return.seurat = TRUE, group.by = c('cell_type', 'day', 'replicate'))
```

Set up labels to use for comparisons
```{r, eval=FALSE}
pseudo_catch_so$day.celltype = paste(pseudo_catch_so$day, pseudo_catch_so$cell_type, sep = '_')

Idents(pseudo_catch_so) = 'day.celltype'
```

Run comparison between Day 21 and Day 0:
```{r, eval=FALSE}
de_pseudo_monocytes_D21_vs_D0 = FindMarkers(
    object = pseudo_catch_so, 
    ident.1 = 'Day21_Monocyte', ident.2 = 'Day0_Monocyte', 
    test.use = 'DESeq2')
```

Look at results,  using thresholds of adjusted p-value < 0.05 and logFC > 1.5. Note - this fold-change threshold is more stringent so may need to tailor to specific dataset
```{r, eval=FALSE}
table(de_pseudo_monocytes_D21_vs_D0$p_val_adj < 0.05 & abs(de_pseudo_monocytes_D21_vs_D0$avg_log2FC) > 1.5)
```
~~~
FALSE  TRUE 
14209    32
~~~

Output results
```{r, eval=FALSE}
write_csv(de_pseudo_monocytes_D21_vs_D0, file = 'ISC_R/results/tables/de_pseudo_monocytes_D21_vs_D0.csv')
```

Since we're working with pseudobulk data (so no percentag =e of )
```{r, eval=FALSE}
pseudo_monocytes_D21_vs_D0_volcano = ggplot(de_pseudo_monocytes_D21_vs_D0, aes(x = avg_log2FC, y = -log10(p_val))) + geom_point()
ggsave(filename = 'ISC_R/results/figures/volcano_de_pseudo_monocytes_D21_vs_D0.png', plot = pseudo_monocytes_D21_vs_D0_volcano, width = 7, height = 7, units = 'in')
```




# Downstream approaches

<!--- [Ouyang Lab's Trajectory analysis](https://ouyanglab.com/singlecell/dimrd.html) --->

# Summary

Now that we have ... 



----

These materials have been adapted and extended from materials listed above. These are open access materials distributed under the terms of the [Creative Commons Attribution license (CC BY 4.0)](http://creativecommons.org/licenses/by/4.0/), which permits unrestricted use, distribution, and reproduction in any medium, provided the original author and source are credited.

<br/>
<br/>
<hr/>
| [Previous lesson](07-CellTypeAnnos.html) | [Top of this lesson](#top) | [Workshop Wrap Up](workshop_wrap_up.html) |
| :--- | :----: | ---: |

