---
title: "Clusters, PCs, and Resolutions: FAQ"
author: "UM Bioinformatics Core Workshop Team"
date: "`r Sys.Date()`"
lang: "en"
output:
  html_document:
      includes:
          in_header: header-clusters_faq.html
      theme: paper
      toc: false
      toc_depth: 4
      toc_float: true
      number_sections: false
      fig_caption: true
      markdown: GFM
      code_download: false
---

<style type="text/css">

body, td {
   font-size: 18px;
}
code.r{
  font-size: 12px;
}
pre {
  font-size: 12px
}


table.fig, th.fig, td.fig {
  border: 1px solid black;
  border-collapse: collapse;
  padding: 15px;
}

table{
   width:100%;
}
</style>

```{r setup, include=FALSE}
# Note that a table ande several images referenced by this page are generated in
# a separate script /images/cluster_matrix/matrix.R

library(tidyverse)
library(kableExtra)
```
---

### Why do we cluster on Principle Components (PC) instead of genes?

* Most genes show little variance at all; those that do often behave 
  in concert (regulating up and down in a coherent experssion program). Because
  of this behavior, **clustering directly on genes doesn't work very well 
  mathematically.** (Because in 20k dimensions, the notion of distance breaks down;
  everything is too far from everything else.)
* PCs naturally diminish noise and aggregate signal in expression patterns 
   across related genes; this makes patterns clearer.
* It's **much much faster** to cluster 15 PCs than 20k genes.

---

### How can such a small number of PCs faithfully represent so many genes? Aren't we throwing away a ton of information?

* Selecting a few PCs helps by bringing the most important parts of the picture 
  into focus. Remember that each PCs represents a gene expression **pattern** 
   across many genes.


| Genes look like this | A Principle Component shows this | 
|:-:|:-:|
| ![](images/clusters_faq/PC-before.png) | ![](images/clusters_faq/PC-after.png) |

* Recall from the earlier [elbow plot](04-PCAandIntegration.html#Visualizing_relative_contributions_of_each_PC) 
  that almost all the variance (signal) in the dataset is captured in the first 
  handful of PCs. In our example dataset, the first 50 PCs capture over 99% of the
  cumulative variance.

---

### So, what's the correct resolution and number of PCs?

* That depends on **how many cell types/cell states do you expect.** 
  And the answer to that depends on:

  -  the tissue/bio-fluid of origin
  - the quality of the sample / suspension
  - your experimental design
  - your research question
  - the number of cells profiled
  - the number of genes detected
  - the range and specificity of cellular expression programs

* For our sample data (bone biopsies from mice), we anticipate about 20 cell types.
  
---

### Couldn't the computer figure out the best PCs and resolution?

* The computer could try a bunch of PCs and resolutions to get the right number of
clusters - and those clusters might look good numerically. But they wouldn't be 
the best representation of the different cells types **because the algorithm 
doesn't understand the cell biology**.

---

### So how do we pick the right PCs and resolution?

* You will want to iterate through a few combinations of PCs and resolutions.
* It helps to know how PCs and resolution work together to influence the number 
  of clusters.
* We applied the clustering and UMAP code blocks from [earlier](05-ProjectionAndClustering.html) 
  and tried a bunch of clusterings each with a different number of PCs and a 
  different resolution.

<table class='fig'>
<tr><th class='fig'>Different cluster counts across resolutions and PCs</th></tr>
<tr><td class='fig'>![](images/clusters_faq/pc_res_cluster.png)</td></tr>
<tr><td class='fig'>
* Each dot represents the number of clusters produced with a specific resolution and number of PCs.
* Each colored line is a different resolution.
* Each column is a different num of PCs.
* Note that PCs have a modest effect on the number of clusters while 
  resolution has a strong effect on the number of clusters.
</td></tr>
</table>

---

### What do the cluster UMAPs look like at different PCs and resolutions?

- Each UMAP below was made with a different num of PCs and different resolution. (You can click on any UMAP to make it larger.)
- Each column is a different PC. Each row is a different resolution.
- The first row (in black and white) shows the UMAP projection 
  without any clustering.
- Note that as PCs increase, the UMAPs go from "blobby" to"spindly". (They gain local structure.)
- As the resolution increases, the UMAP projection is the same for a given PC, 
  but that projection is cut up into more and more clusters.


```{r render_matrix, echo=FALSE, warning=FALSE}
base_dir = "images/clusters_faq"
cluster_stats <- read_csv(sprintf("%s/pc_res_cluster.csv", base_dir), show_col_types = FALSE) %>%
  arrange(included_pcs, resolution)

# This gnarly block generates a the html to render a table (grid) of UMAPS
# across different PCs and resolutions. We use the raw HTML for the images.
# The hover-shadow and dbox tags enable the click-to-expand lightbox effect.
df = cluster_stats %>%
    mutate(img_title=sprintf('clusters=%s : pcs=%s res=%s ', cluster_count, included_pcs, resolution)) %>%
    mutate(img_src=sprintf('%s/%s', base_dir, umap_filename)) %>%
    mutate(img=sprintf("<img alt='%s' title='%s' src='%s' class='hover-shadow'/>", img_title, img_title, img_src)) %>%
    mutate(anchor=sprintf("<a class='dbox' href='%s'>%s</a>", img_src, img)) %>%
    rename(pcs=included_pcs) %>%
    select(pcs, resolution, anchor) %>%
    pivot_wider(names_from=pcs, values_from=anchor)

# Setting escape = FALSE allows kable to pass through the raw HTML so those cells
# render as images.
df %>%
  kbl(booktabs = T, align = "c", format = "html", escape = FALSE) %>%
  column_spec(1,extra_css = "vertical-align:middle;", bold=TRUE) %>%
  kable_styling()
```

```{=html}
<dialog id="js-dbox" class="dbox-dialog">
  <div class="dbox-dialog--content">
    <div class="dbox-dialog--container">
      <p class="dbox-dialog--caption" style="font-size:x-large"></p>
        <button type="button"  class="dbox-dialog--close" style="font-size:x-large">&times;</button>
    </div>
    <img src="" alt="" class="dbox-dialog--image">
  </div>
</dialog>
```

```{js, echo=FALSE}
    document.addEventListener("DOMContentLoaded", function() {
        let lightbox = new Dbox();
        lightbox.run();
    });
```

### But these are all still just ... clusters - how do I know if they are biologically accurate?

* We need two more tools to validate these are good biological clusters:

  - We need to identify **gene markers** that behave differently across clusters.
  - We need to **annotate** the clusters with prospective cell types.

* We will cover both of topics in the next sections.

--------

| [Previous lesson](05-ProjectionAndClustering.html) | [Top of this lesson](#top) | [Next lesson](06-MarkerVisualization.html) |
| :--- | :----: | ---: |