---
title: "Running Seurat on UMich Great Lakes HPC"
author: "UM Bioinformatics Core Workshop Team"
date: "`r Sys.Date()`"
lang: "en"
output:
        html_document:
            includes:
                in_header: header.html
            theme: paper
            toc: true
            toc_depth: 4
            toc_float: true
            number_sections: false
            fig_caption: true
            markdown: GFM
            code_download: false
---

<style type="text/css">
body, td {
   font-size: 18px;
}
code.r{
  font-size: 12px;
}
pre {
  font-size: 12px
}
</style>

```{r klippy, echo=FALSE, include=TRUE}
klippy::klippy(lang = c("r", "markdown", "bash"), position = c("top", "right"))
```

```{r, include = FALSE}
source("../bin/chunk-options.R")
source("../bin/set_values.R")
```

<br/>
<br/>

It's possible to run Seurat and several complementary libraries using 
University of Michigan's **Great Lakes** High Performance Cluster (HPC). This is
a nice approach because:

- Sometimes scRNA-Seq analysis requires a LOT of memory (RAM). It's easy to ask 
  for more memory on Great Lakes and Great Lakes nodes have more memory 
  available than a typical laptop/workstation.
- From Cell Ranger to RStudio to Seurat, much of the software is pre-installed.
  Installing bioinformatics software is typically nuanced and often 
  unpleasantly complex - so having it ready is a _big win_.

That said, using Great Lakes is different than using your laptop/workstation.
A few key ideas:

-  Great Lakes is provided by [UM Advanced Research Computing](https://its.umich.edu/advanced-research-computing){target="_blank"}. It is only available to folks affiliated with UM.
- Great Lakes uses **Open OnDemand** to reserve a part of the cluster to run 
  your analysis in RStudio.  
-  Great Lakes is a requires some setup.
-  Great Lakes is a recharge service, BUT it is heavily subsidized and extremely 
   affordable for most researchers. (Many researchers who are doing a low to 
   moderate amount of data analysis may not have to pay anything.)

---

## You will need

1. You need a a UM core-imaged workstation and a web browser.

2. You need to be on UM campus or connected to the UM VPN or Michigan Medicine 
   VPN using Cisco Secure Client. (This will require Duo 2-factor authentication.)

3. You need a user account on GreatLakes. The first time, you need to request 
   this from ARC; it can take 1-2 days to get set up:
   - see [Getting Started with the Great Lakes Cluster](https://its.umich.edu/advanced-research-computing/high-performance-computing/great-lakes/getting-started){target="_blank"} to get setup.
   - also, to save money, request the [UM Research Compute Package (UMRCP)](https://its.umich.edu/advanced-research-computing/research-computing-package){target="_blank"} to offset costs with subsidy from UM.

4. You will need enough storage space. How much depends on the experimental design and what analyses you want to do. Using    the workshop inputs as a guide:

    | Data | Storage required (Gb) |
    |------|-----------------------|
    | Total raw FASTQ file size | 546 |
    | Full Cell Ranger outputs | 8 |
    | Minimal inputs to run workshop analyses | 3 |
    | Minimal space to store outputs | 6 |
  
    If you focus on the minimal inputs and space to execute the workshop analyses, 
    10Gb is a good working number. Each user is allocated 80Gb of storage in their 
    home dir. As a new user, you can execute this tutorial from your home dir. As 
    you experiments and analyses grow, you can check your available space from the 
    command line (`homw-quota`) and request more storage from ARC as necessary. 


5. You will need data. Below, we'll show you how to download the input data used
   in the workshop, but you can adapt that to use your own data.

---

## Launch the RStudio Session

Once you've got everything you need from above, you're going to launch a 
RStudio session that runs on Great Lakes and is accessed through your browser.

1. If off-campus, connect to the VPN using Cisco Secure Client.

2. In your browser, go to Great Lakes Open On Demand:   
   [greatlakes.arc-ts.umich.edu](https://greatlakes.arc-ts.umich.edu){target="_blank"} and 
   login with your uniqname and password. 

3. In the menu at the top of the screen, click **Interactive Apps** and select 
   **RStudio**. ![](images/seuratOnGreatLakes/ood-interactive-apps.png)

4. The previous step will display a _launch configuration page_ with several fields, e.g.
       ![](images/seuratOnGreatLakes/ood-launch-config.png)
    ...
  
    Enter values below in the corresponding fields:
  
    | Field | Value |
    |-------|-------|
    | R Version | **Rtidyverse/4.4.0** |
    | RStudio version | **RStudio/2024.04.1** |
    | Slurm account | _(This is in the email from ARC)_ |
    | Partition | **standard** |
    | Number of hours | **4** <br/> _(Enough to get started, adjust if you need more time)_|
    | Number of cores | **4** |
    | Memory (GB) | **16** <br/> _(Enough to get started, ok to boost to 24 or 32 if you ever run out)_ |
    | Module commands | **load Bioinformatics r-seurat/5.1.0-R-4.4.1-c3m7yfq BPCells presto  scCATCH RglmGamPoi RDESeq2** |

5. After you've entered values above, at the bottom of this page click **Launch**. 
   (Conveniently, the values you entered above will now be your default for 
   launching an RStudio session.) The screen will update to show Great Lakes is 
   preparing your session. 
   ![](images/seuratOnGreatLakes/ood-queue-rstudio.png){width=50%}

6. When the session is ready (usually a few seconds later), the screen will update:
   ![](images/seuratOnGreatLakes/ood-launch-rstudio.png){width=50%}
    - Drag the Compression slider all the way to the right
    - Drag Image Quality slider all the way to the right
    - Click **Launch RStudio**

7. This will open a new browser tab that contains your RStudio session:
![](images/seuratOnGreatLakes/rstudio-startup.png)

8. You may see a prompt asking permission to use the clipboard. Click **Allow**.
![](images/seuratOnGreatLakes/clipboard-permission.png){width=25%}

---

## Using Open OnDemand RStudio

### Exiting
- The RStudio session will last for 4 hours (or however long you specified in 
  launch config above).
- If you close the browser window, you can click **Launch RStudio** again from
  the [Open OnDemand session page](https://greatlakes.arc-ts.umich.edu/pun/sys/dashboard/batch_connect/sessions){target="_blank"} 
- You can end the session several ways. _Make sure your scripts and data are saved because these actions do not prompt to save or confirm; and once the session is closed it is gone for good._
  - In RStudio menu at the top, click File | Quit
  - In RStudio, click Session | Quit Session
  - In RStudio Close the Rstudio window (top right X)
  ![](images/seuratOnGreatLakes/rstudio-close.png){width=25%}
  - In [Open OnDemand session page](https://greatlakes.arc-ts.umich.edu/pun/sys/dashboard/batch_connect/sessions){target="_blank"}, click **Delete**.
  ![](images/seuratOnGreatLakes/ood-delete.png){width=50%}

---  
  
### Keyboards shortcuts and copy-paste

- In the Rstudio window, you can Zoom in and out using _control +_ and _control -_. 
- Running RStudio in the browser inside the Greatlakes HPC complicates copying 
  and pasting from your computer to/from the remote session.
  - Copy-paste functionality doesn't always work perfectly. Sometimes you may 
    need to copy twice to get the content into the buffer. So check your work.
  - Within the RStudio window you can right-click to get a copy or paste context menu.
  - Keyboard shortcuts are _control C_ and _control V_. This is familliar to PC 
    users *but will take some getting used to for Mac-users*.
  - Copy and paste _mostly work_ using Chrome or Edge. Other browsers (e.g. 
    Safari) require extra steps:
    - Open the control panel (left side) 
    ![](images/seuratOnGreatLakes/novnc-controls.png){width=25%}
    - Select the clipboard icon.
    ![](images/seuratOnGreatLakes/novnc-controlpanel.png){width=25%}
    - Copy text into the clipboard area; within that area, use the native 
      keyboard shortcuts, i.e. control-C on PC or command-C on Mac). Close the
      control panel.
      ![](images/seuratOnGreatLakes/novnc-clipboard.png)
    - Right-click (or _control V_) to paste the contents into the console.
    ![](images/seuratOnGreatLakes/novnc-paste.png){width=50%}

- If the copy-paste pattern is too flaky or clumsy for your tastes, you can
  upgrade your experience by switching from using Open OnDemand in the browser
  to a [VNC viewer](https://www.realvnc.com/en/connect/download/viewer/){target="_blank"}. 
  (Macs have a built in VNC viewer, but that viewer doesn't support copy/paste 
  with a Unix system.) Details on how to set this up are outside the scope
  of this guide, but reach out and we can help.

---

### Open a browser

- FYI, you can launch the Firefox browser by clicking on Applications (top left) | 
  Web Browser. This will open another application tab (along the top).
  ![](images/seuratOnGreatLakes/ood-browser.png){width=50%}
- If you want to do any substantial copy/paste, we recommend you use this browser 
  window because it's more reliable to copy/paste within the session than from
  your local computer.

---

## To re-run the workshop analyses

1. The Seurat inputs (cellranger triples and dbcells files) used in the 
    workshop can be installed locally. In the RStudio window, click on the 
    **Terminal** tab. The tab will be blank with a prompt that looks something 
    like this: `[YOUR_UNIQNAME@glXXXX ~]$ `

2. Paste the following block into the Terminal prompt: and hit _Enter/Return_. 
   This will take a minute or two to download and unpack the inputs.

    ```
    # download Seurat inputs --------------------------------------------------
    mkdir -p intro_scrnaseq_workshop/ISC_R
    cd intro_scrnaseq_workshop/ISC_R
    
    # Use curl to download a ~2 Gb tarball
    # We'll use evironment variables to avoid extremely long command lines
    source_url="`r workshop_vars$aws_s3_bucket`"
    source_file="`r workshop_vars$aws_s3_file`"
    curl -o workshop_isc_inputs.tgz ${source_url}/${source_file}

    # tar unpacks the tarball into directories
    tar xzvf workshop_isc_inputs.tgz

    # Since we have unpacked the tarball, we can remove it
    rm workshop_isc_inputs.tgz
    
    ls inputs
    
    ```

    The last line should show two directories:
    
    ```
    10x_cellranger_filtered_triples
    prepared_data
    ```

3. Now you can review the workshop [analysis scripts](analysis_scripts.html){target="_blank"}
   and execute them on these input data. 

---

## Transfering your files

To analyze your own files, you will need to transfer them to Greatlakes. ARC 
provides the Globus tool for moving files. A detailed explanation of how to use 
Globus is outside the scope of this guide but we recommend these links:

- [ARC Globus page](https://documentation.its.umich.edu/node/5019){target="_blank"}
- [ARC Video on using Globus](https://www.mivideo.it.umich.edu/media/t/1_uk9nmxsw/181860561?st=2980){target="_blank"}
- [Globus tutorial from Reproducible Computing workshop](https://umich-brcf-bioinf.github.io/workshop-reproducible-computing/main/html/Module_transferring_data_globus.html){target="_blank"}

