---
title: "Intro to Single Cell RNA-Seq Workshop"
author: "UM Bioinformatics Core Workshop Team"
lang: "en"
output:
        html_document:
            includes:
                in_header: header.html
            theme: paper
            toc: true
            toc_depth: 4
            toc_float: true
            number_sections: false
            fig_caption: true
            markdown: GFM
            code_download: false
---

<style type="text/css">

body, td {
   font-size: 18px;
}
code.r{
  font-size: 12px;
}
pre {
  font-size: 12px
}

</style>

```{r, include = FALSE}
source("../bin/set_values.R")
```

```{r klippy, echo=FALSE, include=TRUE}
klippy::klippy(lang = c("r", "markdown", "bash"), position = c("top", "right"))
```

```{r, include = FALSE}
source("../bin/set_values.R") # can use to store parameters
```

## Wrapping up

We hope you now have more familiarity with key concepts, data types, tools, and how they all
connect to enable single-cell gene expression analysis from RNA-Seq data.

![](images/wayfinder/00-FullWayfinder.png){title="Wayfinder: Wrap-up"}


---

## Housekeeping

- Please take our optional [post-workshop survey](https://michmed.org/7k9Zr){target="_blank"} (5-10 minutes).

- We will email you a link to the final session recordings next week.

- The website/notes for this workshop will be available.

- The [UM Bioinformatics Core Workshop Slack channel](https://umbioinfcoreworkshops.slack.com){target="_blank"} content will be available for 90 days.

---

### Can I continue to use the RStudio environments we used in the workshop?
- The RStudio workshop compute environment will be available through `r workshop_vars$rstudio_server_enddate`.
  - _Please save all your R scripts now_ so that we can "right-size" the compute environment immediately following today's workshop session.

- You can download files from the workshop environment from your terminal/command line window as below. 
  (You will need to substitute your actual workshop username and type workshop password when prompted.)
  ```
  # download workshop files -------------------------------------------------
  mkdir intro_scrnaseq_workshop
  cd intro_scrnaseq_workshop
  scp -r YOUR_USERNAME@`r workshop_vars$ssh_download_dns`:"ISC_R*" .
  ```
  - Note that the full download of the R data is about 8Gb, so depending on your internet 
    speeds it could take a while. (We do not recommend you download the full set of Cell 
    Ranger outputs.)

---

### Where else could I use RStudio?
  - See [Advanced setup instructions](workshop_setup/setup_instructions_advanced.html){target="_blank"}
    for details on how to install RStudio and required packages on your own computer.
    Please note:
    - Installing bioinformatics software is often non-trivial.
    - For typical data, Cell Ranger steps (reviewed Day 1) assume your 
    computer has powerful compute (many CPUs and lots of RAM) and sizable 
    storage capacity. (i.e. it's impractical to run these on your laptop.)
  - [PositCloud](https://posit.cloud/){target="_blank"} allows you to run R-Studio over the web for free. Click here to [login to PositCloud](https://posit.cloud/content/10156998) and launch an R session. (UM users can login using their umich.edu Google account.) *Note: A full single-cell analysis will require you to upgrade beyond the free tier.*
  - UMich users can run RStudio on the [UM Great Lakes HPC](seurat-on-great-lakes.html){target="_blank"}.

---

#### Could I download Seurat inputs used in the workshop?

  - The Seurat inputs (cellranger triples and dbcells files) used in this 
    workshop can be installed locally like so:

    ```
    # download Seurat inputs --------------------------------------------------
    mkdir -p intro_scrnaseq_workshop/ISC_R
    cd intro_scrnaseq_workshop/ISC_R
    
    # Use curl to download a ~2 Gb tarball
    # We'll use evironment variables to avoid extremely long command lines
    source_url="`r workshop_vars$aws_s3_bucket`"
    source_file="`r workshop_vars$aws_s3_file`"
    curl -o workshop_isc_inputs.tgz ${source_url}/${source_file}

    # tar unpacks the tarball into directories
    tar xzvf workshop_isc_inputs.tgz

    # Since we have unpacked the tarball, we can remove it
    rm workshop_isc_inputs.tgz
    ```
<br/>

---

### Can you recommend other relevant workshops or tutorials?
- The [University of Michigan Bioinformatics Core](https://michmed.org/XYQwq){target="_blank"} regularly hosts workshops.
- Intro lessons and workshops in Bash / Git / R / Python : 
  - [Software Carpentry](https://software-carpentry.org/lessons/){target="_blank"}
  - [The Carpentries at the University of Michigan](https://umcarpentries.org/){target="_blank"} host several introductory workshops each year following curricula that are developed collaboratively and using the latest research into best practices for teaching computer science content. All workshops are offered for free to learners. 

---

## How can I learn more about computational research at University of Michigan?

- [UM CoderSpaces "office hours"](https://datascience.isr.umich.edu/events/coderspaces/){target="_blank"} and UM CoderSpaces Slack workspace. _(See "Useful Resources" section in CoderSpaces page for instructions on how to join and access the CoderSpaces Slack workspace.)_
- [Advanced Research Computing](https://its.umich.edu/advanced-research-computing){target="_blank"}
  - [Research Computing Package](https://its.umich.edu/advanced-research-computing/research-computing-package){target="_blank"}
  - Upcoming [UM Advanced Research Computing workshops](https://ttc.iss.lsa.umich.edu/ttc/sessions/tag/hpc/){target="_blank"}
  - Videos on [getting started with Great Lakes](https://www.mivideo.it.umich.edu/playlist/dedicated/1_rmm5rnp1/){target="_blank"} (Requires UM login)


---

## Thank you to our sponsors

![](images/workshop_intro/sponsor_logos.png)

<br/>

## Thank you to/from the workshop team

```{r, child='helper-gallery.Rmd'}
```
<br/>



Thank you for participating in our workshop. We welcome your questions and feedback now and in the future.

Bioinformatics Workshop Team

[bioinformatics-workshops@umich.edu](mailto:bioinformatics-workshops@umich.edu) <br/>
[UM BRCF Bioinformatics Core](https://medresearch.umich.edu/office-research/about-office-research/biomedical-research-core-facilities/bioinformatics-core){target="_blank"}
