---
title: "Intro to Single Cell RNA-Seq Workshop"
author: "UM Bioinformatics Core"
output:
        html_document:
            includes:
                in_header: header.html
            theme: paper
            toc: true
            toc_depth: 4
            toc_float: true
            number_sections: false
            fig_caption: true
            markdown: GFM
            code_download: false
---

<style type="text/css">

body, td {
   font-size: 18px;
}
code.r{
  font-size: 12px;
}
pre {
  font-size: 12px
}

</style>

```{r klippy, echo=FALSE, include=TRUE}
klippy::klippy(lang = c("r", "markdown", "bash"), position = c("top", "right"))
```

## Wrapping up

We hope you now have more familiarity with key concepts, data types, tools, and how they all
connect to enable single-cell gene expression analysis from RNA-Seq data.

![](images/wayfinder/wayfinder_workshop_wrap_up.png)

---

## Housekeeping

- Please take our optional [post-workshop survey](https://forms.gle/St5D6b6mQMg5rmEk9){target="_blank"} (5-10 minutes).

- We will email you a link to the final session recordings next week.

- The website/notes for this workshop will be available.

- The [UM Bioinformatics Core Workshop Slack channel](https://umbioinfcoreworkshops.slack.com){target="_blank"} content will be available for 90 days.

---

## Looking ahead

#### Workshop environment
- RStudio workshop compute environment will be available until **5/1/2024**.
  - _Please save all your R scripts now_ so that we can "right-size" the compute environment immediately following today's workshop session.
- You can download files from the workshop environment from your terminal/command line window as below. 
  (You will need to substitute your actual workshop username and type workshop password when prompted.)
  ```{bash, eval=F, echo=T}
  # download workshop files -------------------------------------------------
  mkdir intro_scrnaseq-workshop
  cd intro_scrnaseq-workshop
  scp -r YOUR_USERNAME@bfx-workshop01.med.umich.edu:"ISC_R*" .
  ```
  - Note that the full download of the R data is about 8Gb, so depending on your internet 
    speeds it could take a while. (We do not recommend you download the full set of Cell 
    Ranger outputs.)

#### Installing software locally
  - You can install necessary programs to run programs locally. Note
    that for typical data, Cell Ranger steps (reviewed Day 1) assume your 
    computer has powerful compute (many CPUs and lots of RAM) and sizable 
    storage capacity. (i.e. it's impractical to run these on your laptop.)
  - Installing bioinformatics software is non-trivial and comprehensive instructions
    to setup a complete compute environment are outside the scope of this workshop.
    (For University of Michigan learners, we are planning to host a Computational Reproducibility workshop later this year 
    that would cover this installation and other related tasks in more detail.) For the intrepid, see relevant links below:
    - [Setting up R/R-Studio](workshop_setup/setup_instructions_advanced.html)

<details>
<summary>*University of Michigan Resources*</summary>
- [UM CoderSpaces "office hours"](https://datascience.isr.umich.edu/events/coderspaces/){target="_blank"} and UM CoderSpaces Slack workspace. _(See "Useful Resources" section of the CoderSpaces page for instructions on how to join the CoderSpaces Slack workspace.)_
- Upcoming [UM Advanced Research Computing workshops](https://arc.umich.edu/events/){target="_blank"}.
- Advanced Research Computing (ARC) at University of Michigan hosts a
  high-performance computing (HPC) platform called _Great Lakes_ which combines
  high-end computers, fast/resilient storage, and pre-installed software.
  Great Lakes may be a good resource for folks who need to run the more compute
  intensive steps and a substantial block of compute and storage is subsidized
  by ARC making it essentially free to many UM researchers.
  - About [Great Lakes](https://arc.umich.edu/greatlakes/){target="_blank"}.
  - About the ARC [Research Computing Package](https://arc.umich.edu/umrcp/){target="_blank"}.
  - Videos on [getting started with Great Lakes](https://www.mivideo.it.umich.edu/channel/ARC-TS%2BTraining/181860561/){target="_blank"}. (Videos are available to UM folks.)


## Resources for continued learning
  - Learning bioinformatic analyses is more like a process than a task. Resources 
    to consider:
    - A CZI NDCN curated list of resources for [self-guided learning](https://ndcn.github.io/self-guided-learning/){target="_blank"}
    - UM Bioinformatics Core links to [training resources/workshops](https://brcf.medicine.umich.edu/cores/bioinformatics-core/training/){target="_blank"}.
    - **Techniques in Bash and R**
      - Lesson plans from the UM Bioinformatics Core's recent [Computational Foundations Workshop](https://umich-brcf-bioinf.github.io/workshop-computational-foundations/main/html/){target="_blank"}.
      - Lessons and workshops in Bash / Git / R / Python : <a href="https://software-carpentry.org/lessons/" target="_blank">
        Software Carpentry</a>

    - **More info on NGS analysis**
      (A decidedly non-exhaustive list in no discernible order.)
      - Lesson plans from UM Bioinformatics Core's recent [RNA-Seq Demystified Workshop](https://umich-brcf-bioinf.github.io/workshop-rnaseq-demystified/main/html/){target="_blank"}. (Note this is for bulk RNA-Seq.)
      - Training materials from [Harvard Chan Bioinformatics Core](https://hbctraining.github.io/main/#advanced-topics-analysis-of-high-throughput-sequencing-ngs-data){target="_blank"}.
      - [Orchestrating Single-Cell Analysis with Bioconductor](https://bioconductor.org/books/3.12/OSCA/index.html){ target="_blank"}.
      - Welcome Sanger Instutute's [Analysis of single cell RNA-seq data](https://www.singlecellcourse.org/index.html){target="_blank"}.
      - In addition to maintaining the Seurat analysis platform, New York Genome
        Center's [Satija lab](https://satijalab.org/){target="_blank"} publishes
        many excellent step-by-step tutorials on single-cell analysis, e.g.
        [Seurat - Guided Clustering Tutorial](https://satijalab.org/seurat/articles/pbmc3k_tutorial){target="_blank"}.
        The Satija lab also hosts [Single Cell Genomics Day](https://satijalab.org/scgd24/){target="_blank"}, 
        a live-streamed annual symposium on emerging techniques in single-cell
        analysis.
      - Nayak, Richa, and Yasha Hasija. **“A hitchhiker's guide to single-cell transcriptomics and data analysis pipelines.”** 
        Genomics vol. 113,2 (2021): 606-619.<br/>
        https://pubmed.ncbi.nlm.nih.gov/33485955/
      - Luecken, Malte D, and Fabian J Theis. **“Current best practices in single-cell RNA-seq analysis: a tutorial.”**
        Molecular systems biology vol. 15,6 e8746. 19 Jun. 2019.<br/>
        https://pubmed.ncbi.nlm.nih.gov/31217225/
      - He, Jiangping et al. **“Practical bioinformatics pipelines for single-cell RNA-seq data analysis.”**
        Biophysics reports vol. 8,3 (2022): 158-169.<br/>
        https://www.ncbi.nlm.nih.gov/pmc/articles/PMC10189648/

    - For more context on the experiments and data we referenced during the workshop, see the source paper: <br/>
      Sorkin, Michael et al. **“Regulation of heterotopic ossification by monocytes in a mouse model of aberrant wound healing.”**
      Nature communications vol. 11,1 722. 5 Feb. 2020.<br/>
      <a class="external" href="https://pubmed.ncbi.nlm.nih.gov/32024825/" target="_blank">https://pubmed.ncbi.nlm.nih.gov/32024825/</a>


---

## Thank you to our sponsors

![](images/workshop_intro/sponsor_logos.png)

<br/>

## Thank you to/from the workshop team

| ![](images/headshots/headshot_cgates.jpg) | ![](images/headshots/headshot_mbradenb.jpg) | ![](images/headshots/headshot_rcavalca.jpg) |
|:-:|:-:|:-:|
| **Chris** | **Marci** | **Raymond** |
![](images/headshots/headshot_damki.jpg) | ![](images/headshots/headshot_trsaari.jpg) | ![](images/headshots/headshot_okoues.jpg) |
| **Dana** | **Travis** | **Olivia** |
![](images/headshots/headshot_rammohan.jpg) | ![](images/headshots/headshot_mkdohert.jpg) | ![](images/headshots/headshot_ncarruth.jpg) |
| **Ram** | **Matt** | **Nick** |

<br/>



Thank you for participating in our workshop. We welcome your questions and feedback now and in the future.

Bioinformatics Workshop Team

[bioinformatics-workshops@umich.edu](mailto:bioinformatics-workshops@umich.edu) <br/>
[UM BRCF Bioinformatics Core](https://medresearch.umich.edu/office-research/about-office-research/biomedical-research-core-facilities/bioinformatics-core){target="_blank"}
